// @metaplex-foundation/genesis/dist/src/allowlist.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isAllowlistActive = exports.computeTreeHeight = exports.getAllowlistLeafBytes = exports.prepareAllowlist = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const merkleTree_1 = require("./merkleTree");
function prepareAllowlist(recipients) {
    const leafs = recipients.map(getAllowlistLeafBytes);
    const tree = (0, merkleTree_1.createMerkleTree)(leafs);
    const merkleRoot = (0, merkleTree_1.getRoot)(tree);
    // Calculate the minimum tree height required for the number of recipients.
    const treeHeight = computeTreeHeight(leafs.length);
    return {
        root: merkleRoot,
        proofs: leafs.map((_, i) => (0, merkleTree_1.getProof)(tree, i)),
        treeHeight,
    };
}
exports.prepareAllowlist = prepareAllowlist;
function getAllowlistLeafBytes(recipient) {
    const pubkey = (0, umi_1.publicKey)(recipient.address);
    const pubkeyBytes = (0, umi_1.publicKeyBytes)(pubkey);
    return pubkeyBytes;
}
exports.getAllowlistLeafBytes = getAllowlistLeafBytes;
/**
 * Computes the minimum Merkle tree height required to accommodate the given
 * number of leaves. The height is defined as the number of internal layers
 * above the leaves (i.e. a tree with a single leaf has height 0).
 */
function computeTreeHeight(leavesCount) {
    if (leavesCount <= 1)
        return 0;
    // Double the capacity at each level. Using log2 to calculate the number of doublings required.
    return Math.ceil(Math.log2(leavesCount));
}
exports.computeTreeHeight = computeTreeHeight;
function isAllowlistActive(presaleBucket) {
    if (!presaleBucket.hasAllowlist) {
        return false;
    }
    if (presaleBucket.allowlistEnd < BigInt(Date.now() * 1000)) {
        return false;
    }
    if (presaleBucket.isAllowlistSolCapReached) {
        return false;
    }
    return true;
}
exports.isAllowlistActive = isAllowlistActive;
