// @metaplex-foundation/genesis/dist/src/auction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.totalBidSol = exports.solPerToken = exports.uSolPerTokenBasis = exports.toSol = exports.toUSol = exports.getBidFillAmount = exports.isBidPartiallyFilled = exports.isBidLosing = exports.isBidWinning = exports.createMicroLamports = exports.fetchAuctionWinners = exports.calculateAuctionWinners = exports.findAuctionBids = exports.getAuctionTreeNode = exports.getAllAuctionTreeNodes = exports.lamportsToUlamports = exports.MICRO_LAMPORTS_PER_LAMPORT = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const generated_1 = require("./generated");
const hooked_1 = require("./hooked");
exports.MICRO_LAMPORTS_PER_LAMPORT = 1000000;
/**
 * Converts lamports to micro-lamports for auction pricing.
 *
 * Auction bids use micro-lamports (1/1,000,000 of a lamport) for sub-lamport precision.
 * This allows for finer granularity in token pricing.
 *
 * @param lamports - The amount in lamports to convert
 * @returns The equivalent amount in micro-lamports
 *
 * @example
 * ```typescript
 * // 0.5 SOL per token = 500,000,000 lamports per token
 * const pricePerToken = lamportsToUlamports(500_000_000);
 *
 * // Use in auction bid
 * await placeAuctionBidV1(umi, {
 *   // ... other params
 *   pricePerToken,
 *   // ...
 * }).sendAndConfirm(umi);
 * ```
 */
function lamportsToUlamports(lamportsInput) {
    return BigInt(lamportsInput) * BigInt(1000000);
}
exports.lamportsToUlamports = lamportsToUlamports;
/**
 * Retrieves all nodes from an auction AVL tree.
 *
 * @param tree - The auction AVL tree to extract nodes from
 * @returns An array of all auction nodes in the tree, in allocation order
 *
 * @example
 * ```typescript
 * const tree = await fetchAuctionAVLTreeV1(umi, treePda);
 * const allNodes = getAllAuctionTreeNodes(tree);
 * console.log(`Tree contains ${allNodes.length} nodes`);
 * ```
 */
function getAllAuctionTreeNodes(tree) {
    const nodes = [];
    for (let i = 0; i < tree.allocator.size; i += 1) {
        nodes.push((0, hooked_1.getAuctionNodeSerializer)().deserialize(tree.rawData, tree.rawDataOffset + i * hooked_1.AUCTION_NODE_SIZE)[0]);
    }
    return nodes;
}
exports.getAllAuctionTreeNodes = getAllAuctionTreeNodes;
/**
 * Retrieves a specific node from an auction AVL tree by its index.
 *
 * @param tree - The auction AVL tree to search in
 * @param index - The 1-based index of the node to retrieve
 * @returns The auction node at the specified index, or null if index is out of bounds
 *
 * @example
 * ```typescript
 * const tree = await fetchAuctionAVLTreeV1(umi, treePda);
 * const node = getAuctionTreeNode(tree, 5);
 * if (node) {
 *   console.log(`Node 5 bid: ${node.bid.pricePerToken}`);
 * }
 * ```
 */
function getAuctionTreeNode(tree, index) {
    if (index < 1 || index > tree.allocator.size) {
        return null;
    }
    return (0, hooked_1.getAuctionNodeSerializer)().deserialize(tree.rawData, tree.rawDataOffset + (index - 1) * hooked_1.AUCTION_NODE_SIZE)[0];
}
exports.getAuctionTreeNode = getAuctionTreeNode;
/**
 * Finds all bids in the auction tree that meet or exceed the clearing price.
 *
 * Uses an in-order traversal of the AVL tree to efficiently collect qualifying bids.
 * The function traverses from highest to lowest prices and stops when prices fall
 * below the clearing price threshold.
 *
 * @param tree - The auction AVL tree containing all bids
 * @param clearingPriceULamports - The minimum price in micro-lamports for a bid to qualify
 * @returns Array of bids that meet or exceed the clearing price, sorted by price (highest first)
 *
 * @example
 * ```typescript
 * const tree = await fetchAuctionAVLTreeV1(umi, treePda);
 * const clearingPrice = lamportsToMicroLamports(1_000_000); // 0.001 SOL
 * const qualifyingBids = findAuctionBids(tree, clearingPrice);
 * ```
 */
function findAuctionBids(tree, clearingPriceULamports) {
    const bids = [];
    // Use in-order traversal to collect winners efficiently
    // Since AVL tree is sorted, we can traverse from highest to lowest
    function collectBids(nodeIndex) {
        if (nodeIndex === hooked_1.AUCTION_SENTINEL) {
            return;
        }
        const [node] = (0, hooked_1.getAuctionNodeSerializer)().deserialize(tree.rawData, tree.rawDataOffset + (nodeIndex - 1) * hooked_1.AUCTION_NODE_SIZE);
        if (!node || node.bid.priceUlamports === 0n) {
            // Skip deleted/inactive nodes
            return;
        }
        // Always traverse right subtree first (higher prices)
        // The right subtree might have matching bids even if current node doesn't
        collectBids(node.right);
        if (node.bid.priceUlamports >= clearingPriceULamports) {
            bids.push(node.bid);
            // Traverse left subtree to find bids of the same price
            collectBids(node.left);
        }
    }
    // Start traversal from root
    collectBids(tree.allocator.root);
    return bids;
}
exports.findAuctionBids = findAuctionBids;
/**
 * Calculates the winning bids for an auction based on clearing price and token allocation.
 *
 * This function determines which bids win tokens in the auction by:
 * 1. Finding all bids at or above the clearing price
 * 2. Sorting them by price (highest first) and bid number (earliest first for ties)
 * 3. Allocating tokens until the supply is exhausted
 *
 * @param tree - The auction AVL tree containing all bids
 * @param clearingPriceULamports - The clearing price in micro-lamports
 * @param tokenAllocation - The total number of tokens available for distribution
 * @returns Array of winning bids in order of priority
 *
 * @example
 * ```typescript
 * const tree = await fetchAuctionAVLTreeV1(umi, treePda);
 * const clearingPrice = lamportsToMicroLamports(500_000_000); // 0.5 SOL
 * const tokenSupply = 1_000_000n;
 * const winners = calculateAuctionWinners(tree, clearingPrice, tokenSupply);
 * ```
 */
function calculateAuctionWinners(tree, clearingPriceULamports, tokenAllocation) {
    const bids = findAuctionBids(tree, clearingPriceULamports);
    bids.sort((a, b) => {
        if (a.priceUlamports === b.priceUlamports) {
            if (a.bidNumber === b.bidNumber)
                return 0;
            return a.bidNumber < b.bidNumber ? -1 : 1;
        }
        return a.priceUlamports > b.priceUlamports ? -1 : 1;
    });
    const winners = [];
    let remainingTokens = tokenAllocation;
    while (remainingTokens > 0n) {
        const bid = bids.shift();
        if (!bid) {
            break;
        }
        winners.push(bid);
        remainingTokens -= bid.tokenQuantity;
    }
    return winners;
}
exports.calculateAuctionWinners = calculateAuctionWinners;
/**
 * Fetches all winning bids from an auction bucket.
 *
 * This async function retrieves the auction tree associated with a bucket
 * and calculates the winning bids based on the bucket's clearing price
 * and token allocation.
 *
 * @param umi - Umi context containing RPC, programs, and eddsa for Solana interaction
 * @param auction - The auction bucket containing clearing price and allocation info
 * @returns Promise resolving to an array of winning bids
 * @throws Error if the auction tree is not found
 *
 * @example
 * ```typescript
 * const auctionBucket = await fetchAuctionBucketV1(umi, bucketPda);
 * const winners = await getAuctionWinners(umi, auctionBucket);
 * console.log(`${winners.length} winning bids found`);
 * ```
 */
async function fetchAuctionWinners(umi, auction) {
    const treePda = (0, generated_1.findAuctionAVLTreeV1Pda)(umi, {
        bucket: auction.publicKey,
    });
    const tree = await (0, generated_1.safeFetchAuctionAVLTreeV1)(umi, treePda[0]);
    if (!tree) {
        throw new Error('Auction tree not found');
    }
    // Use the clearing price from the auction bucket
    const { clearingPriceUlamports, tokenAllocation } = auction;
    return calculateAuctionWinners(tree, clearingPriceUlamports, tokenAllocation);
}
exports.fetchAuctionWinners = fetchAuctionWinners;
/**
 * Creates a MicroSolAmount from either a bigint or an existing MicroSolAmount.
 *
 * @param amount - The amount as bigint or MicroSolAmount
 * @returns A MicroSolAmount object
 * @internal
 */
function createMicroLamports(amount) {
    return typeof amount === 'bigint' ? (0, umi_1.microLamports)(amount) : amount;
}
exports.createMicroLamports = createMicroLamports;
/**
 * Determines if a bid is winning based on the auction's clearing price and bid number.
 *
 * A bid is considered winning if:
 * - Its price exceeds the clearing price, OR
 * - Its price equals the clearing price AND its bid number is at or before the clearing price bid number
 *
 * @param auction - The auction bucket with clearing price and clearing price bid number
 * @param bid - The bid to check with price per token and bid number
 * @returns True if the bid is winning, false otherwise
 *
 * @example
 * ```typescript
 * const auction = await fetchAuctionBucketV1(umi, auctionPda);
 * const bid = await fetchAuctionBidV1(umi, bidPda);
 * const isWinner = isBidWinning(auction, bid); // true (price exceeds clearing)
 * ```
 */
function isBidWinning(auction, bid) {
    const priceUlamports = createMicroLamports(bid.priceUlamports);
    const clearingPrice = createMicroLamports(auction.clearingPriceUlamports);
    if ((0, umi_1.isGreaterThanAmount)(priceUlamports, clearingPrice)) {
        return true;
    }
    if ((0, umi_1.isEqualToAmount)(priceUlamports, clearingPrice) &&
        bid.bidNumber <= auction.clearingPriceBidNumber) {
        return true;
    }
    return false;
}
exports.isBidWinning = isBidWinning;
/**
 * Determines if a bid is losing (not winning) in the auction.
 *
 * This is the inverse of isBidWinning - a bid is losing if it doesn't meet
 * the criteria for winning.
 *
 * @param auction - The auction bucket with clearing price and clearing price bid number
 * @param bid - The bid to check with price per token and bid number
 * @returns True if the bid is losing, false if winning
 *
 * @example
 * ```typescript
 * const auction = await fetchAuctionBucketV1(umi, auctionPda);
 * const bid = await fetchAuctionBidV1(umi, bidPda);
 * const isLoser = isBidLosing(auction, bid); // true (price below clearing)
 * ```
 */
function isBidLosing(auction, bid) {
    return !isBidWinning(auction, bid);
}
exports.isBidLosing = isBidLosing;
/**
 * Determines if a bid is partially filled at the clearing price.
 *
 * A bid is considered partially filled when:
 * - The bid is at the clearing price (bid number matches clearing price bid number)
 * - The bid requests more tokens than are available at the clearing price
 *
 * This occurs when the auction's token supply is exhausted at the clearing price,
 * causing the last bid at that price to receive only a portion of their requested tokens.
 *
 * @param auctionBucket - The auction bucket with clearing price bid number and allocation
 * @param bid - The bid to check with bid number and requested token quantity
 * @returns True if the bid is partially filled, false if fully filled or not at clearing price
 *
 * @example
 * ```typescript
 * const auction = await fetchAuctionBucketV1(umi, auctionPda);
 * const bid = await fetchAuctionBidV1(umi, bidPda);
 * const isPartial = isBidPartiallyFilled(auction, bid); // true (partial fill at clearing price)
 * ```
 */
function isBidPartiallyFilled(auctionBucket, bid) {
    if (auctionBucket.clearingPriceBidNumber === bid.bidNumber &&
        bid.tokenQuantity > auctionBucket.clearingPriceAllocation) {
        return true;
    }
    return false;
}
exports.isBidPartiallyFilled = isBidPartiallyFilled;
/**
 * Calculates the actual token amount a bid will receive.
 *
 * Returns either the partial fill amount (if bid is at clearing price) or
 * the full token quantity requested by the bid.
 *
 * @param bid - The bid with bid number and requested token quantity
 * @param auctionBucket - The auction bucket with clearing price info
 * @returns The actual number of tokens the bid will receive
 *
 * @example
 * ```typescript
 * const bid = await fetchAuctionBidV1(umi, bidPda);
 * const auction = await fetchAuctionBucketV1(umi, auctionPda);
 * const fillAmount = getBidFillAmount(bid, auction); // 50000n (partial fill)
 * ```
 */
function getBidFillAmount(auctionBucket, bid) {
    return isBidPartiallyFilled(auctionBucket, bid)
        ? auctionBucket.clearingPriceAllocation
        : bid.tokenQuantity;
}
exports.getBidFillAmount = getBidFillAmount;
/**
 * Converts SOL amount to micro-SOL (µSOL) amount.
 *
 * @param sol - The amount in SOL (lamports)
 * @returns The equivalent amount in micro-SOL (micro-lamports)
 *
 * @example
 * ```typescript
 * const solAmount = lamports(1_000_000_000); // 1 SOL
 * const microSolAmount = toUSol(solAmount); // 1,000,000,000,000,000 µSOL
 * ```
 */
function toUSol(sol) {
    return (0, umi_1.microLamports)(sol.basisPoints * BigInt(exports.MICRO_LAMPORTS_PER_LAMPORT));
}
exports.toUSol = toUSol;
/**
 * Converts micro-SOL (µSOL) amount to SOL amount.
 *
 * @param uSol - The amount in micro-SOL (micro-lamports)
 * @returns The equivalent amount in SOL (lamports)
 *
 * @example
 * ```typescript
 * const microSolAmount = microLamports(1_000_000_000_000_000n); // 1,000,000,000,000,000 µSOL
 * const solAmount = toSol(microSolAmount); // 1 SOL
 * ```
 */
function toSol(uSol) {
    return (0, umi_1.lamports)((0, umi_1.divideAmount)(uSol, BigInt(exports.MICRO_LAMPORTS_PER_LAMPORT)).basisPoints);
}
exports.toSol = toSol;
/**
 * Calculates the micro-SOL price per token basis unit.
 *
 * Converts a SOL per token price to micro-SOL per smallest token unit,
 * accounting for token decimals. This is used for precise sub-lamport pricing.
 *
 * @param solPerTokenInput - The price in SOL per whole token (or lamports as bigint)
 * @param tokenDecimals - The number of decimal places for the token
 * @returns The price in micro-SOL per token basis unit
 *
 * @example
 * ```typescript
 * // For a token with 6 decimals, priced at 0.5 SOL per token
 * const priceUlamports = lamports(500_000_000); // 0.5 SOL
 * const microSolPerBasis = uSolPerTokenBasis(priceUlamports, 6);
 * // Returns price per smallest token unit (1/1,000,000 of a token)
 * ```
 */
function uSolPerTokenBasis(solPerTokenInput, tokenDecimals) {
    return (0, umi_1.divideAmount)(toUSol(typeof solPerTokenInput === 'bigint'
        ? (0, umi_1.lamports)(solPerTokenInput)
        : solPerTokenInput), 10 ** tokenDecimals);
}
exports.uSolPerTokenBasis = uSolPerTokenBasis;
/**
 * Converts micro-SOL per token basis unit to SOL per whole token.
 *
 * Takes a price in micro-SOL per smallest token unit and converts it to
 * SOL per whole token, accounting for token decimals.
 *
 * @param ulamportsPerTokenBasis - The price in micro-SOL per token basis unit
 * @param tokenDecimals - The number of decimal places for the token
 * @returns The price in SOL per whole token
 *
 * @example
 * ```typescript
 * // For a token with 6 decimals
 * const microSolPerBasis = microLamports(500n); // Price per smallest unit
 * const solPerWholeToken = solPerToken(microSolPerBasis, 6);
 * // Returns price in SOL for one whole token
 * ```
 */
function solPerToken(ulamportsPerTokenBasis, tokenDecimals) {
    return toSol((0, umi_1.multiplyAmount)(createMicroLamports(ulamportsPerTokenBasis), 10 ** tokenDecimals));
}
exports.solPerToken = solPerToken;
/**
 * Calculates the total SOL cost for a bid.
 *
 * Multiplies the micro-SOL price per token basis by the token amount
 * to get the total cost in SOL.
 *
 * @param pricePerTokenBasis - The price in micro-SOL per token basis unit
 * @param tokenAmount - The number of token basis units to purchase
 * @returns The total cost in SOL
 *
 * @example
 * ```typescript
 * // Bidding for 1,000,000 token units at 500 micro-SOL each
 * const pricePerUnit = microLamports(500n);
 * const amount = 1_000_000n;
 * const totalCost = totalBidSol(pricePerUnit, amount);
 * // Returns total SOL needed for the bid
 * ```
 */
function totalBidSol(pricePerTokenBasis, tokenAmount) {
    const microLamportsPerToken = (0, umi_1.multiplyAmount)(createMicroLamports(pricePerTokenBasis), tokenAmount);
    return toSol(microLamportsPerToken);
}
exports.totalBidSol = totalBidSol;
