// @metaplex-foundation/genesis/dist/src/bondingCurveHelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getCurrentPrice = exports.getSwapResult = exports.getSwapAmountInForOut = exports.getSwapAmountOutForIn = void 0;
const fees_1 = require("./fees");
const generated_1 = require("./generated");
const bigInt_1 = require("./utils/bigInt");
/**
 * Calculate the output amount for a given input amount on a bonding curve.
 * This function determines how much of the output asset you'll receive for a given input amount.
 * This function does not take into account fees.
 *
 * @param bucket - The bonding curve bucket containing reserves and parameters
 * @param amountIn - The input amount to swap
 * @param direction - The swap direction (Buy = SOL in, tokens out; Sell = tokens in, SOL out)
 * @returns The output amount you'll receive from the swap
 * @throws Error if the bonding curve type is unsupported or calculation errors occur
 */
function getSwapAmountOutForIn(bucket, amountIn, direction) {
    switch (bucket.bondingCurveType) {
        case generated_1.BondingCurveType.ConstantProduct:
            return getConstantProductSwapAmountOutForIn(amountIn, bucket.bondingCurveParameters.virtualSol, bucket.bondingCurveParameters.virtualTokens, bucket.quoteDepositTotal, bucket.tokenBalance, direction);
        default:
            throw new Error(`Unsupported bonding curve type: ${bucket.bondingCurveType}`);
    }
}
exports.getSwapAmountOutForIn = getSwapAmountOutForIn;
/**
 * Calculate the input amount needed to receive a desired output amount on a bonding curve.
 * This function determines how much of the input asset you need to provide to get a specific output amount.
 * This function does not take into account fees.
 *
 * @param bucket - The bonding curve bucket containing reserves and parameters
 * @param amountOut - The desired output amount to receive
 * @param direction - The swap direction (Buy = SOL in, tokens out; Sell = tokens in, SOL out)
 * @returns The input amount required to achieve the desired output
 * @throws Error if the bonding curve type is unsupported or calculation errors occur
 */
function getSwapAmountInForOut(bucket, amountOut, direction) {
    switch (bucket.bondingCurveType) {
        case generated_1.BondingCurveType.ConstantProduct:
            return getConstantProductSwapAmountInForOut(amountOut, bucket.bondingCurveParameters.virtualSol, bucket.bondingCurveParameters.virtualTokens, bucket.quoteDepositTotal, bucket.tokenBalance, direction);
        default:
            throw new Error(`Unsupported bonding curve type: ${bucket.bondingCurveType}`);
    }
}
exports.getSwapAmountInForOut = getSwapAmountInForOut;
/**
 * Calculates the complete swap result including fees for a bonding curve swap.
 *
 * @param bucket - The bonding curve bucket
 * @param amountIn - The input amount (before fees)
 * @param swapDirection - The swap direction (Buy = SOL in, tokens out; Sell = tokens in, SOL out)
 * @returns An object containing the actual input amount, fee, and output amount
 * @throws Error if the bucket has insufficient reserves or calculation errors occur
 */
function getSwapResult(bucket, amountIn, swapDirection) {
    if (amountIn === 0n) {
        throw new Error('Invalid amount: must be greater than 0');
    }
    // Fee handling logic:
    // We apply the fee to the SOL amount, whether it's the input or output of the swap.
    // If we're swapping SOL for tokens (Buy), we apply the fee to the input amount
    let amountInAfterFee;
    let feeIn;
    if (swapDirection === generated_1.SwapDirection.Buy) {
        feeIn = (0, fees_1.calculateFee)(amountIn, bucket.depositFeeType, bucket.depositFee);
        amountInAfterFee = (0, bigInt_1.maxBigInt)(amountIn - feeIn, 0n);
    }
    else {
        amountInAfterFee = amountIn;
        feeIn = 0n;
    }
    // Calculate the swap amount
    let amountOut = getSwapAmountOutForIn(bucket, amountInAfterFee, swapDirection);
    // If we're swapping tokens for SOL (Sell), we apply the fee to the output amount
    let feeOut;
    if (swapDirection === generated_1.SwapDirection.Sell) {
        feeOut = (0, fees_1.calculateFee)(amountOut, bucket.withdrawFeeType, bucket.withdrawFee);
        amountOut = (0, bigInt_1.maxBigInt)(amountOut - feeOut, 0n);
    }
    else {
        feeOut = 0n;
    }
    // Check if we have enough reserves for the swap
    const reservesOut = swapDirection === generated_1.SwapDirection.Buy
        ? bucket.tokenBalance
        : bucket.quoteDepositTotal;
    if (reservesOut === 0n) {
        throw new Error('Insufficient liquidity: no reserves available');
    }
    // If we're swapping more tokens than the bucket has, we need to recalculate the swap
    if (amountOut + feeOut > reservesOut) {
        // Match Rust implementation lines 234-257
        if (swapDirection === generated_1.SwapDirection.Sell) {
            feeOut = (0, fees_1.calculateFee)(reservesOut, bucket.withdrawFeeType, bucket.withdrawFee);
            // We subtract the fee from the reservesOut, as we are removing SOL from the bucket in the sell case
            amountOut = reservesOut - feeOut;
            // If amountOut is negative, that means that we only have enough SOL to cover the fee.
            // In this case, we set amountOut to 0 and feeOut to the reservesOut.
            if (amountOut < 0n) {
                amountOut = 0n;
                feeOut = reservesOut;
            }
        }
        else {
            amountOut = reservesOut;
            feeOut = 0n;
        }
        // Recalculate the amountIn for the new amountOut
        amountInAfterFee = getSwapAmountInForOut(bucket, amountOut, swapDirection);
        // No need to remove feeIn from amountIn, as we are adding SOL from the payer to the bucket in the buy case
        if (swapDirection === generated_1.SwapDirection.Buy) {
            feeIn = (0, fees_1.calculateFee)(amountInAfterFee, bucket.depositFeeType, bucket.depositFee);
        }
        else {
            feeIn = 0n;
        }
    }
    // Calculate total amount in (including fees) - this is what the user actually pays
    const totalAmountIn = swapDirection === generated_1.SwapDirection.Buy
        ? amountInAfterFee + feeIn
        : amountInAfterFee;
    return {
        amountIn: totalAmountIn,
        fee: feeIn + feeOut,
        amountOut,
    };
}
exports.getSwapResult = getSwapResult;
/**
 * Get the current price of the bonding curve.
 *
 * @param bucket - The bonding curve bucket
 * @returns The current price
 */
function getCurrentPrice(bucket) {
    switch (bucket.bondingCurveType) {
        case generated_1.BondingCurveType.ConstantProduct:
            return bucket.bondingCurveParameters.virtualSol > 0n
                ? (bucket.bondingCurveParameters.virtualTokens + bucket.tokenBalance) /
                    (bucket.bondingCurveParameters.virtualSol +
                        bucket.quoteDepositTotal)
                : 0n;
        default:
            throw new Error(`Unsupported bonding curve type: ${bucket.bondingCurveType}`);
    }
}
exports.getCurrentPrice = getCurrentPrice;
// Constant Product swap helpers
/**
 * Calculate the output amount for a given input amount using constant product AMM formula.
 * This mirrors the Rust implementation in bonding_curve.rs.
 *
 * @param amountIn - The input amount
 * @param virtualSol - Virtual SOL reserves
 * @param virtualTokens - Virtual token reserves
 * @param realSol - Real SOL deposits
 * @param realTokens - Real token balance
 * @param direction - Swap direction (Buy = SOL in, tokens out; Sell = tokens in, SOL out)
 * @returns The output amount
 * @throws Error if calculation would cause overflow or invalid constant product
 */
function getConstantProductSwapAmountOutForIn(amountIn, virtualSol, virtualTokens, realSol, realTokens, direction) {
    // Calculate total reserves
    const totalSol = virtualSol + realSol;
    const totalTokens = virtualTokens + realTokens;
    // Calculate constant product K
    const k = totalSol * totalTokens;
    // Determine input and output reserves based on direction
    // Buy: SOL in, tokens out
    // Sell: tokens in, SOL out
    const [inputReserve, outputReserve] = direction === generated_1.SwapDirection.Buy
        ? [totalSol, totalTokens]
        : [totalTokens, totalSol];
    // Constant product AMM formula:
    // output_amount = output_reserve - K / (input_reserve + amount_in)
    const newInputReserve = inputReserve + amountIn;
    // Use ceiling division to round up (equivalent to div_ceil in Rust)
    const newOutputReserve = divCeil(k, newInputReserve);
    const outputAmount = outputReserve - newOutputReserve;
    // Verify constant product doesn't decrease
    const newK = newInputReserve * (outputReserve - outputAmount);
    if (newK < k) {
        throw new Error('Constant product would decrease');
    }
    if (outputAmount < 0n) {
        throw new Error('Insufficient liquidity');
    }
    return outputAmount;
}
/**
 * Calculate the input amount needed for a given output amount using constant product AMM formula.
 * This mirrors the Rust implementation in bonding_curve.rs.
 *
 * @param amountOut - The desired output amount
 * @param virtualSol - Virtual SOL reserves
 * @param virtualTokens - Virtual token reserves
 * @param realSol - Real SOL deposits
 * @param realTokens - Real token balance
 * @param direction - Swap direction (Buy = SOL in, tokens out; Sell = tokens in, SOL out)
 * @returns The required input amount
 * @throws Error if calculation would cause overflow or invalid constant product
 */
function getConstantProductSwapAmountInForOut(amountOut, virtualSol, virtualTokens, realSol, realTokens, direction) {
    // Calculate total reserves
    const totalSol = virtualSol + realSol;
    const totalTokens = virtualTokens + realTokens;
    // Calculate constant product K
    const k = totalSol * totalTokens;
    // Determine input and output reserves based on direction
    const [inputReserve, outputReserve] = direction === generated_1.SwapDirection.Buy
        ? [totalSol, totalTokens]
        : [totalTokens, totalSol];
    if (amountOut >= outputReserve) {
        throw new Error('Output amount exceeds available liquidity');
    }
    // Constant product AMM formula:
    // input_amount = K / (output_reserve - output_amount) - input_reserve
    const newOutputReserve = outputReserve - amountOut;
    // Use ceiling division to round up
    const newInputReserve = divCeil(k, newOutputReserve);
    const inputAmount = newInputReserve - inputReserve;
    // Verify constant product doesn't decrease
    const newK = newInputReserve * newOutputReserve;
    if (newK < k) {
        throw new Error('Constant product would decrease');
    }
    if (inputAmount < 0n) {
        throw new Error('Invalid input amount calculation');
    }
    return inputAmount;
}
// Miscellaneous
/**
 * Performs ceiling division for bigint values.
 * Equivalent to div_ceil in Rust: rounds up when there's a remainder.
 *
 * @param dividend - The number to be divided
 * @param divisor - The number to divide by
 * @returns The ceiling of the division result
 */
function divCeil(dividend, divisor) {
    return dividend / divisor + (dividend % divisor > 0n ? 1n : 0n);
}
