// @metaplex-foundation/genesis/dist/src/conditions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLinearBpsScheduleV2WithRelativeStart = exports.createLinearBpsScheduleV2WithAbsoluteStart = exports.createNeverClaimSchedule = exports.createClaimSchedule = exports.createNeverCondition = exports.createTimeRelativeCondition = exports.createTimeAbsoluteCondition = exports.isConditionArgs = void 0;
const schedule_1 = require("./schedule");
/**
 * Type guard to narrow a ConditionArgs to a specific variant.
 * @param kind - The discriminant to check for ('TimeAbsolute' or 'TimeRelative')
 * @param value - The ConditionArgs value to narrow
 * @returns True if the value is the specified kind, with type narrowing
 */
function isConditionArgs(kind, value) {
    return value.__kind === kind;
}
exports.isConditionArgs = isConditionArgs;
// ---------------------------------------------------------------------------
// Condition Creators
// ---------------------------------------------------------------------------
/**
 * Creates a TimeAbsolute condition that triggers at a specific Unix timestamp.
 * @param time - Unix timestamp (seconds) when the condition is met
 */
function createTimeAbsoluteCondition(time) {
    return {
        __kind: 'TimeAbsolute',
        padding: Array(47).fill(0),
        time,
        triggeredTimestamp: null,
    };
}
exports.createTimeAbsoluteCondition = createTimeAbsoluteCondition;
/**
 * Creates a TimeRelative condition that triggers relative to another bucket's
 * start or end time.
 * @param referenceBucket - The public key of the bucket to reference
 * @param bucketTime - Which time on the reference bucket (e.g. Start, End)
 * @param timeOffset - Offset in seconds from the reference time (default 0)
 */
function createTimeRelativeCondition(referenceBucket, bucketTime, timeOffset = 0n) {
    return {
        __kind: 'TimeRelative',
        bucketTime,
        padding: Array(14).fill(0),
        timeOffset,
        bucket: referenceBucket,
        triggeredTimestamp: null,
    };
}
exports.createTimeRelativeCondition = createTimeRelativeCondition;
/**
 * Creates a Never condition. The condition will never be met and its
 * `triggeredTimestamp` will never be set. Use this for permanently locked
 * schedules (e.g. LP tokens that should never vest).
 */
function createNeverCondition() {
    return {
        __kind: 'Never',
        padding: Array(55).fill(0),
        triggeredTimestamp: null,
    };
}
exports.createNeverCondition = createNeverCondition;
// ---------------------------------------------------------------------------
// ClaimSchedule Creators
// ---------------------------------------------------------------------------
/**
 * Creates a ClaimSchedule with TimeAbsolute conditions.
 *
 * Conditions start untriggered — call `CrankExtensionConditionsV2`
 * (or `TriggerConditionsV2`) before claiming to trigger them on-chain.
 *
 * @param params.startTime  - Unix timestamp when linear vesting begins
 * @param params.endTime    - Unix timestamp when linear vesting ends
 * @param params.cliffTime  - Unix timestamp for the cliff
 * @param params.cliffAmountBps - Basis points (0-10000) unlocked at cliff
 * @param params.period     - Vesting period length in seconds
 */
function createClaimSchedule(params) {
    return {
        startCondition: createTimeAbsoluteCondition(params.startTime),
        duration: params.endTime - params.startTime,
        period: params.period,
        cliffCondition: createTimeAbsoluteCondition(params.cliffTime),
        cliffAmountBps: params.cliffAmountBps,
        reserved: Array(126).fill(0),
    };
}
exports.createClaimSchedule = createClaimSchedule;
/**
 * Creates a ClaimSchedule with Never conditions for both start and cliff.
 * Represents a **permanently locked** schedule where tokens never vest.
 *
 * Use this for LP tokens that should be locked forever.
 *
 * @example
 * ```ts
 * import { createNeverClaimSchedule } from '@metaplex-foundation/genesis';
 *
 * const schedule = createNeverClaimSchedule();
 * // pass as lpLockSchedule when adding a Raydium CPMM bucket
 * ```
 */
function createNeverClaimSchedule() {
    return {
        startCondition: createNeverCondition(),
        duration: 1n,
        period: 1n,
        cliffCondition: createNeverCondition(),
        cliffAmountBps: 0,
        reserved: Array(126).fill(0),
    };
}
exports.createNeverClaimSchedule = createNeverClaimSchedule;
// ---------------------------------------------------------------------------
// LinearBpsScheduleV2 Creators
// ---------------------------------------------------------------------------
/**
 * Creates a LinearBpsScheduleV2 with a TimeAbsolute start condition.
 *
 * @param params.startTime  - Unix timestamp when the schedule begins
 * @param params.duration   - Duration of the schedule in seconds
 * @param params.point1     - First point on the line (timeBps in [0,10000], bps)
 * @param params.point2     - Second point on the line (timeBps in [0,10000], bps)
 * @param params.maxBps     - Upper bound for the evaluated value
 */
function createLinearBpsScheduleV2WithAbsoluteStart(params) {
    return (0, schedule_1.createLinearBpsScheduleV2)(params.point1, params.point2, {
        startCondition: createTimeAbsoluteCondition(params.startTime),
        duration: params.duration,
        maxBps: params.maxBps,
    });
}
exports.createLinearBpsScheduleV2WithAbsoluteStart = createLinearBpsScheduleV2WithAbsoluteStart;
/**
 * Creates a LinearBpsScheduleV2 with a TimeRelative start condition.
 *
 * @param params.referenceBucket - The bucket to reference
 * @param params.bucketTime      - Which time on the reference bucket
 * @param params.timeOffset      - Optional start offset in seconds relative to the reference
 *                                  bucket time. Passed to `createTimeRelativeCondition` which
 *                                  defaults to `0n` when `undefined`.
 * @param params.duration        - Duration of the schedule in seconds
 * @param params.point1          - First point on the line
 * @param params.point2          - Second point on the line
 * @param params.maxBps          - Upper bound for the evaluated value
 */
function createLinearBpsScheduleV2WithRelativeStart(params) {
    return (0, schedule_1.createLinearBpsScheduleV2)(params.point1, params.point2, {
        startCondition: createTimeRelativeCondition(params.referenceBucket, params.bucketTime, params.timeOffset),
        duration: params.duration,
        maxBps: params.maxBps,
    });
}
exports.createLinearBpsScheduleV2WithRelativeStart = createLinearBpsScheduleV2WithRelativeStart;
