// @metaplex-foundation/genesis/dist/src/fees.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.calculateFee = exports.DEFAULT_FEE = exports.DEFAULT_FEE_TYPE = exports.DEFAULT_UNLOCKED_CLAIM_FEE = exports.DEFAULT_UNLOCKED_CLAIM_FEE_TYPE = exports.DEFAULT_UNLOCKED_FEE = exports.DEFAULT_UNLOCKED_FEE_TYPE = exports.DEFAULT_AUCTION_BID_FEE = exports.DEFAULT_AUCTION_BID_FEE_TYPE = exports.DEFAULT_AUCTION_FEE = exports.DEFAULT_AUCTION_FEE_TYPE = exports.DEFAULT_BONDING_CURVE_WITHDRAW_FEE = exports.DEFAULT_BONDING_CURVE_WITHDRAW_FEE_TYPE = exports.DEFAULT_BONDING_CURVE_DEPOSIT_FEE = exports.DEFAULT_BONDING_CURVE_DEPOSIT_FEE_TYPE = exports.DEFAULT_BONDING_CURVE_FEE = exports.DEFAULT_BONDING_CURVE_FEE_TYPE = exports.DEFAULT_VAULT_WITHDRAW_FEE = exports.DEFAULT_VAULT_WITHDRAW_FEE_TYPE = exports.DEFAULT_VAULT_DEPOSIT_FEE = exports.DEFAULT_VAULT_DEPOSIT_FEE_TYPE = exports.DEFAULT_VAULT_FEE = exports.DEFAULT_VAULT_FEE_TYPE = exports.DEFAULT_PRESALE_WITHDRAW_FEE = exports.DEFAULT_PRESALE_WITHDRAW_FEE_TYPE = exports.DEFAULT_PRESALE_CLAIM_FEE = exports.DEFAULT_PRESALE_CLAIM_FEE_TYPE = exports.DEFAULT_PRESALE_DEPOSIT_FEE = exports.DEFAULT_PRESALE_DEPOSIT_FEE_TYPE = exports.DEFAULT_PRESALE_FEE_V2 = exports.DEFAULT_PRESALE_FEE_TYPE_V2 = exports.DEFAULT_PRESALE_FEE_V1 = exports.DEFAULT_PRESALE_FEE_TYPE_V1 = exports.DEFAULT_LAUNCHPOOL_CLAIM_FEE = exports.DEFAULT_LAUNCHPOOL_CLAIM_FEE_TYPE = exports.DEFAULT_LAUNCHPOOL_WITHDRAW_FEE = exports.DEFAULT_LAUNCHPOOL_WITHDRAW_FEE_TYPE = exports.DEFAULT_LAUNCHPOOL_DEPOSIT_FEE = exports.DEFAULT_LAUNCHPOOL_DEPOSIT_FEE_TYPE = exports.DEFAULT_LAUNCHPOOL_FEE_V2 = exports.DEFAULT_LAUNCHPOOL_FEE_TYPE_V2 = exports.DEFAULT_LAUNCHPOOL_FEE = exports.DEFAULT_LAUNCHPOOL_FEE_TYPE = void 0;
const generated_1 = require("./generated");
const bigInt_1 = require("./utils/bigInt");
// Launch Pool
exports.DEFAULT_LAUNCHPOOL_FEE_TYPE = generated_1.FeeDiscriminants.Flat;
exports.DEFAULT_LAUNCHPOOL_FEE = 0n;
exports.DEFAULT_LAUNCHPOOL_FEE_TYPE_V2 = generated_1.FeeDiscriminants.Flat;
exports.DEFAULT_LAUNCHPOOL_FEE_V2 = 0n;
exports.DEFAULT_LAUNCHPOOL_DEPOSIT_FEE_TYPE = generated_1.FeeDiscriminants.BasisPoints;
exports.DEFAULT_LAUNCHPOOL_DEPOSIT_FEE = 200n;
exports.DEFAULT_LAUNCHPOOL_WITHDRAW_FEE_TYPE = generated_1.FeeDiscriminants.BasisPoints;
exports.DEFAULT_LAUNCHPOOL_WITHDRAW_FEE = 200n;
exports.DEFAULT_LAUNCHPOOL_CLAIM_FEE_TYPE = generated_1.FeeDiscriminants.Flat;
exports.DEFAULT_LAUNCHPOOL_CLAIM_FEE = 0n;
// Presale
exports.DEFAULT_PRESALE_FEE_TYPE_V1 = generated_1.FeeDiscriminants.Flat;
exports.DEFAULT_PRESALE_FEE_V1 = 0n;
exports.DEFAULT_PRESALE_FEE_TYPE_V2 = generated_1.FeeDiscriminants.Flat;
exports.DEFAULT_PRESALE_FEE_V2 = 0n;
exports.DEFAULT_PRESALE_DEPOSIT_FEE_TYPE = generated_1.FeeDiscriminants.BasisPoints;
exports.DEFAULT_PRESALE_DEPOSIT_FEE = 200n;
exports.DEFAULT_PRESALE_CLAIM_FEE_TYPE = generated_1.FeeDiscriminants.Flat;
exports.DEFAULT_PRESALE_CLAIM_FEE = 0n;
exports.DEFAULT_PRESALE_WITHDRAW_FEE_TYPE = generated_1.FeeDiscriminants.BasisPoints;
exports.DEFAULT_PRESALE_WITHDRAW_FEE = 200n;
// Vault
exports.DEFAULT_VAULT_FEE_TYPE = generated_1.FeeDiscriminants.Flat;
exports.DEFAULT_VAULT_FEE = 0n;
exports.DEFAULT_VAULT_DEPOSIT_FEE_TYPE = generated_1.FeeDiscriminants.BasisPoints;
exports.DEFAULT_VAULT_DEPOSIT_FEE = 200n;
exports.DEFAULT_VAULT_WITHDRAW_FEE_TYPE = generated_1.FeeDiscriminants.BasisPoints;
exports.DEFAULT_VAULT_WITHDRAW_FEE = 200n;
// Bonding Curve
exports.DEFAULT_BONDING_CURVE_FEE_TYPE = generated_1.FeeDiscriminants.Flat;
exports.DEFAULT_BONDING_CURVE_FEE = 0n;
exports.DEFAULT_BONDING_CURVE_DEPOSIT_FEE_TYPE = generated_1.FeeDiscriminants.BasisPoints;
exports.DEFAULT_BONDING_CURVE_DEPOSIT_FEE = 200n;
exports.DEFAULT_BONDING_CURVE_WITHDRAW_FEE_TYPE = generated_1.FeeDiscriminants.BasisPoints;
exports.DEFAULT_BONDING_CURVE_WITHDRAW_FEE = 200n;
// Auction
exports.DEFAULT_AUCTION_FEE_TYPE = generated_1.FeeDiscriminants.Flat;
exports.DEFAULT_AUCTION_FEE = 0n;
exports.DEFAULT_AUCTION_BID_FEE_TYPE = generated_1.FeeDiscriminants.Flat;
exports.DEFAULT_AUCTION_BID_FEE = 1000000n; // 0.001 SOL
// Unlocked
exports.DEFAULT_UNLOCKED_FEE_TYPE = generated_1.FeeDiscriminants.Flat;
exports.DEFAULT_UNLOCKED_FEE = 0n;
exports.DEFAULT_UNLOCKED_CLAIM_FEE_TYPE = generated_1.FeeDiscriminants.BasisPoints;
exports.DEFAULT_UNLOCKED_CLAIM_FEE = 500n;
// Default for the rest
exports.DEFAULT_FEE_TYPE = generated_1.FeeDiscriminants.Flat;
exports.DEFAULT_FEE = 0n;
function calculateFee(amount, kind, fee) {
    switch (kind) {
        case generated_1.FeeDiscriminants.Flat:
            return fee;
        case generated_1.FeeDiscriminants.BasisPoints:
            return calculateFeeBasisPoints(amount, fee);
        default:
            throw new Error(`Invalid fee kind: ${kind}`);
    }
}
exports.calculateFee = calculateFee;
function calculateFeeBasisPoints(amount, feeBps) {
    if (feeBps === 0n) {
        return 1n;
    }
    return (0, bigInt_1.maxBigInt)(1n, (amount * feeBps) / 10000n);
}
