// @metaplex-foundation/genesis/dist/src/merkleTree.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifyClaim = exports.verifyProof = exports.getHexProof = exports.getHexRoot = exports.getProof = exports.getRoot = exports.internalHash = exports.nodeHash = exports.createMerkleTree = void 0;
const sha3_1 = require("@noble/hashes/sha3");
const utils_1 = require("@noble/hashes/utils");
function createMerkleTree(leaves) {
    const tree = {
        leaves: leaves.slice(),
        layers: [],
    };
    let hashes = tree.leaves.map(nodeHash);
    while (hashes.length > 1) {
        tree.layers.push(hashes.slice());
        hashes = hashes.reduce((acc, cur, idx, arr) => {
            if (idx % 2 === 0) {
                const nxt = arr[idx + 1];
                acc.push(internalHash(cur, nxt));
            }
            return acc;
        }, Array());
    }
    // Push the final layer if there is a remaining hash.
    if (hashes.length === 1) {
        tree.layers.push(hashes.slice());
    }
    return tree;
}
exports.createMerkleTree = createMerkleTree;
function nodeHash(data) {
    return (0, sha3_1.keccak_256)(data);
}
exports.nodeHash = nodeHash;
function internalHash(first, second) {
    // If the sibling node doesn't exist (i.e. the tree level has an odd number
    // of nodes), we duplicate the current node. This mirrors the behavior on
    // chain where the verifier expects every level of the proof to have a
    // sibling hash, even if that sibling is identical to the current node.
    const sibling = second ?? first;
    const [fst, snd] = [first, sibling].sort(Buffer.compare);
    const prefixed = (0, utils_1.concatBytes)(new Uint8Array([0x01]), new Uint8Array(fst), new Uint8Array(snd));
    return (0, sha3_1.keccak_256)(prefixed);
}
exports.internalHash = internalHash;
function getRoot(tree) {
    return tree.layers[tree.layers.length - 1][0];
}
exports.getRoot = getRoot;
function getProof(tree, idx) {
    return tree.layers.slice(0, -1).reduce((proof, layer) => {
        // The sibling index can be computed by flipping the least-significant bit.
        // eslint-disable-next-line no-bitwise
        const sibling = idx ^ 1;
        if (sibling < layer.length) {
            // Normal case: the sibling exists in the current layer.
            proof.push(layer[sibling]);
        }
        else {
            // When the sibling index is out of bounds (i.e. the current node is the
            // lone node without a pair), we duplicate the current node. This keeps
            // the proof length consistent with the tree height and matches the
            // on-chain verification logic that duplicates solitary nodes when
            // hashing up the tree.
            proof.push(layer[idx]);
        }
        idx = Math.floor(idx / 2);
        return proof;
    }, []);
}
exports.getProof = getProof;
function getHexRoot(tree) {
    return Buffer.from(getRoot(tree)).toString('hex');
}
exports.getHexRoot = getHexRoot;
function getHexProof(tree, idx) {
    return getProof(tree, idx).map((el) => Buffer.from(el).toString('hex'));
}
exports.getHexProof = getHexProof;
function verifyProof(tree, idx, proof, root) {
    const pair = nodeHash(tree.leaves[idx]);
    return proof
        .reduce((acc, item) => internalHash(acc, item), pair)
        .every((el, i) => el === root[i]);
}
exports.verifyProof = verifyProof;
function verifyClaim(leaf, proof, root) {
    const pair = nodeHash(leaf);
    return proof
        .reduce((acc, item) => internalHash(acc, item), pair)
        .every((el, i) => el === root[i]);
}
exports.verifyClaim = verifyClaim;
