// @metaplex-foundation/genesis/dist/src/meteora.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findVaultTokenAccountPda = exports.findVaultPda = exports.findProtocolFeeTokenAccountPda = exports.findLpVaultTokenAccountPda = exports.findVaultLpMintPda = exports.findPoolLpMintPda = exports.findPoolPda = exports.METEORA_VAULT_PROGRAM_ID = exports.METEORA_CONFIG = exports.METEORA_PROGRAM_ID = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
/// Meteora Pools Program ID
exports.METEORA_PROGRAM_ID = (0, umi_1.publicKey)('Eo7WjKq67rjJQSZxS6z3YkapzY3eMj6Xy8X5EQVn5UaB');
/// Config for Meteora pools.
exports.METEORA_CONFIG = (0, umi_1.publicKey)('85t6rsA3Dm8kTW1rSVjVPNJs4DQtZv7uCMjpaRTH9hQA');
// Meteora Vault Program
exports.METEORA_VAULT_PROGRAM_ID = (0, umi_1.publicKey)('24Uqj9JCLxUeoC3hGfh5W3s9FM9uCHDS2SG3LYwBpyTi');
// Base Key
const BASE_KEY = (0, umi_1.publicKey)('HWzXGcGHy4tcpYfaRDCyLNzXqBTv3E6BttpCH2vJxArv');
function findPoolPda(umi, token_a_mint, token_b_mint, config) {
    return umi.eddsa.findPda(exports.METEORA_PROGRAM_ID, [
        (0, serializers_1.publicKey)().serialize(token_a_mint),
        (0, serializers_1.publicKey)().serialize(token_b_mint),
        (0, serializers_1.publicKey)().serialize(config),
    ]);
}
exports.findPoolPda = findPoolPda;
function findPoolLpMintPda(umi, pool) {
    return umi.eddsa.findPda(exports.METEORA_PROGRAM_ID, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('lp_mint'),
        (0, serializers_1.publicKey)().serialize(pool),
    ]);
}
exports.findPoolLpMintPda = findPoolLpMintPda;
function findVaultLpMintPda(umi, vault) {
    return umi.eddsa.findPda(exports.METEORA_VAULT_PROGRAM_ID, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('lp_mint'),
        (0, serializers_1.publicKey)().serialize(vault),
    ]);
}
exports.findVaultLpMintPda = findVaultLpMintPda;
function findLpVaultTokenAccountPda(umi, vault, pool) {
    return umi.eddsa.findPda(exports.METEORA_PROGRAM_ID, [
        (0, serializers_1.publicKey)().serialize(vault),
        (0, serializers_1.publicKey)().serialize(pool),
    ]);
}
exports.findLpVaultTokenAccountPda = findLpVaultTokenAccountPda;
function findProtocolFeeTokenAccountPda(umi, mint, pool) {
    return umi.eddsa.findPda(exports.METEORA_PROGRAM_ID, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('fee'),
        (0, serializers_1.publicKey)().serialize(mint),
        (0, serializers_1.publicKey)().serialize(pool),
    ]);
}
exports.findProtocolFeeTokenAccountPda = findProtocolFeeTokenAccountPda;
function findVaultPda(umi, token_mint) {
    return umi.eddsa.findPda(exports.METEORA_VAULT_PROGRAM_ID, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('vault'),
        (0, serializers_1.publicKey)().serialize(token_mint),
        (0, serializers_1.publicKey)().serialize(BASE_KEY),
    ]);
}
exports.findVaultPda = findVaultPda;
function findVaultTokenAccountPda(umi, vault) {
    return umi.eddsa.findPda(exports.METEORA_VAULT_PROGRAM_ID, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('token_vault'),
        (0, serializers_1.publicKey)().serialize(vault),
    ]);
}
exports.findVaultTokenAccountPda = findVaultTokenAccountPda;
