// @metaplex-foundation/genesis/dist/src/raydium.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deriveRaydiumPDAsV2 = exports.deriveRaydiumPDAs = exports.RAYDIUM_AMM_CONFIG_MAINNET = exports.RAYDIUM_AMM_CONFIG_DEVNET = exports.NATIVE_SOL_MINT = exports.findRaydiumPermissionPda = exports.findObservationStatePda = exports.findTokenVaultPda = exports.findLpMintPda = exports.findPoolStatePda = exports.findAuthorityPda = exports.CREATE_POOL_FEE_RECEIVER_MAINNET = exports.CREATE_POOL_FEE_RECEIVER_DEVNET = exports.RAYDIUM_CP_SWAP_PROGRAM_ID_MAINNET = exports.RAYDIUM_CP_SWAP_PROGRAM_ID_DEVNET = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const generated_1 = require("./generated");
// Raydium CP-Swap Program IDs
exports.RAYDIUM_CP_SWAP_PROGRAM_ID_DEVNET = (0, umi_1.publicKey)('DRaycpLY18LhpbydsBWbVJtxpNv9oXPgjRSfpF2bWpYb');
exports.RAYDIUM_CP_SWAP_PROGRAM_ID_MAINNET = (0, umi_1.publicKey)('CPMMoo8L3F4NbTegBCKVNunggL7H1ZpdTHKxQB5qKP1C');
// Raydium CP-Swap PDA seeds (from the Rust constants)
const AUTH_SEED = 'vault_and_lp_mint_auth_seed';
const POOL_SEED = 'pool';
const POOL_LP_MINT_SEED = 'pool_lp_mint';
const POOL_VAULT_SEED = 'pool_vault';
const OBSERVATION_SEED = 'observation';
const PERMISSION_SEED = 'permission';
// Create pool fee receivers (based on Raydium's feature flags)
exports.CREATE_POOL_FEE_RECEIVER_DEVNET = (0, umi_1.publicKey)('3oE58BKVt8KuYkGxx8zBojugnymWmBiyafWgMrnb6eYy');
exports.CREATE_POOL_FEE_RECEIVER_MAINNET = (0, umi_1.publicKey)('DNXgeM9EiiaAbaWvwjHj9fQQLAX5ZsfHyvmYUNRAdNC8');
function findAuthorityPda(umi, programId) {
    return umi.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize(AUTH_SEED),
    ]);
}
exports.findAuthorityPda = findAuthorityPda;
function findPoolStatePda(umi, programId, ammConfig, token0Mint, token1Mint) {
    return umi.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize(POOL_SEED),
        (0, serializers_1.publicKey)().serialize(ammConfig),
        (0, serializers_1.publicKey)().serialize(token0Mint),
        (0, serializers_1.publicKey)().serialize(token1Mint),
    ]);
}
exports.findPoolStatePda = findPoolStatePda;
function findLpMintPda(umi, programId, poolState) {
    return umi.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize(POOL_LP_MINT_SEED),
        (0, serializers_1.publicKey)().serialize(poolState),
    ]);
}
exports.findLpMintPda = findLpMintPda;
function findTokenVaultPda(umi, programId, poolState, tokenMint) {
    return umi.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize(POOL_VAULT_SEED),
        (0, serializers_1.publicKey)().serialize(poolState),
        (0, serializers_1.publicKey)().serialize(tokenMint),
    ]);
}
exports.findTokenVaultPda = findTokenVaultPda;
function findObservationStatePda(umi, programId, poolState) {
    return umi.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize(OBSERVATION_SEED),
        (0, serializers_1.publicKey)().serialize(poolState),
    ]);
}
exports.findObservationStatePda = findObservationStatePda;
/**
 * Find the Raydium Permission PDA
 * This is the permission account derived from the payer's pubkey in the Raydium program
 */
function findRaydiumPermissionPda(umi, raydiumProgramId, payer) {
    return umi.eddsa.findPda(raydiumProgramId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize(PERMISSION_SEED),
        (0, serializers_1.publicKey)().serialize(payer),
    ]);
}
exports.findRaydiumPermissionPda = findRaydiumPermissionPda;
// Native SOL mint address
exports.NATIVE_SOL_MINT = (0, umi_1.publicKey)('So11111111111111111111111111111111111111112');
// Raydium AMM configs - these are the only configs allowed by the Genesis protocol
exports.RAYDIUM_AMM_CONFIG_DEVNET = (0, umi_1.publicKey)('G7YfJJp1TX1VtzN4V2yhPNSU23AKPSy1U2miRdwAByK5');
exports.RAYDIUM_AMM_CONFIG_MAINNET = (0, umi_1.publicKey)('Bq9YmSkZ6f13L4rPsx9MxSXuyPfKCBbvrbvarKti1LdG');
/**
 * Compares two PublicKeys lexicographically (as bytes)
 * Returns true if a < b lexicographically
 */
function comparePublicKeysLexicographically(a, b) {
    const aBytes = (0, serializers_1.publicKey)().serialize(a);
    const bBytes = (0, serializers_1.publicKey)().serialize(b);
    for (let i = 0; i < Math.min(aBytes.length, bBytes.length); i += 1) {
        if (aBytes[i] < bBytes[i])
            return true;
        if (aBytes[i] > bBytes[i])
            return false;
    }
    return aBytes.length < bBytes.length;
}
/**
 * Derives all Raydium PDAs with correct token ordering for graduateToRaydiumV1
 * Handles the lexicographic ordering requirement (token_0 < token_1)
 *
 * @param umi - Umi instance
 * @param projectMint - The project's token mint
 * @param options - Optional configuration
 * @param options.solMint - The SOL/quote token mint (defaults to wSOL)
 * @param options.env - Environment ('mainnet' or 'devnet'), defaults to 'mainnet'
 */
function deriveRaydiumPDAs(umi, projectMint, options = {}) {
    const solMint = options.solMint ?? exports.NATIVE_SOL_MINT;
    const env = options.env ?? 'mainnet';
    // Select the appropriate program ID and AMM config based on environment
    const programId = env === 'mainnet'
        ? exports.RAYDIUM_CP_SWAP_PROGRAM_ID_MAINNET
        : exports.RAYDIUM_CP_SWAP_PROGRAM_ID_DEVNET;
    const ammConfig = env === 'mainnet' ? exports.RAYDIUM_AMM_CONFIG_MAINNET : exports.RAYDIUM_AMM_CONFIG_DEVNET;
    // Validate inputs
    if (projectMint === solMint) {
        throw new Error('Token mints cannot be the same');
    }
    // Determine token ordering based on lexicographic comparison of bytes
    const isProjectMintToken0 = comparePublicKeysLexicographically(projectMint, solMint);
    const token0Mint = isProjectMintToken0 ? projectMint : solMint;
    const token1Mint = isProjectMintToken0 ? solMint : projectMint;
    // Derive pool state with correct token ordering
    const poolState = findPoolStatePda(umi, programId, ammConfig, token0Mint, token1Mint);
    // Derive all dependent PDAs
    const poolAuthority = findAuthorityPda(umi, programId);
    const lpMint = findLpMintPda(umi, programId, poolState);
    const mintVault = findTokenVaultPda(umi, programId, poolState, projectMint);
    const solVault = findTokenVaultPda(umi, programId, poolState, solMint);
    const observationState = findObservationStatePda(umi, programId, poolState);
    // Select the appropriate create pool fee receiver based on environment
    const createPoolFee = env === 'mainnet'
        ? exports.CREATE_POOL_FEE_RECEIVER_MAINNET
        : exports.CREATE_POOL_FEE_RECEIVER_DEVNET;
    return {
        // Extract PublicKey from Pda tuples for convenience
        poolState: poolState[0],
        poolAuthority: poolAuthority[0],
        lpMint: lpMint[0],
        mintVault: mintVault[0],
        solVault: solVault[0],
        observationState: observationState[0],
        // Configuration accounts
        ammConfig,
        raydiumProgram: programId,
        createPoolFee,
        // Token ordering info
        token0Mint,
        token1Mint,
        isProjectMintToken0,
    };
}
exports.deriveRaydiumPDAs = deriveRaydiumPDAs;
/**
 * Derives Raydium pool PDAs for V2 with property names matching V2 instruction accounts.
 * This is a convenience wrapper around deriveRaydiumPDAs that maps vault names to V2 conventions.
 *
 * @param umi - Umi instance
 * @param baseMint - The base token mint (project mint)
 * @param options - Optional configuration
 * @param options.quoteMint - The quote token mint (defaults to wSOL)
 * @param options.env - Environment ('mainnet' or 'devnet'), defaults to 'mainnet'
 */
function deriveRaydiumPDAsV2(umi, baseMint, options = {}) {
    // Map V2 options to V1 options
    const pdas = deriveRaydiumPDAs(umi, baseMint, {
        solMint: options.quoteMint,
        env: options.env,
    });
    // Derive Genesis Raydium Signer PDA (universal for all pools)
    // Uses the Kinobi-generated function which gets the program ID from context
    const [raydiumSigner] = (0, generated_1.findRaydiumSignerPda)(umi);
    // Derive Permission PDA from Raydium Signer (created by Raydium team)
    const [permission] = findRaydiumPermissionPda(umi, pdas.raydiumProgram, raydiumSigner);
    return {
        poolState: pdas.poolState,
        ammConfig: pdas.ammConfig,
        poolAuthority: pdas.poolAuthority,
        lpMint: pdas.lpMint,
        baseVault: pdas.mintVault,
        quoteVault: pdas.solVault,
        createPoolFee: pdas.createPoolFee,
        observationState: pdas.observationState,
        raydiumProgram: pdas.raydiumProgram,
        // Universal Raydium Signer for InitializeWithPermission
        raydiumSigner,
        permission,
        // Include token ordering info for reference
        token0Mint: pdas.token0Mint,
        token1Mint: pdas.token1Mint,
        isProjectMintToken0: pdas.isProjectMintToken0,
    };
}
exports.deriveRaydiumPDAsV2 = deriveRaydiumPDAsV2;
