// @metaplex-foundation/genesis/dist/src/schedule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBpsFromScheduleV2 = exports.createLinearBpsScheduleV2 = exports.getBpsFromScheduleTimestamps = exports.getBpsFromSchedule = exports.createLinearBpsSchedule = exports.MIN_TIME_BPS = exports.MAX_TIME_BPS = exports.BPS_DIVISOR = void 0;
const umi_1 = require("@metaplex-foundation/umi");
exports.BPS_DIVISOR = 10000n;
exports.MAX_TIME_BPS = 10000n;
exports.MIN_TIME_BPS = 0n;
// Signed 64-bit integer bounds (i64)
const MAX_I64 = 9223372036854775807n;
const MIN_I64 = -9223372036854775808n;
function createLinearBpsSchedule(point1, point2, args) {
    // Convert all inputs to bigint for internal math.
    const p1Time = BigInt(point1.timeBps);
    const p2Time = BigInt(point2.timeBps);
    const p1Bps = BigInt(point1.bps);
    const p2Bps = BigInt(point2.bps);
    if (p1Time > exports.MAX_TIME_BPS || p2Time > exports.MAX_TIME_BPS) {
        throw new Error('Time BPS must be less than or equal to 10,000');
    }
    if (p1Time < exports.MIN_TIME_BPS || p2Time < exports.MIN_TIME_BPS) {
        throw new Error('Time BPS must be greater than or equal to 0');
    }
    // Check if the line is vertical (undefined slope)
    if (p1Time === p2Time) {
        throw new Error('Cannot calculate slope and y-intercept for a vertical line (undefined slope)');
    }
    // Calculate slope scaled by BPS_DIVISOR to avoid fractions:
    // slopeBps = ((y2 - y1) * BPS_DIVISOR) / (x2 - x1)
    const slopeBps = ((p2Bps - p1Bps) * exports.BPS_DIVISOR) / (p2Time - p1Time);
    // Validate slopeBps is within i64 range
    if (slopeBps < MIN_I64 || slopeBps > MAX_I64) {
        throw new Error('slopeBps is outside the supported i64 range');
    }
    // interceptBps = y1 - (slopeBps * x1) / BPS_DIVISOR
    const interceptBps = p1Bps - (slopeBps * p1Time) / exports.BPS_DIVISOR;
    // Validate interceptBps is within i64 range
    if (interceptBps < MIN_I64 || interceptBps > MAX_I64) {
        throw new Error('interceptBps is outside the supported i64 range');
    }
    return {
        interceptBps,
        slopeBps,
        maxBps: BigInt(args.maxBps),
        startTime: BigInt(args.startTime),
        endTime: BigInt(args.endTime),
    };
}
exports.createLinearBpsSchedule = createLinearBpsSchedule;
function getBpsFromSchedule(schedule, timeBps) {
    const { interceptBps, slopeBps, maxBps } = schedule;
    const timeBpsBigInt = BigInt(timeBps);
    const bpsBigInt = interceptBps + (slopeBps * timeBpsBigInt) / exports.BPS_DIVISOR;
    if (bpsBigInt > maxBps) {
        return maxBps;
    }
    if (bpsBigInt < 0n) {
        return 0n;
    }
    return bpsBigInt;
}
exports.getBpsFromSchedule = getBpsFromSchedule;
function getBpsFromScheduleTimestamps(schedule, { now, start, end, }) {
    const nowBigInt = BigInt(now);
    const startBigInt = BigInt(start);
    const endBigInt = BigInt(end);
    if (nowBigInt < schedule.startTime || nowBigInt > schedule.endTime) {
        return 0n;
    }
    if (nowBigInt < startBigInt) {
        throw new Error('Now must be greater than or equal to start');
    }
    if (nowBigInt > endBigInt) {
        throw new Error('Now must be less than or equal to end');
    }
    const duration = endBigInt - startBigInt;
    const timeBps = ((nowBigInt - startBigInt) * exports.BPS_DIVISOR) / duration;
    return getBpsFromSchedule(schedule, timeBps);
}
exports.getBpsFromScheduleTimestamps = getBpsFromScheduleTimestamps;
// ---------------------------------------------------------------------------
// V2 Schedule Helpers (condition-based)
// ---------------------------------------------------------------------------
/**
 * Creates a LinearBpsScheduleV2 from two schedule points and a condition + duration.
 *
 * @param point1 - First point on the schedule line (timeBps, bps)
 * @param point2 - Second point on the schedule line (timeBps, bps)
 * @param args - startCondition, duration, and maxBps
 */
function createLinearBpsScheduleV2(point1, point2, args) {
    // Convert all inputs to bigint for internal math.
    const p1Time = BigInt(point1.timeBps);
    const p2Time = BigInt(point2.timeBps);
    const p1Bps = BigInt(point1.bps);
    const p2Bps = BigInt(point2.bps);
    if (p1Time > exports.MAX_TIME_BPS || p2Time > exports.MAX_TIME_BPS) {
        throw new Error('Time BPS must be less than or equal to 10,000');
    }
    if (p1Time < exports.MIN_TIME_BPS || p2Time < exports.MIN_TIME_BPS) {
        throw new Error('Time BPS must be greater than or equal to 0');
    }
    // Check if the line is vertical (undefined slope)
    if (p1Time === p2Time) {
        throw new Error('Cannot calculate slope and y-intercept for a vertical line (undefined slope)');
    }
    // Calculate slope scaled by BPS_DIVISOR to avoid fractions:
    // slopeBps = ((y2 - y1) * BPS_DIVISOR) / (x2 - x1)
    const slopeBps = ((p2Bps - p1Bps) * exports.BPS_DIVISOR) / (p2Time - p1Time);
    // Validate slopeBps is within i64 range
    if (slopeBps < MIN_I64 || slopeBps > MAX_I64) {
        throw new Error('slopeBps is outside the supported i64 range');
    }
    // interceptBps = y1 - (slopeBps * x1) / BPS_DIVISOR
    const interceptBps = p1Bps - (slopeBps * p1Time) / exports.BPS_DIVISOR;
    // Validate interceptBps is within i64 range
    if (interceptBps < MIN_I64 || interceptBps > MAX_I64) {
        throw new Error('interceptBps is outside the supported i64 range');
    }
    const duration = BigInt(args.duration);
    if (duration <= 0n) {
        throw new Error('Duration must be greater than zero');
    }
    return {
        interceptBps,
        slopeBps,
        maxBps: BigInt(args.maxBps),
        startCondition: args.startCondition,
        duration,
        reserved: new Uint8Array(64),
    };
}
exports.createLinearBpsScheduleV2 = createLinearBpsScheduleV2;
/**
 * Evaluate a V2 schedule at a given timestamp.
 *
 * Reads `startCondition.triggeredTimestamp` as the start time.
 * Returns 0 if the condition hasn't triggered yet or if `now` is outside the schedule window.
 */
function getBpsFromScheduleV2(schedule, now) {
    const nowBigInt = BigInt(now);
    // Check if the condition has been triggered
    const { triggeredTimestamp } = schedule.startCondition;
    if (!(0, umi_1.isSome)(triggeredTimestamp)) {
        return 0n;
    }
    const start = BigInt(triggeredTimestamp.value);
    const { duration } = schedule;
    if (duration <= 0n) {
        return 0n;
    }
    const end = start + duration;
    // Outside schedule window
    if (nowBigInt < start || nowBigInt > end) {
        return 0n;
    }
    const timeBps = ((nowBigInt - start) * exports.BPS_DIVISOR) / duration;
    const { interceptBps, slopeBps, maxBps } = schedule;
    const bpsBigInt = interceptBps + (slopeBps * timeBps) / exports.BPS_DIVISOR;
    if (bpsBigInt > maxBps) {
        return maxBps;
    }
    if (bpsBigInt < 0n) {
        return 0n;
    }
    return bpsBigInt;
}
exports.getBpsFromScheduleV2 = getBpsFromScheduleV2;
