// @metaplex-foundation/genesis/dist/src/streamflow.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findEscrowTokenPda = exports.DEVNET_FEE_ORACLE_ADDRESS = exports.DEVNET_STREAMFLOW_PROGRAM_ID = exports.STREAMFLOW_PROGRAM_ID = exports.FEE_ORACLE_ADDRESS = exports.WITHDRAWOR_ADDRESS = exports.STREAMFLOW_TREASURY = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
/// Streamflow Treasury address, by default receives 0.25% of tokens deposited
exports.STREAMFLOW_TREASURY = (0, umi_1.publicKey)('5SEpbdjFK5FxwTvfsGMXVQTD2v4M2c5tyRTxhdsPkgDw');
/// Streamflow Withdrawor address, this account will process withdrawals
exports.WITHDRAWOR_ADDRESS = (0, umi_1.publicKey)('wdrwhnCv4pzW8beKsbPa4S2UDZrXenjg16KJdKSpb5u');
/// Address of Fee Oracle that stores information about fees for speficic partners
exports.FEE_ORACLE_ADDRESS = (0, umi_1.publicKey)('B743wFVk2pCYhV91cn287e1xY7f1vt4gdY48hhNiuQmT');
/// Streamflow Program address
exports.STREAMFLOW_PROGRAM_ID = (0, umi_1.publicKey)('strmRqUCoQUgGUan5YhzUZa6KqdzwX5L6FpUxfmKg5m');
exports.DEVNET_STREAMFLOW_PROGRAM_ID = (0, umi_1.publicKey)('HqDGZjaVRXJ9MGRQEw7qDc2rAr6iH1n1kAQdCZaCMfMZ');
/// Address of Fee Oracle on devnet
exports.DEVNET_FEE_ORACLE_ADDRESS = (0, umi_1.publicKey)('Aa2JJfFzUN3V54DXUHRBJowFw416xfZHpPk9DaNy3iYs');
function findEscrowTokenPda(umi, metadata) {
    return umi.eddsa.findPda(exports.STREAMFLOW_PROGRAM_ID, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('strm'),
        (0, serializers_1.publicKey)().serialize(metadata),
    ]);
}
exports.findEscrowTokenPda = findEscrowTokenPda;
