// @metaplex-foundation/genesis/dist/src/generated/types/auctionBucketState.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAuctionBucketStateSerializer = exports.AuctionBucketState = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var AuctionBucketState;
(function (AuctionBucketState) {
    AuctionBucketState[AuctionBucketState["Uninitialized"] = 0] = "Uninitialized";
    AuctionBucketState[AuctionBucketState["Initialized"] = 1] = "Initialized";
    AuctionBucketState[AuctionBucketState["Started"] = 2] = "Started";
    AuctionBucketState[AuctionBucketState["Ended"] = 3] = "Ended";
    AuctionBucketState[AuctionBucketState["Stored"] = 4] = "Stored";
})(AuctionBucketState = exports.AuctionBucketState || (exports.AuctionBucketState = {}));
function getAuctionBucketStateSerializer() {
    return (0, serializers_1.scalarEnum)(AuctionBucketState, {
        description: 'AuctionBucketState',
    });
}
exports.getAuctionBucketStateSerializer = getAuctionBucketStateSerializer;
