// @metaplex-foundation/genesis/dist/src/generated/types/behaviorTypes.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBehaviorTypesSerializer = exports.BehaviorTypes = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var BehaviorTypes;
(function (BehaviorTypes) {
    BehaviorTypes[BehaviorTypes["None"] = 0] = "None";
    BehaviorTypes[BehaviorTypes["SendQuoteTokenPercentage"] = 1] = "SendQuoteTokenPercentage";
    BehaviorTypes[BehaviorTypes["BaseTokenRollover"] = 2] = "BaseTokenRollover";
    BehaviorTypes[BehaviorTypes["SendStartPrice"] = 3] = "SendStartPrice";
    BehaviorTypes[BehaviorTypes["ReallocateBaseTokensOnFailure"] = 4] = "ReallocateBaseTokensOnFailure";
})(BehaviorTypes = exports.BehaviorTypes || (exports.BehaviorTypes = {}));
function getBehaviorTypesSerializer() {
    return (0, serializers_1.scalarEnum)(BehaviorTypes, {
        description: 'BehaviorTypes',
    });
}
exports.getBehaviorTypesSerializer = getBehaviorTypesSerializer;
