// @metaplex-foundation/genesis/dist/src/generated/types/bondingCurveBucketState.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBondingCurveBucketStateSerializer = exports.BondingCurveBucketState = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var BondingCurveBucketState;
(function (BondingCurveBucketState) {
    BondingCurveBucketState[BondingCurveBucketState["Uninitialized"] = 0] = "Uninitialized";
    BondingCurveBucketState[BondingCurveBucketState["Initialized"] = 1] = "Initialized";
    BondingCurveBucketState[BondingCurveBucketState["Started"] = 2] = "Started";
    BondingCurveBucketState[BondingCurveBucketState["Ended"] = 3] = "Ended";
    BondingCurveBucketState[BondingCurveBucketState["Stored"] = 4] = "Stored";
})(BondingCurveBucketState = exports.BondingCurveBucketState || (exports.BondingCurveBucketState = {}));
function getBondingCurveBucketStateSerializer() {
    return (0, serializers_1.scalarEnum)(BondingCurveBucketState, {
        description: 'BondingCurveBucketState',
    });
}
exports.getBondingCurveBucketStateSerializer = getBondingCurveBucketStateSerializer;
