// @metaplex-foundation/genesis/dist/src/generated/types/bondingCurveV2ExtensionType.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBondingCurveV2ExtensionTypeSerializer = exports.BondingCurveV2ExtensionType = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var BondingCurveV2ExtensionType;
(function (BondingCurveV2ExtensionType) {
    BondingCurveV2ExtensionType[BondingCurveV2ExtensionType["BackendSigner"] = 0] = "BackendSigner";
    BondingCurveV2ExtensionType[BondingCurveV2ExtensionType["ActionsAuthority"] = 1] = "ActionsAuthority";
    BondingCurveV2ExtensionType[BondingCurveV2ExtensionType["Paused"] = 2] = "Paused";
})(BondingCurveV2ExtensionType = exports.BondingCurveV2ExtensionType || (exports.BondingCurveV2ExtensionType = {}));
function getBondingCurveV2ExtensionTypeSerializer() {
    return (0, serializers_1.scalarEnum)(BondingCurveV2ExtensionType, {
        description: 'BondingCurveV2ExtensionType',
    });
}
exports.getBondingCurveV2ExtensionTypeSerializer = getBondingCurveV2ExtensionTypeSerializer;
