// @metaplex-foundation/genesis/dist/src/generated/types/bucketTimes.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBucketTimesSerializer = exports.BucketTimes = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var BucketTimes;
(function (BucketTimes) {
    BucketTimes[BucketTimes["DepositStart"] = 0] = "DepositStart";
    BucketTimes[BucketTimes["DepositEnd"] = 1] = "DepositEnd";
    BucketTimes[BucketTimes["ClaimStart"] = 2] = "ClaimStart";
    BucketTimes[BucketTimes["ClaimEnd"] = 3] = "ClaimEnd";
    BucketTimes[BucketTimes["SwapStart"] = 4] = "SwapStart";
    BucketTimes[BucketTimes["SwapEnd"] = 5] = "SwapEnd";
    BucketTimes[BucketTimes["LockStart"] = 6] = "LockStart";
    BucketTimes[BucketTimes["LockEnd"] = 7] = "LockEnd";
    BucketTimes[BucketTimes["GraduateStart"] = 8] = "GraduateStart";
    BucketTimes[BucketTimes["Graduate"] = 9] = "Graduate";
})(BucketTimes = exports.BucketTimes || (exports.BucketTimes = {}));
function getBucketTimesSerializer() {
    return (0, serializers_1.scalarEnum)(BucketTimes, {
        description: 'BucketTimes',
    });
}
exports.getBucketTimesSerializer = getBucketTimesSerializer;
