// @metaplex-foundation/genesis/dist/src/generated/types/condition.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.isCondition = exports.condition = exports.getConditionSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const _1 = require(".");
function getConditionSerializer() {
    return (0, serializers_1.dataEnum)([
        [
            'Uninitialized',
            (0, serializers_1.struct)([
                ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 55 })],
                [
                    'triggeredTimestamp',
                    (() => {
                        const inner = (0, serializers_1.i64)();
                        const sentinel = new Uint8Array([
                            255, 255, 255, 255, 255, 255, 255, 127,
                        ]);
                        if (inner.fixedSize == null) {
                            throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                        }
                        if (inner.fixedSize !== sentinel.length) {
                            throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                        }
                        const { fixedSize } = inner;
                        const normalize = (input) => {
                            if (input == null) {
                                return null;
                            }
                            if ((0, umi_1.isOption)(input)) {
                                return input.__option === 'None' ? null : input.value;
                            }
                            return input;
                        };
                        return {
                            description: 'Option<bigint>',
                            fixedSize,
                            maxSize: fixedSize,
                            serialize: (value) => {
                                const normalized = normalize(value);
                                return normalized == null
                                    ? sentinel.slice()
                                    : inner.serialize(normalized);
                            },
                            deserialize: (bytes, offset = 0) => {
                                const slice = bytes.slice(offset, offset + fixedSize);
                                const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                                if (isSentinel) {
                                    return [(0, umi_1.none)(), offset + fixedSize];
                                }
                                const [value, newOffset] = inner.deserialize(bytes, offset);
                                return [(0, umi_1.some)(value), newOffset];
                            },
                        };
                    })(),
                ],
            ]),
        ],
        [
            'TimeAbsolute',
            (0, serializers_1.struct)([
                ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 47 })],
                ['time', (0, serializers_1.i64)()],
                [
                    'triggeredTimestamp',
                    (() => {
                        const inner = (0, serializers_1.i64)();
                        const sentinel = new Uint8Array([
                            255, 255, 255, 255, 255, 255, 255, 127,
                        ]);
                        if (inner.fixedSize == null) {
                            throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                        }
                        if (inner.fixedSize !== sentinel.length) {
                            throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                        }
                        const { fixedSize } = inner;
                        const normalize = (input) => {
                            if (input == null) {
                                return null;
                            }
                            if ((0, umi_1.isOption)(input)) {
                                return input.__option === 'None' ? null : input.value;
                            }
                            return input;
                        };
                        return {
                            description: 'Option<bigint>',
                            fixedSize,
                            maxSize: fixedSize,
                            serialize: (value) => {
                                const normalized = normalize(value);
                                return normalized == null
                                    ? sentinel.slice()
                                    : inner.serialize(normalized);
                            },
                            deserialize: (bytes, offset = 0) => {
                                const slice = bytes.slice(offset, offset + fixedSize);
                                const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                                if (isSentinel) {
                                    return [(0, umi_1.none)(), offset + fixedSize];
                                }
                                const [value, newOffset] = inner.deserialize(bytes, offset);
                                return [(0, umi_1.some)(value), newOffset];
                            },
                        };
                    })(),
                ],
            ]),
        ],
        [
            'TimeRelative',
            (0, serializers_1.struct)([
                ['bucketTime', (0, _1.getBucketTimesSerializer)()],
                ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 14 })],
                ['timeOffset', (0, serializers_1.i64)()],
                ['bucket', (0, serializers_1.publicKey)()],
                [
                    'triggeredTimestamp',
                    (() => {
                        const inner = (0, serializers_1.i64)();
                        const sentinel = new Uint8Array([
                            255, 255, 255, 255, 255, 255, 255, 127,
                        ]);
                        if (inner.fixedSize == null) {
                            throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                        }
                        if (inner.fixedSize !== sentinel.length) {
                            throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                        }
                        const { fixedSize } = inner;
                        const normalize = (input) => {
                            if (input == null) {
                                return null;
                            }
                            if ((0, umi_1.isOption)(input)) {
                                return input.__option === 'None' ? null : input.value;
                            }
                            return input;
                        };
                        return {
                            description: 'Option<bigint>',
                            fixedSize,
                            maxSize: fixedSize,
                            serialize: (value) => {
                                const normalized = normalize(value);
                                return normalized == null
                                    ? sentinel.slice()
                                    : inner.serialize(normalized);
                            },
                            deserialize: (bytes, offset = 0) => {
                                const slice = bytes.slice(offset, offset + fixedSize);
                                const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                                if (isSentinel) {
                                    return [(0, umi_1.none)(), offset + fixedSize];
                                }
                                const [value, newOffset] = inner.deserialize(bytes, offset);
                                return [(0, umi_1.some)(value), newOffset];
                            },
                        };
                    })(),
                ],
            ]),
        ],
        [
            'Triggered',
            (0, serializers_1.struct)([
                ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 55 })],
                [
                    'triggeredTimestamp',
                    (() => {
                        const inner = (0, serializers_1.i64)();
                        const sentinel = new Uint8Array([
                            255, 255, 255, 255, 255, 255, 255, 127,
                        ]);
                        if (inner.fixedSize == null) {
                            throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                        }
                        if (inner.fixedSize !== sentinel.length) {
                            throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                        }
                        const { fixedSize } = inner;
                        const normalize = (input) => {
                            if (input == null) {
                                return null;
                            }
                            if ((0, umi_1.isOption)(input)) {
                                return input.__option === 'None' ? null : input.value;
                            }
                            return input;
                        };
                        return {
                            description: 'Option<bigint>',
                            fixedSize,
                            maxSize: fixedSize,
                            serialize: (value) => {
                                const normalized = normalize(value);
                                return normalized == null
                                    ? sentinel.slice()
                                    : inner.serialize(normalized);
                            },
                            deserialize: (bytes, offset = 0) => {
                                const slice = bytes.slice(offset, offset + fixedSize);
                                const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                                if (isSentinel) {
                                    return [(0, umi_1.none)(), offset + fixedSize];
                                }
                                const [value, newOffset] = inner.deserialize(bytes, offset);
                                return [(0, umi_1.some)(value), newOffset];
                            },
                        };
                    })(),
                ],
            ]),
        ],
        [
            'Never',
            (0, serializers_1.struct)([
                ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 55 })],
                [
                    'triggeredTimestamp',
                    (() => {
                        const inner = (0, serializers_1.i64)();
                        const sentinel = new Uint8Array([
                            255, 255, 255, 255, 255, 255, 255, 127,
                        ]);
                        if (inner.fixedSize == null) {
                            throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                        }
                        if (inner.fixedSize !== sentinel.length) {
                            throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                        }
                        const { fixedSize } = inner;
                        const normalize = (input) => {
                            if (input == null) {
                                return null;
                            }
                            if ((0, umi_1.isOption)(input)) {
                                return input.__option === 'None' ? null : input.value;
                            }
                            return input;
                        };
                        return {
                            description: 'Option<bigint>',
                            fixedSize,
                            maxSize: fixedSize,
                            serialize: (value) => {
                                const normalized = normalize(value);
                                return normalized == null
                                    ? sentinel.slice()
                                    : inner.serialize(normalized);
                            },
                            deserialize: (bytes, offset = 0) => {
                                const slice = bytes.slice(offset, offset + fixedSize);
                                const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                                if (isSentinel) {
                                    return [(0, umi_1.none)(), offset + fixedSize];
                                }
                                const [value, newOffset] = inner.deserialize(bytes, offset);
                                return [(0, umi_1.some)(value), newOffset];
                            },
                        };
                    })(),
                ],
            ]),
        ],
        [
            'Always',
            (0, serializers_1.struct)([
                ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 55 })],
                [
                    'triggeredTimestamp',
                    (() => {
                        const inner = (0, serializers_1.i64)();
                        const sentinel = new Uint8Array([
                            255, 255, 255, 255, 255, 255, 255, 127,
                        ]);
                        if (inner.fixedSize == null) {
                            throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                        }
                        if (inner.fixedSize !== sentinel.length) {
                            throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                        }
                        const { fixedSize } = inner;
                        const normalize = (input) => {
                            if (input == null) {
                                return null;
                            }
                            if ((0, umi_1.isOption)(input)) {
                                return input.__option === 'None' ? null : input.value;
                            }
                            return input;
                        };
                        return {
                            description: 'Option<bigint>',
                            fixedSize,
                            maxSize: fixedSize,
                            serialize: (value) => {
                                const normalized = normalize(value);
                                return normalized == null
                                    ? sentinel.slice()
                                    : inner.serialize(normalized);
                            },
                            deserialize: (bytes, offset = 0) => {
                                const slice = bytes.slice(offset, offset + fixedSize);
                                const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                                if (isSentinel) {
                                    return [(0, umi_1.none)(), offset + fixedSize];
                                }
                                const [value, newOffset] = inner.deserialize(bytes, offset);
                                return [(0, umi_1.some)(value), newOffset];
                            },
                        };
                    })(),
                ],
            ]),
        ],
    ], { description: 'Condition' });
}
exports.getConditionSerializer = getConditionSerializer;
function condition(kind, data) {
    return Array.isArray(data)
        ? { __kind: kind, fields: data }
        : { __kind: kind, ...(data ?? {}) };
}
exports.condition = condition;
function isCondition(kind, value) {
    return value.__kind === kind;
}
exports.isCondition = isCondition;
