// @metaplex-foundation/genesis/dist/src/generated/types/conditionTypes.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConditionTypesSerializer = exports.ConditionTypes = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var ConditionTypes;
(function (ConditionTypes) {
    ConditionTypes[ConditionTypes["Uninitialized"] = 0] = "Uninitialized";
    ConditionTypes[ConditionTypes["TimeAbsolute"] = 1] = "TimeAbsolute";
    ConditionTypes[ConditionTypes["TimeRelative"] = 2] = "TimeRelative";
    ConditionTypes[ConditionTypes["Triggered"] = 3] = "Triggered";
    ConditionTypes[ConditionTypes["Never"] = 4] = "Never";
    ConditionTypes[ConditionTypes["Always"] = 5] = "Always";
})(ConditionTypes = exports.ConditionTypes || (exports.ConditionTypes = {}));
function getConditionTypesSerializer() {
    return (0, serializers_1.scalarEnum)(ConditionTypes, {
        description: 'ConditionTypes',
    });
}
exports.getConditionTypesSerializer = getConditionTypesSerializer;
