// @metaplex-foundation/genesis/dist/src/generated/types/distributionBucketState.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDistributionBucketStateSerializer = exports.DistributionBucketState = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var DistributionBucketState;
(function (DistributionBucketState) {
    DistributionBucketState[DistributionBucketState["Uninitialized"] = 0] = "Uninitialized";
    DistributionBucketState[DistributionBucketState["Initialized"] = 1] = "Initialized";
    DistributionBucketState[DistributionBucketState["Deposited"] = 2] = "Deposited";
    DistributionBucketState[DistributionBucketState["Funded"] = 3] = "Funded";
})(DistributionBucketState = exports.DistributionBucketState || (exports.DistributionBucketState = {}));
function getDistributionBucketStateSerializer() {
    return (0, serializers_1.scalarEnum)(DistributionBucketState, {
        description: 'DistributionBucketState',
    });
}
exports.getDistributionBucketStateSerializer = getDistributionBucketStateSerializer;
