// @metaplex-foundation/genesis/dist/src/generated/types/feeOverrideType.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getFeeOverrideTypeSerializer = exports.FeeOverrideType = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var FeeOverrideType;
(function (FeeOverrideType) {
    FeeOverrideType[FeeOverrideType["Bucket"] = 0] = "Bucket";
    FeeOverrideType[FeeOverrideType["Deposit"] = 1] = "Deposit";
    FeeOverrideType[FeeOverrideType["Withdraw"] = 2] = "Withdraw";
    FeeOverrideType[FeeOverrideType["Bid"] = 3] = "Bid";
    FeeOverrideType[FeeOverrideType["Claim"] = 4] = "Claim";
})(FeeOverrideType = exports.FeeOverrideType || (exports.FeeOverrideType = {}));
function getFeeOverrideTypeSerializer() {
    return (0, serializers_1.scalarEnum)(FeeOverrideType, {
        description: 'FeeOverrideType',
    });
}
exports.getFeeOverrideTypeSerializer = getFeeOverrideTypeSerializer;
