// @metaplex-foundation/genesis/dist/src/generated/types/key.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getKeySerializer = exports.Key = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var Key;
(function (Key) {
    Key[Key["Uninitialized"] = 0] = "Uninitialized";
    Key[Key["GenesisAccount"] = 1] = "GenesisAccount";
    Key[Key["LaunchPoolBucket"] = 2] = "LaunchPoolBucket";
    Key[Key["LaunchPoolDeposit"] = 3] = "LaunchPoolDeposit";
    Key[Key["StreamflowBucket"] = 4] = "StreamflowBucket";
    Key[Key["UnlockedBucket"] = 5] = "UnlockedBucket";
    Key[Key["MeteoraBucket"] = 6] = "MeteoraBucket";
    Key[Key["PumpBucket"] = 7] = "PumpBucket";
    Key[Key["DistributionBucket"] = 8] = "DistributionBucket";
    Key[Key["PresaleBucket"] = 9] = "PresaleBucket";
    Key[Key["PresaleDeposit"] = 10] = "PresaleDeposit";
    Key[Key["VaultBucket"] = 11] = "VaultBucket";
    Key[Key["VaultDeposit"] = 12] = "VaultDeposit";
    Key[Key["BondingCurveBucket"] = 13] = "BondingCurveBucket";
    Key[Key["AuctionBucket"] = 14] = "AuctionBucket";
    Key[Key["AuctionBid"] = 15] = "AuctionBid";
    Key[Key["AuctionTree"] = 16] = "AuctionTree";
    Key[Key["RaydiumCpmmBucket"] = 17] = "RaydiumCpmmBucket";
    Key[Key["GenesisAccountV2"] = 18] = "GenesisAccountV2";
    Key[Key["PresaleBucketV2"] = 19] = "PresaleBucketV2";
    Key[Key["PresaleDepositV2"] = 20] = "PresaleDepositV2";
    Key[Key["UnlockedBucketV2"] = 21] = "UnlockedBucketV2";
    Key[Key["RaydiumCpmmBucketV2"] = 22] = "RaydiumCpmmBucketV2";
    Key[Key["VaultBucketV2"] = 23] = "VaultBucketV2";
    Key[Key["VaultDepositV2"] = 24] = "VaultDepositV2";
    Key[Key["BondingCurveBucketV2"] = 25] = "BondingCurveBucketV2";
    Key[Key["LaunchPoolBucketV2"] = 26] = "LaunchPoolBucketV2";
    Key[Key["LaunchPoolDepositV2"] = 27] = "LaunchPoolDepositV2";
    Key[Key["StreamflowBucketV2"] = 28] = "StreamflowBucketV2";
})(Key = exports.Key || (exports.Key = {}));
function getKeySerializer() {
    return (0, serializers_1.scalarEnum)(Key, { description: 'Key' });
}
exports.getKeySerializer = getKeySerializer;
