// @metaplex-foundation/genesis/dist/src/generated/types/launchPoolV2ExtensionType.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLaunchPoolV2ExtensionTypeSerializer = exports.LaunchPoolV2ExtensionType = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var LaunchPoolV2ExtensionType;
(function (LaunchPoolV2ExtensionType) {
    LaunchPoolV2ExtensionType[LaunchPoolV2ExtensionType["BackendSigner"] = 0] = "BackendSigner";
    LaunchPoolV2ExtensionType[LaunchPoolV2ExtensionType["DepositPenalty"] = 1] = "DepositPenalty";
    LaunchPoolV2ExtensionType[LaunchPoolV2ExtensionType["WithdrawPenalty"] = 2] = "WithdrawPenalty";
    LaunchPoolV2ExtensionType[LaunchPoolV2ExtensionType["BonusSchedule"] = 3] = "BonusSchedule";
    LaunchPoolV2ExtensionType[LaunchPoolV2ExtensionType["DepositLimit"] = 4] = "DepositLimit";
    LaunchPoolV2ExtensionType[LaunchPoolV2ExtensionType["Allowlist"] = 5] = "Allowlist";
    LaunchPoolV2ExtensionType[LaunchPoolV2ExtensionType["ClaimSchedule"] = 6] = "ClaimSchedule";
    LaunchPoolV2ExtensionType[LaunchPoolV2ExtensionType["MinimumDepositAmount"] = 7] = "MinimumDepositAmount";
    LaunchPoolV2ExtensionType[LaunchPoolV2ExtensionType["MinimumQuoteTokenThreshold"] = 8] = "MinimumQuoteTokenThreshold";
})(LaunchPoolV2ExtensionType = exports.LaunchPoolV2ExtensionType || (exports.LaunchPoolV2ExtensionType = {}));
function getLaunchPoolV2ExtensionTypeSerializer() {
    return (0, serializers_1.scalarEnum)(LaunchPoolV2ExtensionType, {
        description: 'LaunchPoolV2ExtensionType',
    });
}
exports.getLaunchPoolV2ExtensionTypeSerializer = getLaunchPoolV2ExtensionTypeSerializer;
