// @metaplex-foundation/genesis/dist/src/generated/types/launchPoolBucketState.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLaunchPoolBucketStateSerializer = exports.LaunchPoolBucketState = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var LaunchPoolBucketState;
(function (LaunchPoolBucketState) {
    LaunchPoolBucketState[LaunchPoolBucketState["Uninitialized"] = 0] = "Uninitialized";
    LaunchPoolBucketState[LaunchPoolBucketState["Initialized"] = 1] = "Initialized";
    LaunchPoolBucketState[LaunchPoolBucketState["Started"] = 2] = "Started";
    LaunchPoolBucketState[LaunchPoolBucketState["Ended"] = 3] = "Ended";
    LaunchPoolBucketState[LaunchPoolBucketState["Stored"] = 4] = "Stored";
})(LaunchPoolBucketState = exports.LaunchPoolBucketState || (exports.LaunchPoolBucketState = {}));
function getLaunchPoolBucketStateSerializer() {
    return (0, serializers_1.scalarEnum)(LaunchPoolBucketState, {
        description: 'LaunchPoolBucketState',
    });
}
exports.getLaunchPoolBucketStateSerializer = getLaunchPoolBucketStateSerializer;
