// @metaplex-foundation/genesis/dist/src/generated/types/meteoraBucketState.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMeteoraBucketStateSerializer = exports.MeteoraBucketState = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var MeteoraBucketState;
(function (MeteoraBucketState) {
    MeteoraBucketState[MeteoraBucketState["Uninitialized"] = 0] = "Uninitialized";
    MeteoraBucketState[MeteoraBucketState["Initialized"] = 1] = "Initialized";
    MeteoraBucketState[MeteoraBucketState["Funded"] = 2] = "Funded";
})(MeteoraBucketState = exports.MeteoraBucketState || (exports.MeteoraBucketState = {}));
function getMeteoraBucketStateSerializer() {
    return (0, serializers_1.scalarEnum)(MeteoraBucketState, {
        description: 'MeteoraBucketState',
    });
}
exports.getMeteoraBucketStateSerializer = getMeteoraBucketStateSerializer;
