// @metaplex-foundation/genesis/dist/src/generated/types/presaleV2ExtensionType.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPresaleV2ExtensionTypeSerializer = exports.PresaleV2ExtensionType = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var PresaleV2ExtensionType;
(function (PresaleV2ExtensionType) {
    PresaleV2ExtensionType[PresaleV2ExtensionType["BackendSigner"] = 0] = "BackendSigner";
    PresaleV2ExtensionType[PresaleV2ExtensionType["DepositLimit"] = 1] = "DepositLimit";
    PresaleV2ExtensionType[PresaleV2ExtensionType["DepositCooldown"] = 2] = "DepositCooldown";
    PresaleV2ExtensionType[PresaleV2ExtensionType["PerCooldownDepositLimit"] = 3] = "PerCooldownDepositLimit";
    PresaleV2ExtensionType[PresaleV2ExtensionType["Allowlist"] = 4] = "Allowlist";
    PresaleV2ExtensionType[PresaleV2ExtensionType["SteppedDepositLimit"] = 5] = "SteppedDepositLimit";
    PresaleV2ExtensionType[PresaleV2ExtensionType["ClaimSchedule"] = 6] = "ClaimSchedule";
    PresaleV2ExtensionType[PresaleV2ExtensionType["MinimumDepositAmount"] = 7] = "MinimumDepositAmount";
})(PresaleV2ExtensionType = exports.PresaleV2ExtensionType || (exports.PresaleV2ExtensionType = {}));
function getPresaleV2ExtensionTypeSerializer() {
    return (0, serializers_1.scalarEnum)(PresaleV2ExtensionType, {
        description: 'PresaleV2ExtensionType',
    });
}
exports.getPresaleV2ExtensionTypeSerializer = getPresaleV2ExtensionTypeSerializer;
