// @metaplex-foundation/genesis/dist/src/generated/types/presaleBucketState.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPresaleBucketStateSerializer = exports.PresaleBucketState = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var PresaleBucketState;
(function (PresaleBucketState) {
    PresaleBucketState[PresaleBucketState["Uninitialized"] = 0] = "Uninitialized";
    PresaleBucketState[PresaleBucketState["Initialized"] = 1] = "Initialized";
    PresaleBucketState[PresaleBucketState["Started"] = 2] = "Started";
    PresaleBucketState[PresaleBucketState["Ended"] = 3] = "Ended";
    PresaleBucketState[PresaleBucketState["Stored"] = 4] = "Stored";
})(PresaleBucketState = exports.PresaleBucketState || (exports.PresaleBucketState = {}));
function getPresaleBucketStateSerializer() {
    return (0, serializers_1.scalarEnum)(PresaleBucketState, {
        description: 'PresaleBucketState',
    });
}
exports.getPresaleBucketStateSerializer = getPresaleBucketStateSerializer;
