// @metaplex-foundation/genesis/dist/src/generated/types/pumpBucketState.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPumpBucketStateSerializer = exports.PumpBucketState = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var PumpBucketState;
(function (PumpBucketState) {
    PumpBucketState[PumpBucketState["Uninitialized"] = 0] = "Uninitialized";
    PumpBucketState[PumpBucketState["Initialized"] = 1] = "Initialized";
    PumpBucketState[PumpBucketState["Funded"] = 2] = "Funded";
})(PumpBucketState = exports.PumpBucketState || (exports.PumpBucketState = {}));
function getPumpBucketStateSerializer() {
    return (0, serializers_1.scalarEnum)(PumpBucketState, {
        description: 'PumpBucketState',
    });
}
exports.getPumpBucketStateSerializer = getPumpBucketStateSerializer;
