// @metaplex-foundation/genesis/dist/src/generated/types/raydiumCpmmBucketState.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRaydiumCpmmBucketStateSerializer = exports.RaydiumCpmmBucketState = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var RaydiumCpmmBucketState;
(function (RaydiumCpmmBucketState) {
    RaydiumCpmmBucketState[RaydiumCpmmBucketState["Uninitialized"] = 0] = "Uninitialized";
    RaydiumCpmmBucketState[RaydiumCpmmBucketState["Initialized"] = 1] = "Initialized";
    RaydiumCpmmBucketState[RaydiumCpmmBucketState["Funded"] = 2] = "Funded";
    RaydiumCpmmBucketState[RaydiumCpmmBucketState["Graduated"] = 3] = "Graduated";
})(RaydiumCpmmBucketState = exports.RaydiumCpmmBucketState || (exports.RaydiumCpmmBucketState = {}));
function getRaydiumCpmmBucketStateSerializer() {
    return (0, serializers_1.scalarEnum)(RaydiumCpmmBucketState, {
        description: 'RaydiumCpmmBucketState',
    });
}
exports.getRaydiumCpmmBucketStateSerializer = getRaydiumCpmmBucketStateSerializer;
