// @metaplex-foundation/genesis/dist/src/generated/types/streamflowBucketState.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStreamflowBucketStateSerializer = exports.StreamflowBucketState = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var StreamflowBucketState;
(function (StreamflowBucketState) {
    StreamflowBucketState[StreamflowBucketState["Uninitialized"] = 0] = "Uninitialized";
    StreamflowBucketState[StreamflowBucketState["Initialized"] = 1] = "Initialized";
    StreamflowBucketState[StreamflowBucketState["Locked"] = 2] = "Locked";
    StreamflowBucketState[StreamflowBucketState["Funded"] = 3] = "Funded";
})(StreamflowBucketState = exports.StreamflowBucketState || (exports.StreamflowBucketState = {}));
function getStreamflowBucketStateSerializer() {
    return (0, serializers_1.scalarEnum)(StreamflowBucketState, {
        description: 'StreamflowBucketState',
    });
}
exports.getStreamflowBucketStateSerializer = getStreamflowBucketStateSerializer;
