// @metaplex-foundation/genesis/dist/src/generated/types/unlockedBucketState.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getUnlockedBucketStateSerializer = exports.UnlockedBucketState = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var UnlockedBucketState;
(function (UnlockedBucketState) {
    UnlockedBucketState[UnlockedBucketState["Uninitialized"] = 0] = "Uninitialized";
    UnlockedBucketState[UnlockedBucketState["Initialized"] = 1] = "Initialized";
    UnlockedBucketState[UnlockedBucketState["Claimed"] = 2] = "Claimed";
    UnlockedBucketState[UnlockedBucketState["Funded"] = 3] = "Funded";
})(UnlockedBucketState = exports.UnlockedBucketState || (exports.UnlockedBucketState = {}));
function getUnlockedBucketStateSerializer() {
    return (0, serializers_1.scalarEnum)(UnlockedBucketState, {
        description: 'UnlockedBucketState',
    });
}
exports.getUnlockedBucketStateSerializer = getUnlockedBucketStateSerializer;
