// @metaplex-foundation/genesis/dist/src/generated/types/vaultBucketState.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVaultBucketStateSerializer = exports.VaultBucketState = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var VaultBucketState;
(function (VaultBucketState) {
    VaultBucketState[VaultBucketState["Uninitialized"] = 0] = "Uninitialized";
    VaultBucketState[VaultBucketState["Initialized"] = 1] = "Initialized";
    VaultBucketState[VaultBucketState["Started"] = 2] = "Started";
    VaultBucketState[VaultBucketState["Ended"] = 3] = "Ended";
    VaultBucketState[VaultBucketState["Stored"] = 4] = "Stored";
})(VaultBucketState = exports.VaultBucketState || (exports.VaultBucketState = {}));
function getVaultBucketStateSerializer() {
    return (0, serializers_1.scalarEnum)(VaultBucketState, {
        description: 'VaultBucketState',
    });
}
exports.getVaultBucketStateSerializer = getVaultBucketStateSerializer;
