// @metaplex-foundation/genesis/dist/src/generated/types/vaultV2ExtensionType.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVaultV2ExtensionTypeSerializer = exports.VaultV2ExtensionType = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var VaultV2ExtensionType;
(function (VaultV2ExtensionType) {
    VaultV2ExtensionType[VaultV2ExtensionType["BackendSigner"] = 0] = "BackendSigner";
    VaultV2ExtensionType[VaultV2ExtensionType["ActionsAuthority"] = 1] = "ActionsAuthority";
    VaultV2ExtensionType[VaultV2ExtensionType["AllowWithdrawals"] = 2] = "AllowWithdrawals";
})(VaultV2ExtensionType = exports.VaultV2ExtensionType || (exports.VaultV2ExtensionType = {}));
function getVaultV2ExtensionTypeSerializer() {
    return (0, serializers_1.scalarEnum)(VaultV2ExtensionType, {
        description: 'VaultV2ExtensionType',
    });
}
exports.getVaultV2ExtensionTypeSerializer = getVaultV2ExtensionTypeSerializer;
