// @metaplex-foundation/genesis/dist/src/generated/errors/genesis.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.DepositAlreadyClaimedError = exports.InvalidBucketIndexError = exports.InvalidGenesisAccountError = exports.InvalidBucketStateError = exports.InvalidDistributionAddressError = exports.GenesisAccountNotFinalizedError = exports.GenesisAccountFinalizedError = exports.LockNotStartedError = exports.InvalidTimelockProgramError = exports.InvalidRentError = exports.RecipientMismatchError = exports.InvalidTokenAccountOwnerError = exports.InvalidMintForTokenAccountError = exports.InvalidBucketForDepositPdaError = exports.InvalidGenesisAccountForBucketError = exports.InvalidMintForGenesisAccountError = exports.InvalidClaimStartTimeError = exports.InvalidEndTimeError = exports.InvalidAssociatedTokenProgramError = exports.MissingTokenAccountError = exports.ExpectedOutflowBucketError = exports.ExpectedInflowBucketError = exports.InvalidBucketError = exports.InsufficientFundsError = exports.LaunchPoolNotEndedError = exports.InvalidUnlockTimeError = exports.InvalidTokenMetadataProgramError = exports.ClaimNotStartedError = exports.LaunchPoolEndedError = exports.LaunchPoolNotStartedError = exports.InvalidStartTimeError = exports.InvalidFeeWalletError = exports.InvalidDepositPdaDerivationError = exports.InstructionNotImplementedError = exports.InsufficientAmountRemainingError = exports.InvalidAuthorityError = exports.InvalidBucketAccountDerivationError = exports.InvalidGenesisAccountDerivationError = exports.NameTooLongError = exports.MissingAtaProgramError = exports.MissingSystemProgramError = exports.InvalidAtaProgramError = exports.InvalidSplTokenProgramError = exports.DeserializationInstructionErrorError = exports.InvalidInstructionError = exports.InvalidAccountError = exports.NumericalOverflowError = exports.SerializationErrorError = exports.DeserializationErrorError = exports.InvalidSystemProgramError = void 0;
exports.InvalidAuctionTreeError = exports.BucketPausedError = exports.VaultDepositConsumedError = exports.VaultWithdrawNotAllowedError = exports.InvalidActionsAuthorityError = exports.InvalidOrdinalError = exports.BondingCurveInsufficientFundsError = exports.BondingCurveConstantProductDecreasedError = exports.BondingCurveEndedError = exports.BondingCurveNotStartedError = exports.InvalidConstantProductError = exports.BondingCurveNotEndedError = exports.InsufficientOutputAmountError = exports.InvalidAmountError = exports.VaultEndedError = exports.VaultNotStartedError = exports.VaultNotEndedError = exports.InvalidScheduleTimingError = exports.InsufficientFundsForPenaltyFeeError = exports.InvalidLinearScheduleError = exports.InvalidPenaltyWalletError = exports.FeeTransferFailedError = exports.InvalidFeeOverrideError = exports.InvalidFeeOverrideTypeError = exports.InvalidFeeAuthorityError = exports.InvalidMerkleProofError = exports.PresalePerCooldownDepositLimitExceededError = exports.PresaleDepositCooldownActiveError = exports.PresaleDepositLimitExceededError = exports.InvalidPresaleDepositAmountError = exports.InvalidBackendSignerError = exports.PresaleBucketWithdrawNotAllowedError = exports.PresaleBucketNotEndedError = exports.PresaleBucketAlreadyFundedError = exports.PresaleSaleEndedError = exports.PresaleSaleNotStartedError = exports.DistributionEndedError = exports.InvalidMintForDistributionError = exports.InvalidSolAllocationBpsError = exports.GenesisAccountAlreadyFinalizedError = exports.DistributionNotStartedError = exports.InvalidAtaDerivationError = exports.InflowsMustBeCompletedError = exports.BucketDoesNotSupportFundingError = exports.UnlockedBucketNotInitializedError = exports.StreamflowNotInitializedError = exports.BucketIsNotYetFundedError = exports.InvalidProgramOwnerForBucketError = exports.UnlockedBucketAlreadyClaimedError = exports.InvalidBucketAddressError = void 0;
exports.InvalidPrimaryBucketError = exports.InvalidPrimaryBucketOwnerError = exports.DestinationBucketNotFoundError = exports.PresaleClaimEndedError = exports.MissingBackendSignerError = exports.InvalidDepositPdaAccountError = exports.InvalidQuoteTokenAccountOwnerError = exports.InvalidQuoteTokenAccountMintError = exports.InvalidGenesisAccountIndexError = exports.InvalidQuoteMintForGenesisAccountError = exports.InvalidQuoteMintError = exports.InvalidNativeSolMintError = exports.InvalidRaydiumProgramError = exports.InvalidRaydiumBucketSignerError = exports.RaydiumBucketNotFundedError = exports.MissingRaydiumBucketSignerError = exports.RaydiumBucketNotInitializedError = exports.RefundNotClaimedError = exports.TokensNotClaimedError = exports.TokensAlreadyClaimedError = exports.InvalidAuctionCapacityError = exports.LosingBidderCannotRebidError = exports.BidNotWinningError = exports.RefundAlreadyClaimedError = exports.CannotCloseWinningBidError = exports.BidQuantityCannotDecreaseError = exports.BidPriceCannotDecreaseError = exports.BidNotCompetitiveError = exports.BidQuantityExceedsTotalAllocationError = exports.BidQuantityExceedsMaximumError = exports.BidQuantityBelowMinimumError = exports.BidPriceBelowMinimumError = exports.BidQuantityCannotBeZeroError = exports.BidPriceCannotBeZeroError = exports.UnauthorizedBidderError = exports.InvalidBidIndexError = exports.InvalidBucketForAuctionBidError = exports.InvalidAuctionBidPdaDerivationError = exports.InvalidAuctionBidError = exports.AuctionNotEndedError = exports.AuctionEndedError = exports.AuctionNotStartedError = exports.InvalidAuctionExtensionConfigError = exports.MaximumBidQuantityExceedsAllocationError = exports.InvalidMaximumBidQuantityError = exports.InvalidMinimumPriceError = exports.InvalidTokenQuantityBpsError = exports.InvalidBidIncrementError = exports.InvalidTokenAllocationError = exports.InvalidAuctionTreeStateError = void 0;
exports.InsufficientUnlockedAmountError = exports.InvalidRemainingAccountsLengthError = exports.InvalidConditionTypeError = exports.InvalidAuthorityDestinationTokenAccountError = exports.MissingAuthorityDestinationTokenAccountError = exports.SupplyBelowAllocatedError = exports.InvalidMintAndBurnError = exports.InvalidMintAuthorityError = exports.InvalidSupplyDeltaError = exports.InvalidStreamflowMetadataAccountError = exports.GraduationNotStartedError = exports.BucketAlreadyGraduatedError = exports.TotalSupplyNotFullyAllocatedError = exports.InvalidStreamflowConfigError = exports.LockEndedError = exports.StreamflowBucketAlreadyLockedError = exports.MissingFinalizeBucketAccountsError = exports.LaunchPoolMissingGraduationBehaviorError = exports.InvalidRaydiumPermissionPdaError = exports.InvalidRaydiumSignerError = exports.InvalidAllocationQuoteTokenCapError = exports.InvalidBaseTokenAllocationError = exports.DepositAlreadyRefundedError = exports.MissingPenaltyQuoteTokenAccountError = exports.BehaviorNotSupportedForBucketTypeError = exports.InvalidMinimumQuoteTokenThresholdError = exports.LaunchPoolThresholdMetError = exports.LaunchPoolFundingThresholdNotMetError = exports.LaunchPoolClaimEndedError = exports.InvalidMetadataAccountError = exports.InvalidTokenAccountProgramOwnerError = exports.BehaviorPercentageOverflowError = exports.PresaleDepositBelowMinimumError = exports.InvalidMinimumDepositAmountError = exports.SourceBucketNotFoundError = exports.PresaleNotFullyClaimedError = exports.BucketDoesNotSupportUnsoldWithdrawalError = exports.InvalidUnsoldWithdrawalAmountError = exports.NoUnsoldTokensError = exports.SaleNotEndedError = exports.ClaimScheduleDurationTooShortError = exports.InvalidClaimScheduleCliffAmountError = exports.InvalidClaimScheduleDurationError = exports.InvalidClaimSchedulePeriodError = exports.InvalidPresaleDepositError = exports.InvalidAuthoritySourceTokenAccountError = exports.MissingAuthoritySourceTokenAccountError = exports.UnlockedBucketClaimEndedError = exports.InvalidFundingModeError = exports.InvalidSecondaryBucketError = void 0;
exports.getGenesisErrorFromName = exports.getGenesisErrorFromCode = exports.InvalidLaunchTypeError = exports.InvalidRaydiumPoolStateError = exports.InstructionDeprecatedError = exports.InvalidConditionTypeForSlotError = exports.UninitializedConditionError = exports.InvalidRaydiumAmmConfigError = exports.InvalidLpClaimAuthorityError = exports.NoLpTokensToClaimError = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const codeToErrorMap = new Map();
const nameToErrorMap = new Map();
/** InvalidSystemProgram: Invalid System Program */
class InvalidSystemProgramError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid System Program', program, cause);
        this.name = 'InvalidSystemProgram';
        this.code = 0x0; // 0
    }
}
exports.InvalidSystemProgramError = InvalidSystemProgramError;
codeToErrorMap.set(0x0, InvalidSystemProgramError);
nameToErrorMap.set('InvalidSystemProgram', InvalidSystemProgramError);
/** DeserializationError: Error deserializing account */
class DeserializationErrorError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Error deserializing account', program, cause);
        this.name = 'DeserializationError';
        this.code = 0x1; // 1
    }
}
exports.DeserializationErrorError = DeserializationErrorError;
codeToErrorMap.set(0x1, DeserializationErrorError);
nameToErrorMap.set('DeserializationError', DeserializationErrorError);
/** SerializationError: Error serializing account */
class SerializationErrorError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Error serializing account', program, cause);
        this.name = 'SerializationError';
        this.code = 0x2; // 2
    }
}
exports.SerializationErrorError = SerializationErrorError;
codeToErrorMap.set(0x2, SerializationErrorError);
nameToErrorMap.set('SerializationError', SerializationErrorError);
/** NumericalOverflow: Numerical overflow */
class NumericalOverflowError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Numerical overflow', program, cause);
        this.name = 'NumericalOverflow';
        this.code = 0x3; // 3
    }
}
exports.NumericalOverflowError = NumericalOverflowError;
codeToErrorMap.set(0x3, NumericalOverflowError);
nameToErrorMap.set('NumericalOverflow', NumericalOverflowError);
/** InvalidAccount: Invalid account */
class InvalidAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid account', program, cause);
        this.name = 'InvalidAccount';
        this.code = 0x4; // 4
    }
}
exports.InvalidAccountError = InvalidAccountError;
codeToErrorMap.set(0x4, InvalidAccountError);
nameToErrorMap.set('InvalidAccount', InvalidAccountError);
/** InvalidInstruction: Invalid instruction */
class InvalidInstructionError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid instruction', program, cause);
        this.name = 'InvalidInstruction';
        this.code = 0x5; // 5
    }
}
exports.InvalidInstructionError = InvalidInstructionError;
codeToErrorMap.set(0x5, InvalidInstructionError);
nameToErrorMap.set('InvalidInstruction', InvalidInstructionError);
/** DeserializationInstructionError: Error deserializing instruction */
class DeserializationInstructionErrorError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Error deserializing instruction', program, cause);
        this.name = 'DeserializationInstructionError';
        this.code = 0x6; // 6
    }
}
exports.DeserializationInstructionErrorError = DeserializationInstructionErrorError;
codeToErrorMap.set(0x6, DeserializationInstructionErrorError);
nameToErrorMap.set('DeserializationInstructionError', DeserializationInstructionErrorError);
/** InvalidSplTokenProgram: Invalid SPL Token Program */
class InvalidSplTokenProgramError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid SPL Token Program', program, cause);
        this.name = 'InvalidSplTokenProgram';
        this.code = 0x7; // 7
    }
}
exports.InvalidSplTokenProgramError = InvalidSplTokenProgramError;
codeToErrorMap.set(0x7, InvalidSplTokenProgramError);
nameToErrorMap.set('InvalidSplTokenProgram', InvalidSplTokenProgramError);
/** InvalidAtaProgram: Invalid Associated Token Account Program */
class InvalidAtaProgramError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Associated Token Account Program', program, cause);
        this.name = 'InvalidAtaProgram';
        this.code = 0x8; // 8
    }
}
exports.InvalidAtaProgramError = InvalidAtaProgramError;
codeToErrorMap.set(0x8, InvalidAtaProgramError);
nameToErrorMap.set('InvalidAtaProgram', InvalidAtaProgramError);
/** MissingSystemProgram: Missing System Program */
class MissingSystemProgramError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Missing System Program', program, cause);
        this.name = 'MissingSystemProgram';
        this.code = 0x9; // 9
    }
}
exports.MissingSystemProgramError = MissingSystemProgramError;
codeToErrorMap.set(0x9, MissingSystemProgramError);
nameToErrorMap.set('MissingSystemProgram', MissingSystemProgramError);
/** MissingAtaProgram: Missing Associated Token Account Program */
class MissingAtaProgramError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Missing Associated Token Account Program', program, cause);
        this.name = 'MissingAtaProgram';
        this.code = 0xa; // 10
    }
}
exports.MissingAtaProgramError = MissingAtaProgramError;
codeToErrorMap.set(0xa, MissingAtaProgramError);
nameToErrorMap.set('MissingAtaProgram', MissingAtaProgramError);
/** NameTooLong: Name must be less than 32 bytes */
class NameTooLongError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Name must be less than 32 bytes', program, cause);
        this.name = 'NameTooLong';
        this.code = 0xb; // 11
    }
}
exports.NameTooLongError = NameTooLongError;
codeToErrorMap.set(0xb, NameTooLongError);
nameToErrorMap.set('NameTooLong', NameTooLongError);
/** InvalidGenesisAccountDerivation: Invalid derivation for Genesis Account */
class InvalidGenesisAccountDerivationError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid derivation for Genesis Account', program, cause);
        this.name = 'InvalidGenesisAccountDerivation';
        this.code = 0xc; // 12
    }
}
exports.InvalidGenesisAccountDerivationError = InvalidGenesisAccountDerivationError;
codeToErrorMap.set(0xc, InvalidGenesisAccountDerivationError);
nameToErrorMap.set('InvalidGenesisAccountDerivation', InvalidGenesisAccountDerivationError);
/** InvalidBucketAccountDerivation: Invalid derivation for Bucket Account */
class InvalidBucketAccountDerivationError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid derivation for Bucket Account', program, cause);
        this.name = 'InvalidBucketAccountDerivation';
        this.code = 0xd; // 13
    }
}
exports.InvalidBucketAccountDerivationError = InvalidBucketAccountDerivationError;
codeToErrorMap.set(0xd, InvalidBucketAccountDerivationError);
nameToErrorMap.set('InvalidBucketAccountDerivation', InvalidBucketAccountDerivationError);
/** InvalidAuthority: Invalid authority */
class InvalidAuthorityError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid authority', program, cause);
        this.name = 'InvalidAuthority';
        this.code = 0xe; // 14
    }
}
exports.InvalidAuthorityError = InvalidAuthorityError;
codeToErrorMap.set(0xe, InvalidAuthorityError);
nameToErrorMap.set('InvalidAuthority', InvalidAuthorityError);
/** InsufficientAmountRemaining: Insufficient amount remaining */
class InsufficientAmountRemainingError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Insufficient amount remaining', program, cause);
        this.name = 'InsufficientAmountRemaining';
        this.code = 0xf; // 15
    }
}
exports.InsufficientAmountRemainingError = InsufficientAmountRemainingError;
codeToErrorMap.set(0xf, InsufficientAmountRemainingError);
nameToErrorMap.set('InsufficientAmountRemaining', InsufficientAmountRemainingError);
/** InstructionNotImplemented: Instruction Not Implemented */
class InstructionNotImplementedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Instruction Not Implemented', program, cause);
        this.name = 'InstructionNotImplemented';
        this.code = 0x10; // 16
    }
}
exports.InstructionNotImplementedError = InstructionNotImplementedError;
codeToErrorMap.set(0x10, InstructionNotImplementedError);
nameToErrorMap.set('InstructionNotImplemented', InstructionNotImplementedError);
/** InvalidDepositPdaDerivation: Invalid Deposit PDA Derivation */
class InvalidDepositPdaDerivationError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Deposit PDA Derivation', program, cause);
        this.name = 'InvalidDepositPdaDerivation';
        this.code = 0x11; // 17
    }
}
exports.InvalidDepositPdaDerivationError = InvalidDepositPdaDerivationError;
codeToErrorMap.set(0x11, InvalidDepositPdaDerivationError);
nameToErrorMap.set('InvalidDepositPdaDerivation', InvalidDepositPdaDerivationError);
/** InvalidFeeWallet: Invalid Fee Wallet */
class InvalidFeeWalletError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Fee Wallet', program, cause);
        this.name = 'InvalidFeeWallet';
        this.code = 0x12; // 18
    }
}
exports.InvalidFeeWalletError = InvalidFeeWalletError;
codeToErrorMap.set(0x12, InvalidFeeWalletError);
nameToErrorMap.set('InvalidFeeWallet', InvalidFeeWalletError);
/** InvalidStartTime: Start Time cannot be in the past */
class InvalidStartTimeError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Start Time cannot be in the past', program, cause);
        this.name = 'InvalidStartTime';
        this.code = 0x13; // 19
    }
}
exports.InvalidStartTimeError = InvalidStartTimeError;
codeToErrorMap.set(0x13, InvalidStartTimeError);
nameToErrorMap.set('InvalidStartTime', InvalidStartTimeError);
/** LaunchPoolNotStarted: LaunchPool period has not started */
class LaunchPoolNotStartedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('LaunchPool period has not started', program, cause);
        this.name = 'LaunchPoolNotStarted';
        this.code = 0x14; // 20
    }
}
exports.LaunchPoolNotStartedError = LaunchPoolNotStartedError;
codeToErrorMap.set(0x14, LaunchPoolNotStartedError);
nameToErrorMap.set('LaunchPoolNotStarted', LaunchPoolNotStartedError);
/** LaunchPoolEnded: LaunchPool period has ended */
class LaunchPoolEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('LaunchPool period has ended', program, cause);
        this.name = 'LaunchPoolEnded';
        this.code = 0x15; // 21
    }
}
exports.LaunchPoolEndedError = LaunchPoolEndedError;
codeToErrorMap.set(0x15, LaunchPoolEndedError);
nameToErrorMap.set('LaunchPoolEnded', LaunchPoolEndedError);
/** ClaimNotStarted: Claim period has not started */
class ClaimNotStartedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Claim period has not started', program, cause);
        this.name = 'ClaimNotStarted';
        this.code = 0x16; // 22
    }
}
exports.ClaimNotStartedError = ClaimNotStartedError;
codeToErrorMap.set(0x16, ClaimNotStartedError);
nameToErrorMap.set('ClaimNotStarted', ClaimNotStartedError);
/** InvalidTokenMetadataProgram: Invalid Token Metadata Program */
class InvalidTokenMetadataProgramError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Token Metadata Program', program, cause);
        this.name = 'InvalidTokenMetadataProgram';
        this.code = 0x17; // 23
    }
}
exports.InvalidTokenMetadataProgramError = InvalidTokenMetadataProgramError;
codeToErrorMap.set(0x17, InvalidTokenMetadataProgramError);
nameToErrorMap.set('InvalidTokenMetadataProgram', InvalidTokenMetadataProgramError);
/** InvalidUnlockTime: Unlock Time cannot be in the past */
class InvalidUnlockTimeError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Unlock Time cannot be in the past', program, cause);
        this.name = 'InvalidUnlockTime';
        this.code = 0x18; // 24
    }
}
exports.InvalidUnlockTimeError = InvalidUnlockTimeError;
codeToErrorMap.set(0x18, InvalidUnlockTimeError);
nameToErrorMap.set('InvalidUnlockTime', InvalidUnlockTimeError);
/** LaunchPoolNotEnded: LaunchPool has not ended */
class LaunchPoolNotEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('LaunchPool has not ended', program, cause);
        this.name = 'LaunchPoolNotEnded';
        this.code = 0x19; // 25
    }
}
exports.LaunchPoolNotEndedError = LaunchPoolNotEndedError;
codeToErrorMap.set(0x19, LaunchPoolNotEndedError);
nameToErrorMap.set('LaunchPoolNotEnded', LaunchPoolNotEndedError);
/** InsufficientFunds: Insufficient funds */
class InsufficientFundsError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Insufficient funds', program, cause);
        this.name = 'InsufficientFunds';
        this.code = 0x1a; // 26
    }
}
exports.InsufficientFundsError = InsufficientFundsError;
codeToErrorMap.set(0x1a, InsufficientFundsError);
nameToErrorMap.set('InsufficientFunds', InsufficientFundsError);
/** InvalidBucket: Invalid Bucket passed in */
class InvalidBucketError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Bucket passed in', program, cause);
        this.name = 'InvalidBucket';
        this.code = 0x1b; // 27
    }
}
exports.InvalidBucketError = InvalidBucketError;
codeToErrorMap.set(0x1b, InvalidBucketError);
nameToErrorMap.set('InvalidBucket', InvalidBucketError);
/** ExpectedInflowBucket: Expected inflow bucket, received outflow bucket */
class ExpectedInflowBucketError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Expected inflow bucket, received outflow bucket', program, cause);
        this.name = 'ExpectedInflowBucket';
        this.code = 0x1c; // 28
    }
}
exports.ExpectedInflowBucketError = ExpectedInflowBucketError;
codeToErrorMap.set(0x1c, ExpectedInflowBucketError);
nameToErrorMap.set('ExpectedInflowBucket', ExpectedInflowBucketError);
/** ExpectedOutflowBucket: Expected outflow bucket, received inflow bucket */
class ExpectedOutflowBucketError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Expected outflow bucket, received inflow bucket', program, cause);
        this.name = 'ExpectedOutflowBucket';
        this.code = 0x1d; // 29
    }
}
exports.ExpectedOutflowBucketError = ExpectedOutflowBucketError;
codeToErrorMap.set(0x1d, ExpectedOutflowBucketError);
nameToErrorMap.set('ExpectedOutflowBucket', ExpectedOutflowBucketError);
/** MissingTokenAccount: Missing token account */
class MissingTokenAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Missing token account', program, cause);
        this.name = 'MissingTokenAccount';
        this.code = 0x1e; // 30
    }
}
exports.MissingTokenAccountError = MissingTokenAccountError;
codeToErrorMap.set(0x1e, MissingTokenAccountError);
nameToErrorMap.set('MissingTokenAccount', MissingTokenAccountError);
/** InvalidAssociatedTokenProgram: Invalid Associated Token Program */
class InvalidAssociatedTokenProgramError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Associated Token Program', program, cause);
        this.name = 'InvalidAssociatedTokenProgram';
        this.code = 0x1f; // 31
    }
}
exports.InvalidAssociatedTokenProgramError = InvalidAssociatedTokenProgramError;
codeToErrorMap.set(0x1f, InvalidAssociatedTokenProgramError);
nameToErrorMap.set('InvalidAssociatedTokenProgram', InvalidAssociatedTokenProgramError);
/** InvalidEndTime: End Time must be after Start Time */
class InvalidEndTimeError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('End Time must be after Start Time', program, cause);
        this.name = 'InvalidEndTime';
        this.code = 0x20; // 32
    }
}
exports.InvalidEndTimeError = InvalidEndTimeError;
codeToErrorMap.set(0x20, InvalidEndTimeError);
nameToErrorMap.set('InvalidEndTime', InvalidEndTimeError);
/** InvalidClaimStartTime: Claim Start Time must be after End Time */
class InvalidClaimStartTimeError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Claim Start Time must be after End Time', program, cause);
        this.name = 'InvalidClaimStartTime';
        this.code = 0x21; // 33
    }
}
exports.InvalidClaimStartTimeError = InvalidClaimStartTimeError;
codeToErrorMap.set(0x21, InvalidClaimStartTimeError);
nameToErrorMap.set('InvalidClaimStartTime', InvalidClaimStartTimeError);
/** InvalidMintForGenesisAccount: The mint account does not match the Genesis Account Mint */
class InvalidMintForGenesisAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The mint account does not match the Genesis Account Mint', program, cause);
        this.name = 'InvalidMintForGenesisAccount';
        this.code = 0x22; // 34
    }
}
exports.InvalidMintForGenesisAccountError = InvalidMintForGenesisAccountError;
codeToErrorMap.set(0x22, InvalidMintForGenesisAccountError);
nameToErrorMap.set('InvalidMintForGenesisAccount', InvalidMintForGenesisAccountError);
/** InvalidGenesisAccountForBucket: The Genesis Account does not match the Bucket's Genesis Account */
class InvalidGenesisAccountForBucketError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("The Genesis Account does not match the Bucket's Genesis Account", program, cause);
        this.name = 'InvalidGenesisAccountForBucket';
        this.code = 0x23; // 35
    }
}
exports.InvalidGenesisAccountForBucketError = InvalidGenesisAccountForBucketError;
codeToErrorMap.set(0x23, InvalidGenesisAccountForBucketError);
nameToErrorMap.set('InvalidGenesisAccountForBucket', InvalidGenesisAccountForBucketError);
/** InvalidBucketForDepositPda: The Bucket does not match the Deposit PDA's Bucket */
class InvalidBucketForDepositPdaError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("The Bucket does not match the Deposit PDA's Bucket", program, cause);
        this.name = 'InvalidBucketForDepositPda';
        this.code = 0x24; // 36
    }
}
exports.InvalidBucketForDepositPdaError = InvalidBucketForDepositPdaError;
codeToErrorMap.set(0x24, InvalidBucketForDepositPdaError);
nameToErrorMap.set('InvalidBucketForDepositPda', InvalidBucketForDepositPdaError);
/** InvalidMintForTokenAccount: The Mint does not match the Token Account's Mint */
class InvalidMintForTokenAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("The Mint does not match the Token Account's Mint", program, cause);
        this.name = 'InvalidMintForTokenAccount';
        this.code = 0x25; // 37
    }
}
exports.InvalidMintForTokenAccountError = InvalidMintForTokenAccountError;
codeToErrorMap.set(0x25, InvalidMintForTokenAccountError);
nameToErrorMap.set('InvalidMintForTokenAccount', InvalidMintForTokenAccountError);
/** InvalidTokenAccountOwner: The Token Account's owner does not match the recipient */
class InvalidTokenAccountOwnerError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("The Token Account's owner does not match the recipient", program, cause);
        this.name = 'InvalidTokenAccountOwner';
        this.code = 0x26; // 38
    }
}
exports.InvalidTokenAccountOwnerError = InvalidTokenAccountOwnerError;
codeToErrorMap.set(0x26, InvalidTokenAccountOwnerError);
nameToErrorMap.set('InvalidTokenAccountOwner', InvalidTokenAccountOwnerError);
/** RecipientMismatch: The recipient does not match the Bucket's recipient */
class RecipientMismatchError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("The recipient does not match the Bucket's recipient", program, cause);
        this.name = 'RecipientMismatch';
        this.code = 0x27; // 39
    }
}
exports.RecipientMismatchError = RecipientMismatchError;
codeToErrorMap.set(0x27, RecipientMismatchError);
nameToErrorMap.set('RecipientMismatch', RecipientMismatchError);
/** InvalidRent: The Rent Sysvar is invalid */
class InvalidRentError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Rent Sysvar is invalid', program, cause);
        this.name = 'InvalidRent';
        this.code = 0x28; // 40
    }
}
exports.InvalidRentError = InvalidRentError;
codeToErrorMap.set(0x28, InvalidRentError);
nameToErrorMap.set('InvalidRent', InvalidRentError);
/** InvalidTimelockProgram: The Timelock Program is invalid */
class InvalidTimelockProgramError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Timelock Program is invalid', program, cause);
        this.name = 'InvalidTimelockProgram';
        this.code = 0x29; // 41
    }
}
exports.InvalidTimelockProgramError = InvalidTimelockProgramError;
codeToErrorMap.set(0x29, InvalidTimelockProgramError);
nameToErrorMap.set('InvalidTimelockProgram', InvalidTimelockProgramError);
/** LockNotStarted: The lock period has not started */
class LockNotStartedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The lock period has not started', program, cause);
        this.name = 'LockNotStarted';
        this.code = 0x2a; // 42
    }
}
exports.LockNotStartedError = LockNotStartedError;
codeToErrorMap.set(0x2a, LockNotStartedError);
nameToErrorMap.set('LockNotStarted', LockNotStartedError);
/** GenesisAccountFinalized: The Genesis Account is already finalized and no new buckets can be added */
class GenesisAccountFinalizedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Genesis Account is already finalized and no new buckets can be added', program, cause);
        this.name = 'GenesisAccountFinalized';
        this.code = 0x2b; // 43
    }
}
exports.GenesisAccountFinalizedError = GenesisAccountFinalizedError;
codeToErrorMap.set(0x2b, GenesisAccountFinalizedError);
nameToErrorMap.set('GenesisAccountFinalized', GenesisAccountFinalizedError);
/** GenesisAccountNotFinalized: The Genesis Account is not finalized */
class GenesisAccountNotFinalizedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Genesis Account is not finalized', program, cause);
        this.name = 'GenesisAccountNotFinalized';
        this.code = 0x2c; // 44
    }
}
exports.GenesisAccountNotFinalizedError = GenesisAccountNotFinalizedError;
codeToErrorMap.set(0x2c, GenesisAccountNotFinalizedError);
nameToErrorMap.set('GenesisAccountNotFinalized', GenesisAccountNotFinalizedError);
/** InvalidDistributionAddress: The Distribution Address is invalid */
class InvalidDistributionAddressError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Distribution Address is invalid', program, cause);
        this.name = 'InvalidDistributionAddress';
        this.code = 0x2d; // 45
    }
}
exports.InvalidDistributionAddressError = InvalidDistributionAddressError;
codeToErrorMap.set(0x2d, InvalidDistributionAddressError);
nameToErrorMap.set('InvalidDistributionAddress', InvalidDistributionAddressError);
/** InvalidBucketState: The Bucket is in an invalid state for this operation */
class InvalidBucketStateError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Bucket is in an invalid state for this operation', program, cause);
        this.name = 'InvalidBucketState';
        this.code = 0x2e; // 46
    }
}
exports.InvalidBucketStateError = InvalidBucketStateError;
codeToErrorMap.set(0x2e, InvalidBucketStateError);
nameToErrorMap.set('InvalidBucketState', InvalidBucketStateError);
/** InvalidGenesisAccount: The Genesis Account is invalid */
class InvalidGenesisAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Genesis Account is invalid', program, cause);
        this.name = 'InvalidGenesisAccount';
        this.code = 0x2f; // 47
    }
}
exports.InvalidGenesisAccountError = InvalidGenesisAccountError;
codeToErrorMap.set(0x2f, InvalidGenesisAccountError);
nameToErrorMap.set('InvalidGenesisAccount', InvalidGenesisAccountError);
/** InvalidBucketIndex: The Bucket index is invalid */
class InvalidBucketIndexError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Bucket index is invalid', program, cause);
        this.name = 'InvalidBucketIndex';
        this.code = 0x30; // 48
    }
}
exports.InvalidBucketIndexError = InvalidBucketIndexError;
codeToErrorMap.set(0x30, InvalidBucketIndexError);
nameToErrorMap.set('InvalidBucketIndex', InvalidBucketIndexError);
/** DepositAlreadyClaimed: Deposit has already been claimed */
class DepositAlreadyClaimedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Deposit has already been claimed', program, cause);
        this.name = 'DepositAlreadyClaimed';
        this.code = 0x31; // 49
    }
}
exports.DepositAlreadyClaimedError = DepositAlreadyClaimedError;
codeToErrorMap.set(0x31, DepositAlreadyClaimedError);
nameToErrorMap.set('DepositAlreadyClaimed', DepositAlreadyClaimedError);
/** InvalidBucketAddress: The Bucket address doesn't match the expected PDA derivation */
class InvalidBucketAddressError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("The Bucket address doesn't match the expected PDA derivation", program, cause);
        this.name = 'InvalidBucketAddress';
        this.code = 0x32; // 50
    }
}
exports.InvalidBucketAddressError = InvalidBucketAddressError;
codeToErrorMap.set(0x32, InvalidBucketAddressError);
nameToErrorMap.set('InvalidBucketAddress', InvalidBucketAddressError);
/** UnlockedBucketAlreadyClaimed: Unlocked bucket has already been claimed */
class UnlockedBucketAlreadyClaimedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Unlocked bucket has already been claimed', program, cause);
        this.name = 'UnlockedBucketAlreadyClaimed';
        this.code = 0x33; // 51
    }
}
exports.UnlockedBucketAlreadyClaimedError = UnlockedBucketAlreadyClaimedError;
codeToErrorMap.set(0x33, UnlockedBucketAlreadyClaimedError);
nameToErrorMap.set('UnlockedBucketAlreadyClaimed', UnlockedBucketAlreadyClaimedError);
/** InvalidProgramOwnerForBucket: The Bucket's program owner is invalid */
class InvalidProgramOwnerForBucketError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("The Bucket's program owner is invalid", program, cause);
        this.name = 'InvalidProgramOwnerForBucket';
        this.code = 0x34; // 52
    }
}
exports.InvalidProgramOwnerForBucketError = InvalidProgramOwnerForBucketError;
codeToErrorMap.set(0x34, InvalidProgramOwnerForBucketError);
nameToErrorMap.set('InvalidProgramOwnerForBucket', InvalidProgramOwnerForBucketError);
/** BucketIsNotYetFunded: Bucket is not yet funded */
class BucketIsNotYetFundedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Bucket is not yet funded', program, cause);
        this.name = 'BucketIsNotYetFunded';
        this.code = 0x35; // 53
    }
}
exports.BucketIsNotYetFundedError = BucketIsNotYetFundedError;
codeToErrorMap.set(0x35, BucketIsNotYetFundedError);
nameToErrorMap.set('BucketIsNotYetFunded', BucketIsNotYetFundedError);
/** StreamflowNotInitialized: Streamflow bucket is not initialized */
class StreamflowNotInitializedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Streamflow bucket is not initialized', program, cause);
        this.name = 'StreamflowNotInitialized';
        this.code = 0x36; // 54
    }
}
exports.StreamflowNotInitializedError = StreamflowNotInitializedError;
codeToErrorMap.set(0x36, StreamflowNotInitializedError);
nameToErrorMap.set('StreamflowNotInitialized', StreamflowNotInitializedError);
/** UnlockedBucketNotInitialized: Unlocked bucket is not initialized */
class UnlockedBucketNotInitializedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Unlocked bucket is not initialized', program, cause);
        this.name = 'UnlockedBucketNotInitialized';
        this.code = 0x37; // 55
    }
}
exports.UnlockedBucketNotInitializedError = UnlockedBucketNotInitializedError;
codeToErrorMap.set(0x37, UnlockedBucketNotInitializedError);
nameToErrorMap.set('UnlockedBucketNotInitialized', UnlockedBucketNotInitializedError);
/** BucketDoesNotSupportFunding: Bucket does not support funding */
class BucketDoesNotSupportFundingError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Bucket does not support funding', program, cause);
        this.name = 'BucketDoesNotSupportFunding';
        this.code = 0x38; // 56
    }
}
exports.BucketDoesNotSupportFundingError = BucketDoesNotSupportFundingError;
codeToErrorMap.set(0x38, BucketDoesNotSupportFundingError);
nameToErrorMap.set('BucketDoesNotSupportFunding', BucketDoesNotSupportFundingError);
/** InflowsMustBeCompleted: Inflows must be completed */
class InflowsMustBeCompletedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Inflows must be completed', program, cause);
        this.name = 'InflowsMustBeCompleted';
        this.code = 0x39; // 57
    }
}
exports.InflowsMustBeCompletedError = InflowsMustBeCompletedError;
codeToErrorMap.set(0x39, InflowsMustBeCompletedError);
nameToErrorMap.set('InflowsMustBeCompleted', InflowsMustBeCompletedError);
/** InvalidAtaDerivation: Invalid Associated Token Account derivation */
class InvalidAtaDerivationError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Associated Token Account derivation', program, cause);
        this.name = 'InvalidAtaDerivation';
        this.code = 0x3a; // 58
    }
}
exports.InvalidAtaDerivationError = InvalidAtaDerivationError;
codeToErrorMap.set(0x3a, InvalidAtaDerivationError);
nameToErrorMap.set('InvalidAtaDerivation', InvalidAtaDerivationError);
/** DistributionNotStarted: Distribution not started */
class DistributionNotStartedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Distribution not started', program, cause);
        this.name = 'DistributionNotStarted';
        this.code = 0x3b; // 59
    }
}
exports.DistributionNotStartedError = DistributionNotStartedError;
codeToErrorMap.set(0x3b, DistributionNotStartedError);
nameToErrorMap.set('DistributionNotStarted', DistributionNotStartedError);
/** GenesisAccountAlreadyFinalized: The Genesis Account is already finalized */
class GenesisAccountAlreadyFinalizedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Genesis Account is already finalized', program, cause);
        this.name = 'GenesisAccountAlreadyFinalized';
        this.code = 0x3c; // 60
    }
}
exports.GenesisAccountAlreadyFinalizedError = GenesisAccountAlreadyFinalizedError;
codeToErrorMap.set(0x3c, GenesisAccountAlreadyFinalizedError);
nameToErrorMap.set('GenesisAccountAlreadyFinalized', GenesisAccountAlreadyFinalizedError);
/** InvalidSolAllocationBps: Invalid SOL Allocation BPS */
class InvalidSolAllocationBpsError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid SOL Allocation BPS', program, cause);
        this.name = 'InvalidSolAllocationBps';
        this.code = 0x3d; // 61
    }
}
exports.InvalidSolAllocationBpsError = InvalidSolAllocationBpsError;
codeToErrorMap.set(0x3d, InvalidSolAllocationBpsError);
nameToErrorMap.set('InvalidSolAllocationBps', InvalidSolAllocationBpsError);
/** InvalidMintForDistribution: The Mint does not match the Distribution's Mint */
class InvalidMintForDistributionError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("The Mint does not match the Distribution's Mint", program, cause);
        this.name = 'InvalidMintForDistribution';
        this.code = 0x3e; // 62
    }
}
exports.InvalidMintForDistributionError = InvalidMintForDistributionError;
codeToErrorMap.set(0x3e, InvalidMintForDistributionError);
nameToErrorMap.set('InvalidMintForDistribution', InvalidMintForDistributionError);
/** DistributionEnded: The Distribution has ended */
class DistributionEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Distribution has ended', program, cause);
        this.name = 'DistributionEnded';
        this.code = 0x3f; // 63
    }
}
exports.DistributionEndedError = DistributionEndedError;
codeToErrorMap.set(0x3f, DistributionEndedError);
nameToErrorMap.set('DistributionEnded', DistributionEndedError);
/** PresaleSaleNotStarted: The Presale has not started */
class PresaleSaleNotStartedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Presale has not started', program, cause);
        this.name = 'PresaleSaleNotStarted';
        this.code = 0x40; // 64
    }
}
exports.PresaleSaleNotStartedError = PresaleSaleNotStartedError;
codeToErrorMap.set(0x40, PresaleSaleNotStartedError);
nameToErrorMap.set('PresaleSaleNotStarted', PresaleSaleNotStartedError);
/** PresaleSaleEnded: The Presale has ended */
class PresaleSaleEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Presale has ended', program, cause);
        this.name = 'PresaleSaleEnded';
        this.code = 0x41; // 65
    }
}
exports.PresaleSaleEndedError = PresaleSaleEndedError;
codeToErrorMap.set(0x41, PresaleSaleEndedError);
nameToErrorMap.set('PresaleSaleEnded', PresaleSaleEndedError);
/** PresaleBucketAlreadyFunded: The Presale Bucket has already been fully funded */
class PresaleBucketAlreadyFundedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Presale Bucket has already been fully funded', program, cause);
        this.name = 'PresaleBucketAlreadyFunded';
        this.code = 0x42; // 66
    }
}
exports.PresaleBucketAlreadyFundedError = PresaleBucketAlreadyFundedError;
codeToErrorMap.set(0x42, PresaleBucketAlreadyFundedError);
nameToErrorMap.set('PresaleBucketAlreadyFunded', PresaleBucketAlreadyFundedError);
/** PresaleBucketNotEnded: The Presale Bucket has not ended */
class PresaleBucketNotEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Presale Bucket has not ended', program, cause);
        this.name = 'PresaleBucketNotEnded';
        this.code = 0x43; // 67
    }
}
exports.PresaleBucketNotEndedError = PresaleBucketNotEndedError;
codeToErrorMap.set(0x43, PresaleBucketNotEndedError);
nameToErrorMap.set('PresaleBucketNotEnded', PresaleBucketNotEndedError);
/** PresaleBucketWithdrawNotAllowed: The Presale Bucket does not allow withdrawals */
class PresaleBucketWithdrawNotAllowedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Presale Bucket does not allow withdrawals', program, cause);
        this.name = 'PresaleBucketWithdrawNotAllowed';
        this.code = 0x44; // 68
    }
}
exports.PresaleBucketWithdrawNotAllowedError = PresaleBucketWithdrawNotAllowedError;
codeToErrorMap.set(0x44, PresaleBucketWithdrawNotAllowedError);
nameToErrorMap.set('PresaleBucketWithdrawNotAllowed', PresaleBucketWithdrawNotAllowedError);
/** InvalidBackendSigner: The backend signer is missing or invalid */
class InvalidBackendSignerError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The backend signer is missing or invalid', program, cause);
        this.name = 'InvalidBackendSigner';
        this.code = 0x45; // 69
    }
}
exports.InvalidBackendSignerError = InvalidBackendSignerError;
codeToErrorMap.set(0x45, InvalidBackendSignerError);
nameToErrorMap.set('InvalidBackendSigner', InvalidBackendSignerError);
/** InvalidPresaleDepositAmount: The deposit amount is invalid */
class InvalidPresaleDepositAmountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The deposit amount is invalid', program, cause);
        this.name = 'InvalidPresaleDepositAmount';
        this.code = 0x46; // 70
    }
}
exports.InvalidPresaleDepositAmountError = InvalidPresaleDepositAmountError;
codeToErrorMap.set(0x46, InvalidPresaleDepositAmountError);
nameToErrorMap.set('InvalidPresaleDepositAmount', InvalidPresaleDepositAmountError);
/** PresaleDepositLimitExceeded: The deposit amount exceeds the deposit limit */
class PresaleDepositLimitExceededError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The deposit amount exceeds the deposit limit', program, cause);
        this.name = 'PresaleDepositLimitExceeded';
        this.code = 0x47; // 71
    }
}
exports.PresaleDepositLimitExceededError = PresaleDepositLimitExceededError;
codeToErrorMap.set(0x47, PresaleDepositLimitExceededError);
nameToErrorMap.set('PresaleDepositLimitExceeded', PresaleDepositLimitExceededError);
/** PresaleDepositCooldownActive: The deposit cooldown is still active */
class PresaleDepositCooldownActiveError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The deposit cooldown is still active', program, cause);
        this.name = 'PresaleDepositCooldownActive';
        this.code = 0x48; // 72
    }
}
exports.PresaleDepositCooldownActiveError = PresaleDepositCooldownActiveError;
codeToErrorMap.set(0x48, PresaleDepositCooldownActiveError);
nameToErrorMap.set('PresaleDepositCooldownActive', PresaleDepositCooldownActiveError);
/** PresalePerCooldownDepositLimitExceeded: The deposit amount exceeds the per-cooldown deposit limit */
class PresalePerCooldownDepositLimitExceededError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The deposit amount exceeds the per-cooldown deposit limit', program, cause);
        this.name = 'PresalePerCooldownDepositLimitExceeded';
        this.code = 0x49; // 73
    }
}
exports.PresalePerCooldownDepositLimitExceededError = PresalePerCooldownDepositLimitExceededError;
codeToErrorMap.set(0x49, PresalePerCooldownDepositLimitExceededError);
nameToErrorMap.set('PresalePerCooldownDepositLimitExceeded', PresalePerCooldownDepositLimitExceededError);
/** InvalidMerkleProof: The merkle proof is invalid */
class InvalidMerkleProofError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The merkle proof is invalid', program, cause);
        this.name = 'InvalidMerkleProof';
        this.code = 0x4a; // 74
    }
}
exports.InvalidMerkleProofError = InvalidMerkleProofError;
codeToErrorMap.set(0x4a, InvalidMerkleProofError);
nameToErrorMap.set('InvalidMerkleProof', InvalidMerkleProofError);
/** InvalidFeeAuthority: The Fee Authority is invalid */
class InvalidFeeAuthorityError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Fee Authority is invalid', program, cause);
        this.name = 'InvalidFeeAuthority';
        this.code = 0x4b; // 75
    }
}
exports.InvalidFeeAuthorityError = InvalidFeeAuthorityError;
codeToErrorMap.set(0x4b, InvalidFeeAuthorityError);
nameToErrorMap.set('InvalidFeeAuthority', InvalidFeeAuthorityError);
/** InvalidFeeOverrideType: Invalid fee override type for this bucket */
class InvalidFeeOverrideTypeError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid fee override type for this bucket', program, cause);
        this.name = 'InvalidFeeOverrideType';
        this.code = 0x4c; // 76
    }
}
exports.InvalidFeeOverrideTypeError = InvalidFeeOverrideTypeError;
codeToErrorMap.set(0x4c, InvalidFeeOverrideTypeError);
nameToErrorMap.set('InvalidFeeOverrideType', InvalidFeeOverrideTypeError);
/** InvalidFeeOverride: Invalid fee override */
class InvalidFeeOverrideError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid fee override', program, cause);
        this.name = 'InvalidFeeOverride';
        this.code = 0x4d; // 77
    }
}
exports.InvalidFeeOverrideError = InvalidFeeOverrideError;
codeToErrorMap.set(0x4d, InvalidFeeOverrideError);
nameToErrorMap.set('InvalidFeeOverride', InvalidFeeOverrideError);
/** FeeTransferFailed: Dynamic penalty transfer failed */
class FeeTransferFailedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Dynamic penalty transfer failed', program, cause);
        this.name = 'FeeTransferFailed';
        this.code = 0x4e; // 78
    }
}
exports.FeeTransferFailedError = FeeTransferFailedError;
codeToErrorMap.set(0x4e, FeeTransferFailedError);
nameToErrorMap.set('FeeTransferFailed', FeeTransferFailedError);
/** InvalidPenaltyWallet: Invalid or missing penalty wallet */
class InvalidPenaltyWalletError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid or missing penalty wallet', program, cause);
        this.name = 'InvalidPenaltyWallet';
        this.code = 0x4f; // 79
    }
}
exports.InvalidPenaltyWalletError = InvalidPenaltyWalletError;
codeToErrorMap.set(0x4f, InvalidPenaltyWalletError);
nameToErrorMap.set('InvalidPenaltyWallet', InvalidPenaltyWalletError);
/** InvalidLinearSchedule: Invalid linear schedule parameters */
class InvalidLinearScheduleError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid linear schedule parameters', program, cause);
        this.name = 'InvalidLinearSchedule';
        this.code = 0x50; // 80
    }
}
exports.InvalidLinearScheduleError = InvalidLinearScheduleError;
codeToErrorMap.set(0x50, InvalidLinearScheduleError);
nameToErrorMap.set('InvalidLinearSchedule', InvalidLinearScheduleError);
/** InsufficientFundsForPenaltyFee: Payer has insufficient funds to cover penalty fee and maintain minimum balance */
class InsufficientFundsForPenaltyFeeError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Payer has insufficient funds to cover penalty fee and maintain minimum balance', program, cause);
        this.name = 'InsufficientFundsForPenaltyFee';
        this.code = 0x51; // 81
    }
}
exports.InsufficientFundsForPenaltyFeeError = InsufficientFundsForPenaltyFeeError;
codeToErrorMap.set(0x51, InsufficientFundsForPenaltyFeeError);
nameToErrorMap.set('InsufficientFundsForPenaltyFee', InsufficientFundsForPenaltyFeeError);
/** InvalidScheduleTiming: Schedule start/end times are invalid or outside bucket bounds */
class InvalidScheduleTimingError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Schedule start/end times are invalid or outside bucket bounds', program, cause);
        this.name = 'InvalidScheduleTiming';
        this.code = 0x52; // 82
    }
}
exports.InvalidScheduleTimingError = InvalidScheduleTimingError;
codeToErrorMap.set(0x52, InvalidScheduleTimingError);
nameToErrorMap.set('InvalidScheduleTiming', InvalidScheduleTimingError);
/** VaultNotEnded: Vault bucket is not ended */
class VaultNotEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Vault bucket is not ended', program, cause);
        this.name = 'VaultNotEnded';
        this.code = 0x53; // 83
    }
}
exports.VaultNotEndedError = VaultNotEndedError;
codeToErrorMap.set(0x53, VaultNotEndedError);
nameToErrorMap.set('VaultNotEnded', VaultNotEndedError);
/** VaultNotStarted: Vault bucket is not started */
class VaultNotStartedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Vault bucket is not started', program, cause);
        this.name = 'VaultNotStarted';
        this.code = 0x54; // 84
    }
}
exports.VaultNotStartedError = VaultNotStartedError;
codeToErrorMap.set(0x54, VaultNotStartedError);
nameToErrorMap.set('VaultNotStarted', VaultNotStartedError);
/** VaultEnded: Vault bucket has ended */
class VaultEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Vault bucket has ended', program, cause);
        this.name = 'VaultEnded';
        this.code = 0x55; // 85
    }
}
exports.VaultEndedError = VaultEndedError;
codeToErrorMap.set(0x55, VaultEndedError);
nameToErrorMap.set('VaultEnded', VaultEndedError);
/** InvalidAmount: The amount is invalid */
class InvalidAmountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The amount is invalid', program, cause);
        this.name = 'InvalidAmount';
        this.code = 0x56; // 86
    }
}
exports.InvalidAmountError = InvalidAmountError;
codeToErrorMap.set(0x56, InvalidAmountError);
nameToErrorMap.set('InvalidAmount', InvalidAmountError);
/** InsufficientOutputAmount: Insufficient output amount */
class InsufficientOutputAmountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Insufficient output amount', program, cause);
        this.name = 'InsufficientOutputAmount';
        this.code = 0x57; // 87
    }
}
exports.InsufficientOutputAmountError = InsufficientOutputAmountError;
codeToErrorMap.set(0x57, InsufficientOutputAmountError);
nameToErrorMap.set('InsufficientOutputAmount', InsufficientOutputAmountError);
/** BondingCurveNotEnded: Bonding Curve not ended */
class BondingCurveNotEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Bonding Curve not ended', program, cause);
        this.name = 'BondingCurveNotEnded';
        this.code = 0x58; // 88
    }
}
exports.BondingCurveNotEndedError = BondingCurveNotEndedError;
codeToErrorMap.set(0x58, BondingCurveNotEndedError);
nameToErrorMap.set('BondingCurveNotEnded', BondingCurveNotEndedError);
/** InvalidConstantProduct: The constant product of the bonding curve is invalid (liquidities on both sides must always be greater than zero) */
class InvalidConstantProductError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The constant product of the bonding curve is invalid (liquidities on both sides must always be greater than zero)', program, cause);
        this.name = 'InvalidConstantProduct';
        this.code = 0x59; // 89
    }
}
exports.InvalidConstantProductError = InvalidConstantProductError;
codeToErrorMap.set(0x59, InvalidConstantProductError);
nameToErrorMap.set('InvalidConstantProduct', InvalidConstantProductError);
/** BondingCurveNotStarted: Bonding Curve not started */
class BondingCurveNotStartedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Bonding Curve not started', program, cause);
        this.name = 'BondingCurveNotStarted';
        this.code = 0x5a; // 90
    }
}
exports.BondingCurveNotStartedError = BondingCurveNotStartedError;
codeToErrorMap.set(0x5a, BondingCurveNotStartedError);
nameToErrorMap.set('BondingCurveNotStarted', BondingCurveNotStartedError);
/** BondingCurveEnded: Bonding Curve ended */
class BondingCurveEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Bonding Curve ended', program, cause);
        this.name = 'BondingCurveEnded';
        this.code = 0x5b; // 91
    }
}
exports.BondingCurveEndedError = BondingCurveEndedError;
codeToErrorMap.set(0x5b, BondingCurveEndedError);
nameToErrorMap.set('BondingCurveEnded', BondingCurveEndedError);
/** BondingCurveConstantProductDecreased: Bonding Curve constant product decreased */
class BondingCurveConstantProductDecreasedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Bonding Curve constant product decreased', program, cause);
        this.name = 'BondingCurveConstantProductDecreased';
        this.code = 0x5c; // 92
    }
}
exports.BondingCurveConstantProductDecreasedError = BondingCurveConstantProductDecreasedError;
codeToErrorMap.set(0x5c, BondingCurveConstantProductDecreasedError);
nameToErrorMap.set('BondingCurveConstantProductDecreased', BondingCurveConstantProductDecreasedError);
/** BondingCurveInsufficientFunds: Bonding Curve has insufficient funds */
class BondingCurveInsufficientFundsError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Bonding Curve has insufficient funds', program, cause);
        this.name = 'BondingCurveInsufficientFunds';
        this.code = 0x5d; // 93
    }
}
exports.BondingCurveInsufficientFundsError = BondingCurveInsufficientFundsError;
codeToErrorMap.set(0x5d, BondingCurveInsufficientFundsError);
nameToErrorMap.set('BondingCurveInsufficientFunds', BondingCurveInsufficientFundsError);
/** InvalidOrdinal: Invalid Ordinal */
class InvalidOrdinalError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Ordinal', program, cause);
        this.name = 'InvalidOrdinal';
        this.code = 0x5e; // 94
    }
}
exports.InvalidOrdinalError = InvalidOrdinalError;
codeToErrorMap.set(0x5e, InvalidOrdinalError);
nameToErrorMap.set('InvalidOrdinal', InvalidOrdinalError);
/** InvalidActionsAuthority: The actions authority is invalid */
class InvalidActionsAuthorityError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The actions authority is invalid', program, cause);
        this.name = 'InvalidActionsAuthority';
        this.code = 0x5f; // 95
    }
}
exports.InvalidActionsAuthorityError = InvalidActionsAuthorityError;
codeToErrorMap.set(0x5f, InvalidActionsAuthorityError);
nameToErrorMap.set('InvalidActionsAuthority', InvalidActionsAuthorityError);
/** VaultWithdrawNotAllowed: Vault bucket does not allow withdrawals */
class VaultWithdrawNotAllowedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Vault bucket does not allow withdrawals', program, cause);
        this.name = 'VaultWithdrawNotAllowed';
        this.code = 0x60; // 96
    }
}
exports.VaultWithdrawNotAllowedError = VaultWithdrawNotAllowedError;
codeToErrorMap.set(0x60, VaultWithdrawNotAllowedError);
nameToErrorMap.set('VaultWithdrawNotAllowed', VaultWithdrawNotAllowedError);
/** VaultDepositConsumed: Vault deposit is consumed */
class VaultDepositConsumedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Vault deposit is consumed', program, cause);
        this.name = 'VaultDepositConsumed';
        this.code = 0x61; // 97
    }
}
exports.VaultDepositConsumedError = VaultDepositConsumedError;
codeToErrorMap.set(0x61, VaultDepositConsumedError);
nameToErrorMap.set('VaultDepositConsumed', VaultDepositConsumedError);
/** BucketPaused: Bucket is paused */
class BucketPausedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Bucket is paused', program, cause);
        this.name = 'BucketPaused';
        this.code = 0x62; // 98
    }
}
exports.BucketPausedError = BucketPausedError;
codeToErrorMap.set(0x62, BucketPausedError);
nameToErrorMap.set('BucketPaused', BucketPausedError);
/** InvalidAuctionTree: Invalid auction tree - does not match the bucket's tree account */
class InvalidAuctionTreeError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("Invalid auction tree - does not match the bucket's tree account", program, cause);
        this.name = 'InvalidAuctionTree';
        this.code = 0x63; // 99
    }
}
exports.InvalidAuctionTreeError = InvalidAuctionTreeError;
codeToErrorMap.set(0x63, InvalidAuctionTreeError);
nameToErrorMap.set('InvalidAuctionTree', InvalidAuctionTreeError);
/** InvalidAuctionTreeState: Invalid auction tree state - tree is inconsistent with expected data */
class InvalidAuctionTreeStateError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid auction tree state - tree is inconsistent with expected data', program, cause);
        this.name = 'InvalidAuctionTreeState';
        this.code = 0x64; // 100
    }
}
exports.InvalidAuctionTreeStateError = InvalidAuctionTreeStateError;
codeToErrorMap.set(0x64, InvalidAuctionTreeStateError);
nameToErrorMap.set('InvalidAuctionTreeState', InvalidAuctionTreeStateError);
/** InvalidTokenAllocation: Token allocation must be greater than 0 */
class InvalidTokenAllocationError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Token allocation must be greater than 0', program, cause);
        this.name = 'InvalidTokenAllocation';
        this.code = 0x65; // 101
    }
}
exports.InvalidTokenAllocationError = InvalidTokenAllocationError;
codeToErrorMap.set(0x65, InvalidTokenAllocationError);
nameToErrorMap.set('InvalidTokenAllocation', InvalidTokenAllocationError);
/** InvalidBidIncrement: Minimum bid increment must be greater than 0 */
class InvalidBidIncrementError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Minimum bid increment must be greater than 0', program, cause);
        this.name = 'InvalidBidIncrement';
        this.code = 0x66; // 102
    }
}
exports.InvalidBidIncrementError = InvalidBidIncrementError;
codeToErrorMap.set(0x66, InvalidBidIncrementError);
nameToErrorMap.set('InvalidBidIncrement', InvalidBidIncrementError);
/** InvalidTokenQuantityBps: Minimum token quantity basis points must be between 1 and 10000 */
class InvalidTokenQuantityBpsError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Minimum token quantity basis points must be between 1 and 10000', program, cause);
        this.name = 'InvalidTokenQuantityBps';
        this.code = 0x67; // 103
    }
}
exports.InvalidTokenQuantityBpsError = InvalidTokenQuantityBpsError;
codeToErrorMap.set(0x67, InvalidTokenQuantityBpsError);
nameToErrorMap.set('InvalidTokenQuantityBps', InvalidTokenQuantityBpsError);
/** InvalidMinimumPrice: Minimum price must be greater than 0 */
class InvalidMinimumPriceError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Minimum price must be greater than 0', program, cause);
        this.name = 'InvalidMinimumPrice';
        this.code = 0x68; // 104
    }
}
exports.InvalidMinimumPriceError = InvalidMinimumPriceError;
codeToErrorMap.set(0x68, InvalidMinimumPriceError);
nameToErrorMap.set('InvalidMinimumPrice', InvalidMinimumPriceError);
/** InvalidMaximumBidQuantity: Maximum bid quantity must be greater than 0 */
class InvalidMaximumBidQuantityError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Maximum bid quantity must be greater than 0', program, cause);
        this.name = 'InvalidMaximumBidQuantity';
        this.code = 0x69; // 105
    }
}
exports.InvalidMaximumBidQuantityError = InvalidMaximumBidQuantityError;
codeToErrorMap.set(0x69, InvalidMaximumBidQuantityError);
nameToErrorMap.set('InvalidMaximumBidQuantity', InvalidMaximumBidQuantityError);
/** MaximumBidQuantityExceedsAllocation: Maximum bid quantity cannot exceed token allocation */
class MaximumBidQuantityExceedsAllocationError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Maximum bid quantity cannot exceed token allocation', program, cause);
        this.name = 'MaximumBidQuantityExceedsAllocation';
        this.code = 0x6a; // 106
    }
}
exports.MaximumBidQuantityExceedsAllocationError = MaximumBidQuantityExceedsAllocationError;
codeToErrorMap.set(0x6a, MaximumBidQuantityExceedsAllocationError);
nameToErrorMap.set('MaximumBidQuantityExceedsAllocation', MaximumBidQuantityExceedsAllocationError);
/** InvalidAuctionExtensionConfig: Invalid auction extension configuration */
class InvalidAuctionExtensionConfigError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid auction extension configuration', program, cause);
        this.name = 'InvalidAuctionExtensionConfig';
        this.code = 0x6b; // 107
    }
}
exports.InvalidAuctionExtensionConfigError = InvalidAuctionExtensionConfigError;
codeToErrorMap.set(0x6b, InvalidAuctionExtensionConfigError);
nameToErrorMap.set('InvalidAuctionExtensionConfig', InvalidAuctionExtensionConfigError);
/** AuctionNotStarted: Auction has not started yet - bids cannot be placed */
class AuctionNotStartedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Auction has not started yet - bids cannot be placed', program, cause);
        this.name = 'AuctionNotStarted';
        this.code = 0x6c; // 108
    }
}
exports.AuctionNotStartedError = AuctionNotStartedError;
codeToErrorMap.set(0x6c, AuctionNotStartedError);
nameToErrorMap.set('AuctionNotStarted', AuctionNotStartedError);
/** AuctionEnded: Auction has ended - no more bids can be placed */
class AuctionEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Auction has ended - no more bids can be placed', program, cause);
        this.name = 'AuctionEnded';
        this.code = 0x6d; // 109
    }
}
exports.AuctionEndedError = AuctionEndedError;
codeToErrorMap.set(0x6d, AuctionEndedError);
nameToErrorMap.set('AuctionEnded', AuctionEndedError);
/** AuctionNotEnded: The auction has not ended yet */
class AuctionNotEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The auction has not ended yet', program, cause);
        this.name = 'AuctionNotEnded';
        this.code = 0x6e; // 110
    }
}
exports.AuctionNotEndedError = AuctionNotEndedError;
codeToErrorMap.set(0x6e, AuctionNotEndedError);
nameToErrorMap.set('AuctionNotEnded', AuctionNotEndedError);
/** InvalidAuctionBid: Invalid auction bid */
class InvalidAuctionBidError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid auction bid', program, cause);
        this.name = 'InvalidAuctionBid';
        this.code = 0x6f; // 111
    }
}
exports.InvalidAuctionBidError = InvalidAuctionBidError;
codeToErrorMap.set(0x6f, InvalidAuctionBidError);
nameToErrorMap.set('InvalidAuctionBid', InvalidAuctionBidError);
/** InvalidAuctionBidPdaDerivation: Invalid Auction Bid PDA Derivation */
class InvalidAuctionBidPdaDerivationError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Auction Bid PDA Derivation', program, cause);
        this.name = 'InvalidAuctionBidPdaDerivation';
        this.code = 0x70; // 112
    }
}
exports.InvalidAuctionBidPdaDerivationError = InvalidAuctionBidPdaDerivationError;
codeToErrorMap.set(0x70, InvalidAuctionBidPdaDerivationError);
nameToErrorMap.set('InvalidAuctionBidPdaDerivation', InvalidAuctionBidPdaDerivationError);
/** InvalidBucketForAuctionBid: The Bucket does not match the Auction Bid's Bucket */
class InvalidBucketForAuctionBidError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("The Bucket does not match the Auction Bid's Bucket", program, cause);
        this.name = 'InvalidBucketForAuctionBid';
        this.code = 0x71; // 113
    }
}
exports.InvalidBucketForAuctionBidError = InvalidBucketForAuctionBidError;
codeToErrorMap.set(0x71, InvalidBucketForAuctionBidError);
nameToErrorMap.set('InvalidBucketForAuctionBid', InvalidBucketForAuctionBidError);
/** InvalidBidIndex: The Bid index is invalid */
class InvalidBidIndexError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Bid index is invalid', program, cause);
        this.name = 'InvalidBidIndex';
        this.code = 0x72; // 114
    }
}
exports.InvalidBidIndexError = InvalidBidIndexError;
codeToErrorMap.set(0x72, InvalidBidIndexError);
nameToErrorMap.set('InvalidBidIndex', InvalidBidIndexError);
/** UnauthorizedBidder: Unauthorized bidder */
class UnauthorizedBidderError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Unauthorized bidder', program, cause);
        this.name = 'UnauthorizedBidder';
        this.code = 0x73; // 115
    }
}
exports.UnauthorizedBidderError = UnauthorizedBidderError;
codeToErrorMap.set(0x73, UnauthorizedBidderError);
nameToErrorMap.set('UnauthorizedBidder', UnauthorizedBidderError);
/** BidPriceCannotBeZero: Bid price must be greater than 0 */
class BidPriceCannotBeZeroError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Bid price must be greater than 0', program, cause);
        this.name = 'BidPriceCannotBeZero';
        this.code = 0x74; // 116
    }
}
exports.BidPriceCannotBeZeroError = BidPriceCannotBeZeroError;
codeToErrorMap.set(0x74, BidPriceCannotBeZeroError);
nameToErrorMap.set('BidPriceCannotBeZero', BidPriceCannotBeZeroError);
/** BidQuantityCannotBeZero: Bid quantity must be greater than 0 */
class BidQuantityCannotBeZeroError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Bid quantity must be greater than 0', program, cause);
        this.name = 'BidQuantityCannotBeZero';
        this.code = 0x75; // 117
    }
}
exports.BidQuantityCannotBeZeroError = BidQuantityCannotBeZeroError;
codeToErrorMap.set(0x75, BidQuantityCannotBeZeroError);
nameToErrorMap.set('BidQuantityCannotBeZero', BidQuantityCannotBeZeroError);
/** BidPriceBelowMinimum: Bid price is below the auction's minimum price */
class BidPriceBelowMinimumError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("Bid price is below the auction's minimum price", program, cause);
        this.name = 'BidPriceBelowMinimum';
        this.code = 0x76; // 118
    }
}
exports.BidPriceBelowMinimumError = BidPriceBelowMinimumError;
codeToErrorMap.set(0x76, BidPriceBelowMinimumError);
nameToErrorMap.set('BidPriceBelowMinimum', BidPriceBelowMinimumError);
/** BidQuantityBelowMinimum: Bid quantity is below the auction's minimum token quantity threshold */
class BidQuantityBelowMinimumError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("Bid quantity is below the auction's minimum token quantity threshold", program, cause);
        this.name = 'BidQuantityBelowMinimum';
        this.code = 0x77; // 119
    }
}
exports.BidQuantityBelowMinimumError = BidQuantityBelowMinimumError;
codeToErrorMap.set(0x77, BidQuantityBelowMinimumError);
nameToErrorMap.set('BidQuantityBelowMinimum', BidQuantityBelowMinimumError);
/** BidQuantityExceedsMaximum: Bid quantity exceeds the auction's maximum bid quantity */
class BidQuantityExceedsMaximumError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("Bid quantity exceeds the auction's maximum bid quantity", program, cause);
        this.name = 'BidQuantityExceedsMaximum';
        this.code = 0x78; // 120
    }
}
exports.BidQuantityExceedsMaximumError = BidQuantityExceedsMaximumError;
codeToErrorMap.set(0x78, BidQuantityExceedsMaximumError);
nameToErrorMap.set('BidQuantityExceedsMaximum', BidQuantityExceedsMaximumError);
/** BidQuantityExceedsTotalAllocation: Bid quantity cannot exceed the auction's total token allocation */
class BidQuantityExceedsTotalAllocationError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("Bid quantity cannot exceed the auction's total token allocation", program, cause);
        this.name = 'BidQuantityExceedsTotalAllocation';
        this.code = 0x79; // 121
    }
}
exports.BidQuantityExceedsTotalAllocationError = BidQuantityExceedsTotalAllocationError;
codeToErrorMap.set(0x79, BidQuantityExceedsTotalAllocationError);
nameToErrorMap.set('BidQuantityExceedsTotalAllocation', BidQuantityExceedsTotalAllocationError);
/** BidNotCompetitive: Bid rejected - not competitive enough to replace existing bids */
class BidNotCompetitiveError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Bid rejected - not competitive enough to replace existing bids', program, cause);
        this.name = 'BidNotCompetitive';
        this.code = 0x7a; // 122
    }
}
exports.BidNotCompetitiveError = BidNotCompetitiveError;
codeToErrorMap.set(0x7a, BidNotCompetitiveError);
nameToErrorMap.set('BidNotCompetitive', BidNotCompetitiveError);
/** BidPriceCannotDecrease: Winning bids cannot decrease their price per token */
class BidPriceCannotDecreaseError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Winning bids cannot decrease their price per token', program, cause);
        this.name = 'BidPriceCannotDecrease';
        this.code = 0x7b; // 123
    }
}
exports.BidPriceCannotDecreaseError = BidPriceCannotDecreaseError;
codeToErrorMap.set(0x7b, BidPriceCannotDecreaseError);
nameToErrorMap.set('BidPriceCannotDecrease', BidPriceCannotDecreaseError);
/** BidQuantityCannotDecrease: Winning bids cannot decrease their token quantity */
class BidQuantityCannotDecreaseError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Winning bids cannot decrease their token quantity', program, cause);
        this.name = 'BidQuantityCannotDecrease';
        this.code = 0x7c; // 124
    }
}
exports.BidQuantityCannotDecreaseError = BidQuantityCannotDecreaseError;
codeToErrorMap.set(0x7c, BidQuantityCannotDecreaseError);
nameToErrorMap.set('BidQuantityCannotDecrease', BidQuantityCannotDecreaseError);
/** CannotCloseWinningBid: Cannot close winning bid during active auction */
class CannotCloseWinningBidError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Cannot close winning bid during active auction', program, cause);
        this.name = 'CannotCloseWinningBid';
        this.code = 0x7d; // 125
    }
}
exports.CannotCloseWinningBidError = CannotCloseWinningBidError;
codeToErrorMap.set(0x7d, CannotCloseWinningBidError);
nameToErrorMap.set('CannotCloseWinningBid', CannotCloseWinningBidError);
/** RefundAlreadyClaimed: Refund already claimed for this bid */
class RefundAlreadyClaimedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Refund already claimed for this bid', program, cause);
        this.name = 'RefundAlreadyClaimed';
        this.code = 0x7e; // 126
    }
}
exports.RefundAlreadyClaimedError = RefundAlreadyClaimedError;
codeToErrorMap.set(0x7e, RefundAlreadyClaimedError);
nameToErrorMap.set('RefundAlreadyClaimed', RefundAlreadyClaimedError);
/** BidNotWinning: Bid is not winning */
class BidNotWinningError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Bid is not winning', program, cause);
        this.name = 'BidNotWinning';
        this.code = 0x7f; // 127
    }
}
exports.BidNotWinningError = BidNotWinningError;
codeToErrorMap.set(0x7f, BidNotWinningError);
nameToErrorMap.set('BidNotWinning', BidNotWinningError);
/** LosingBidderCannotRebid: Losing bidders cannot rebid */
class LosingBidderCannotRebidError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Losing bidders cannot rebid', program, cause);
        this.name = 'LosingBidderCannotRebid';
        this.code = 0x80; // 128
    }
}
exports.LosingBidderCannotRebidError = LosingBidderCannotRebidError;
codeToErrorMap.set(0x80, LosingBidderCannotRebidError);
nameToErrorMap.set('LosingBidderCannotRebid', LosingBidderCannotRebidError);
/** InvalidAuctionCapacity: Auction capacity must be between minimum required by decibps and MAX_AUCTION_CAPACITY */
class InvalidAuctionCapacityError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Auction capacity must be between minimum required by decibps and MAX_AUCTION_CAPACITY', program, cause);
        this.name = 'InvalidAuctionCapacity';
        this.code = 0x81; // 129
    }
}
exports.InvalidAuctionCapacityError = InvalidAuctionCapacityError;
codeToErrorMap.set(0x81, InvalidAuctionCapacityError);
nameToErrorMap.set('InvalidAuctionCapacity', InvalidAuctionCapacityError);
/** TokensAlreadyClaimed: Tokens already claimed for this bid */
class TokensAlreadyClaimedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Tokens already claimed for this bid', program, cause);
        this.name = 'TokensAlreadyClaimed';
        this.code = 0x82; // 130
    }
}
exports.TokensAlreadyClaimedError = TokensAlreadyClaimedError;
codeToErrorMap.set(0x82, TokensAlreadyClaimedError);
nameToErrorMap.set('TokensAlreadyClaimed', TokensAlreadyClaimedError);
/** TokensNotClaimed: Tokens must be claimed before closing bid */
class TokensNotClaimedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Tokens must be claimed before closing bid', program, cause);
        this.name = 'TokensNotClaimed';
        this.code = 0x83; // 131
    }
}
exports.TokensNotClaimedError = TokensNotClaimedError;
codeToErrorMap.set(0x83, TokensNotClaimedError);
nameToErrorMap.set('TokensNotClaimed', TokensNotClaimedError);
/** RefundNotClaimed: Refund must be claimed before closing bid */
class RefundNotClaimedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Refund must be claimed before closing bid', program, cause);
        this.name = 'RefundNotClaimed';
        this.code = 0x84; // 132
    }
}
exports.RefundNotClaimedError = RefundNotClaimedError;
codeToErrorMap.set(0x84, RefundNotClaimedError);
nameToErrorMap.set('RefundNotClaimed', RefundNotClaimedError);
/** RaydiumBucketNotInitialized: Raydium bucket is not initialized */
class RaydiumBucketNotInitializedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Raydium bucket is not initialized', program, cause);
        this.name = 'RaydiumBucketNotInitialized';
        this.code = 0x85; // 133
    }
}
exports.RaydiumBucketNotInitializedError = RaydiumBucketNotInitializedError;
codeToErrorMap.set(0x85, RaydiumBucketNotInitializedError);
nameToErrorMap.set('RaydiumBucketNotInitialized', RaydiumBucketNotInitializedError);
/** MissingRaydiumBucketSigner: Missing Raydium bucket signer account */
class MissingRaydiumBucketSignerError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Missing Raydium bucket signer account', program, cause);
        this.name = 'MissingRaydiumBucketSigner';
        this.code = 0x86; // 134
    }
}
exports.MissingRaydiumBucketSignerError = MissingRaydiumBucketSignerError;
codeToErrorMap.set(0x86, MissingRaydiumBucketSignerError);
nameToErrorMap.set('MissingRaydiumBucketSigner', MissingRaydiumBucketSignerError);
/** RaydiumBucketNotFunded: Raydium bucket is not funded */
class RaydiumBucketNotFundedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Raydium bucket is not funded', program, cause);
        this.name = 'RaydiumBucketNotFunded';
        this.code = 0x87; // 135
    }
}
exports.RaydiumBucketNotFundedError = RaydiumBucketNotFundedError;
codeToErrorMap.set(0x87, RaydiumBucketNotFundedError);
nameToErrorMap.set('RaydiumBucketNotFunded', RaydiumBucketNotFundedError);
/** InvalidRaydiumBucketSigner: Invalid Raydium bucket signer */
class InvalidRaydiumBucketSignerError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Raydium bucket signer', program, cause);
        this.name = 'InvalidRaydiumBucketSigner';
        this.code = 0x88; // 136
    }
}
exports.InvalidRaydiumBucketSignerError = InvalidRaydiumBucketSignerError;
codeToErrorMap.set(0x88, InvalidRaydiumBucketSignerError);
nameToErrorMap.set('InvalidRaydiumBucketSigner', InvalidRaydiumBucketSignerError);
/** InvalidRaydiumProgram: Invalid Raydium program */
class InvalidRaydiumProgramError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Raydium program', program, cause);
        this.name = 'InvalidRaydiumProgram';
        this.code = 0x89; // 137
    }
}
exports.InvalidRaydiumProgramError = InvalidRaydiumProgramError;
codeToErrorMap.set(0x89, InvalidRaydiumProgramError);
nameToErrorMap.set('InvalidRaydiumProgram', InvalidRaydiumProgramError);
/** InvalidNativeSolMint: Invalid native SOL mint */
class InvalidNativeSolMintError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid native SOL mint', program, cause);
        this.name = 'InvalidNativeSolMint';
        this.code = 0x8a; // 138
    }
}
exports.InvalidNativeSolMintError = InvalidNativeSolMintError;
codeToErrorMap.set(0x8a, InvalidNativeSolMintError);
nameToErrorMap.set('InvalidNativeSolMint', InvalidNativeSolMintError);
/** InvalidQuoteMint: The quote mint is invalid */
class InvalidQuoteMintError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The quote mint is invalid', program, cause);
        this.name = 'InvalidQuoteMint';
        this.code = 0x8b; // 139
    }
}
exports.InvalidQuoteMintError = InvalidQuoteMintError;
codeToErrorMap.set(0x8b, InvalidQuoteMintError);
nameToErrorMap.set('InvalidQuoteMint', InvalidQuoteMintError);
/** InvalidQuoteMintForGenesisAccount: The quote mint does not match the Genesis Account's quote mint */
class InvalidQuoteMintForGenesisAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("The quote mint does not match the Genesis Account's quote mint", program, cause);
        this.name = 'InvalidQuoteMintForGenesisAccount';
        this.code = 0x8c; // 140
    }
}
exports.InvalidQuoteMintForGenesisAccountError = InvalidQuoteMintForGenesisAccountError;
codeToErrorMap.set(0x8c, InvalidQuoteMintForGenesisAccountError);
nameToErrorMap.set('InvalidQuoteMintForGenesisAccount', InvalidQuoteMintForGenesisAccountError);
/** InvalidGenesisAccountIndex: The Genesis Account index is invalid */
class InvalidGenesisAccountIndexError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Genesis Account index is invalid', program, cause);
        this.name = 'InvalidGenesisAccountIndex';
        this.code = 0x8d; // 141
    }
}
exports.InvalidGenesisAccountIndexError = InvalidGenesisAccountIndexError;
codeToErrorMap.set(0x8d, InvalidGenesisAccountIndexError);
nameToErrorMap.set('InvalidGenesisAccountIndex', InvalidGenesisAccountIndexError);
/** InvalidQuoteTokenAccountMint: The quote token account mint is invalid */
class InvalidQuoteTokenAccountMintError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The quote token account mint is invalid', program, cause);
        this.name = 'InvalidQuoteTokenAccountMint';
        this.code = 0x8e; // 142
    }
}
exports.InvalidQuoteTokenAccountMintError = InvalidQuoteTokenAccountMintError;
codeToErrorMap.set(0x8e, InvalidQuoteTokenAccountMintError);
nameToErrorMap.set('InvalidQuoteTokenAccountMint', InvalidQuoteTokenAccountMintError);
/** InvalidQuoteTokenAccountOwner: The quote token account owner is invalid */
class InvalidQuoteTokenAccountOwnerError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The quote token account owner is invalid', program, cause);
        this.name = 'InvalidQuoteTokenAccountOwner';
        this.code = 0x8f; // 143
    }
}
exports.InvalidQuoteTokenAccountOwnerError = InvalidQuoteTokenAccountOwnerError;
codeToErrorMap.set(0x8f, InvalidQuoteTokenAccountOwnerError);
nameToErrorMap.set('InvalidQuoteTokenAccountOwner', InvalidQuoteTokenAccountOwnerError);
/** InvalidDepositPdaAccount: The deposit PDA account is invalid */
class InvalidDepositPdaAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The deposit PDA account is invalid', program, cause);
        this.name = 'InvalidDepositPdaAccount';
        this.code = 0x90; // 144
    }
}
exports.InvalidDepositPdaAccountError = InvalidDepositPdaAccountError;
codeToErrorMap.set(0x90, InvalidDepositPdaAccountError);
nameToErrorMap.set('InvalidDepositPdaAccount', InvalidDepositPdaAccountError);
/** MissingBackendSigner: The backend signer is missing */
class MissingBackendSignerError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The backend signer is missing', program, cause);
        this.name = 'MissingBackendSigner';
        this.code = 0x91; // 145
    }
}
exports.MissingBackendSignerError = MissingBackendSignerError;
codeToErrorMap.set(0x91, MissingBackendSignerError);
nameToErrorMap.set('MissingBackendSigner', MissingBackendSignerError);
/** PresaleClaimEnded: The Presale claim has ended */
class PresaleClaimEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Presale claim has ended', program, cause);
        this.name = 'PresaleClaimEnded';
        this.code = 0x92; // 146
    }
}
exports.PresaleClaimEndedError = PresaleClaimEndedError;
codeToErrorMap.set(0x92, PresaleClaimEndedError);
nameToErrorMap.set('PresaleClaimEnded', PresaleClaimEndedError);
/** DestinationBucketNotFound: The destination bucket was not found */
class DestinationBucketNotFoundError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The destination bucket was not found', program, cause);
        this.name = 'DestinationBucketNotFound';
        this.code = 0x93; // 147
    }
}
exports.DestinationBucketNotFoundError = DestinationBucketNotFoundError;
codeToErrorMap.set(0x93, DestinationBucketNotFoundError);
nameToErrorMap.set('DestinationBucketNotFound', DestinationBucketNotFoundError);
/** InvalidPrimaryBucketOwner: The Primary Bucket's owner is invalid */
class InvalidPrimaryBucketOwnerError extends umi_1.ProgramError {
    constructor(program, cause) {
        super("The Primary Bucket's owner is invalid", program, cause);
        this.name = 'InvalidPrimaryBucketOwner';
        this.code = 0x94; // 148
    }
}
exports.InvalidPrimaryBucketOwnerError = InvalidPrimaryBucketOwnerError;
codeToErrorMap.set(0x94, InvalidPrimaryBucketOwnerError);
nameToErrorMap.set('InvalidPrimaryBucketOwner', InvalidPrimaryBucketOwnerError);
/** InvalidPrimaryBucket: The Primary Bucket is invalid */
class InvalidPrimaryBucketError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Primary Bucket is invalid', program, cause);
        this.name = 'InvalidPrimaryBucket';
        this.code = 0x95; // 149
    }
}
exports.InvalidPrimaryBucketError = InvalidPrimaryBucketError;
codeToErrorMap.set(0x95, InvalidPrimaryBucketError);
nameToErrorMap.set('InvalidPrimaryBucket', InvalidPrimaryBucketError);
/** InvalidSecondaryBucket: The Secondary Bucket is invalid */
class InvalidSecondaryBucketError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Secondary Bucket is invalid', program, cause);
        this.name = 'InvalidSecondaryBucket';
        this.code = 0x96; // 150
    }
}
exports.InvalidSecondaryBucketError = InvalidSecondaryBucketError;
codeToErrorMap.set(0x96, InvalidSecondaryBucketError);
nameToErrorMap.set('InvalidSecondaryBucket', InvalidSecondaryBucketError);
/** InvalidFundingMode: Invalid funding mode; expected 0 (Mint) or 1 (Transfer) */
class InvalidFundingModeError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid funding mode; expected 0 (Mint) or 1 (Transfer)', program, cause);
        this.name = 'InvalidFundingMode';
        this.code = 0x97; // 151
    }
}
exports.InvalidFundingModeError = InvalidFundingModeError;
codeToErrorMap.set(0x97, InvalidFundingModeError);
nameToErrorMap.set('InvalidFundingMode', InvalidFundingModeError);
/** UnlockedBucketClaimEnded: The Unlocked Bucket claim has ended */
class UnlockedBucketClaimEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Unlocked Bucket claim has ended', program, cause);
        this.name = 'UnlockedBucketClaimEnded';
        this.code = 0x98; // 152
    }
}
exports.UnlockedBucketClaimEndedError = UnlockedBucketClaimEndedError;
codeToErrorMap.set(0x98, UnlockedBucketClaimEndedError);
nameToErrorMap.set('UnlockedBucketClaimEnded', UnlockedBucketClaimEndedError);
/** MissingAuthoritySourceTokenAccount: The authority source token account is missing */
class MissingAuthoritySourceTokenAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The authority source token account is missing', program, cause);
        this.name = 'MissingAuthoritySourceTokenAccount';
        this.code = 0x99; // 153
    }
}
exports.MissingAuthoritySourceTokenAccountError = MissingAuthoritySourceTokenAccountError;
codeToErrorMap.set(0x99, MissingAuthoritySourceTokenAccountError);
nameToErrorMap.set('MissingAuthoritySourceTokenAccount', MissingAuthoritySourceTokenAccountError);
/** InvalidAuthoritySourceTokenAccount: The authority source token account is invalid */
class InvalidAuthoritySourceTokenAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The authority source token account is invalid', program, cause);
        this.name = 'InvalidAuthoritySourceTokenAccount';
        this.code = 0x9a; // 154
    }
}
exports.InvalidAuthoritySourceTokenAccountError = InvalidAuthoritySourceTokenAccountError;
codeToErrorMap.set(0x9a, InvalidAuthoritySourceTokenAccountError);
nameToErrorMap.set('InvalidAuthoritySourceTokenAccount', InvalidAuthoritySourceTokenAccountError);
/** InvalidPresaleDeposit: The Presale Deposit is invalid */
class InvalidPresaleDepositError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Presale Deposit is invalid', program, cause);
        this.name = 'InvalidPresaleDeposit';
        this.code = 0x9b; // 155
    }
}
exports.InvalidPresaleDepositError = InvalidPresaleDepositError;
codeToErrorMap.set(0x9b, InvalidPresaleDepositError);
nameToErrorMap.set('InvalidPresaleDeposit', InvalidPresaleDepositError);
/** InvalidClaimSchedulePeriod: The claim schedule period is invalid */
class InvalidClaimSchedulePeriodError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The claim schedule period is invalid', program, cause);
        this.name = 'InvalidClaimSchedulePeriod';
        this.code = 0x9c; // 156
    }
}
exports.InvalidClaimSchedulePeriodError = InvalidClaimSchedulePeriodError;
codeToErrorMap.set(0x9c, InvalidClaimSchedulePeriodError);
nameToErrorMap.set('InvalidClaimSchedulePeriod', InvalidClaimSchedulePeriodError);
/** InvalidClaimScheduleDuration: The claim schedule duration must be greater than zero */
class InvalidClaimScheduleDurationError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The claim schedule duration must be greater than zero', program, cause);
        this.name = 'InvalidClaimScheduleDuration';
        this.code = 0x9d; // 157
    }
}
exports.InvalidClaimScheduleDurationError = InvalidClaimScheduleDurationError;
codeToErrorMap.set(0x9d, InvalidClaimScheduleDurationError);
nameToErrorMap.set('InvalidClaimScheduleDuration', InvalidClaimScheduleDurationError);
/** InvalidClaimScheduleCliffAmount: The claim schedule cliff amount is invalid */
class InvalidClaimScheduleCliffAmountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The claim schedule cliff amount is invalid', program, cause);
        this.name = 'InvalidClaimScheduleCliffAmount';
        this.code = 0x9e; // 158
    }
}
exports.InvalidClaimScheduleCliffAmountError = InvalidClaimScheduleCliffAmountError;
codeToErrorMap.set(0x9e, InvalidClaimScheduleCliffAmountError);
nameToErrorMap.set('InvalidClaimScheduleCliffAmount', InvalidClaimScheduleCliffAmountError);
/** ClaimScheduleDurationTooShort: The claim schedule duration is too short; it must be longer than the claim schedule period */
class ClaimScheduleDurationTooShortError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The claim schedule duration is too short; it must be longer than the claim schedule period', program, cause);
        this.name = 'ClaimScheduleDurationTooShort';
        this.code = 0x9f; // 159
    }
}
exports.ClaimScheduleDurationTooShortError = ClaimScheduleDurationTooShortError;
codeToErrorMap.set(0x9f, ClaimScheduleDurationTooShortError);
nameToErrorMap.set('ClaimScheduleDurationTooShort', ClaimScheduleDurationTooShortError);
/** SaleNotEnded: The bucket sale has not ended */
class SaleNotEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The bucket sale has not ended', program, cause);
        this.name = 'SaleNotEnded';
        this.code = 0xa0; // 160
    }
}
exports.SaleNotEndedError = SaleNotEndedError;
codeToErrorMap.set(0xa0, SaleNotEndedError);
nameToErrorMap.set('SaleNotEnded', SaleNotEndedError);
/** NoUnsoldTokens: There are no unsold tokens available to withdraw */
class NoUnsoldTokensError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('There are no unsold tokens available to withdraw', program, cause);
        this.name = 'NoUnsoldTokens';
        this.code = 0xa1; // 161
    }
}
exports.NoUnsoldTokensError = NoUnsoldTokensError;
codeToErrorMap.set(0xa1, NoUnsoldTokensError);
nameToErrorMap.set('NoUnsoldTokens', NoUnsoldTokensError);
/** InvalidUnsoldWithdrawalAmount: The unsold withdrawal amount is invalid */
class InvalidUnsoldWithdrawalAmountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The unsold withdrawal amount is invalid', program, cause);
        this.name = 'InvalidUnsoldWithdrawalAmount';
        this.code = 0xa2; // 162
    }
}
exports.InvalidUnsoldWithdrawalAmountError = InvalidUnsoldWithdrawalAmountError;
codeToErrorMap.set(0xa2, InvalidUnsoldWithdrawalAmountError);
nameToErrorMap.set('InvalidUnsoldWithdrawalAmount', InvalidUnsoldWithdrawalAmountError);
/** BucketDoesNotSupportUnsoldWithdrawal: This bucket does not support unsold token withdrawals */
class BucketDoesNotSupportUnsoldWithdrawalError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('This bucket does not support unsold token withdrawals', program, cause);
        this.name = 'BucketDoesNotSupportUnsoldWithdrawal';
        this.code = 0xa3; // 163
    }
}
exports.BucketDoesNotSupportUnsoldWithdrawalError = BucketDoesNotSupportUnsoldWithdrawalError;
codeToErrorMap.set(0xa3, BucketDoesNotSupportUnsoldWithdrawalError);
nameToErrorMap.set('BucketDoesNotSupportUnsoldWithdrawal', BucketDoesNotSupportUnsoldWithdrawalError);
/** PresaleNotFullyClaimed: All presale deposits must be claimed before withdrawing unsold tokens */
class PresaleNotFullyClaimedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('All presale deposits must be claimed before withdrawing unsold tokens', program, cause);
        this.name = 'PresaleNotFullyClaimed';
        this.code = 0xa4; // 164
    }
}
exports.PresaleNotFullyClaimedError = PresaleNotFullyClaimedError;
codeToErrorMap.set(0xa4, PresaleNotFullyClaimedError);
nameToErrorMap.set('PresaleNotFullyClaimed', PresaleNotFullyClaimedError);
/** SourceBucketNotFound: Source Bucket Not Found */
class SourceBucketNotFoundError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Source Bucket Not Found', program, cause);
        this.name = 'SourceBucketNotFound';
        this.code = 0xa5; // 165
    }
}
exports.SourceBucketNotFoundError = SourceBucketNotFoundError;
codeToErrorMap.set(0xa5, SourceBucketNotFoundError);
nameToErrorMap.set('SourceBucketNotFound', SourceBucketNotFoundError);
/** InvalidMinimumDepositAmount: The minimum deposit amount must be greater than zero */
class InvalidMinimumDepositAmountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The minimum deposit amount must be greater than zero', program, cause);
        this.name = 'InvalidMinimumDepositAmount';
        this.code = 0xa6; // 166
    }
}
exports.InvalidMinimumDepositAmountError = InvalidMinimumDepositAmountError;
codeToErrorMap.set(0xa6, InvalidMinimumDepositAmountError);
nameToErrorMap.set('InvalidMinimumDepositAmount', InvalidMinimumDepositAmountError);
/** PresaleDepositBelowMinimum: The deposit amount is below the minimum allowed */
class PresaleDepositBelowMinimumError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The deposit amount is below the minimum allowed', program, cause);
        this.name = 'PresaleDepositBelowMinimum';
        this.code = 0xa7; // 167
    }
}
exports.PresaleDepositBelowMinimumError = PresaleDepositBelowMinimumError;
codeToErrorMap.set(0xa7, PresaleDepositBelowMinimumError);
nameToErrorMap.set('PresaleDepositBelowMinimum', PresaleDepositBelowMinimumError);
/** BehaviorPercentageOverflow: Total SendQuoteTokenPercentage basis points cannot exceed 10000 */
class BehaviorPercentageOverflowError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Total SendQuoteTokenPercentage basis points cannot exceed 10000', program, cause);
        this.name = 'BehaviorPercentageOverflow';
        this.code = 0xa8; // 168
    }
}
exports.BehaviorPercentageOverflowError = BehaviorPercentageOverflowError;
codeToErrorMap.set(0xa8, BehaviorPercentageOverflowError);
nameToErrorMap.set('BehaviorPercentageOverflow', BehaviorPercentageOverflowError);
/** InvalidTokenAccountProgramOwner: The token account is not owned by the SPL Token program */
class InvalidTokenAccountProgramOwnerError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The token account is not owned by the SPL Token program', program, cause);
        this.name = 'InvalidTokenAccountProgramOwner';
        this.code = 0xa9; // 169
    }
}
exports.InvalidTokenAccountProgramOwnerError = InvalidTokenAccountProgramOwnerError;
codeToErrorMap.set(0xa9, InvalidTokenAccountProgramOwnerError);
nameToErrorMap.set('InvalidTokenAccountProgramOwner', InvalidTokenAccountProgramOwnerError);
/** InvalidMetadataAccount: The metadata account is invalid */
class InvalidMetadataAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The metadata account is invalid', program, cause);
        this.name = 'InvalidMetadataAccount';
        this.code = 0xaa; // 170
    }
}
exports.InvalidMetadataAccountError = InvalidMetadataAccountError;
codeToErrorMap.set(0xaa, InvalidMetadataAccountError);
nameToErrorMap.set('InvalidMetadataAccount', InvalidMetadataAccountError);
/** LaunchPoolClaimEnded: The LaunchPool claim has ended */
class LaunchPoolClaimEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The LaunchPool claim has ended', program, cause);
        this.name = 'LaunchPoolClaimEnded';
        this.code = 0xab; // 171
    }
}
exports.LaunchPoolClaimEndedError = LaunchPoolClaimEndedError;
codeToErrorMap.set(0xab, LaunchPoolClaimEndedError);
nameToErrorMap.set('LaunchPoolClaimEnded', LaunchPoolClaimEndedError);
/** LaunchPoolFundingThresholdNotMet: The launchpool funding threshold was not met */
class LaunchPoolFundingThresholdNotMetError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The launchpool funding threshold was not met', program, cause);
        this.name = 'LaunchPoolFundingThresholdNotMet';
        this.code = 0xac; // 172
    }
}
exports.LaunchPoolFundingThresholdNotMetError = LaunchPoolFundingThresholdNotMetError;
codeToErrorMap.set(0xac, LaunchPoolFundingThresholdNotMetError);
nameToErrorMap.set('LaunchPoolFundingThresholdNotMet', LaunchPoolFundingThresholdNotMetError);
/** LaunchPoolThresholdMet: The launchpool threshold was met, refunds are not allowed */
class LaunchPoolThresholdMetError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The launchpool threshold was met, refunds are not allowed', program, cause);
        this.name = 'LaunchPoolThresholdMet';
        this.code = 0xad; // 173
    }
}
exports.LaunchPoolThresholdMetError = LaunchPoolThresholdMetError;
codeToErrorMap.set(0xad, LaunchPoolThresholdMetError);
nameToErrorMap.set('LaunchPoolThresholdMet', LaunchPoolThresholdMetError);
/** InvalidMinimumQuoteTokenThreshold: The minimum quote token threshold must be greater than zero */
class InvalidMinimumQuoteTokenThresholdError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The minimum quote token threshold must be greater than zero', program, cause);
        this.name = 'InvalidMinimumQuoteTokenThreshold';
        this.code = 0xae; // 174
    }
}
exports.InvalidMinimumQuoteTokenThresholdError = InvalidMinimumQuoteTokenThresholdError;
codeToErrorMap.set(0xae, InvalidMinimumQuoteTokenThresholdError);
nameToErrorMap.set('InvalidMinimumQuoteTokenThreshold', InvalidMinimumQuoteTokenThresholdError);
/** BehaviorNotSupportedForBucketType: This behavior is not supported for this bucket type */
class BehaviorNotSupportedForBucketTypeError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('This behavior is not supported for this bucket type', program, cause);
        this.name = 'BehaviorNotSupportedForBucketType';
        this.code = 0xaf; // 175
    }
}
exports.BehaviorNotSupportedForBucketTypeError = BehaviorNotSupportedForBucketTypeError;
codeToErrorMap.set(0xaf, BehaviorNotSupportedForBucketTypeError);
nameToErrorMap.set('BehaviorNotSupportedForBucketType', BehaviorNotSupportedForBucketTypeError);
/** MissingPenaltyQuoteTokenAccount: Penalty quote token account is required when penalties are enabled */
class MissingPenaltyQuoteTokenAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Penalty quote token account is required when penalties are enabled', program, cause);
        this.name = 'MissingPenaltyQuoteTokenAccount';
        this.code = 0xb0; // 176
    }
}
exports.MissingPenaltyQuoteTokenAccountError = MissingPenaltyQuoteTokenAccountError;
codeToErrorMap.set(0xb0, MissingPenaltyQuoteTokenAccountError);
nameToErrorMap.set('MissingPenaltyQuoteTokenAccount', MissingPenaltyQuoteTokenAccountError);
/** DepositAlreadyRefunded: Deposit has already been refunded */
class DepositAlreadyRefundedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Deposit has already been refunded', program, cause);
        this.name = 'DepositAlreadyRefunded';
        this.code = 0xb1; // 177
    }
}
exports.DepositAlreadyRefundedError = DepositAlreadyRefundedError;
codeToErrorMap.set(0xb1, DepositAlreadyRefundedError);
nameToErrorMap.set('DepositAlreadyRefunded', DepositAlreadyRefundedError);
/** InvalidBaseTokenAllocation: The base token allocation must be greater than zero */
class InvalidBaseTokenAllocationError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The base token allocation must be greater than zero', program, cause);
        this.name = 'InvalidBaseTokenAllocation';
        this.code = 0xb2; // 178
    }
}
exports.InvalidBaseTokenAllocationError = InvalidBaseTokenAllocationError;
codeToErrorMap.set(0xb2, InvalidBaseTokenAllocationError);
nameToErrorMap.set('InvalidBaseTokenAllocation', InvalidBaseTokenAllocationError);
/** InvalidAllocationQuoteTokenCap: The allocation quote token cap must be greater than zero */
class InvalidAllocationQuoteTokenCapError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The allocation quote token cap must be greater than zero', program, cause);
        this.name = 'InvalidAllocationQuoteTokenCap';
        this.code = 0xb3; // 179
    }
}
exports.InvalidAllocationQuoteTokenCapError = InvalidAllocationQuoteTokenCapError;
codeToErrorMap.set(0xb3, InvalidAllocationQuoteTokenCapError);
nameToErrorMap.set('InvalidAllocationQuoteTokenCap', InvalidAllocationQuoteTokenCapError);
/** InvalidRaydiumSigner: Invalid Raydium signer PDA */
class InvalidRaydiumSignerError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Raydium signer PDA', program, cause);
        this.name = 'InvalidRaydiumSigner';
        this.code = 0xb4; // 180
    }
}
exports.InvalidRaydiumSignerError = InvalidRaydiumSignerError;
codeToErrorMap.set(0xb4, InvalidRaydiumSignerError);
nameToErrorMap.set('InvalidRaydiumSigner', InvalidRaydiumSignerError);
/** InvalidRaydiumPermissionPda: Invalid Raydium Permission PDA */
class InvalidRaydiumPermissionPdaError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Raydium Permission PDA', program, cause);
        this.name = 'InvalidRaydiumPermissionPda';
        this.code = 0xb5; // 181
    }
}
exports.InvalidRaydiumPermissionPdaError = InvalidRaydiumPermissionPdaError;
codeToErrorMap.set(0xb5, InvalidRaydiumPermissionPdaError);
nameToErrorMap.set('InvalidRaydiumPermissionPda', InvalidRaydiumPermissionPdaError);
/** LaunchPoolMissingGraduationBehavior: The LaunchPool bucket is missing a required graduation behavior */
class LaunchPoolMissingGraduationBehaviorError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The LaunchPool bucket is missing a required graduation behavior', program, cause);
        this.name = 'LaunchPoolMissingGraduationBehavior';
        this.code = 0xb6; // 182
    }
}
exports.LaunchPoolMissingGraduationBehaviorError = LaunchPoolMissingGraduationBehaviorError;
codeToErrorMap.set(0xb6, LaunchPoolMissingGraduationBehaviorError);
nameToErrorMap.set('LaunchPoolMissingGraduationBehavior', LaunchPoolMissingGraduationBehaviorError);
/** MissingFinalizeBucketAccounts: Missing bucket accounts for finalize */
class MissingFinalizeBucketAccountsError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Missing bucket accounts for finalize', program, cause);
        this.name = 'MissingFinalizeBucketAccounts';
        this.code = 0xb7; // 183
    }
}
exports.MissingFinalizeBucketAccountsError = MissingFinalizeBucketAccountsError;
codeToErrorMap.set(0xb7, MissingFinalizeBucketAccountsError);
nameToErrorMap.set('MissingFinalizeBucketAccounts', MissingFinalizeBucketAccountsError);
/** StreamflowBucketAlreadyLocked: Streamflow bucket has already been locked */
class StreamflowBucketAlreadyLockedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Streamflow bucket has already been locked', program, cause);
        this.name = 'StreamflowBucketAlreadyLocked';
        this.code = 0xb8; // 184
    }
}
exports.StreamflowBucketAlreadyLockedError = StreamflowBucketAlreadyLockedError;
codeToErrorMap.set(0xb8, StreamflowBucketAlreadyLockedError);
nameToErrorMap.set('StreamflowBucketAlreadyLocked', StreamflowBucketAlreadyLockedError);
/** LockEnded: The lock period has ended */
class LockEndedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The lock period has ended', program, cause);
        this.name = 'LockEnded';
        this.code = 0xb9; // 185
    }
}
exports.LockEndedError = LockEndedError;
codeToErrorMap.set(0xb9, LockEndedError);
nameToErrorMap.set('LockEnded', LockEndedError);
/** InvalidStreamflowConfig: The streamflow config is invalid */
class InvalidStreamflowConfigError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The streamflow config is invalid', program, cause);
        this.name = 'InvalidStreamflowConfig';
        this.code = 0xba; // 186
    }
}
exports.InvalidStreamflowConfigError = InvalidStreamflowConfigError;
codeToErrorMap.set(0xba, InvalidStreamflowConfigError);
nameToErrorMap.set('InvalidStreamflowConfig', InvalidStreamflowConfigError);
/** TotalSupplyNotFullyAllocated: Total supply must be fully allocated before finalize */
class TotalSupplyNotFullyAllocatedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Total supply must be fully allocated before finalize', program, cause);
        this.name = 'TotalSupplyNotFullyAllocated';
        this.code = 0xbb; // 187
    }
}
exports.TotalSupplyNotFullyAllocatedError = TotalSupplyNotFullyAllocatedError;
codeToErrorMap.set(0xbb, TotalSupplyNotFullyAllocatedError);
nameToErrorMap.set('TotalSupplyNotFullyAllocated', TotalSupplyNotFullyAllocatedError);
/** BucketAlreadyGraduated: The bucket has already graduated */
class BucketAlreadyGraduatedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The bucket has already graduated', program, cause);
        this.name = 'BucketAlreadyGraduated';
        this.code = 0xbc; // 188
    }
}
exports.BucketAlreadyGraduatedError = BucketAlreadyGraduatedError;
codeToErrorMap.set(0xbc, BucketAlreadyGraduatedError);
nameToErrorMap.set('BucketAlreadyGraduated', BucketAlreadyGraduatedError);
/** GraduationNotStarted: The graduation period has not started */
class GraduationNotStartedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The graduation period has not started', program, cause);
        this.name = 'GraduationNotStarted';
        this.code = 0xbd; // 189
    }
}
exports.GraduationNotStartedError = GraduationNotStartedError;
codeToErrorMap.set(0xbd, GraduationNotStartedError);
nameToErrorMap.set('GraduationNotStarted', GraduationNotStartedError);
/** InvalidStreamflowMetadataAccount: Invalid Streamflow metadata account */
class InvalidStreamflowMetadataAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Streamflow metadata account', program, cause);
        this.name = 'InvalidStreamflowMetadataAccount';
        this.code = 0xbe; // 190
    }
}
exports.InvalidStreamflowMetadataAccountError = InvalidStreamflowMetadataAccountError;
codeToErrorMap.set(0xbe, InvalidStreamflowMetadataAccountError);
nameToErrorMap.set('InvalidStreamflowMetadataAccount', InvalidStreamflowMetadataAccountError);
/** InvalidSupplyDelta: Invalid supply delta */
class InvalidSupplyDeltaError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid supply delta', program, cause);
        this.name = 'InvalidSupplyDelta';
        this.code = 0xbf; // 191
    }
}
exports.InvalidSupplyDeltaError = InvalidSupplyDeltaError;
codeToErrorMap.set(0xbf, InvalidSupplyDeltaError);
nameToErrorMap.set('InvalidSupplyDelta', InvalidSupplyDeltaError);
/** InvalidMintAuthority: Invalid mint authority */
class InvalidMintAuthorityError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid mint authority', program, cause);
        this.name = 'InvalidMintAuthority';
        this.code = 0xc0; // 192
    }
}
exports.InvalidMintAuthorityError = InvalidMintAuthorityError;
codeToErrorMap.set(0xc0, InvalidMintAuthorityError);
nameToErrorMap.set('InvalidMintAuthority', InvalidMintAuthorityError);
/** InvalidMintAndBurn: Cannot mint and burn in the same instruction */
class InvalidMintAndBurnError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Cannot mint and burn in the same instruction', program, cause);
        this.name = 'InvalidMintAndBurn';
        this.code = 0xc1; // 193
    }
}
exports.InvalidMintAndBurnError = InvalidMintAndBurnError;
codeToErrorMap.set(0xc1, InvalidMintAndBurnError);
nameToErrorMap.set('InvalidMintAndBurn', InvalidMintAndBurnError);
/** SupplyBelowAllocated: Total allocated supply exceeds total supply */
class SupplyBelowAllocatedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Total allocated supply exceeds total supply', program, cause);
        this.name = 'SupplyBelowAllocated';
        this.code = 0xc2; // 194
    }
}
exports.SupplyBelowAllocatedError = SupplyBelowAllocatedError;
codeToErrorMap.set(0xc2, SupplyBelowAllocatedError);
nameToErrorMap.set('SupplyBelowAllocated', SupplyBelowAllocatedError);
/** MissingAuthorityDestinationTokenAccount: The authority destination token account is missing */
class MissingAuthorityDestinationTokenAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The authority destination token account is missing', program, cause);
        this.name = 'MissingAuthorityDestinationTokenAccount';
        this.code = 0xc3; // 195
    }
}
exports.MissingAuthorityDestinationTokenAccountError = MissingAuthorityDestinationTokenAccountError;
codeToErrorMap.set(0xc3, MissingAuthorityDestinationTokenAccountError);
nameToErrorMap.set('MissingAuthorityDestinationTokenAccount', MissingAuthorityDestinationTokenAccountError);
/** InvalidAuthorityDestinationTokenAccount: The authority destination token account is invalid */
class InvalidAuthorityDestinationTokenAccountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The authority destination token account is invalid', program, cause);
        this.name = 'InvalidAuthorityDestinationTokenAccount';
        this.code = 0xc4; // 196
    }
}
exports.InvalidAuthorityDestinationTokenAccountError = InvalidAuthorityDestinationTokenAccountError;
codeToErrorMap.set(0xc4, InvalidAuthorityDestinationTokenAccountError);
nameToErrorMap.set('InvalidAuthorityDestinationTokenAccount', InvalidAuthorityDestinationTokenAccountError);
/** InvalidConditionType: The condition type is invalid */
class InvalidConditionTypeError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The condition type is invalid', program, cause);
        this.name = 'InvalidConditionType';
        this.code = 0xc5; // 197
    }
}
exports.InvalidConditionTypeError = InvalidConditionTypeError;
codeToErrorMap.set(0xc5, InvalidConditionTypeError);
nameToErrorMap.set('InvalidConditionType', InvalidConditionTypeError);
/** InvalidRemainingAccountsLength: Remaining accounts must be provided in pairs (bucket, quote_token_account) */
class InvalidRemainingAccountsLengthError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Remaining accounts must be provided in pairs (bucket, quote_token_account)', program, cause);
        this.name = 'InvalidRemainingAccountsLength';
        this.code = 0xc6; // 198
    }
}
exports.InvalidRemainingAccountsLengthError = InvalidRemainingAccountsLengthError;
codeToErrorMap.set(0xc6, InvalidRemainingAccountsLengthError);
nameToErrorMap.set('InvalidRemainingAccountsLength', InvalidRemainingAccountsLengthError);
/** InsufficientUnlockedAmount: The requested amount exceeds the unlocked amount */
class InsufficientUnlockedAmountError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The requested amount exceeds the unlocked amount', program, cause);
        this.name = 'InsufficientUnlockedAmount';
        this.code = 0xc7; // 199
    }
}
exports.InsufficientUnlockedAmountError = InsufficientUnlockedAmountError;
codeToErrorMap.set(0xc7, InsufficientUnlockedAmountError);
nameToErrorMap.set('InsufficientUnlockedAmount', InsufficientUnlockedAmountError);
/** NoLpTokensToClaim: The bucket has no LP tokens to claim (not graduated) */
class NoLpTokensToClaimError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The bucket has no LP tokens to claim (not graduated)', program, cause);
        this.name = 'NoLpTokensToClaim';
        this.code = 0xc8; // 200
    }
}
exports.NoLpTokensToClaimError = NoLpTokensToClaimError;
codeToErrorMap.set(0xc8, NoLpTokensToClaimError);
nameToErrorMap.set('NoLpTokensToClaim', NoLpTokensToClaimError);
/** InvalidLpClaimAuthority: The LP claim authority is invalid or missing */
class InvalidLpClaimAuthorityError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The LP claim authority is invalid or missing', program, cause);
        this.name = 'InvalidLpClaimAuthority';
        this.code = 0xc9; // 201
    }
}
exports.InvalidLpClaimAuthorityError = InvalidLpClaimAuthorityError;
codeToErrorMap.set(0xc9, InvalidLpClaimAuthorityError);
nameToErrorMap.set('InvalidLpClaimAuthority', InvalidLpClaimAuthorityError);
/** InvalidRaydiumAmmConfig: The Raydium AMM config must be one of the approved configs */
class InvalidRaydiumAmmConfigError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The Raydium AMM config must be one of the approved configs', program, cause);
        this.name = 'InvalidRaydiumAmmConfig';
        this.code = 0xca; // 202
    }
}
exports.InvalidRaydiumAmmConfigError = InvalidRaydiumAmmConfigError;
codeToErrorMap.set(0xca, InvalidRaydiumAmmConfigError);
nameToErrorMap.set('InvalidRaydiumAmmConfig', InvalidRaydiumAmmConfigError);
/** UninitializedCondition: Condition is uninitialized — all conditions must be explicitly set */
class UninitializedConditionError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Condition is uninitialized — all conditions must be explicitly set', program, cause);
        this.name = 'UninitializedCondition';
        this.code = 0xcb; // 203
    }
}
exports.UninitializedConditionError = UninitializedConditionError;
codeToErrorMap.set(0xcb, UninitializedConditionError);
nameToErrorMap.set('UninitializedCondition', UninitializedConditionError);
/** InvalidConditionTypeForSlot: The condition type is not valid for this slot */
class InvalidConditionTypeForSlotError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('The condition type is not valid for this slot', program, cause);
        this.name = 'InvalidConditionTypeForSlot';
        this.code = 0xcc; // 204
    }
}
exports.InvalidConditionTypeForSlotError = InvalidConditionTypeForSlotError;
codeToErrorMap.set(0xcc, InvalidConditionTypeForSlotError);
nameToErrorMap.set('InvalidConditionTypeForSlot', InvalidConditionTypeForSlotError);
/** InstructionDeprecated: This instruction has been deprecated and is no longer supported */
class InstructionDeprecatedError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('This instruction has been deprecated and is no longer supported', program, cause);
        this.name = 'InstructionDeprecated';
        this.code = 0xcd; // 205
    }
}
exports.InstructionDeprecatedError = InstructionDeprecatedError;
codeToErrorMap.set(0xcd, InstructionDeprecatedError);
nameToErrorMap.set('InstructionDeprecated', InstructionDeprecatedError);
/** InvalidRaydiumPoolState: Invalid Raydium pool state account data */
class InvalidRaydiumPoolStateError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid Raydium pool state account data', program, cause);
        this.name = 'InvalidRaydiumPoolState';
        this.code = 0xce; // 206
    }
}
exports.InvalidRaydiumPoolStateError = InvalidRaydiumPoolStateError;
codeToErrorMap.set(0xce, InvalidRaydiumPoolStateError);
nameToErrorMap.set('InvalidRaydiumPoolState', InvalidRaydiumPoolStateError);
/** InvalidLaunchType: Invalid launch type; expected 0 (Project) or 1 (Meme) */
class InvalidLaunchTypeError extends umi_1.ProgramError {
    constructor(program, cause) {
        super('Invalid launch type; expected 0 (Project) or 1 (Meme)', program, cause);
        this.name = 'InvalidLaunchType';
        this.code = 0xcf; // 207
    }
}
exports.InvalidLaunchTypeError = InvalidLaunchTypeError;
codeToErrorMap.set(0xcf, InvalidLaunchTypeError);
nameToErrorMap.set('InvalidLaunchType', InvalidLaunchTypeError);
/**
 * Attempts to resolve a custom program error from the provided error code.
 * @category Errors
 */
function getGenesisErrorFromCode(code, program, cause) {
    const constructor = codeToErrorMap.get(code);
    return constructor ? new constructor(program, cause) : null;
}
exports.getGenesisErrorFromCode = getGenesisErrorFromCode;
/**
 * Attempts to resolve a custom program error from the provided error name, i.e. 'Unauthorized'.
 * @category Errors
 */
function getGenesisErrorFromName(name, program, cause) {
    const constructor = nameToErrorMap.get(name);
    return constructor ? new constructor(program, cause) : null;
}
exports.getGenesisErrorFromName = getGenesisErrorFromName;
