// @metaplex-foundation/genesis/dist/src/api/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createAndRegisterLaunch = exports.registerLaunch = exports.createLaunch = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const payloadBuilder_1 = require("./payloadBuilder");
const transactionHelper_1 = require("./transactionHelper");
const errors_1 = require("./errors");
const DEFAULT_BASE_URL = 'https://api.metaplex.com';
function resolveBaseUrl(config) {
    return config.baseUrl ?? DEFAULT_BASE_URL;
}
function toKeyString(key) {
    return typeof key === 'string' ? key : key.toString();
}
/**
 * Creates a project launch via the Genesis API.
 *
 * Returns deserialized Umi transactions that can be signed and sent
 * using standard Umi methods before calling {@link registerLaunch}.
 *
 * @example
 * ```ts
 * const result = await createLaunch({}, umi, {
 *   wallet: umi.identity.publicKey,
 *   token: { name: 'My Token', symbol: 'MTK', image: 'https://gateway.irys.xyz/...' },
 *   launchpool: {
 *     tokenAllocation: 500_000_000,
 *     depositStartTime: new Date('2026-03-01T00:00:00Z'),
 *     raiseGoal: 250,
 *     raydiumLiquidityBps: 5000,
 *     fundsRecipient: 'RecipientWallet...',
 *   },
 * });
 *
 * // Sign and send using Umi directly
 * for (const tx of result.transactions) {
 *   const signed = await umi.identity.signTransaction(tx);
 *   await umi.rpc.sendTransaction(signed);
 * }
 * ```
 */
async function createLaunch(umi, config, input) {
    const payload = (0, payloadBuilder_1.buildCreateLaunchPayload)(input);
    const fetchFn = config.fetch ?? globalThis.fetch;
    const baseUrl = resolveBaseUrl(config);
    let response;
    try {
        response = await fetchFn(`${baseUrl}/v1/launches/create`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload),
        });
    }
    catch (err) {
        throw (0, errors_1.genesisApiNetworkError)(`Network error calling createLaunch: ${err.message}`, err);
    }
    let body;
    try {
        body = await response.json();
    }
    catch (parseError) {
        const rawText = await response.text().catch(() => '<unable to read body>');
        throw (0, errors_1.genesisApiError)(`Failed to parse createLaunch response as JSON: ${parseError.message} — raw body: ${rawText}`, response.status, { rawText });
    }
    if (!response.ok || !body.success) {
        throw (0, errors_1.genesisApiError)(body.error ?? `API returned status ${response.status}`, response.status, body);
    }
    const transactions = body.transactions.map((txBase64) => {
        const txBytes = serializers_1.base64.serialize(txBase64);
        return umi.transactions.deserialize(txBytes);
    });
    return {
        transactions,
        blockhash: body.blockhash,
        mintAddress: typeof body.mintAddress === 'string'
            ? body.mintAddress
            : body.mintAddress.toString(),
        genesisAccount: typeof body.genesisAccount === 'string'
            ? body.genesisAccount
            : body.genesisAccount.toString(),
    };
}
exports.createLaunch = createLaunch;
/**
 * Registers a launched genesis account.
 *
 * Call this after the create transactions have been signed and confirmed
 * on-chain.
 *
 * @example
 * ```ts
 * const registered = await registerLaunch(
 *   {},
 *   { genesisAccount: createResult.genesisAccount }
 * );
 * console.log(`Launch live at: ${registered.launch.link}`);
 * ```
 */
async function registerLaunch(_umi, config, input) {
    const fetchFn = config.fetch ?? globalThis.fetch;
    const baseUrl = resolveBaseUrl(config);
    const payload = (0, payloadBuilder_1.buildCreateLaunchPayload)(input.createLaunchInput);
    const requestBody = {
        genesisAccount: toKeyString(input.genesisAccount),
        network: input.createLaunchInput.network ?? 'solana-mainnet',
        ...payload,
    };
    let response;
    try {
        response = await fetchFn(`${baseUrl}/v1/launches/register`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(requestBody),
        });
    }
    catch (err) {
        throw (0, errors_1.genesisApiNetworkError)(`Network error calling registerLaunch: ${err.message}`, err);
    }
    let body;
    try {
        body = await response.json();
    }
    catch (parseError) {
        const rawText = await response.text().catch(() => '<unable to read body>');
        throw (0, errors_1.genesisApiError)(`Failed to parse registerLaunch response as JSON: ${parseError.message} — raw body: ${rawText}`, response.status, { rawText });
    }
    if (!response.ok || !body.success) {
        throw (0, errors_1.genesisApiError)(body.error ?? `API returned status ${response.status}`, response.status, body);
    }
    return {
        existing: body.existing,
        launch: body.launch,
        token: body.token,
    };
}
exports.registerLaunch = registerLaunch;
/**
 * High-level convenience method that orchestrates the full launch flow:
 *
 * 1. Creates the launch via the Genesis API (get unsigned transactions)
 * 2. Signs and sends all transactions to the Solana RPC via Umi
 * 3. Registers the launch
 *
 * @example
 * ```ts
 * const result = await createAndRegisterLaunch({}, umi, {
 *   wallet: umi.identity.publicKey,
 *   token: { name: 'My Token', symbol: 'MTK', image: 'https://gateway.irys.xyz/...' },
 *   launchpool: {
 *     tokenAllocation: 500_000_000,
 *     depositStartTime: new Date('2026-03-01T00:00:00Z'),
 *     raiseGoal: 250,
 *     raydiumLiquidityBps: 5000,
 *     fundsRecipient: 'RecipientWallet...',
 *   },
 * });
 * console.log(`Launch live at: ${result.launch.link}`);
 * ```
 */
async function createAndRegisterLaunch(umi, config, input, signAndSendOptions) {
    const createResult = await createLaunch(umi, config, input);
    let signatures = [];
    if (signAndSendOptions?.txSender) {
        signatures = await signAndSendOptions.txSender(createResult.transactions);
    }
    else {
        signatures = await (0, transactionHelper_1.signAndSendLaunchTransactions)(umi, createResult, {
            commitment: signAndSendOptions?.commitment ?? 'confirmed',
            preflightCommitment: signAndSendOptions?.preflightCommitment ?? 'confirmed',
            skipPreflight: signAndSendOptions?.skipPreflight ?? false,
        });
    }
    const registerResult = await registerLaunch(umi, config, {
        genesisAccount: createResult.genesisAccount,
        createLaunchInput: input,
    });
    return {
        signatures,
        mintAddress: createResult.mintAddress,
        genesisAccount: createResult.genesisAccount,
        launch: registerResult.launch,
        token: registerResult.token,
    };
}
exports.createAndRegisterLaunch = createAndRegisterLaunch;
