// @metaplex-foundation/genesis/dist/src/api/errors.js
"use strict";
// ─── API Error ───────────────────────────────────────────────────────────────
Object.defineProperty(exports, "__esModule", { value: true });
exports.isGenesisValidationError = exports.genesisValidationError = exports.isGenesisApiNetworkError = exports.genesisApiNetworkError = exports.isGenesisApiError = exports.genesisApiError = void 0;
function genesisApiError(message, statusCode, responseBody) {
    const error = new Error(message);
    error.name = 'GenesisApiError';
    error.statusCode = statusCode;
    error.responseBody = responseBody;
    return error;
}
exports.genesisApiError = genesisApiError;
function isGenesisApiError(err) {
    return err instanceof Error && err.name === 'GenesisApiError';
}
exports.isGenesisApiError = isGenesisApiError;
function genesisApiNetworkError(message, cause) {
    const error = new Error(message);
    error.name = 'GenesisApiNetworkError';
    error.cause = cause;
    return error;
}
exports.genesisApiNetworkError = genesisApiNetworkError;
function isGenesisApiNetworkError(err) {
    return err instanceof Error && err.name === 'GenesisApiNetworkError';
}
exports.isGenesisApiNetworkError = isGenesisApiNetworkError;
function genesisValidationError(message, field) {
    const error = new Error(message);
    error.name = 'GenesisValidationError';
    error.field = field;
    return error;
}
exports.genesisValidationError = genesisValidationError;
function isGenesisValidationError(err) {
    return err instanceof Error && err.name === 'GenesisValidationError';
}
exports.isGenesisValidationError = isGenesisValidationError;
