// @metaplex-foundation/genesis/dist/src/api/payloadBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildCreateLaunchPayload = void 0;
const errors_1 = require("./errors");
// ─── Constants ──────────────────────────────────────────────────────────────
const TOTAL_SUPPLY = 1000000000;
const MAX_VALID_JS_DATE_ISO = '9999-12-31T23:59:59.999Z';
const TOTAL_BPS = 10000;
// Project constants
const PROJECT_DEPOSIT_DURATION_MS = 48 * 60 * 60 * 1000; // 48 hours
const RAYDIUM_MIN_BPS = 2000;
const DEFAULT_UNLOCKED_BUCKET_NAME = 'Creator Allocation';
// Memecoin constants
const MEMECOIN_DEPOSIT_DURATION_MS = 1 * 60 * 60 * 1000; // 1 hour
const MEMECOIN_ALLOCATION_SUPPLY = 500000000; // 50% of 1B
const MEMECOIN_UNLOCKED_BPS = 100; // 1%
const MEMECOIN_UNLOCKED_SUPPLY = 10000000; // 1% of 1B
const MEMECOIN_RAYDIUM_BPS = 9800; // 98%
const MEMECOIN_RAYDIUM_SUPPLY = 490000000; // 98% of 500M
const MEMECOIN_FEE_BPS = 100; // 1%
const FEE_WALLET = '9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb';
const MEMECOIN_MIN_RAISE_GOAL = {
    SOL: 50,
    USDC: 5000,
};
// ─── Quote Mint Resolution ──────────────────────────────────────────────────
const USDC_MINTS = {
    'solana-mainnet': 'EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v',
    'solana-devnet': '4zMMC9srt5Ri5X14GAgXhaHii3GnPAEERYPJgZJDncDU',
};
function resolveQuoteMint(input, network) {
    if (!input || input === 'SOL')
        return undefined;
    if (input === 'USDC')
        return USDC_MINTS[network];
    throw (0, errors_1.genesisValidationError)(`Unsupported quote mint: '${input}'. Must be 'SOL' or 'USDC'.`, 'quoteMint');
}
// ─── Helpers ────────────────────────────────────────────────────────────────
function toIsoString(time) {
    if (typeof time === 'string') {
        const d = new Date(time);
        if (Number.isNaN(d.getTime())) {
            throw (0, errors_1.genesisValidationError)(`Invalid date string: ${time}`, 'time');
        }
        return d.toISOString();
    }
    if (Number.isNaN(time.getTime())) {
        throw (0, errors_1.genesisValidationError)(`Invalid date: ${time}`, 'time');
    }
    return time.toISOString();
}
function toMs(time) {
    const d = typeof time === 'string' ? new Date(time) : time;
    return d.getTime();
}
function toKeyString(key) {
    return typeof key === 'string' ? key : key.toString();
}
function timeAbsolute(time) {
    return { type: 'TimeAbsolute', timestamp: toIsoString(time) };
}
// ─── Validation ─────────────────────────────────────────────────────────────
function validateTokenMetadata(input) {
    const { token } = input;
    if (!token.name || token.name.length < 1 || token.name.length > 32) {
        throw (0, errors_1.genesisValidationError)('Token name must be between 1 and 32 characters', 'token.name');
    }
    if (!token.symbol || token.symbol.length < 1 || token.symbol.length > 10) {
        throw (0, errors_1.genesisValidationError)('Token symbol must be between 1 and 10 characters', 'token.symbol');
    }
    try {
        const url = new URL(token.image);
        if (url.protocol !== 'https:' && url.protocol !== 'http:') {
            throw new Error();
        }
    }
    catch {
        throw (0, errors_1.genesisValidationError)('Token image must be a valid URL', 'token.image');
    }
    if (token.description && token.description.length > 250) {
        throw (0, errors_1.genesisValidationError)('Token description must be at most 250 characters', 'token.description');
    }
}
function validateProjectInput(input) {
    validateTokenMetadata(input);
    const { launchpool, lockedAllocations } = input.launch;
    if (launchpool.tokenAllocation <= 0) {
        throw (0, errors_1.genesisValidationError)('Launchpool token allocation must be greater than 0', 'launchpool.tokenAllocation');
    }
    if (launchpool.raiseGoal <= 0) {
        throw (0, errors_1.genesisValidationError)('Raise goal must be greater than 0', 'launchpool.raiseGoal');
    }
    if (launchpool.raydiumLiquidityBps < RAYDIUM_MIN_BPS ||
        launchpool.raydiumLiquidityBps > TOTAL_BPS) {
        throw (0, errors_1.genesisValidationError)(`Raydium liquidity must be between ${RAYDIUM_MIN_BPS} and ${TOTAL_BPS} basis points`, 'launchpool.raydiumLiquidityBps');
    }
    let totalAllocated = launchpool.tokenAllocation;
    if (lockedAllocations) {
        for (const alloc of lockedAllocations) {
            if (alloc.tokenAmount <= 0) {
                throw (0, errors_1.genesisValidationError)('Locked allocation token amount must be > 0', 'lockedAllocations.tokenAmount');
            }
            totalAllocated += alloc.tokenAmount;
        }
    }
    if (totalAllocated > TOTAL_SUPPLY) {
        throw (0, errors_1.genesisValidationError)(`Total allocated tokens (${totalAllocated}) exceeds total supply (${TOTAL_SUPPLY})`, 'allocations');
    }
}
function validateMemecoinInput(input) {
    validateTokenMetadata(input);
}
// ─── Payload Builder ────────────────────────────────────────────────────────
function buildCommonPayloadFields(input) {
    const network = input.network ?? 'solana-mainnet';
    const quoteMint = resolveQuoteMint(input.quoteMint, network);
    const walletStr = toKeyString(input.wallet);
    return { network, quoteMint, walletStr };
}
function assemblePayload(input, walletStr, network, quoteMint, allocations) {
    return {
        wallet: walletStr,
        launch: {
            type: input.launchType,
            name: input.token.name,
            symbol: input.token.symbol,
            image: input.token.image,
            ...(input.token.description
                ? { description: input.token.description }
                : {}),
            ...(input.token.externalLinks
                ? { externalLinks: input.token.externalLinks }
                : {}),
            publicKey: walletStr,
            ...(network !== 'solana-mainnet' ? { network } : {}),
            ...(quoteMint ? { quoteMint } : {}),
            allocations,
        },
    };
}
function buildProjectPayload(input) {
    validateProjectInput(input);
    const { launch } = input;
    const { launchpool, lockedAllocations } = launch;
    const { network, quoteMint, walletStr } = buildCommonPayloadFields(input);
    // ── Launchpool timing ─────────────────────────────────────────────────
    const depositStartMs = toMs(launchpool.depositStartTime);
    const depositEndMs = depositStartMs + PROJECT_DEPOSIT_DURATION_MS;
    const depositStartCondition = timeAbsolute(new Date(depositStartMs));
    const depositEndCondition = timeAbsolute(new Date(depositEndMs));
    const claimEndCondition = timeAbsolute(MAX_VALID_JS_DATE_ISO);
    // Claim start is relative to Raydium graduation (not an absolute time).
    // The server resolves the bucket PDA address at transaction build time.
    const claimStartCondition = {
        type: 'TimeRelativeToUnknownAddress',
        bucketType: 'raydiumV2',
        bucketIndex: 0,
        bucketTime: 'Graduate',
        timeOffset: 0,
    };
    // ── Fund flows ────────────────────────────────────────────────────────
    const raydiumBps = launchpool.raydiumLiquidityBps;
    const unlockedBps = TOTAL_BPS - raydiumBps;
    const raydiumLpSupply = (launchpool.tokenAllocation * raydiumBps) / TOTAL_BPS;
    const fundFlows = [
        {
            type: 'UnlockedFunds',
            percentageBps: unlockedBps,
            recipient: toKeyString(launchpool.fundsRecipient),
            claimStartCondition,
            claimEndCondition,
        },
        {
            type: 'RaydiumLP',
            percentageBps: raydiumBps,
            supply: raydiumLpSupply,
            startCondition: depositEndCondition,
            lpLockSchedule: {
                startCondition: depositEndCondition,
                cliffCondition: depositEndCondition,
            },
        },
    ];
    // ── Allocations ───────────────────────────────────────────────────────
    const allocations = [];
    // Launchpool (always first)
    allocations.push({
        type: 'launchpoolV2',
        name: 'Launchpool',
        supply: launchpool.tokenAllocation,
        launchpoolV2: {
            depositStartCondition,
            depositEndCondition,
            claimStartCondition,
            claimEndCondition,
            minimumQuoteTokenThreshold: launchpool.raiseGoal,
            depositBonus: {},
            withdrawPenalty: {},
            fundFlows,
        },
    });
    // Locked/vesting allocations (Streamflow)
    if (lockedAllocations) {
        for (const alloc of lockedAllocations) {
            allocations.push({
                type: 'lockedV2',
                name: alloc.name,
                supply: alloc.tokenAmount,
                lockedV2: {
                    type: 'streamflow',
                    vesting: {
                        recipient: toKeyString(alloc.recipient),
                        startTime: toIsoString(alloc.vestingStartTime),
                        durationValue: alloc.vestingDuration.value,
                        durationUnit: alloc.vestingDuration.unit,
                        unlockSchedule: alloc.unlockSchedule,
                        cliff: alloc.cliff
                            ? {
                                enabled: true,
                                durationValue: alloc.cliff.duration.value,
                                durationUnit: alloc.cliff.duration.unit,
                                ...(alloc.cliff.unlockAmount !== undefined
                                    ? { unlockAmount: alloc.cliff.unlockAmount }
                                    : {}),
                            }
                            : { enabled: false },
                    },
                },
            });
        }
    }
    // Unlocked allocation for remaining supply
    let totalAllocated = launchpool.tokenAllocation + raydiumLpSupply;
    if (lockedAllocations) {
        for (const alloc of lockedAllocations) {
            totalAllocated += alloc.tokenAmount;
        }
    }
    const remainingSupply = TOTAL_SUPPLY - totalAllocated;
    if (remainingSupply > 0) {
        allocations.push({
            type: 'unlockedV2',
            name: DEFAULT_UNLOCKED_BUCKET_NAME,
            supply: remainingSupply,
            unlockedV2: {
                recipient: walletStr,
                claimStartCondition,
                claimEndCondition,
            },
        });
    }
    return assemblePayload(input, walletStr, network, quoteMint, allocations);
}
function buildMemecoinPayload(input) {
    validateMemecoinInput(input);
    const { launch, wallet } = input;
    const { network, quoteMint, walletStr } = buildCommonPayloadFields(input);
    // ── Launchpool timing (1 hour deposit window) ─────────────────────────
    const depositStartMs = toMs(launch.depositStartTime);
    const depositEndMs = depositStartMs + MEMECOIN_DEPOSIT_DURATION_MS;
    const depositStartCondition = timeAbsolute(new Date(depositStartMs));
    const depositEndCondition = timeAbsolute(new Date(depositEndMs));
    const claimEndCondition = timeAbsolute(MAX_VALID_JS_DATE_ISO);
    const claimStartCondition = {
        type: 'TimeRelativeToUnknownAddress',
        bucketType: 'raydiumV2',
        bucketIndex: 0,
        bucketTime: 'Graduate',
        timeOffset: 0,
    };
    // ── Fund flows (fixed: 1% unlocked, 98% raydium, 1% fees) ────────────
    const fundFlows = [
        {
            type: 'UnlockedFunds',
            percentageBps: MEMECOIN_UNLOCKED_BPS,
            name: 'Unlocked',
            supply: MEMECOIN_UNLOCKED_SUPPLY,
            recipient: toKeyString(wallet),
            claimStartCondition,
            claimEndCondition,
        },
        {
            type: 'RaydiumLP',
            percentageBps: MEMECOIN_RAYDIUM_BPS,
            name: 'Liquidity Pool',
            supply: MEMECOIN_RAYDIUM_SUPPLY,
            startCondition: depositEndCondition,
            lpLockSchedule: {
                enabled: true,
                startCondition: { type: 'Never' },
                duration: 1,
                period: 1,
                cliffCondition: { type: 'Never' },
                cliffAmountBps: 0,
            },
        },
        {
            type: 'UnlockedFees',
            percentageBps: MEMECOIN_FEE_BPS,
            name: 'Unlocked Fees',
            supply: 0,
            recipient: FEE_WALLET,
            claimStartCondition,
            claimEndCondition,
        },
    ];
    // ── Single launchpool allocation (fixed 500M supply) ──────────────────
    const allocations = [
        {
            type: 'launchpoolV2',
            name: 'Launchpool',
            supply: MEMECOIN_ALLOCATION_SUPPLY,
            launchpoolV2: {
                depositStartCondition,
                depositEndCondition,
                claimStartCondition,
                claimEndCondition,
                minimumQuoteTokenThreshold: MEMECOIN_MIN_RAISE_GOAL[input.quoteMint ?? 'SOL'],
                depositBonus: { enabled: false },
                withdrawPenalty: { enabled: false },
                fundFlows,
            },
        },
    ];
    return assemblePayload(input, walletStr, network, quoteMint, allocations);
}
/**
 * Builds the API payload for `/v1/launches/create` from simplified input.
 */
function buildCreateLaunchPayload(input) {
    switch (input.launchType) {
        case 'project':
            return buildProjectPayload(input);
        case 'memecoin':
            return buildMemecoinPayload(input);
        default:
            throw (0, errors_1.genesisValidationError)(`Unsupported launch type: ${input.launchType}`, 'launchType');
    }
}
exports.buildCreateLaunchPayload = buildCreateLaunchPayload;
