// @metaplex-foundation/genesis/dist/src/hooked/auctionAVLTreeV1.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAuctionAVLTreeV1AccountDataSerializer = exports.getAuctionNodeSerializer = exports.getTreeBidSerializer = exports.getAllocatorSerializer = exports.AUCTION_NODE_SIZE = exports.AUCTION_SENTINEL = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const auctionAVLTreeV1AccountData_1 = require("../generated/types/auctionAVLTreeV1AccountData");
const generated_1 = require("../generated");
exports.AUCTION_SENTINEL = 0;
exports.AUCTION_NODE_SIZE = 80; // Complete node size in bytes
// Serializers for Allocator type
function getAllocatorSerializer() {
    return (0, serializers_1.struct)([
        ['root', (0, serializers_1.u32)()],
        ['size', (0, serializers_1.u32)()],
        ['capacity', (0, serializers_1.u32)()],
        ['freeListHead', (0, serializers_1.u32)()],
        ['sequence', (0, serializers_1.u32)()],
        ['leftRotateCount', (0, serializers_1.u32)()],
        ['rightRotateCount', (0, serializers_1.u32)()],
        // Note: 4 bytes padding after rightRotateCount is handled automatically
    ], { description: 'Allocator' });
}
exports.getAllocatorSerializer = getAllocatorSerializer;
// Serializer for Bid type
function getTreeBidSerializer() {
    return (0, serializers_1.struct)([
        ['bidIndex', (0, serializers_1.u16)()],
        ['_padding1', (0, serializers_1.u16)()],
        ['priceUlamports', (0, serializers_1.u64)()],
        ['bidNumber', (0, serializers_1.u64)()],
        ['bidder', (0, serializers_1.publicKey)()],
        ['tokenQuantity', (0, serializers_1.u64)()],
    ], { description: 'Bid' });
}
exports.getTreeBidSerializer = getTreeBidSerializer;
function getAuctionNodeSerializer() {
    return (0, serializers_1.struct)([
        ['left', (0, serializers_1.u32)()],
        ['right', (0, serializers_1.u32)()],
        ['height', (0, serializers_1.u32)()],
        ['bid', getTreeBidSerializer()],
        ['subtreeTokens', (0, serializers_1.u64)()],
    ], { description: 'AuctionNode' });
}
exports.getAuctionNodeSerializer = getAuctionNodeSerializer;
const getAuctionAVLTreeV1AccountDataSerializer = () => ({
    description: 'AuctionAVLTreeV1AccountData',
    fixedSize: null,
    maxSize: null,
    serialize: () => {
        throw new Error('Operation not supported.');
    },
    deserialize: (buffer, offset = 0) => {
        const [auctionTree, auctionTreeOffset] = (0, auctionAVLTreeV1AccountData_1.getAuctionAVLTreeV1AccountDataSerializer)().deserialize(buffer, offset);
        if (auctionTree.key !== generated_1.Key.AuctionTree) {
            throw new Error(`Expected an AuctionTree account, got key: ${auctionTree.key}`);
        }
        // Deserialize the allocator using the new serializer
        const allocatorBytes = (0, generated_1.getBaseAllocatorSerializer)().serialize(auctionTree.allocator);
        const [allocator] = getAllocatorSerializer().deserialize(allocatorBytes);
        return [
            {
                ...auctionTree,
                rawData: buffer,
                rawDataOffset: auctionTreeOffset,
                allocator,
            },
            auctionTreeOffset,
        ];
    },
});
exports.getAuctionAVLTreeV1AccountDataSerializer = getAuctionAVLTreeV1AccountDataSerializer;
