// @metaplex-foundation/genesis/dist/src/hooked/bondingCurveBucketV1AccountData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBondingCurveBucketV1AccountDataSerializer = void 0;
const generated_1 = require("../generated");
const bondingCurveBucketV1AccountData_1 = require("../generated/types/bondingCurveBucketV1AccountData");
const getBondingCurveBucketV1AccountDataSerializer = () => ({
    description: 'BondingCurveBucketV1AccountData',
    fixedSize: (0, generated_1.getBondingCurveBucketV1Size)(),
    maxSize: null,
    serialize: (value) => (0, bondingCurveBucketV1AccountData_1.getBondingCurveBucketV1AccountDataSerializer)().serialize(value),
    deserialize: (buffer, offset = 0) => {
        const [asset, assetOfset] = (0, bondingCurveBucketV1AccountData_1.getBondingCurveBucketV1AccountDataSerializer)().deserialize(buffer, offset);
        if (asset.bondingCurveType === generated_1.BondingCurveType.ConstantProduct) {
            const [bondingCurveParameters, parametersOffset] = (0, generated_1.getConstantProductBondingCurveParametersSerializer)().deserialize(buffer, assetOfset);
            return [
                {
                    ...asset,
                    bondingCurveType: generated_1.BondingCurveType.ConstantProduct,
                    bondingCurveParameters,
                },
                parametersOffset,
            ];
        }
        throw new Error(`Unknown bonding curve type: ${asset.bondingCurveType}`);
    },
});
exports.getBondingCurveBucketV1AccountDataSerializer = getBondingCurveBucketV1AccountDataSerializer;
