// @metaplex-foundation/genesis/dist/src/generated/programs/genesis.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGenesisProgramId = exports.getGenesisProgram = exports.createGenesisProgram = exports.GENESIS_PROGRAM_ID = void 0;
const errors_1 = require("../errors");
exports.GENESIS_PROGRAM_ID = 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B';
function createGenesisProgram() {
    return {
        name: 'genesis',
        publicKey: exports.GENESIS_PROGRAM_ID,
        getErrorFromCode(code, cause) {
            return (0, errors_1.getGenesisErrorFromCode)(code, this, cause);
        },
        getErrorFromName(name, cause) {
            return (0, errors_1.getGenesisErrorFromName)(name, this, cause);
        },
        isOnCluster() {
            return true;
        },
    };
}
exports.createGenesisProgram = createGenesisProgram;
function getGenesisProgram(context, clusterFilter) {
    return context.programs.get('genesis', clusterFilter);
}
exports.getGenesisProgram = getGenesisProgram;
function getGenesisProgramId(context, clusterFilter) {
    return context.programs.getPublicKey('genesis', exports.GENESIS_PROGRAM_ID, clusterFilter);
}
exports.getGenesisProgramId = getGenesisProgramId;
