// @metaplex-foundation/genesis/dist/src/generated/accounts/auctionBucketV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchAuctionBucketV1FromSeeds = exports.fetchAuctionBucketV1FromSeeds = exports.findAuctionBucketV1Pda = exports.getAuctionBucketV1Size = exports.getAuctionBucketV1GpaBuilder = exports.safeFetchAllAuctionBucketV1 = exports.fetchAllAuctionBucketV1 = exports.safeFetchAuctionBucketV1 = exports.fetchAuctionBucketV1 = exports.deserializeAuctionBucketV1 = exports.getAuctionBucketV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getAuctionBucketV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['state', (0, types_1.getAuctionBucketStateSerializer)()],
        ['index', (0, serializers_1.u8)()],
        ['feeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['bidFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['hasPartialFill', (0, serializers_1.bool)()],
        ['extended', (0, serializers_1.u8)()],
        ['genesis', (0, serializers_1.publicKey)()],
        ['tokenAllocation', (0, serializers_1.u64)()],
        ['tokenBalance', (0, serializers_1.u64)()],
        ['solBalance', (0, serializers_1.u64)()],
        ['bidNumber', (0, serializers_1.u64)()],
        ['refundCount', (0, serializers_1.u64)()],
        ['claimCount', (0, serializers_1.u64)()],
        ['start', (0, serializers_1.i64)()],
        ['end', (0, serializers_1.i64)()],
        ['claimStart', (0, serializers_1.i64)()],
        ['backendSigner', (0, serializers_1.publicKey)()],
        ['fee', (0, serializers_1.u64)()],
        ['bidFee', (0, serializers_1.u64)()],
        ['treeAccount', (0, serializers_1.publicKey)()],
        ['minBidIncrementBps', (0, serializers_1.u32)()],
        ['minTokenQuantityDecibps', (0, serializers_1.u32)()],
        ['minPriceUlamports', (0, serializers_1.u64)()],
        ['maxBidQuantity', (0, serializers_1.u64)()],
        ['maxAuctionCapacity', (0, serializers_1.u32)()],
        ['extensionThresholdSeconds', (0, serializers_1.u32)()],
        ['extensionDurationSeconds', (0, serializers_1.u32)()],
        ['extensionMultiplierBps', (0, serializers_1.u32)()],
        ['clearingPriceUlamports', (0, serializers_1.u64)()],
        ['tokensCovered', (0, serializers_1.u64)()],
        ['clearingPriceBidder', (0, serializers_1.publicKey)()],
        ['clearingPriceBidNumber', (0, serializers_1.u64)()],
        ['clearingPriceAllocation', (0, serializers_1.u64)()],
        ['leftRotateCount', (0, serializers_1.u32)()],
        ['rightRotateCount', (0, serializers_1.u32)()],
        ['reserved', (0, serializers_1.bytes)({ size: 32 })],
    ], { description: 'AuctionBucketV1AccountData' });
}
exports.getAuctionBucketV1AccountDataSerializer = getAuctionBucketV1AccountDataSerializer;
function deserializeAuctionBucketV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getAuctionBucketV1AccountDataSerializer());
}
exports.deserializeAuctionBucketV1 = deserializeAuctionBucketV1;
async function fetchAuctionBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'AuctionBucketV1');
    return deserializeAuctionBucketV1(maybeAccount);
}
exports.fetchAuctionBucketV1 = fetchAuctionBucketV1;
async function safeFetchAuctionBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeAuctionBucketV1(maybeAccount) : null;
}
exports.safeFetchAuctionBucketV1 = safeFetchAuctionBucketV1;
async function fetchAllAuctionBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'AuctionBucketV1');
        return deserializeAuctionBucketV1(maybeAccount);
    });
}
exports.fetchAllAuctionBucketV1 = fetchAllAuctionBucketV1;
async function safeFetchAllAuctionBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeAuctionBucketV1(maybeAccount));
}
exports.safeFetchAllAuctionBucketV1 = safeFetchAllAuctionBucketV1;
function getAuctionBucketV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        state: [2, (0, types_1.getAuctionBucketStateSerializer)()],
        index: [3, (0, serializers_1.u8)()],
        feeType: [4, (0, types_1.getFeeDiscriminantsSerializer)()],
        bidFeeType: [5, (0, types_1.getFeeDiscriminantsSerializer)()],
        hasPartialFill: [6, (0, serializers_1.bool)()],
        extended: [7, (0, serializers_1.u8)()],
        genesis: [8, (0, serializers_1.publicKey)()],
        tokenAllocation: [40, (0, serializers_1.u64)()],
        tokenBalance: [48, (0, serializers_1.u64)()],
        solBalance: [56, (0, serializers_1.u64)()],
        bidNumber: [64, (0, serializers_1.u64)()],
        refundCount: [72, (0, serializers_1.u64)()],
        claimCount: [80, (0, serializers_1.u64)()],
        start: [88, (0, serializers_1.i64)()],
        end: [96, (0, serializers_1.i64)()],
        claimStart: [104, (0, serializers_1.i64)()],
        backendSigner: [112, (0, serializers_1.publicKey)()],
        fee: [144, (0, serializers_1.u64)()],
        bidFee: [152, (0, serializers_1.u64)()],
        treeAccount: [160, (0, serializers_1.publicKey)()],
        minBidIncrementBps: [192, (0, serializers_1.u32)()],
        minTokenQuantityDecibps: [196, (0, serializers_1.u32)()],
        minPriceUlamports: [200, (0, serializers_1.u64)()],
        maxBidQuantity: [208, (0, serializers_1.u64)()],
        maxAuctionCapacity: [216, (0, serializers_1.u32)()],
        extensionThresholdSeconds: [220, (0, serializers_1.u32)()],
        extensionDurationSeconds: [224, (0, serializers_1.u32)()],
        extensionMultiplierBps: [228, (0, serializers_1.u32)()],
        clearingPriceUlamports: [232, (0, serializers_1.u64)()],
        tokensCovered: [240, (0, serializers_1.u64)()],
        clearingPriceBidder: [248, (0, serializers_1.publicKey)()],
        clearingPriceBidNumber: [280, (0, serializers_1.u64)()],
        clearingPriceAllocation: [288, (0, serializers_1.u64)()],
        leftRotateCount: [296, (0, serializers_1.u32)()],
        rightRotateCount: [300, (0, serializers_1.u32)()],
        reserved: [304, (0, serializers_1.bytes)({ size: 32 })],
    })
        .deserializeUsing((account) => deserializeAuctionBucketV1(account));
}
exports.getAuctionBucketV1GpaBuilder = getAuctionBucketV1GpaBuilder;
function getAuctionBucketV1Size() {
    return 336;
}
exports.getAuctionBucketV1Size = getAuctionBucketV1Size;
function findAuctionBucketV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('auction'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.index),
    ]);
}
exports.findAuctionBucketV1Pda = findAuctionBucketV1Pda;
async function fetchAuctionBucketV1FromSeeds(context, seeds, options) {
    return fetchAuctionBucketV1(context, findAuctionBucketV1Pda(context, seeds), options);
}
exports.fetchAuctionBucketV1FromSeeds = fetchAuctionBucketV1FromSeeds;
async function safeFetchAuctionBucketV1FromSeeds(context, seeds, options) {
    return safeFetchAuctionBucketV1(context, findAuctionBucketV1Pda(context, seeds), options);
}
exports.safeFetchAuctionBucketV1FromSeeds = safeFetchAuctionBucketV1FromSeeds;
