// @metaplex-foundation/genesis/dist/src/generated/accounts/auctionBidV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchAuctionBidV1FromSeeds = exports.fetchAuctionBidV1FromSeeds = exports.findAuctionBidV1Pda = exports.getAuctionBidV1Size = exports.getAuctionBidV1GpaBuilder = exports.safeFetchAllAuctionBidV1 = exports.fetchAllAuctionBidV1 = exports.safeFetchAuctionBidV1 = exports.fetchAuctionBidV1 = exports.deserializeAuctionBidV1 = exports.getAuctionBidV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getAuctionBidV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u16)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        ['bucket', (0, serializers_1.publicKey)()],
        ['bidder', (0, serializers_1.publicKey)()],
        ['priceUlamports', (0, serializers_1.u64)()],
        ['tokenQuantity', (0, serializers_1.u64)()],
        ['solAmount', (0, serializers_1.u64)()],
        ['timestamp', (0, serializers_1.i64)()],
        ['bidNumber', (0, serializers_1.u64)()],
        ['tokensClaimed', (0, serializers_1.bool)()],
        ['refundClaimed', (0, serializers_1.bool)()],
        ['padding2', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
    ], { description: 'AuctionBidV1AccountData' });
}
exports.getAuctionBidV1AccountDataSerializer = getAuctionBidV1AccountDataSerializer;
function deserializeAuctionBidV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getAuctionBidV1AccountDataSerializer());
}
exports.deserializeAuctionBidV1 = deserializeAuctionBidV1;
async function fetchAuctionBidV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'AuctionBidV1');
    return deserializeAuctionBidV1(maybeAccount);
}
exports.fetchAuctionBidV1 = fetchAuctionBidV1;
async function safeFetchAuctionBidV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeAuctionBidV1(maybeAccount) : null;
}
exports.safeFetchAuctionBidV1 = safeFetchAuctionBidV1;
async function fetchAllAuctionBidV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'AuctionBidV1');
        return deserializeAuctionBidV1(maybeAccount);
    });
}
exports.fetchAllAuctionBidV1 = fetchAllAuctionBidV1;
async function safeFetchAllAuctionBidV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeAuctionBidV1(maybeAccount));
}
exports.safeFetchAllAuctionBidV1 = safeFetchAllAuctionBidV1;
function getAuctionBidV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        index: [2, (0, serializers_1.u16)()],
        padding: [4, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        bucket: [8, (0, serializers_1.publicKey)()],
        bidder: [40, (0, serializers_1.publicKey)()],
        priceUlamports: [72, (0, serializers_1.u64)()],
        tokenQuantity: [80, (0, serializers_1.u64)()],
        solAmount: [88, (0, serializers_1.u64)()],
        timestamp: [96, (0, serializers_1.i64)()],
        bidNumber: [104, (0, serializers_1.u64)()],
        tokensClaimed: [112, (0, serializers_1.bool)()],
        refundClaimed: [113, (0, serializers_1.bool)()],
        padding2: [114, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
    })
        .deserializeUsing((account) => deserializeAuctionBidV1(account));
}
exports.getAuctionBidV1GpaBuilder = getAuctionBidV1GpaBuilder;
function getAuctionBidV1Size() {
    return 120;
}
exports.getAuctionBidV1Size = getAuctionBidV1Size;
function findAuctionBidV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('auction_bid'),
        (0, serializers_1.publicKey)().serialize(seeds.bucket),
        (0, serializers_1.publicKey)().serialize(seeds.payer),
        (0, serializers_1.u16)().serialize(seeds.index),
    ]);
}
exports.findAuctionBidV1Pda = findAuctionBidV1Pda;
async function fetchAuctionBidV1FromSeeds(context, seeds, options) {
    return fetchAuctionBidV1(context, findAuctionBidV1Pda(context, seeds), options);
}
exports.fetchAuctionBidV1FromSeeds = fetchAuctionBidV1FromSeeds;
async function safeFetchAuctionBidV1FromSeeds(context, seeds, options) {
    return safeFetchAuctionBidV1(context, findAuctionBidV1Pda(context, seeds), options);
}
exports.safeFetchAuctionBidV1FromSeeds = safeFetchAuctionBidV1FromSeeds;
