// @metaplex-foundation/genesis/dist/src/generated/accounts/auctionAVLTreeV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchAuctionAVLTreeV1FromSeeds = exports.fetchAuctionAVLTreeV1FromSeeds = exports.findAuctionAVLTreeV1Pda = exports.getAuctionAVLTreeV1Size = exports.getAuctionAVLTreeV1GpaBuilder = exports.safeFetchAllAuctionAVLTreeV1 = exports.fetchAllAuctionAVLTreeV1 = exports.safeFetchAuctionAVLTreeV1 = exports.fetchAuctionAVLTreeV1 = exports.deserializeAuctionAVLTreeV1 = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const types_1 = require("../types");
function deserializeAuctionAVLTreeV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, (0, hooked_1.getAuctionAVLTreeV1AccountDataSerializer)());
}
exports.deserializeAuctionAVLTreeV1 = deserializeAuctionAVLTreeV1;
async function fetchAuctionAVLTreeV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'AuctionAVLTreeV1');
    return deserializeAuctionAVLTreeV1(maybeAccount);
}
exports.fetchAuctionAVLTreeV1 = fetchAuctionAVLTreeV1;
async function safeFetchAuctionAVLTreeV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeAuctionAVLTreeV1(maybeAccount) : null;
}
exports.safeFetchAuctionAVLTreeV1 = safeFetchAuctionAVLTreeV1;
async function fetchAllAuctionAVLTreeV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'AuctionAVLTreeV1');
        return deserializeAuctionAVLTreeV1(maybeAccount);
    });
}
exports.fetchAllAuctionAVLTreeV1 = fetchAllAuctionAVLTreeV1;
async function safeFetchAllAuctionAVLTreeV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeAuctionAVLTreeV1(maybeAccount));
}
exports.safeFetchAllAuctionAVLTreeV1 = safeFetchAllAuctionAVLTreeV1;
function getAuctionAVLTreeV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, serializers_1.u8)()],
        bump: [1, (0, serializers_1.u8)()],
        padding: [2, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        allocator: [8, (0, types_1.getBaseAllocatorSerializer)()],
    })
        .deserializeUsing((account) => deserializeAuctionAVLTreeV1(account));
}
exports.getAuctionAVLTreeV1GpaBuilder = getAuctionAVLTreeV1GpaBuilder;
function getAuctionAVLTreeV1Size() {
    return 40;
}
exports.getAuctionAVLTreeV1Size = getAuctionAVLTreeV1Size;
function findAuctionAVLTreeV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('auction_tree'),
        (0, serializers_1.publicKey)().serialize(seeds.bucket),
    ]);
}
exports.findAuctionAVLTreeV1Pda = findAuctionAVLTreeV1Pda;
async function fetchAuctionAVLTreeV1FromSeeds(context, seeds, options) {
    return fetchAuctionAVLTreeV1(context, findAuctionAVLTreeV1Pda(context, seeds), options);
}
exports.fetchAuctionAVLTreeV1FromSeeds = fetchAuctionAVLTreeV1FromSeeds;
async function safeFetchAuctionAVLTreeV1FromSeeds(context, seeds, options) {
    return safeFetchAuctionAVLTreeV1(context, findAuctionAVLTreeV1Pda(context, seeds), options);
}
exports.safeFetchAuctionAVLTreeV1FromSeeds = safeFetchAuctionAVLTreeV1FromSeeds;
