// @metaplex-foundation/genesis/dist/src/generated/accounts/bondingCurveBucketV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchBondingCurveBucketV1FromSeeds = exports.fetchBondingCurveBucketV1FromSeeds = exports.findBondingCurveBucketV1Pda = exports.getBondingCurveBucketV1Size = exports.getBondingCurveBucketV1GpaBuilder = exports.safeFetchAllBondingCurveBucketV1 = exports.fetchAllBondingCurveBucketV1 = exports.safeFetchBondingCurveBucketV1 = exports.fetchBondingCurveBucketV1 = exports.deserializeBondingCurveBucketV1 = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const types_1 = require("../types");
function deserializeBondingCurveBucketV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, (0, hooked_1.getBondingCurveBucketV1AccountDataSerializer)());
}
exports.deserializeBondingCurveBucketV1 = deserializeBondingCurveBucketV1;
async function fetchBondingCurveBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'BondingCurveBucketV1');
    return deserializeBondingCurveBucketV1(maybeAccount);
}
exports.fetchBondingCurveBucketV1 = fetchBondingCurveBucketV1;
async function safeFetchBondingCurveBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists
        ? deserializeBondingCurveBucketV1(maybeAccount)
        : null;
}
exports.safeFetchBondingCurveBucketV1 = safeFetchBondingCurveBucketV1;
async function fetchAllBondingCurveBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'BondingCurveBucketV1');
        return deserializeBondingCurveBucketV1(maybeAccount);
    });
}
exports.fetchAllBondingCurveBucketV1 = fetchAllBondingCurveBucketV1;
async function safeFetchAllBondingCurveBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeBondingCurveBucketV1(maybeAccount));
}
exports.safeFetchAllBondingCurveBucketV1 = safeFetchAllBondingCurveBucketV1;
function getBondingCurveBucketV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        state: [2, (0, types_1.getBondingCurveBucketStateSerializer)()],
        index: [3, (0, serializers_1.u8)()],
        feeType: [4, (0, types_1.getFeeDiscriminantsSerializer)()],
        depositFeeType: [5, (0, types_1.getFeeDiscriminantsSerializer)()],
        withdrawFeeType: [6, (0, types_1.getFeeDiscriminantsSerializer)()],
        paused: [7, (0, serializers_1.bool)()],
        genesis: [8, (0, serializers_1.publicKey)()],
        tokenAllocation: [40, (0, serializers_1.u64)()],
        tokenBalance: [48, (0, serializers_1.u64)()],
        quoteDepositTotal: [56, (0, serializers_1.u64)()],
        start: [64, (0, serializers_1.i64)()],
        end: [72, (0, serializers_1.i64)()],
        bondingCurveType: [80, (0, types_1.getBondingCurveTypeSerializer)()],
        padding2: [81, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 7 })],
        fee: [88, (0, serializers_1.u64)()],
        depositFee: [96, (0, serializers_1.u64)()],
        withdrawFee: [104, (0, serializers_1.u64)()],
        backendSigner: [112, (0, serializers_1.publicKey)()],
        actionsAuthority: [144, (0, serializers_1.publicKey)()],
    })
        .deserializeUsing((account) => deserializeBondingCurveBucketV1(account));
}
exports.getBondingCurveBucketV1GpaBuilder = getBondingCurveBucketV1GpaBuilder;
function getBondingCurveBucketV1Size() {
    return 176;
}
exports.getBondingCurveBucketV1Size = getBondingCurveBucketV1Size;
function findBondingCurveBucketV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('bonding_curve'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.index),
    ]);
}
exports.findBondingCurveBucketV1Pda = findBondingCurveBucketV1Pda;
async function fetchBondingCurveBucketV1FromSeeds(context, seeds, options) {
    return fetchBondingCurveBucketV1(context, findBondingCurveBucketV1Pda(context, seeds), options);
}
exports.fetchBondingCurveBucketV1FromSeeds = fetchBondingCurveBucketV1FromSeeds;
async function safeFetchBondingCurveBucketV1FromSeeds(context, seeds, options) {
    return safeFetchBondingCurveBucketV1(context, findBondingCurveBucketV1Pda(context, seeds), options);
}
exports.safeFetchBondingCurveBucketV1FromSeeds = safeFetchBondingCurveBucketV1FromSeeds;
