// @metaplex-foundation/genesis/dist/src/generated/accounts/bondingCurveBucketV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchBondingCurveBucketV2FromSeeds = exports.fetchBondingCurveBucketV2FromSeeds = exports.findBondingCurveBucketV2Pda = exports.getBondingCurveBucketV2GpaBuilder = exports.safeFetchAllBondingCurveBucketV2 = exports.fetchAllBondingCurveBucketV2 = exports.safeFetchBondingCurveBucketV2 = exports.fetchBondingCurveBucketV2 = exports.deserializeBondingCurveBucketV2 = exports.getBondingCurveBucketV2AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getBondingCurveBucketV2AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['depositFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['withdrawFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['bondingCurveType', (0, types_1.getBondingCurveTypeSerializer)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        ['depositFee', (0, serializers_1.u64)()],
        ['withdrawFee', (0, serializers_1.u64)()],
        ['quoteTokenDepositTotal', (0, serializers_1.u64)()],
        ['bucket', (0, types_1.getBucketBaseSerializer)()],
        ['swapStartCondition', (0, types_1.getConditionSerializer)()],
        ['swapEndCondition', (0, types_1.getConditionSerializer)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        ['extensions', (0, types_1.getBondingCurveV2ExtensionsSerializer)()],
        ['padding2', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        ['endBehaviors', (0, serializers_1.array)((0, types_1.getBehaviorSerializer)())],
        [
            'constantProductParams',
            (0, serializers_1.struct)([
                ['virtualSol', (0, serializers_1.u64)()],
                ['virtualTokens', (0, serializers_1.u64)()],
            ]),
        ],
    ], { description: 'BondingCurveBucketV2AccountData' });
}
exports.getBondingCurveBucketV2AccountDataSerializer = getBondingCurveBucketV2AccountDataSerializer;
function deserializeBondingCurveBucketV2(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getBondingCurveBucketV2AccountDataSerializer());
}
exports.deserializeBondingCurveBucketV2 = deserializeBondingCurveBucketV2;
async function fetchBondingCurveBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'BondingCurveBucketV2');
    return deserializeBondingCurveBucketV2(maybeAccount);
}
exports.fetchBondingCurveBucketV2 = fetchBondingCurveBucketV2;
async function safeFetchBondingCurveBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists
        ? deserializeBondingCurveBucketV2(maybeAccount)
        : null;
}
exports.safeFetchBondingCurveBucketV2 = safeFetchBondingCurveBucketV2;
async function fetchAllBondingCurveBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'BondingCurveBucketV2');
        return deserializeBondingCurveBucketV2(maybeAccount);
    });
}
exports.fetchAllBondingCurveBucketV2 = fetchAllBondingCurveBucketV2;
async function safeFetchAllBondingCurveBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeBondingCurveBucketV2(maybeAccount));
}
exports.safeFetchAllBondingCurveBucketV2 = safeFetchAllBondingCurveBucketV2;
function getBondingCurveBucketV2GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        depositFeeType: [1, (0, types_1.getFeeDiscriminantsSerializer)()],
        withdrawFeeType: [2, (0, types_1.getFeeDiscriminantsSerializer)()],
        bondingCurveType: [3, (0, types_1.getBondingCurveTypeSerializer)()],
        padding: [4, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        depositFee: [8, (0, serializers_1.u64)()],
        withdrawFee: [16, (0, serializers_1.u64)()],
        quoteTokenDepositTotal: [24, (0, serializers_1.u64)()],
        bucket: [32, (0, types_1.getBucketBaseSerializer)()],
        swapStartCondition: [232, (0, types_1.getConditionSerializer)()],
        swapEndCondition: [296, (0, types_1.getConditionSerializer)()],
        reserved: [360, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        extensions: [488, (0, types_1.getBondingCurveV2ExtensionsSerializer)()],
        padding2: [696, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        endBehaviors: [700, (0, serializers_1.array)((0, types_1.getBehaviorSerializer)())],
        constantProductParams: [
            null,
            (0, serializers_1.struct)([
                ['virtualSol', (0, serializers_1.u64)()],
                ['virtualTokens', (0, serializers_1.u64)()],
            ]),
        ],
    })
        .deserializeUsing((account) => deserializeBondingCurveBucketV2(account));
}
exports.getBondingCurveBucketV2GpaBuilder = getBondingCurveBucketV2GpaBuilder;
function findBondingCurveBucketV2Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('bonding_curve_v2'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.bucketIndex),
    ]);
}
exports.findBondingCurveBucketV2Pda = findBondingCurveBucketV2Pda;
async function fetchBondingCurveBucketV2FromSeeds(context, seeds, options) {
    return fetchBondingCurveBucketV2(context, findBondingCurveBucketV2Pda(context, seeds), options);
}
exports.fetchBondingCurveBucketV2FromSeeds = fetchBondingCurveBucketV2FromSeeds;
async function safeFetchBondingCurveBucketV2FromSeeds(context, seeds, options) {
    return safeFetchBondingCurveBucketV2(context, findBondingCurveBucketV2Pda(context, seeds), options);
}
exports.safeFetchBondingCurveBucketV2FromSeeds = safeFetchBondingCurveBucketV2FromSeeds;
