// @metaplex-foundation/genesis/dist/src/generated/accounts/distributionBucketV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchDistributionBucketV1FromSeeds = exports.fetchDistributionBucketV1FromSeeds = exports.findDistributionBucketV1Pda = exports.getDistributionBucketV1Size = exports.getDistributionBucketV1GpaBuilder = exports.safeFetchAllDistributionBucketV1 = exports.fetchAllDistributionBucketV1 = exports.safeFetchDistributionBucketV1 = exports.fetchDistributionBucketV1 = exports.deserializeDistributionBucketV1 = exports.getDistributionBucketV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getDistributionBucketV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['state', (0, types_1.getDistributionBucketStateSerializer)()],
        ['index', (0, serializers_1.u8)()],
        ['feeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 3 })],
        ['genesis', (0, serializers_1.publicKey)()],
        ['tokenAllocation', (0, serializers_1.u64)()],
        ['solBalance', (0, serializers_1.u64)()],
        ['tokenBalance', (0, serializers_1.u64)()],
        ['start', (0, serializers_1.i64)()],
        ['distributionAddress', (0, serializers_1.publicKey)()],
        ['backendSigner', (0, serializers_1.publicKey)()],
        ['fee', (0, serializers_1.u64)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    ], { description: 'DistributionBucketV1AccountData' });
}
exports.getDistributionBucketV1AccountDataSerializer = getDistributionBucketV1AccountDataSerializer;
function deserializeDistributionBucketV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getDistributionBucketV1AccountDataSerializer());
}
exports.deserializeDistributionBucketV1 = deserializeDistributionBucketV1;
async function fetchDistributionBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'DistributionBucketV1');
    return deserializeDistributionBucketV1(maybeAccount);
}
exports.fetchDistributionBucketV1 = fetchDistributionBucketV1;
async function safeFetchDistributionBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists
        ? deserializeDistributionBucketV1(maybeAccount)
        : null;
}
exports.safeFetchDistributionBucketV1 = safeFetchDistributionBucketV1;
async function fetchAllDistributionBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'DistributionBucketV1');
        return deserializeDistributionBucketV1(maybeAccount);
    });
}
exports.fetchAllDistributionBucketV1 = fetchAllDistributionBucketV1;
async function safeFetchAllDistributionBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeDistributionBucketV1(maybeAccount));
}
exports.safeFetchAllDistributionBucketV1 = safeFetchAllDistributionBucketV1;
function getDistributionBucketV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        state: [2, (0, types_1.getDistributionBucketStateSerializer)()],
        index: [3, (0, serializers_1.u8)()],
        feeType: [4, (0, types_1.getFeeDiscriminantsSerializer)()],
        padding: [5, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 3 })],
        genesis: [8, (0, serializers_1.publicKey)()],
        tokenAllocation: [40, (0, serializers_1.u64)()],
        solBalance: [48, (0, serializers_1.u64)()],
        tokenBalance: [56, (0, serializers_1.u64)()],
        start: [64, (0, serializers_1.i64)()],
        distributionAddress: [72, (0, serializers_1.publicKey)()],
        backendSigner: [104, (0, serializers_1.publicKey)()],
        fee: [136, (0, serializers_1.u64)()],
        reserved: [144, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    })
        .deserializeUsing((account) => deserializeDistributionBucketV1(account));
}
exports.getDistributionBucketV1GpaBuilder = getDistributionBucketV1GpaBuilder;
function getDistributionBucketV1Size() {
    return 272;
}
exports.getDistributionBucketV1Size = getDistributionBucketV1Size;
function findDistributionBucketV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('distribution'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.index),
    ]);
}
exports.findDistributionBucketV1Pda = findDistributionBucketV1Pda;
async function fetchDistributionBucketV1FromSeeds(context, seeds, options) {
    return fetchDistributionBucketV1(context, findDistributionBucketV1Pda(context, seeds), options);
}
exports.fetchDistributionBucketV1FromSeeds = fetchDistributionBucketV1FromSeeds;
async function safeFetchDistributionBucketV1FromSeeds(context, seeds, options) {
    return safeFetchDistributionBucketV1(context, findDistributionBucketV1Pda(context, seeds), options);
}
exports.safeFetchDistributionBucketV1FromSeeds = safeFetchDistributionBucketV1FromSeeds;
