// @metaplex-foundation/genesis/dist/src/generated/accounts/genesisAccountV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchGenesisAccountV1FromSeeds = exports.fetchGenesisAccountV1FromSeeds = exports.findGenesisAccountV1Pda = exports.getGenesisAccountV1GpaBuilder = exports.safeFetchAllGenesisAccountV1 = exports.fetchAllGenesisAccountV1 = exports.safeFetchGenesisAccountV1 = exports.fetchGenesisAccountV1 = exports.deserializeGenesisAccountV1 = exports.getGenesisAccountV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getGenesisAccountV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['finalized', (0, serializers_1.bool)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
        ['authority', (0, serializers_1.publicKey)()],
        ['mint', (0, serializers_1.publicKey)()],
        ['totalSupply', (0, serializers_1.u64)()],
        ['totalProceeds', (0, serializers_1.u64)()],
        ['endPadding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        ['buckets', (0, serializers_1.array)((0, types_1.getBucketEntrySerializer)())],
    ], { description: 'GenesisAccountV1AccountData' });
}
exports.getGenesisAccountV1AccountDataSerializer = getGenesisAccountV1AccountDataSerializer;
function deserializeGenesisAccountV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getGenesisAccountV1AccountDataSerializer());
}
exports.deserializeGenesisAccountV1 = deserializeGenesisAccountV1;
async function fetchGenesisAccountV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'GenesisAccountV1');
    return deserializeGenesisAccountV1(maybeAccount);
}
exports.fetchGenesisAccountV1 = fetchGenesisAccountV1;
async function safeFetchGenesisAccountV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeGenesisAccountV1(maybeAccount) : null;
}
exports.safeFetchGenesisAccountV1 = safeFetchGenesisAccountV1;
async function fetchAllGenesisAccountV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'GenesisAccountV1');
        return deserializeGenesisAccountV1(maybeAccount);
    });
}
exports.fetchAllGenesisAccountV1 = fetchAllGenesisAccountV1;
async function safeFetchAllGenesisAccountV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeGenesisAccountV1(maybeAccount));
}
exports.safeFetchAllGenesisAccountV1 = safeFetchAllGenesisAccountV1;
function getGenesisAccountV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        finalized: [2, (0, serializers_1.bool)()],
        padding: [3, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
        authority: [8, (0, serializers_1.publicKey)()],
        mint: [40, (0, serializers_1.publicKey)()],
        totalSupply: [72, (0, serializers_1.u64)()],
        totalProceeds: [80, (0, serializers_1.u64)()],
        endPadding: [88, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        buckets: [92, (0, serializers_1.array)((0, types_1.getBucketEntrySerializer)())],
    })
        .deserializeUsing((account) => deserializeGenesisAccountV1(account));
}
exports.getGenesisAccountV1GpaBuilder = getGenesisAccountV1GpaBuilder;
function findGenesisAccountV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('genesis'),
        (0, serializers_1.publicKey)().serialize(seeds.mint),
    ]);
}
exports.findGenesisAccountV1Pda = findGenesisAccountV1Pda;
async function fetchGenesisAccountV1FromSeeds(context, seeds, options) {
    return fetchGenesisAccountV1(context, findGenesisAccountV1Pda(context, seeds), options);
}
exports.fetchGenesisAccountV1FromSeeds = fetchGenesisAccountV1FromSeeds;
async function safeFetchGenesisAccountV1FromSeeds(context, seeds, options) {
    return safeFetchGenesisAccountV1(context, findGenesisAccountV1Pda(context, seeds), options);
}
exports.safeFetchGenesisAccountV1FromSeeds = safeFetchGenesisAccountV1FromSeeds;
