// @metaplex-foundation/genesis/dist/src/generated/accounts/genesisAccountV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchGenesisAccountV2FromSeeds = exports.fetchGenesisAccountV2FromSeeds = exports.findGenesisAccountV2Pda = exports.getGenesisAccountV2Size = exports.getGenesisAccountV2GpaBuilder = exports.safeFetchAllGenesisAccountV2 = exports.fetchAllGenesisAccountV2 = exports.safeFetchGenesisAccountV2 = exports.fetchGenesisAccountV2 = exports.deserializeGenesisAccountV2 = exports.getGenesisAccountV2AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getGenesisAccountV2AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u16)()],
        ['finalized', (0, serializers_1.bool)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 3 })],
        ['authority', (0, serializers_1.publicKey)()],
        ['baseMint', (0, serializers_1.publicKey)()],
        ['quoteMint', (0, serializers_1.publicKey)()],
        ['totalSupplyBaseToken', (0, serializers_1.u64)()],
        ['totalAllocatedSupplyBaseToken', (0, serializers_1.u64)()],
        ['totalProceedsQuoteToken', (0, serializers_1.u64)()],
        ['fundingMode', (0, serializers_1.u8)()],
        ['launchType', (0, serializers_1.u8)()],
        ['endPadding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 2 })],
        ['bucketCount', (0, serializers_1.u32)()],
    ], { description: 'GenesisAccountV2AccountData' });
}
exports.getGenesisAccountV2AccountDataSerializer = getGenesisAccountV2AccountDataSerializer;
function deserializeGenesisAccountV2(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getGenesisAccountV2AccountDataSerializer());
}
exports.deserializeGenesisAccountV2 = deserializeGenesisAccountV2;
async function fetchGenesisAccountV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'GenesisAccountV2');
    return deserializeGenesisAccountV2(maybeAccount);
}
exports.fetchGenesisAccountV2 = fetchGenesisAccountV2;
async function safeFetchGenesisAccountV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeGenesisAccountV2(maybeAccount) : null;
}
exports.safeFetchGenesisAccountV2 = safeFetchGenesisAccountV2;
async function fetchAllGenesisAccountV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'GenesisAccountV2');
        return deserializeGenesisAccountV2(maybeAccount);
    });
}
exports.fetchAllGenesisAccountV2 = fetchAllGenesisAccountV2;
async function safeFetchAllGenesisAccountV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeGenesisAccountV2(maybeAccount));
}
exports.safeFetchAllGenesisAccountV2 = safeFetchAllGenesisAccountV2;
function getGenesisAccountV2GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        index: [2, (0, serializers_1.u16)()],
        finalized: [4, (0, serializers_1.bool)()],
        padding: [5, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 3 })],
        authority: [8, (0, serializers_1.publicKey)()],
        baseMint: [40, (0, serializers_1.publicKey)()],
        quoteMint: [72, (0, serializers_1.publicKey)()],
        totalSupplyBaseToken: [104, (0, serializers_1.u64)()],
        totalAllocatedSupplyBaseToken: [112, (0, serializers_1.u64)()],
        totalProceedsQuoteToken: [120, (0, serializers_1.u64)()],
        fundingMode: [128, (0, serializers_1.u8)()],
        launchType: [129, (0, serializers_1.u8)()],
        endPadding: [130, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 2 })],
        bucketCount: [132, (0, serializers_1.u32)()],
    })
        .deserializeUsing((account) => deserializeGenesisAccountV2(account));
}
exports.getGenesisAccountV2GpaBuilder = getGenesisAccountV2GpaBuilder;
function getGenesisAccountV2Size() {
    return 136;
}
exports.getGenesisAccountV2Size = getGenesisAccountV2Size;
function findGenesisAccountV2Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('genesis_v2'),
        (0, serializers_1.publicKey)().serialize(seeds.baseMint),
        (0, serializers_1.u16)().serialize(seeds.genesisIndex),
    ]);
}
exports.findGenesisAccountV2Pda = findGenesisAccountV2Pda;
async function fetchGenesisAccountV2FromSeeds(context, seeds, options) {
    return fetchGenesisAccountV2(context, findGenesisAccountV2Pda(context, seeds), options);
}
exports.fetchGenesisAccountV2FromSeeds = fetchGenesisAccountV2FromSeeds;
async function safeFetchGenesisAccountV2FromSeeds(context, seeds, options) {
    return safeFetchGenesisAccountV2(context, findGenesisAccountV2Pda(context, seeds), options);
}
exports.safeFetchGenesisAccountV2FromSeeds = safeFetchGenesisAccountV2FromSeeds;
