// @metaplex-foundation/genesis/dist/src/generated/accounts/launchPoolBucketV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchLaunchPoolBucketV1FromSeeds = exports.fetchLaunchPoolBucketV1FromSeeds = exports.findLaunchPoolBucketV1Pda = exports.getLaunchPoolBucketV1Size = exports.getLaunchPoolBucketV1GpaBuilder = exports.safeFetchAllLaunchPoolBucketV1 = exports.fetchAllLaunchPoolBucketV1 = exports.safeFetchLaunchPoolBucketV1 = exports.fetchLaunchPoolBucketV1 = exports.deserializeLaunchPoolBucketV1 = exports.getLaunchPoolBucketV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getLaunchPoolBucketV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['state', (0, types_1.getLaunchPoolBucketStateSerializer)()],
        ['index', (0, serializers_1.u8)()],
        ['feeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['depositFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['withdrawFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 1 })],
        ['genesis', (0, serializers_1.publicKey)()],
        ['tokenAllocation', (0, serializers_1.u64)()],
        ['tokenBalance', (0, serializers_1.u64)()],
        ['solDepositTotal', (0, serializers_1.u64)()],
        ['depositCount', (0, serializers_1.u64)()],
        ['claimCount', (0, serializers_1.u64)()],
        ['start', (0, serializers_1.i64)()],
        ['end', (0, serializers_1.i64)()],
        ['claimStart', (0, serializers_1.i64)()],
        ['backendSigner', (0, serializers_1.publicKey)()],
        ['fee', (0, serializers_1.u64)()],
        ['depositFee', (0, serializers_1.u64)()],
        ['withdrawFee', (0, serializers_1.u64)()],
        ['weightedSolTotal', (0, serializers_1.u64)()],
        ['depositPenalty', (0, types_1.getLinearBpsScheduleSerializer)()],
        ['withdrawPenalty', (0, types_1.getLinearBpsScheduleSerializer)()],
        ['bonusSchedule', (0, types_1.getLinearBpsScheduleSerializer)()],
        ['penaltyWallet', (0, serializers_1.publicKey)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    ], { description: 'LaunchPoolBucketV1AccountData' });
}
exports.getLaunchPoolBucketV1AccountDataSerializer = getLaunchPoolBucketV1AccountDataSerializer;
function deserializeLaunchPoolBucketV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getLaunchPoolBucketV1AccountDataSerializer());
}
exports.deserializeLaunchPoolBucketV1 = deserializeLaunchPoolBucketV1;
async function fetchLaunchPoolBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'LaunchPoolBucketV1');
    return deserializeLaunchPoolBucketV1(maybeAccount);
}
exports.fetchLaunchPoolBucketV1 = fetchLaunchPoolBucketV1;
async function safeFetchLaunchPoolBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists
        ? deserializeLaunchPoolBucketV1(maybeAccount)
        : null;
}
exports.safeFetchLaunchPoolBucketV1 = safeFetchLaunchPoolBucketV1;
async function fetchAllLaunchPoolBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'LaunchPoolBucketV1');
        return deserializeLaunchPoolBucketV1(maybeAccount);
    });
}
exports.fetchAllLaunchPoolBucketV1 = fetchAllLaunchPoolBucketV1;
async function safeFetchAllLaunchPoolBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeLaunchPoolBucketV1(maybeAccount));
}
exports.safeFetchAllLaunchPoolBucketV1 = safeFetchAllLaunchPoolBucketV1;
function getLaunchPoolBucketV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        state: [2, (0, types_1.getLaunchPoolBucketStateSerializer)()],
        index: [3, (0, serializers_1.u8)()],
        feeType: [4, (0, types_1.getFeeDiscriminantsSerializer)()],
        depositFeeType: [5, (0, types_1.getFeeDiscriminantsSerializer)()],
        withdrawFeeType: [6, (0, types_1.getFeeDiscriminantsSerializer)()],
        padding: [7, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 1 })],
        genesis: [8, (0, serializers_1.publicKey)()],
        tokenAllocation: [40, (0, serializers_1.u64)()],
        tokenBalance: [48, (0, serializers_1.u64)()],
        solDepositTotal: [56, (0, serializers_1.u64)()],
        depositCount: [64, (0, serializers_1.u64)()],
        claimCount: [72, (0, serializers_1.u64)()],
        start: [80, (0, serializers_1.i64)()],
        end: [88, (0, serializers_1.i64)()],
        claimStart: [96, (0, serializers_1.i64)()],
        backendSigner: [104, (0, serializers_1.publicKey)()],
        fee: [136, (0, serializers_1.u64)()],
        depositFee: [144, (0, serializers_1.u64)()],
        withdrawFee: [152, (0, serializers_1.u64)()],
        weightedSolTotal: [160, (0, serializers_1.u64)()],
        depositPenalty: [168, (0, types_1.getLinearBpsScheduleSerializer)()],
        withdrawPenalty: [208, (0, types_1.getLinearBpsScheduleSerializer)()],
        bonusSchedule: [248, (0, types_1.getLinearBpsScheduleSerializer)()],
        penaltyWallet: [288, (0, serializers_1.publicKey)()],
        reserved: [320, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    })
        .deserializeUsing((account) => deserializeLaunchPoolBucketV1(account));
}
exports.getLaunchPoolBucketV1GpaBuilder = getLaunchPoolBucketV1GpaBuilder;
function getLaunchPoolBucketV1Size() {
    return 448;
}
exports.getLaunchPoolBucketV1Size = getLaunchPoolBucketV1Size;
function findLaunchPoolBucketV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('launch_pool'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.index),
    ]);
}
exports.findLaunchPoolBucketV1Pda = findLaunchPoolBucketV1Pda;
async function fetchLaunchPoolBucketV1FromSeeds(context, seeds, options) {
    return fetchLaunchPoolBucketV1(context, findLaunchPoolBucketV1Pda(context, seeds), options);
}
exports.fetchLaunchPoolBucketV1FromSeeds = fetchLaunchPoolBucketV1FromSeeds;
async function safeFetchLaunchPoolBucketV1FromSeeds(context, seeds, options) {
    return safeFetchLaunchPoolBucketV1(context, findLaunchPoolBucketV1Pda(context, seeds), options);
}
exports.safeFetchLaunchPoolBucketV1FromSeeds = safeFetchLaunchPoolBucketV1FromSeeds;
