// @metaplex-foundation/genesis/dist/src/generated/accounts/launchPoolDepositV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchLaunchPoolDepositV1FromSeeds = exports.fetchLaunchPoolDepositV1FromSeeds = exports.findLaunchPoolDepositV1Pda = exports.getLaunchPoolDepositV1Size = exports.getLaunchPoolDepositV1GpaBuilder = exports.safeFetchAllLaunchPoolDepositV1 = exports.fetchAllLaunchPoolDepositV1 = exports.safeFetchLaunchPoolDepositV1 = exports.fetchLaunchPoolDepositV1 = exports.deserializeLaunchPoolDepositV1 = exports.getLaunchPoolDepositV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getLaunchPoolDepositV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['claimed', (0, serializers_1.bool)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 1 })],
        ['index', (0, serializers_1.u32)()],
        ['depositor', (0, serializers_1.publicKey)()],
        ['bucket', (0, serializers_1.publicKey)()],
        ['timestamp', (0, serializers_1.i64)()],
        ['amount', (0, serializers_1.u64)()],
        ['weightedSol', (0, serializers_1.u64)()],
    ], { description: 'LaunchPoolDepositV1AccountData' });
}
exports.getLaunchPoolDepositV1AccountDataSerializer = getLaunchPoolDepositV1AccountDataSerializer;
function deserializeLaunchPoolDepositV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getLaunchPoolDepositV1AccountDataSerializer());
}
exports.deserializeLaunchPoolDepositV1 = deserializeLaunchPoolDepositV1;
async function fetchLaunchPoolDepositV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'LaunchPoolDepositV1');
    return deserializeLaunchPoolDepositV1(maybeAccount);
}
exports.fetchLaunchPoolDepositV1 = fetchLaunchPoolDepositV1;
async function safeFetchLaunchPoolDepositV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists
        ? deserializeLaunchPoolDepositV1(maybeAccount)
        : null;
}
exports.safeFetchLaunchPoolDepositV1 = safeFetchLaunchPoolDepositV1;
async function fetchAllLaunchPoolDepositV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'LaunchPoolDepositV1');
        return deserializeLaunchPoolDepositV1(maybeAccount);
    });
}
exports.fetchAllLaunchPoolDepositV1 = fetchAllLaunchPoolDepositV1;
async function safeFetchAllLaunchPoolDepositV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeLaunchPoolDepositV1(maybeAccount));
}
exports.safeFetchAllLaunchPoolDepositV1 = safeFetchAllLaunchPoolDepositV1;
function getLaunchPoolDepositV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        claimed: [2, (0, serializers_1.bool)()],
        padding: [3, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 1 })],
        index: [4, (0, serializers_1.u32)()],
        depositor: [8, (0, serializers_1.publicKey)()],
        bucket: [40, (0, serializers_1.publicKey)()],
        timestamp: [72, (0, serializers_1.i64)()],
        amount: [80, (0, serializers_1.u64)()],
        weightedSol: [88, (0, serializers_1.u64)()],
    })
        .deserializeUsing((account) => deserializeLaunchPoolDepositV1(account));
}
exports.getLaunchPoolDepositV1GpaBuilder = getLaunchPoolDepositV1GpaBuilder;
function getLaunchPoolDepositV1Size() {
    return 96;
}
exports.getLaunchPoolDepositV1Size = getLaunchPoolDepositV1Size;
function findLaunchPoolDepositV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('launch_pool_deposit'),
        (0, serializers_1.publicKey)().serialize(seeds.bucket),
        (0, serializers_1.publicKey)().serialize(seeds.recipient),
    ]);
}
exports.findLaunchPoolDepositV1Pda = findLaunchPoolDepositV1Pda;
async function fetchLaunchPoolDepositV1FromSeeds(context, seeds, options) {
    return fetchLaunchPoolDepositV1(context, findLaunchPoolDepositV1Pda(context, seeds), options);
}
exports.fetchLaunchPoolDepositV1FromSeeds = fetchLaunchPoolDepositV1FromSeeds;
async function safeFetchLaunchPoolDepositV1FromSeeds(context, seeds, options) {
    return safeFetchLaunchPoolDepositV1(context, findLaunchPoolDepositV1Pda(context, seeds), options);
}
exports.safeFetchLaunchPoolDepositV1FromSeeds = safeFetchLaunchPoolDepositV1FromSeeds;
