// @metaplex-foundation/genesis/dist/src/generated/accounts/launchPoolBucketV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchLaunchPoolBucketV2FromSeeds = exports.fetchLaunchPoolBucketV2FromSeeds = exports.findLaunchPoolBucketV2Pda = exports.getLaunchPoolBucketV2GpaBuilder = exports.safeFetchAllLaunchPoolBucketV2 = exports.fetchAllLaunchPoolBucketV2 = exports.safeFetchLaunchPoolBucketV2 = exports.fetchLaunchPoolBucketV2 = exports.deserializeLaunchPoolBucketV2 = exports.getLaunchPoolBucketV2AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getLaunchPoolBucketV2AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['depositFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['withdrawFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['claimFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        ['depositFee', (0, serializers_1.u64)()],
        ['depositId', (0, serializers_1.u64)()],
        ['depositCount', (0, serializers_1.u64)()],
        ['quoteTokenDepositTotal', (0, serializers_1.u64)()],
        ['weightedQuoteTokenTotal', (0, serializers_1.u64)()],
        ['withdrawFee', (0, serializers_1.u64)()],
        ['claimFee', (0, serializers_1.u64)()],
        ['claimCount', (0, serializers_1.u64)()],
        ['refundCount', (0, serializers_1.u64)()],
        ['bucket', (0, types_1.getBucketBaseSerializer)()],
        ['depositStartCondition', (0, types_1.getConditionSerializer)()],
        ['depositEndCondition', (0, types_1.getConditionSerializer)()],
        ['claimStartCondition', (0, types_1.getConditionSerializer)()],
        ['claimEndCondition', (0, types_1.getConditionSerializer)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        ['extensions', (0, types_1.getLaunchPoolV2ExtensionsSerializer)()],
        ['padding2', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        ['endBehaviors', (0, serializers_1.array)((0, types_1.getBehaviorSerializer)())],
    ], { description: 'LaunchPoolBucketV2AccountData' });
}
exports.getLaunchPoolBucketV2AccountDataSerializer = getLaunchPoolBucketV2AccountDataSerializer;
function deserializeLaunchPoolBucketV2(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getLaunchPoolBucketV2AccountDataSerializer());
}
exports.deserializeLaunchPoolBucketV2 = deserializeLaunchPoolBucketV2;
async function fetchLaunchPoolBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'LaunchPoolBucketV2');
    return deserializeLaunchPoolBucketV2(maybeAccount);
}
exports.fetchLaunchPoolBucketV2 = fetchLaunchPoolBucketV2;
async function safeFetchLaunchPoolBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists
        ? deserializeLaunchPoolBucketV2(maybeAccount)
        : null;
}
exports.safeFetchLaunchPoolBucketV2 = safeFetchLaunchPoolBucketV2;
async function fetchAllLaunchPoolBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'LaunchPoolBucketV2');
        return deserializeLaunchPoolBucketV2(maybeAccount);
    });
}
exports.fetchAllLaunchPoolBucketV2 = fetchAllLaunchPoolBucketV2;
async function safeFetchAllLaunchPoolBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeLaunchPoolBucketV2(maybeAccount));
}
exports.safeFetchAllLaunchPoolBucketV2 = safeFetchAllLaunchPoolBucketV2;
function getLaunchPoolBucketV2GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        depositFeeType: [1, (0, types_1.getFeeDiscriminantsSerializer)()],
        withdrawFeeType: [2, (0, types_1.getFeeDiscriminantsSerializer)()],
        claimFeeType: [3, (0, types_1.getFeeDiscriminantsSerializer)()],
        padding: [4, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        depositFee: [8, (0, serializers_1.u64)()],
        depositId: [16, (0, serializers_1.u64)()],
        depositCount: [24, (0, serializers_1.u64)()],
        quoteTokenDepositTotal: [32, (0, serializers_1.u64)()],
        weightedQuoteTokenTotal: [40, (0, serializers_1.u64)()],
        withdrawFee: [48, (0, serializers_1.u64)()],
        claimFee: [56, (0, serializers_1.u64)()],
        claimCount: [64, (0, serializers_1.u64)()],
        refundCount: [72, (0, serializers_1.u64)()],
        bucket: [80, (0, types_1.getBucketBaseSerializer)()],
        depositStartCondition: [280, (0, types_1.getConditionSerializer)()],
        depositEndCondition: [344, (0, types_1.getConditionSerializer)()],
        claimStartCondition: [408, (0, types_1.getConditionSerializer)()],
        claimEndCondition: [472, (0, types_1.getConditionSerializer)()],
        reserved: [536, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        extensions: [664, (0, types_1.getLaunchPoolV2ExtensionsSerializer)()],
        padding2: [1704, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        endBehaviors: [1708, (0, serializers_1.array)((0, types_1.getBehaviorSerializer)())],
    })
        .deserializeUsing((account) => deserializeLaunchPoolBucketV2(account));
}
exports.getLaunchPoolBucketV2GpaBuilder = getLaunchPoolBucketV2GpaBuilder;
function findLaunchPoolBucketV2Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('launch_pool_v2'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.bucketIndex),
    ]);
}
exports.findLaunchPoolBucketV2Pda = findLaunchPoolBucketV2Pda;
async function fetchLaunchPoolBucketV2FromSeeds(context, seeds, options) {
    return fetchLaunchPoolBucketV2(context, findLaunchPoolBucketV2Pda(context, seeds), options);
}
exports.fetchLaunchPoolBucketV2FromSeeds = fetchLaunchPoolBucketV2FromSeeds;
async function safeFetchLaunchPoolBucketV2FromSeeds(context, seeds, options) {
    return safeFetchLaunchPoolBucketV2(context, findLaunchPoolBucketV2Pda(context, seeds), options);
}
exports.safeFetchLaunchPoolBucketV2FromSeeds = safeFetchLaunchPoolBucketV2FromSeeds;
