// @metaplex-foundation/genesis/dist/src/generated/accounts/launchPoolDepositV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchLaunchPoolDepositV2FromSeeds = exports.fetchLaunchPoolDepositV2FromSeeds = exports.findLaunchPoolDepositV2Pda = exports.getLaunchPoolDepositV2Size = exports.getLaunchPoolDepositV2GpaBuilder = exports.safeFetchAllLaunchPoolDepositV2 = exports.fetchAllLaunchPoolDepositV2 = exports.safeFetchLaunchPoolDepositV2 = exports.fetchLaunchPoolDepositV2 = exports.deserializeLaunchPoolDepositV2 = exports.getLaunchPoolDepositV2AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getLaunchPoolDepositV2AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['claimed', (0, serializers_1.bool)()],
        ['refunded', (0, serializers_1.bool)()],
        ['index', (0, serializers_1.u32)()],
        ['amountClaimed', (0, serializers_1.u64)()],
        ['depositor', (0, serializers_1.publicKey)()],
        ['bucket', (0, serializers_1.publicKey)()],
        ['timestamp', (0, serializers_1.i64)()],
        ['amountQuoteToken', (0, serializers_1.u64)()],
        ['weightedQuoteToken', (0, serializers_1.u64)()],
    ], { description: 'LaunchPoolDepositV2AccountData' });
}
exports.getLaunchPoolDepositV2AccountDataSerializer = getLaunchPoolDepositV2AccountDataSerializer;
function deserializeLaunchPoolDepositV2(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getLaunchPoolDepositV2AccountDataSerializer());
}
exports.deserializeLaunchPoolDepositV2 = deserializeLaunchPoolDepositV2;
async function fetchLaunchPoolDepositV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'LaunchPoolDepositV2');
    return deserializeLaunchPoolDepositV2(maybeAccount);
}
exports.fetchLaunchPoolDepositV2 = fetchLaunchPoolDepositV2;
async function safeFetchLaunchPoolDepositV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists
        ? deserializeLaunchPoolDepositV2(maybeAccount)
        : null;
}
exports.safeFetchLaunchPoolDepositV2 = safeFetchLaunchPoolDepositV2;
async function fetchAllLaunchPoolDepositV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'LaunchPoolDepositV2');
        return deserializeLaunchPoolDepositV2(maybeAccount);
    });
}
exports.fetchAllLaunchPoolDepositV2 = fetchAllLaunchPoolDepositV2;
async function safeFetchAllLaunchPoolDepositV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeLaunchPoolDepositV2(maybeAccount));
}
exports.safeFetchAllLaunchPoolDepositV2 = safeFetchAllLaunchPoolDepositV2;
function getLaunchPoolDepositV2GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        claimed: [2, (0, serializers_1.bool)()],
        refunded: [3, (0, serializers_1.bool)()],
        index: [4, (0, serializers_1.u32)()],
        amountClaimed: [8, (0, serializers_1.u64)()],
        depositor: [16, (0, serializers_1.publicKey)()],
        bucket: [48, (0, serializers_1.publicKey)()],
        timestamp: [80, (0, serializers_1.i64)()],
        amountQuoteToken: [88, (0, serializers_1.u64)()],
        weightedQuoteToken: [96, (0, serializers_1.u64)()],
    })
        .deserializeUsing((account) => deserializeLaunchPoolDepositV2(account));
}
exports.getLaunchPoolDepositV2GpaBuilder = getLaunchPoolDepositV2GpaBuilder;
function getLaunchPoolDepositV2Size() {
    return 104;
}
exports.getLaunchPoolDepositV2Size = getLaunchPoolDepositV2Size;
function findLaunchPoolDepositV2Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('launch_pool_deposit_v2'),
        (0, serializers_1.publicKey)().serialize(seeds.bucket),
        (0, serializers_1.publicKey)().serialize(seeds.recipient),
    ]);
}
exports.findLaunchPoolDepositV2Pda = findLaunchPoolDepositV2Pda;
async function fetchLaunchPoolDepositV2FromSeeds(context, seeds, options) {
    return fetchLaunchPoolDepositV2(context, findLaunchPoolDepositV2Pda(context, seeds), options);
}
exports.fetchLaunchPoolDepositV2FromSeeds = fetchLaunchPoolDepositV2FromSeeds;
async function safeFetchLaunchPoolDepositV2FromSeeds(context, seeds, options) {
    return safeFetchLaunchPoolDepositV2(context, findLaunchPoolDepositV2Pda(context, seeds), options);
}
exports.safeFetchLaunchPoolDepositV2FromSeeds = safeFetchLaunchPoolDepositV2FromSeeds;
