// @metaplex-foundation/genesis/dist/src/generated/accounts/meteoraBucketV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchMeteoraBucketV1FromSeeds = exports.fetchMeteoraBucketV1FromSeeds = exports.findMeteoraBucketV1Pda = exports.getMeteoraBucketV1Size = exports.getMeteoraBucketV1GpaBuilder = exports.safeFetchAllMeteoraBucketV1 = exports.fetchAllMeteoraBucketV1 = exports.safeFetchMeteoraBucketV1 = exports.fetchMeteoraBucketV1 = exports.deserializeMeteoraBucketV1 = exports.getMeteoraBucketV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getMeteoraBucketV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['state', (0, types_1.getMeteoraBucketStateSerializer)()],
        ['index', (0, serializers_1.u8)()],
        ['feeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 3 })],
        ['genesis', (0, serializers_1.publicKey)()],
        ['tokenAllocation', (0, serializers_1.u64)()],
        ['solBalance', (0, serializers_1.u64)()],
        ['tokenBalance', (0, serializers_1.u64)()],
        ['start', (0, serializers_1.i64)()],
        ['backendSigner', (0, serializers_1.publicKey)()],
        ['fee', (0, serializers_1.u64)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    ], { description: 'MeteoraBucketV1AccountData' });
}
exports.getMeteoraBucketV1AccountDataSerializer = getMeteoraBucketV1AccountDataSerializer;
function deserializeMeteoraBucketV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getMeteoraBucketV1AccountDataSerializer());
}
exports.deserializeMeteoraBucketV1 = deserializeMeteoraBucketV1;
async function fetchMeteoraBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'MeteoraBucketV1');
    return deserializeMeteoraBucketV1(maybeAccount);
}
exports.fetchMeteoraBucketV1 = fetchMeteoraBucketV1;
async function safeFetchMeteoraBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeMeteoraBucketV1(maybeAccount) : null;
}
exports.safeFetchMeteoraBucketV1 = safeFetchMeteoraBucketV1;
async function fetchAllMeteoraBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'MeteoraBucketV1');
        return deserializeMeteoraBucketV1(maybeAccount);
    });
}
exports.fetchAllMeteoraBucketV1 = fetchAllMeteoraBucketV1;
async function safeFetchAllMeteoraBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeMeteoraBucketV1(maybeAccount));
}
exports.safeFetchAllMeteoraBucketV1 = safeFetchAllMeteoraBucketV1;
function getMeteoraBucketV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        state: [2, (0, types_1.getMeteoraBucketStateSerializer)()],
        index: [3, (0, serializers_1.u8)()],
        feeType: [4, (0, types_1.getFeeDiscriminantsSerializer)()],
        padding: [5, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 3 })],
        genesis: [8, (0, serializers_1.publicKey)()],
        tokenAllocation: [40, (0, serializers_1.u64)()],
        solBalance: [48, (0, serializers_1.u64)()],
        tokenBalance: [56, (0, serializers_1.u64)()],
        start: [64, (0, serializers_1.i64)()],
        backendSigner: [72, (0, serializers_1.publicKey)()],
        fee: [104, (0, serializers_1.u64)()],
        reserved: [112, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    })
        .deserializeUsing((account) => deserializeMeteoraBucketV1(account));
}
exports.getMeteoraBucketV1GpaBuilder = getMeteoraBucketV1GpaBuilder;
function getMeteoraBucketV1Size() {
    return 240;
}
exports.getMeteoraBucketV1Size = getMeteoraBucketV1Size;
function findMeteoraBucketV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('meteora'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.index),
    ]);
}
exports.findMeteoraBucketV1Pda = findMeteoraBucketV1Pda;
async function fetchMeteoraBucketV1FromSeeds(context, seeds, options) {
    return fetchMeteoraBucketV1(context, findMeteoraBucketV1Pda(context, seeds), options);
}
exports.fetchMeteoraBucketV1FromSeeds = fetchMeteoraBucketV1FromSeeds;
async function safeFetchMeteoraBucketV1FromSeeds(context, seeds, options) {
    return safeFetchMeteoraBucketV1(context, findMeteoraBucketV1Pda(context, seeds), options);
}
exports.safeFetchMeteoraBucketV1FromSeeds = safeFetchMeteoraBucketV1FromSeeds;
