// @metaplex-foundation/genesis/dist/src/generated/accounts/presaleBucketV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchPresaleBucketV1FromSeeds = exports.fetchPresaleBucketV1FromSeeds = exports.findPresaleBucketV1Pda = exports.getPresaleBucketV1Size = exports.getPresaleBucketV1GpaBuilder = exports.safeFetchAllPresaleBucketV1 = exports.fetchAllPresaleBucketV1 = exports.safeFetchPresaleBucketV1 = exports.fetchPresaleBucketV1 = exports.deserializePresaleBucketV1 = exports.getPresaleBucketV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getPresaleBucketV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['state', (0, types_1.getPresaleBucketStateSerializer)()],
        ['index', (0, serializers_1.u8)()],
        ['allowWithdraw', (0, serializers_1.bool)()],
        ['feeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['depositFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['withdrawFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['genesis', (0, serializers_1.publicKey)()],
        ['tokenAllocation', (0, serializers_1.u64)()],
        ['tokenBalance', (0, serializers_1.u64)()],
        ['solDepositTotal', (0, serializers_1.u64)()],
        ['depositCount', (0, serializers_1.u64)()],
        ['claimCount', (0, serializers_1.u64)()],
        ['start', (0, serializers_1.i64)()],
        ['end', (0, serializers_1.i64)()],
        ['claimStart', (0, serializers_1.i64)()],
        ['allocationSolCap', (0, serializers_1.u64)()],
        ['backendSigner', (0, serializers_1.publicKey)()],
        ['depositLimit', (0, serializers_1.u64)()],
        ['depositCooldown', (0, serializers_1.u64)()],
        ['perCooldownDepositLimit', (0, serializers_1.u64)()],
        ['hasAllowlist', (0, serializers_1.bool)()],
        ['allowlistMerkleTreeHeight', (0, serializers_1.u8)()],
        ['isAllowlistSolCapReached', (0, serializers_1.bool)()],
        ['padding2', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
        ['allowlistMerkleRoot', (0, serializers_1.bytes)({ size: 32 })],
        ['allowlistEnd', (0, serializers_1.i64)()],
        ['allowlistSolCap', (0, serializers_1.u64)()],
        ['allowlistSolDepositTotal', (0, serializers_1.u64)()],
        ['fee', (0, serializers_1.u64)()],
        ['depositFee', (0, serializers_1.u64)()],
        ['withdrawFee', (0, serializers_1.u64)()],
        ['unsoldWithdrawn', (0, serializers_1.u64)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 120 })],
    ], { description: 'PresaleBucketV1AccountData' });
}
exports.getPresaleBucketV1AccountDataSerializer = getPresaleBucketV1AccountDataSerializer;
function deserializePresaleBucketV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getPresaleBucketV1AccountDataSerializer());
}
exports.deserializePresaleBucketV1 = deserializePresaleBucketV1;
async function fetchPresaleBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'PresaleBucketV1');
    return deserializePresaleBucketV1(maybeAccount);
}
exports.fetchPresaleBucketV1 = fetchPresaleBucketV1;
async function safeFetchPresaleBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializePresaleBucketV1(maybeAccount) : null;
}
exports.safeFetchPresaleBucketV1 = safeFetchPresaleBucketV1;
async function fetchAllPresaleBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'PresaleBucketV1');
        return deserializePresaleBucketV1(maybeAccount);
    });
}
exports.fetchAllPresaleBucketV1 = fetchAllPresaleBucketV1;
async function safeFetchAllPresaleBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializePresaleBucketV1(maybeAccount));
}
exports.safeFetchAllPresaleBucketV1 = safeFetchAllPresaleBucketV1;
function getPresaleBucketV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        state: [2, (0, types_1.getPresaleBucketStateSerializer)()],
        index: [3, (0, serializers_1.u8)()],
        allowWithdraw: [4, (0, serializers_1.bool)()],
        feeType: [5, (0, types_1.getFeeDiscriminantsSerializer)()],
        depositFeeType: [6, (0, types_1.getFeeDiscriminantsSerializer)()],
        withdrawFeeType: [7, (0, types_1.getFeeDiscriminantsSerializer)()],
        genesis: [8, (0, serializers_1.publicKey)()],
        tokenAllocation: [40, (0, serializers_1.u64)()],
        tokenBalance: [48, (0, serializers_1.u64)()],
        solDepositTotal: [56, (0, serializers_1.u64)()],
        depositCount: [64, (0, serializers_1.u64)()],
        claimCount: [72, (0, serializers_1.u64)()],
        start: [80, (0, serializers_1.i64)()],
        end: [88, (0, serializers_1.i64)()],
        claimStart: [96, (0, serializers_1.i64)()],
        allocationSolCap: [104, (0, serializers_1.u64)()],
        backendSigner: [112, (0, serializers_1.publicKey)()],
        depositLimit: [144, (0, serializers_1.u64)()],
        depositCooldown: [152, (0, serializers_1.u64)()],
        perCooldownDepositLimit: [160, (0, serializers_1.u64)()],
        hasAllowlist: [168, (0, serializers_1.bool)()],
        allowlistMerkleTreeHeight: [169, (0, serializers_1.u8)()],
        isAllowlistSolCapReached: [170, (0, serializers_1.bool)()],
        padding2: [171, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
        allowlistMerkleRoot: [176, (0, serializers_1.bytes)({ size: 32 })],
        allowlistEnd: [208, (0, serializers_1.i64)()],
        allowlistSolCap: [216, (0, serializers_1.u64)()],
        allowlistSolDepositTotal: [224, (0, serializers_1.u64)()],
        fee: [232, (0, serializers_1.u64)()],
        depositFee: [240, (0, serializers_1.u64)()],
        withdrawFee: [248, (0, serializers_1.u64)()],
        unsoldWithdrawn: [256, (0, serializers_1.u64)()],
        reserved: [264, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 120 })],
    })
        .deserializeUsing((account) => deserializePresaleBucketV1(account));
}
exports.getPresaleBucketV1GpaBuilder = getPresaleBucketV1GpaBuilder;
function getPresaleBucketV1Size() {
    return 384;
}
exports.getPresaleBucketV1Size = getPresaleBucketV1Size;
function findPresaleBucketV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('presale'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.index),
    ]);
}
exports.findPresaleBucketV1Pda = findPresaleBucketV1Pda;
async function fetchPresaleBucketV1FromSeeds(context, seeds, options) {
    return fetchPresaleBucketV1(context, findPresaleBucketV1Pda(context, seeds), options);
}
exports.fetchPresaleBucketV1FromSeeds = fetchPresaleBucketV1FromSeeds;
async function safeFetchPresaleBucketV1FromSeeds(context, seeds, options) {
    return safeFetchPresaleBucketV1(context, findPresaleBucketV1Pda(context, seeds), options);
}
exports.safeFetchPresaleBucketV1FromSeeds = safeFetchPresaleBucketV1FromSeeds;
