// @metaplex-foundation/genesis/dist/src/generated/accounts/presaleBucketV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchPresaleBucketV2FromSeeds = exports.fetchPresaleBucketV2FromSeeds = exports.findPresaleBucketV2Pda = exports.getPresaleBucketV2GpaBuilder = exports.safeFetchAllPresaleBucketV2 = exports.fetchAllPresaleBucketV2 = exports.safeFetchPresaleBucketV2 = exports.fetchPresaleBucketV2 = exports.deserializePresaleBucketV2 = exports.getPresaleBucketV2AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getPresaleBucketV2AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['depositFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['claimFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
        ['depositFee', (0, serializers_1.u64)()],
        ['depositCount', (0, serializers_1.u64)()],
        ['quoteTokenDepositTotal', (0, serializers_1.u64)()],
        ['claimFee', (0, serializers_1.u64)()],
        ['claimCount', (0, serializers_1.u64)()],
        ['bucket', (0, types_1.getBucketBaseSerializer)()],
        ['allocationQuoteTokenCap', (0, serializers_1.u64)()],
        ['depositStartCondition', (0, types_1.getConditionSerializer)()],
        ['depositEndCondition', (0, types_1.getConditionSerializer)()],
        ['claimStartCondition', (0, types_1.getConditionSerializer)()],
        ['claimEndCondition', (0, types_1.getConditionSerializer)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        ['extensions', (0, types_1.getPresaleV2ExtensionsSerializer)()],
        ['padding2', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        ['endBehaviors', (0, serializers_1.array)((0, types_1.getBehaviorSerializer)())],
    ], { description: 'PresaleBucketV2AccountData' });
}
exports.getPresaleBucketV2AccountDataSerializer = getPresaleBucketV2AccountDataSerializer;
function deserializePresaleBucketV2(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getPresaleBucketV2AccountDataSerializer());
}
exports.deserializePresaleBucketV2 = deserializePresaleBucketV2;
async function fetchPresaleBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'PresaleBucketV2');
    return deserializePresaleBucketV2(maybeAccount);
}
exports.fetchPresaleBucketV2 = fetchPresaleBucketV2;
async function safeFetchPresaleBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializePresaleBucketV2(maybeAccount) : null;
}
exports.safeFetchPresaleBucketV2 = safeFetchPresaleBucketV2;
async function fetchAllPresaleBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'PresaleBucketV2');
        return deserializePresaleBucketV2(maybeAccount);
    });
}
exports.fetchAllPresaleBucketV2 = fetchAllPresaleBucketV2;
async function safeFetchAllPresaleBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializePresaleBucketV2(maybeAccount));
}
exports.safeFetchAllPresaleBucketV2 = safeFetchAllPresaleBucketV2;
function getPresaleBucketV2GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        depositFeeType: [1, (0, types_1.getFeeDiscriminantsSerializer)()],
        claimFeeType: [2, (0, types_1.getFeeDiscriminantsSerializer)()],
        padding: [3, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
        depositFee: [8, (0, serializers_1.u64)()],
        depositCount: [16, (0, serializers_1.u64)()],
        quoteTokenDepositTotal: [24, (0, serializers_1.u64)()],
        claimFee: [32, (0, serializers_1.u64)()],
        claimCount: [40, (0, serializers_1.u64)()],
        bucket: [48, (0, types_1.getBucketBaseSerializer)()],
        allocationQuoteTokenCap: [248, (0, serializers_1.u64)()],
        depositStartCondition: [256, (0, types_1.getConditionSerializer)()],
        depositEndCondition: [320, (0, types_1.getConditionSerializer)()],
        claimStartCondition: [384, (0, types_1.getConditionSerializer)()],
        claimEndCondition: [448, (0, types_1.getConditionSerializer)()],
        reserved: [512, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        extensions: [640, (0, types_1.getPresaleV2ExtensionsSerializer)()],
        padding2: [1256, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        endBehaviors: [1260, (0, serializers_1.array)((0, types_1.getBehaviorSerializer)())],
    })
        .deserializeUsing((account) => deserializePresaleBucketV2(account));
}
exports.getPresaleBucketV2GpaBuilder = getPresaleBucketV2GpaBuilder;
function findPresaleBucketV2Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('presale_v2'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.bucketIndex),
    ]);
}
exports.findPresaleBucketV2Pda = findPresaleBucketV2Pda;
async function fetchPresaleBucketV2FromSeeds(context, seeds, options) {
    return fetchPresaleBucketV2(context, findPresaleBucketV2Pda(context, seeds), options);
}
exports.fetchPresaleBucketV2FromSeeds = fetchPresaleBucketV2FromSeeds;
async function safeFetchPresaleBucketV2FromSeeds(context, seeds, options) {
    return safeFetchPresaleBucketV2(context, findPresaleBucketV2Pda(context, seeds), options);
}
exports.safeFetchPresaleBucketV2FromSeeds = safeFetchPresaleBucketV2FromSeeds;
