// @metaplex-foundation/genesis/dist/src/generated/accounts/presaleDepositV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchPresaleDepositV1FromSeeds = exports.fetchPresaleDepositV1FromSeeds = exports.findPresaleDepositV1Pda = exports.getPresaleDepositV1Size = exports.getPresaleDepositV1GpaBuilder = exports.safeFetchAllPresaleDepositV1 = exports.fetchAllPresaleDepositV1 = exports.safeFetchPresaleDepositV1 = exports.fetchPresaleDepositV1 = exports.deserializePresaleDepositV1 = exports.getPresaleDepositV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getPresaleDepositV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['claimed', (0, serializers_1.bool)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 1 })],
        ['index', (0, serializers_1.u32)()],
        ['depositor', (0, serializers_1.publicKey)()],
        ['bucket', (0, serializers_1.publicKey)()],
        ['timestamp', (0, serializers_1.i64)()],
        ['amount', (0, serializers_1.u64)()],
    ], { description: 'PresaleDepositV1AccountData' });
}
exports.getPresaleDepositV1AccountDataSerializer = getPresaleDepositV1AccountDataSerializer;
function deserializePresaleDepositV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getPresaleDepositV1AccountDataSerializer());
}
exports.deserializePresaleDepositV1 = deserializePresaleDepositV1;
async function fetchPresaleDepositV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'PresaleDepositV1');
    return deserializePresaleDepositV1(maybeAccount);
}
exports.fetchPresaleDepositV1 = fetchPresaleDepositV1;
async function safeFetchPresaleDepositV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializePresaleDepositV1(maybeAccount) : null;
}
exports.safeFetchPresaleDepositV1 = safeFetchPresaleDepositV1;
async function fetchAllPresaleDepositV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'PresaleDepositV1');
        return deserializePresaleDepositV1(maybeAccount);
    });
}
exports.fetchAllPresaleDepositV1 = fetchAllPresaleDepositV1;
async function safeFetchAllPresaleDepositV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializePresaleDepositV1(maybeAccount));
}
exports.safeFetchAllPresaleDepositV1 = safeFetchAllPresaleDepositV1;
function getPresaleDepositV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        claimed: [2, (0, serializers_1.bool)()],
        padding: [3, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 1 })],
        index: [4, (0, serializers_1.u32)()],
        depositor: [8, (0, serializers_1.publicKey)()],
        bucket: [40, (0, serializers_1.publicKey)()],
        timestamp: [72, (0, serializers_1.i64)()],
        amount: [80, (0, serializers_1.u64)()],
    })
        .deserializeUsing((account) => deserializePresaleDepositV1(account));
}
exports.getPresaleDepositV1GpaBuilder = getPresaleDepositV1GpaBuilder;
function getPresaleDepositV1Size() {
    return 88;
}
exports.getPresaleDepositV1Size = getPresaleDepositV1Size;
function findPresaleDepositV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('presale_deposit'),
        (0, serializers_1.publicKey)().serialize(seeds.bucket),
        (0, serializers_1.publicKey)().serialize(seeds.recipient),
    ]);
}
exports.findPresaleDepositV1Pda = findPresaleDepositV1Pda;
async function fetchPresaleDepositV1FromSeeds(context, seeds, options) {
    return fetchPresaleDepositV1(context, findPresaleDepositV1Pda(context, seeds), options);
}
exports.fetchPresaleDepositV1FromSeeds = fetchPresaleDepositV1FromSeeds;
async function safeFetchPresaleDepositV1FromSeeds(context, seeds, options) {
    return safeFetchPresaleDepositV1(context, findPresaleDepositV1Pda(context, seeds), options);
}
exports.safeFetchPresaleDepositV1FromSeeds = safeFetchPresaleDepositV1FromSeeds;
