// @metaplex-foundation/genesis/dist/src/generated/accounts/presaleDepositV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchPresaleDepositV2FromSeeds = exports.fetchPresaleDepositV2FromSeeds = exports.findPresaleDepositV2Pda = exports.getPresaleDepositV2Size = exports.getPresaleDepositV2GpaBuilder = exports.safeFetchAllPresaleDepositV2 = exports.fetchAllPresaleDepositV2 = exports.safeFetchPresaleDepositV2 = exports.fetchPresaleDepositV2 = exports.deserializePresaleDepositV2 = exports.getPresaleDepositV2AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getPresaleDepositV2AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['claimed', (0, serializers_1.bool)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 1 })],
        ['index', (0, serializers_1.u32)()],
        ['amountClaimed', (0, serializers_1.u64)()],
        ['depositor', (0, serializers_1.publicKey)()],
        ['bucket', (0, serializers_1.publicKey)()],
        ['timestamp', (0, serializers_1.i64)()],
        ['amountQuoteToken', (0, serializers_1.u64)()],
    ], { description: 'PresaleDepositV2AccountData' });
}
exports.getPresaleDepositV2AccountDataSerializer = getPresaleDepositV2AccountDataSerializer;
function deserializePresaleDepositV2(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getPresaleDepositV2AccountDataSerializer());
}
exports.deserializePresaleDepositV2 = deserializePresaleDepositV2;
async function fetchPresaleDepositV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'PresaleDepositV2');
    return deserializePresaleDepositV2(maybeAccount);
}
exports.fetchPresaleDepositV2 = fetchPresaleDepositV2;
async function safeFetchPresaleDepositV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializePresaleDepositV2(maybeAccount) : null;
}
exports.safeFetchPresaleDepositV2 = safeFetchPresaleDepositV2;
async function fetchAllPresaleDepositV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'PresaleDepositV2');
        return deserializePresaleDepositV2(maybeAccount);
    });
}
exports.fetchAllPresaleDepositV2 = fetchAllPresaleDepositV2;
async function safeFetchAllPresaleDepositV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializePresaleDepositV2(maybeAccount));
}
exports.safeFetchAllPresaleDepositV2 = safeFetchAllPresaleDepositV2;
function getPresaleDepositV2GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        claimed: [2, (0, serializers_1.bool)()],
        padding: [3, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 1 })],
        index: [4, (0, serializers_1.u32)()],
        amountClaimed: [8, (0, serializers_1.u64)()],
        depositor: [16, (0, serializers_1.publicKey)()],
        bucket: [48, (0, serializers_1.publicKey)()],
        timestamp: [80, (0, serializers_1.i64)()],
        amountQuoteToken: [88, (0, serializers_1.u64)()],
    })
        .deserializeUsing((account) => deserializePresaleDepositV2(account));
}
exports.getPresaleDepositV2GpaBuilder = getPresaleDepositV2GpaBuilder;
function getPresaleDepositV2Size() {
    return 96;
}
exports.getPresaleDepositV2Size = getPresaleDepositV2Size;
function findPresaleDepositV2Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('presale_deposit_v2'),
        (0, serializers_1.publicKey)().serialize(seeds.bucket),
        (0, serializers_1.publicKey)().serialize(seeds.recipient),
    ]);
}
exports.findPresaleDepositV2Pda = findPresaleDepositV2Pda;
async function fetchPresaleDepositV2FromSeeds(context, seeds, options) {
    return fetchPresaleDepositV2(context, findPresaleDepositV2Pda(context, seeds), options);
}
exports.fetchPresaleDepositV2FromSeeds = fetchPresaleDepositV2FromSeeds;
async function safeFetchPresaleDepositV2FromSeeds(context, seeds, options) {
    return safeFetchPresaleDepositV2(context, findPresaleDepositV2Pda(context, seeds), options);
}
exports.safeFetchPresaleDepositV2FromSeeds = safeFetchPresaleDepositV2FromSeeds;
