// @metaplex-foundation/genesis/dist/src/generated/accounts/pumpBucketV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchPumpBucketV1FromSeeds = exports.fetchPumpBucketV1FromSeeds = exports.findPumpBucketV1Pda = exports.getPumpBucketV1Size = exports.getPumpBucketV1GpaBuilder = exports.safeFetchAllPumpBucketV1 = exports.fetchAllPumpBucketV1 = exports.safeFetchPumpBucketV1 = exports.fetchPumpBucketV1 = exports.deserializePumpBucketV1 = exports.getPumpBucketV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getPumpBucketV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['state', (0, types_1.getPumpBucketStateSerializer)()],
        ['index', (0, serializers_1.u8)()],
        ['feeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 3 })],
        ['genesis', (0, serializers_1.publicKey)()],
        ['tokenAllocation', (0, serializers_1.u64)()],
        ['solBalance', (0, serializers_1.u64)()],
        ['tokenBalance', (0, serializers_1.u64)()],
        ['start', (0, serializers_1.i64)()],
        ['backendSigner', (0, serializers_1.publicKey)()],
        ['fee', (0, serializers_1.u64)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    ], { description: 'PumpBucketV1AccountData' });
}
exports.getPumpBucketV1AccountDataSerializer = getPumpBucketV1AccountDataSerializer;
function deserializePumpBucketV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getPumpBucketV1AccountDataSerializer());
}
exports.deserializePumpBucketV1 = deserializePumpBucketV1;
async function fetchPumpBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'PumpBucketV1');
    return deserializePumpBucketV1(maybeAccount);
}
exports.fetchPumpBucketV1 = fetchPumpBucketV1;
async function safeFetchPumpBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializePumpBucketV1(maybeAccount) : null;
}
exports.safeFetchPumpBucketV1 = safeFetchPumpBucketV1;
async function fetchAllPumpBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'PumpBucketV1');
        return deserializePumpBucketV1(maybeAccount);
    });
}
exports.fetchAllPumpBucketV1 = fetchAllPumpBucketV1;
async function safeFetchAllPumpBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializePumpBucketV1(maybeAccount));
}
exports.safeFetchAllPumpBucketV1 = safeFetchAllPumpBucketV1;
function getPumpBucketV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        state: [2, (0, types_1.getPumpBucketStateSerializer)()],
        index: [3, (0, serializers_1.u8)()],
        feeType: [4, (0, types_1.getFeeDiscriminantsSerializer)()],
        padding: [5, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 3 })],
        genesis: [8, (0, serializers_1.publicKey)()],
        tokenAllocation: [40, (0, serializers_1.u64)()],
        solBalance: [48, (0, serializers_1.u64)()],
        tokenBalance: [56, (0, serializers_1.u64)()],
        start: [64, (0, serializers_1.i64)()],
        backendSigner: [72, (0, serializers_1.publicKey)()],
        fee: [104, (0, serializers_1.u64)()],
        reserved: [112, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    })
        .deserializeUsing((account) => deserializePumpBucketV1(account));
}
exports.getPumpBucketV1GpaBuilder = getPumpBucketV1GpaBuilder;
function getPumpBucketV1Size() {
    return 240;
}
exports.getPumpBucketV1Size = getPumpBucketV1Size;
function findPumpBucketV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('pump'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.index),
    ]);
}
exports.findPumpBucketV1Pda = findPumpBucketV1Pda;
async function fetchPumpBucketV1FromSeeds(context, seeds, options) {
    return fetchPumpBucketV1(context, findPumpBucketV1Pda(context, seeds), options);
}
exports.fetchPumpBucketV1FromSeeds = fetchPumpBucketV1FromSeeds;
async function safeFetchPumpBucketV1FromSeeds(context, seeds, options) {
    return safeFetchPumpBucketV1(context, findPumpBucketV1Pda(context, seeds), options);
}
exports.safeFetchPumpBucketV1FromSeeds = safeFetchPumpBucketV1FromSeeds;
