// @metaplex-foundation/genesis/dist/src/generated/accounts/raydiumCpmmBucketV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchRaydiumCpmmBucketV2FromSeeds = exports.fetchRaydiumCpmmBucketV2FromSeeds = exports.findRaydiumCpmmBucketV2Pda = exports.getRaydiumCpmmBucketV2GpaBuilder = exports.safeFetchAllRaydiumCpmmBucketV2 = exports.fetchAllRaydiumCpmmBucketV2 = exports.safeFetchRaydiumCpmmBucketV2 = exports.fetchRaydiumCpmmBucketV2 = exports.deserializeRaydiumCpmmBucketV2 = exports.getRaydiumCpmmBucketV2AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getRaydiumCpmmBucketV2AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bucketSignerBump', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        ['bucket', (0, types_1.getBucketBaseSerializer)()],
        ['bucketSigner', (0, serializers_1.publicKey)()],
        ['quoteTokenInSigner', (0, serializers_1.u64)()],
        ['priceNumerator', (0, serializers_1.u64)()],
        ['priceDenominator', (0, serializers_1.u64)()],
        ['graduatedBaseTokenAmount', (0, serializers_1.u64)()],
        ['lpTokenBalance', (0, serializers_1.u64)()],
        [
            'lpClaimAuthority',
            (() => {
                const inner = (0, serializers_1.publicKey)();
                const sentinel = new Uint8Array([
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<PublicKey>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        ['lpTokensClaimed', (0, serializers_1.u64)()],
        ['startCondition', (0, types_1.getConditionSerializer)()],
        ['graduateCondition', (0, types_1.getConditionSerializer)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        ['extensions', (0, types_1.getRaydiumCpmmV2ExtensionsSerializer)()],
        ['padding2', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        ['endBehaviors', (0, serializers_1.array)((0, types_1.getBehaviorSerializer)())],
    ], { description: 'RaydiumCpmmBucketV2AccountData' });
}
exports.getRaydiumCpmmBucketV2AccountDataSerializer = getRaydiumCpmmBucketV2AccountDataSerializer;
function deserializeRaydiumCpmmBucketV2(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getRaydiumCpmmBucketV2AccountDataSerializer());
}
exports.deserializeRaydiumCpmmBucketV2 = deserializeRaydiumCpmmBucketV2;
async function fetchRaydiumCpmmBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'RaydiumCpmmBucketV2');
    return deserializeRaydiumCpmmBucketV2(maybeAccount);
}
exports.fetchRaydiumCpmmBucketV2 = fetchRaydiumCpmmBucketV2;
async function safeFetchRaydiumCpmmBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists
        ? deserializeRaydiumCpmmBucketV2(maybeAccount)
        : null;
}
exports.safeFetchRaydiumCpmmBucketV2 = safeFetchRaydiumCpmmBucketV2;
async function fetchAllRaydiumCpmmBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'RaydiumCpmmBucketV2');
        return deserializeRaydiumCpmmBucketV2(maybeAccount);
    });
}
exports.fetchAllRaydiumCpmmBucketV2 = fetchAllRaydiumCpmmBucketV2;
async function safeFetchAllRaydiumCpmmBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeRaydiumCpmmBucketV2(maybeAccount));
}
exports.safeFetchAllRaydiumCpmmBucketV2 = safeFetchAllRaydiumCpmmBucketV2;
function getRaydiumCpmmBucketV2GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bucketSignerBump: [1, (0, serializers_1.u8)()],
        padding: [2, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        bucket: [8, (0, types_1.getBucketBaseSerializer)()],
        bucketSigner: [208, (0, serializers_1.publicKey)()],
        quoteTokenInSigner: [240, (0, serializers_1.u64)()],
        priceNumerator: [248, (0, serializers_1.u64)()],
        priceDenominator: [256, (0, serializers_1.u64)()],
        graduatedBaseTokenAmount: [264, (0, serializers_1.u64)()],
        lpTokenBalance: [272, (0, serializers_1.u64)()],
        lpClaimAuthority: [
            280,
            (() => {
                const inner = (0, serializers_1.publicKey)();
                const sentinel = new Uint8Array([
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<PublicKey>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        lpTokensClaimed: [312, (0, serializers_1.u64)()],
        startCondition: [320, (0, types_1.getConditionSerializer)()],
        graduateCondition: [384, (0, types_1.getConditionSerializer)()],
        reserved: [448, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        extensions: [576, (0, types_1.getRaydiumCpmmV2ExtensionsSerializer)()],
        padding2: [1016, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        endBehaviors: [1020, (0, serializers_1.array)((0, types_1.getBehaviorSerializer)())],
    })
        .deserializeUsing((account) => deserializeRaydiumCpmmBucketV2(account));
}
exports.getRaydiumCpmmBucketV2GpaBuilder = getRaydiumCpmmBucketV2GpaBuilder;
function findRaydiumCpmmBucketV2Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('raydium_cpmm_v2'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.bucketIndex),
    ]);
}
exports.findRaydiumCpmmBucketV2Pda = findRaydiumCpmmBucketV2Pda;
async function fetchRaydiumCpmmBucketV2FromSeeds(context, seeds, options) {
    return fetchRaydiumCpmmBucketV2(context, findRaydiumCpmmBucketV2Pda(context, seeds), options);
}
exports.fetchRaydiumCpmmBucketV2FromSeeds = fetchRaydiumCpmmBucketV2FromSeeds;
async function safeFetchRaydiumCpmmBucketV2FromSeeds(context, seeds, options) {
    return safeFetchRaydiumCpmmBucketV2(context, findRaydiumCpmmBucketV2Pda(context, seeds), options);
}
exports.safeFetchRaydiumCpmmBucketV2FromSeeds = safeFetchRaydiumCpmmBucketV2FromSeeds;
