// @metaplex-foundation/genesis/dist/src/generated/accounts/raydiumSigner.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchRaydiumSignerFromSeeds = exports.fetchRaydiumSignerFromSeeds = exports.findRaydiumSignerPda = exports.getRaydiumSignerSize = exports.getRaydiumSignerGpaBuilder = exports.safeFetchAllRaydiumSigner = exports.fetchAllRaydiumSigner = exports.safeFetchRaydiumSigner = exports.fetchRaydiumSigner = exports.deserializeRaydiumSigner = exports.getRaydiumSignerAccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
function getRaydiumSignerAccountDataSerializer() {
    return (0, serializers_1.struct)([['data', (0, serializers_1.bytes)()]], {
        description: 'RaydiumSignerAccountData',
    });
}
exports.getRaydiumSignerAccountDataSerializer = getRaydiumSignerAccountDataSerializer;
function deserializeRaydiumSigner(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getRaydiumSignerAccountDataSerializer());
}
exports.deserializeRaydiumSigner = deserializeRaydiumSigner;
async function fetchRaydiumSigner(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'RaydiumSigner');
    return deserializeRaydiumSigner(maybeAccount);
}
exports.fetchRaydiumSigner = fetchRaydiumSigner;
async function safeFetchRaydiumSigner(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeRaydiumSigner(maybeAccount) : null;
}
exports.safeFetchRaydiumSigner = safeFetchRaydiumSigner;
async function fetchAllRaydiumSigner(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'RaydiumSigner');
        return deserializeRaydiumSigner(maybeAccount);
    });
}
exports.fetchAllRaydiumSigner = fetchAllRaydiumSigner;
async function safeFetchAllRaydiumSigner(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeRaydiumSigner(maybeAccount));
}
exports.safeFetchAllRaydiumSigner = safeFetchAllRaydiumSigner;
function getRaydiumSignerGpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({ data: [0, (0, serializers_1.bytes)()] })
        .deserializeUsing((account) => deserializeRaydiumSigner(account));
}
exports.getRaydiumSignerGpaBuilder = getRaydiumSignerGpaBuilder;
function getRaydiumSignerSize() {
    return 0;
}
exports.getRaydiumSignerSize = getRaydiumSignerSize;
function findRaydiumSignerPda(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('raydium_signer'),
    ]);
}
exports.findRaydiumSignerPda = findRaydiumSignerPda;
async function fetchRaydiumSignerFromSeeds(context, options) {
    return fetchRaydiumSigner(context, findRaydiumSignerPda(context), options);
}
exports.fetchRaydiumSignerFromSeeds = fetchRaydiumSignerFromSeeds;
async function safeFetchRaydiumSignerFromSeeds(context, options) {
    return safeFetchRaydiumSigner(context, findRaydiumSignerPda(context), options);
}
exports.safeFetchRaydiumSignerFromSeeds = safeFetchRaydiumSignerFromSeeds;
