// @metaplex-foundation/genesis/dist/src/generated/accounts/raydiumCpmmBucketV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchRaydiumCpmmBucketV1FromSeeds = exports.fetchRaydiumCpmmBucketV1FromSeeds = exports.findRaydiumCpmmBucketV1Pda = exports.getRaydiumCpmmBucketV1Size = exports.getRaydiumCpmmBucketV1GpaBuilder = exports.safeFetchAllRaydiumCpmmBucketV1 = exports.fetchAllRaydiumCpmmBucketV1 = exports.safeFetchRaydiumCpmmBucketV1 = exports.fetchRaydiumCpmmBucketV1 = exports.deserializeRaydiumCpmmBucketV1 = exports.getRaydiumCpmmBucketV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getRaydiumCpmmBucketV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['state', (0, types_1.getRaydiumCpmmBucketStateSerializer)()],
        ['index', (0, serializers_1.u8)()],
        ['feeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['bucketSignerBump', (0, serializers_1.u8)()],
        ['solAllocationBps', (0, serializers_1.u16)()],
        ['genesis', (0, serializers_1.publicKey)()],
        ['bucketSigner', (0, serializers_1.publicKey)()],
        ['tokenAllocation', (0, serializers_1.u64)()],
        ['solInSigner', (0, serializers_1.u64)()],
        ['tokenBalance', (0, serializers_1.u64)()],
        ['start', (0, serializers_1.i64)()],
        ['backendSigner', (0, serializers_1.publicKey)()],
        ['fee', (0, serializers_1.u64)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    ], { description: 'RaydiumCpmmBucketV1AccountData' });
}
exports.getRaydiumCpmmBucketV1AccountDataSerializer = getRaydiumCpmmBucketV1AccountDataSerializer;
function deserializeRaydiumCpmmBucketV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getRaydiumCpmmBucketV1AccountDataSerializer());
}
exports.deserializeRaydiumCpmmBucketV1 = deserializeRaydiumCpmmBucketV1;
async function fetchRaydiumCpmmBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'RaydiumCpmmBucketV1');
    return deserializeRaydiumCpmmBucketV1(maybeAccount);
}
exports.fetchRaydiumCpmmBucketV1 = fetchRaydiumCpmmBucketV1;
async function safeFetchRaydiumCpmmBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists
        ? deserializeRaydiumCpmmBucketV1(maybeAccount)
        : null;
}
exports.safeFetchRaydiumCpmmBucketV1 = safeFetchRaydiumCpmmBucketV1;
async function fetchAllRaydiumCpmmBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'RaydiumCpmmBucketV1');
        return deserializeRaydiumCpmmBucketV1(maybeAccount);
    });
}
exports.fetchAllRaydiumCpmmBucketV1 = fetchAllRaydiumCpmmBucketV1;
async function safeFetchAllRaydiumCpmmBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeRaydiumCpmmBucketV1(maybeAccount));
}
exports.safeFetchAllRaydiumCpmmBucketV1 = safeFetchAllRaydiumCpmmBucketV1;
function getRaydiumCpmmBucketV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        state: [2, (0, types_1.getRaydiumCpmmBucketStateSerializer)()],
        index: [3, (0, serializers_1.u8)()],
        feeType: [4, (0, types_1.getFeeDiscriminantsSerializer)()],
        bucketSignerBump: [5, (0, serializers_1.u8)()],
        solAllocationBps: [6, (0, serializers_1.u16)()],
        genesis: [8, (0, serializers_1.publicKey)()],
        bucketSigner: [40, (0, serializers_1.publicKey)()],
        tokenAllocation: [72, (0, serializers_1.u64)()],
        solInSigner: [80, (0, serializers_1.u64)()],
        tokenBalance: [88, (0, serializers_1.u64)()],
        start: [96, (0, serializers_1.i64)()],
        backendSigner: [104, (0, serializers_1.publicKey)()],
        fee: [136, (0, serializers_1.u64)()],
        reserved: [144, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    })
        .deserializeUsing((account) => deserializeRaydiumCpmmBucketV1(account));
}
exports.getRaydiumCpmmBucketV1GpaBuilder = getRaydiumCpmmBucketV1GpaBuilder;
function getRaydiumCpmmBucketV1Size() {
    return 272;
}
exports.getRaydiumCpmmBucketV1Size = getRaydiumCpmmBucketV1Size;
function findRaydiumCpmmBucketV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('raydium_cpmm'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.index),
    ]);
}
exports.findRaydiumCpmmBucketV1Pda = findRaydiumCpmmBucketV1Pda;
async function fetchRaydiumCpmmBucketV1FromSeeds(context, seeds, options) {
    return fetchRaydiumCpmmBucketV1(context, findRaydiumCpmmBucketV1Pda(context, seeds), options);
}
exports.fetchRaydiumCpmmBucketV1FromSeeds = fetchRaydiumCpmmBucketV1FromSeeds;
async function safeFetchRaydiumCpmmBucketV1FromSeeds(context, seeds, options) {
    return safeFetchRaydiumCpmmBucketV1(context, findRaydiumCpmmBucketV1Pda(context, seeds), options);
}
exports.safeFetchRaydiumCpmmBucketV1FromSeeds = safeFetchRaydiumCpmmBucketV1FromSeeds;
