// @metaplex-foundation/genesis/dist/src/generated/accounts/raydiumBucketSigner.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchRaydiumBucketSignerFromSeeds = exports.fetchRaydiumBucketSignerFromSeeds = exports.findRaydiumBucketSignerPda = exports.getRaydiumBucketSignerSize = exports.getRaydiumBucketSignerGpaBuilder = exports.safeFetchAllRaydiumBucketSigner = exports.fetchAllRaydiumBucketSigner = exports.safeFetchRaydiumBucketSigner = exports.fetchRaydiumBucketSigner = exports.deserializeRaydiumBucketSigner = exports.getRaydiumBucketSignerAccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
function getRaydiumBucketSignerAccountDataSerializer() {
    return (0, serializers_1.struct)([['data', (0, serializers_1.bytes)()]], {
        description: 'RaydiumBucketSignerAccountData',
    });
}
exports.getRaydiumBucketSignerAccountDataSerializer = getRaydiumBucketSignerAccountDataSerializer;
function deserializeRaydiumBucketSigner(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getRaydiumBucketSignerAccountDataSerializer());
}
exports.deserializeRaydiumBucketSigner = deserializeRaydiumBucketSigner;
async function fetchRaydiumBucketSigner(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'RaydiumBucketSigner');
    return deserializeRaydiumBucketSigner(maybeAccount);
}
exports.fetchRaydiumBucketSigner = fetchRaydiumBucketSigner;
async function safeFetchRaydiumBucketSigner(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists
        ? deserializeRaydiumBucketSigner(maybeAccount)
        : null;
}
exports.safeFetchRaydiumBucketSigner = safeFetchRaydiumBucketSigner;
async function fetchAllRaydiumBucketSigner(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'RaydiumBucketSigner');
        return deserializeRaydiumBucketSigner(maybeAccount);
    });
}
exports.fetchAllRaydiumBucketSigner = fetchAllRaydiumBucketSigner;
async function safeFetchAllRaydiumBucketSigner(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeRaydiumBucketSigner(maybeAccount));
}
exports.safeFetchAllRaydiumBucketSigner = safeFetchAllRaydiumBucketSigner;
function getRaydiumBucketSignerGpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({ data: [0, (0, serializers_1.bytes)()] })
        .deserializeUsing((account) => deserializeRaydiumBucketSigner(account));
}
exports.getRaydiumBucketSignerGpaBuilder = getRaydiumBucketSignerGpaBuilder;
function getRaydiumBucketSignerSize() {
    return 0;
}
exports.getRaydiumBucketSignerSize = getRaydiumBucketSignerSize;
function findRaydiumBucketSignerPda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('bucket_signer'),
        (0, serializers_1.publicKey)().serialize(seeds.bucket),
    ]);
}
exports.findRaydiumBucketSignerPda = findRaydiumBucketSignerPda;
async function fetchRaydiumBucketSignerFromSeeds(context, seeds, options) {
    return fetchRaydiumBucketSigner(context, findRaydiumBucketSignerPda(context, seeds), options);
}
exports.fetchRaydiumBucketSignerFromSeeds = fetchRaydiumBucketSignerFromSeeds;
async function safeFetchRaydiumBucketSignerFromSeeds(context, seeds, options) {
    return safeFetchRaydiumBucketSigner(context, findRaydiumBucketSignerPda(context, seeds), options);
}
exports.safeFetchRaydiumBucketSignerFromSeeds = safeFetchRaydiumBucketSignerFromSeeds;
