// @metaplex-foundation/genesis/dist/src/generated/accounts/streamflowMetadata.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchStreamflowMetadataFromSeeds = exports.fetchStreamflowMetadataFromSeeds = exports.findStreamflowMetadataPda = exports.getStreamflowMetadataSize = exports.getStreamflowMetadataGpaBuilder = exports.safeFetchAllStreamflowMetadata = exports.fetchAllStreamflowMetadata = exports.safeFetchStreamflowMetadata = exports.fetchStreamflowMetadata = exports.deserializeStreamflowMetadata = exports.getStreamflowMetadataAccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
function getStreamflowMetadataAccountDataSerializer() {
    return (0, serializers_1.struct)([['data', (0, serializers_1.bytes)()]], {
        description: 'StreamflowMetadataAccountData',
    });
}
exports.getStreamflowMetadataAccountDataSerializer = getStreamflowMetadataAccountDataSerializer;
function deserializeStreamflowMetadata(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getStreamflowMetadataAccountDataSerializer());
}
exports.deserializeStreamflowMetadata = deserializeStreamflowMetadata;
async function fetchStreamflowMetadata(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'StreamflowMetadata');
    return deserializeStreamflowMetadata(maybeAccount);
}
exports.fetchStreamflowMetadata = fetchStreamflowMetadata;
async function safeFetchStreamflowMetadata(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists
        ? deserializeStreamflowMetadata(maybeAccount)
        : null;
}
exports.safeFetchStreamflowMetadata = safeFetchStreamflowMetadata;
async function fetchAllStreamflowMetadata(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'StreamflowMetadata');
        return deserializeStreamflowMetadata(maybeAccount);
    });
}
exports.fetchAllStreamflowMetadata = fetchAllStreamflowMetadata;
async function safeFetchAllStreamflowMetadata(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeStreamflowMetadata(maybeAccount));
}
exports.safeFetchAllStreamflowMetadata = safeFetchAllStreamflowMetadata;
function getStreamflowMetadataGpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({ data: [0, (0, serializers_1.bytes)()] })
        .deserializeUsing((account) => deserializeStreamflowMetadata(account));
}
exports.getStreamflowMetadataGpaBuilder = getStreamflowMetadataGpaBuilder;
function getStreamflowMetadataSize() {
    return 0;
}
exports.getStreamflowMetadataSize = getStreamflowMetadataSize;
function findStreamflowMetadataPda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('streamflow_metadata'),
        (0, serializers_1.publicKey)().serialize(seeds.bucket),
    ]);
}
exports.findStreamflowMetadataPda = findStreamflowMetadataPda;
async function fetchStreamflowMetadataFromSeeds(context, seeds, options) {
    return fetchStreamflowMetadata(context, findStreamflowMetadataPda(context, seeds), options);
}
exports.fetchStreamflowMetadataFromSeeds = fetchStreamflowMetadataFromSeeds;
async function safeFetchStreamflowMetadataFromSeeds(context, seeds, options) {
    return safeFetchStreamflowMetadata(context, findStreamflowMetadataPda(context, seeds), options);
}
exports.safeFetchStreamflowMetadataFromSeeds = safeFetchStreamflowMetadataFromSeeds;
