// @metaplex-foundation/genesis/dist/src/generated/accounts/streamflowBucketV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchStreamflowBucketV2FromSeeds = exports.fetchStreamflowBucketV2FromSeeds = exports.findStreamflowBucketV2Pda = exports.getStreamflowBucketV2Size = exports.getStreamflowBucketV2GpaBuilder = exports.safeFetchAllStreamflowBucketV2 = exports.fetchAllStreamflowBucketV2 = exports.safeFetchStreamflowBucketV2 = exports.fetchStreamflowBucketV2 = exports.deserializeStreamflowBucketV2 = exports.getStreamflowBucketV2AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getStreamflowBucketV2AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['locked', (0, serializers_1.bool)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        ['bucket', (0, types_1.getBucketBaseSerializer)()],
        ['recipient', (0, serializers_1.publicKey)()],
        ['config', (0, types_1.getStreamflowConfigSerializer)()],
        ['lockStartCondition', (0, types_1.getConditionSerializer)()],
        ['lockEndCondition', (0, types_1.getConditionSerializer)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        ['extensions', (0, types_1.getStreamflowV2ExtensionsSerializer)()],
    ], { description: 'StreamflowBucketV2AccountData' });
}
exports.getStreamflowBucketV2AccountDataSerializer = getStreamflowBucketV2AccountDataSerializer;
function deserializeStreamflowBucketV2(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getStreamflowBucketV2AccountDataSerializer());
}
exports.deserializeStreamflowBucketV2 = deserializeStreamflowBucketV2;
async function fetchStreamflowBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'StreamflowBucketV2');
    return deserializeStreamflowBucketV2(maybeAccount);
}
exports.fetchStreamflowBucketV2 = fetchStreamflowBucketV2;
async function safeFetchStreamflowBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists
        ? deserializeStreamflowBucketV2(maybeAccount)
        : null;
}
exports.safeFetchStreamflowBucketV2 = safeFetchStreamflowBucketV2;
async function fetchAllStreamflowBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'StreamflowBucketV2');
        return deserializeStreamflowBucketV2(maybeAccount);
    });
}
exports.fetchAllStreamflowBucketV2 = fetchAllStreamflowBucketV2;
async function safeFetchAllStreamflowBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeStreamflowBucketV2(maybeAccount));
}
exports.safeFetchAllStreamflowBucketV2 = safeFetchAllStreamflowBucketV2;
function getStreamflowBucketV2GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        locked: [1, (0, serializers_1.bool)()],
        padding: [2, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        bucket: [8, (0, types_1.getBucketBaseSerializer)()],
        recipient: [208, (0, serializers_1.publicKey)()],
        config: [240, (0, types_1.getStreamflowConfigSerializer)()],
        lockStartCondition: [360, (0, types_1.getConditionSerializer)()],
        lockEndCondition: [424, (0, types_1.getConditionSerializer)()],
        reserved: [488, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        extensions: [616, (0, types_1.getStreamflowV2ExtensionsSerializer)()],
    })
        .deserializeUsing((account) => deserializeStreamflowBucketV2(account));
}
exports.getStreamflowBucketV2GpaBuilder = getStreamflowBucketV2GpaBuilder;
function getStreamflowBucketV2Size() {
    return 784;
}
exports.getStreamflowBucketV2Size = getStreamflowBucketV2Size;
function findStreamflowBucketV2Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('streamflow_v2'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.bucketIndex),
    ]);
}
exports.findStreamflowBucketV2Pda = findStreamflowBucketV2Pda;
async function fetchStreamflowBucketV2FromSeeds(context, seeds, options) {
    return fetchStreamflowBucketV2(context, findStreamflowBucketV2Pda(context, seeds), options);
}
exports.fetchStreamflowBucketV2FromSeeds = fetchStreamflowBucketV2FromSeeds;
async function safeFetchStreamflowBucketV2FromSeeds(context, seeds, options) {
    return safeFetchStreamflowBucketV2(context, findStreamflowBucketV2Pda(context, seeds), options);
}
exports.safeFetchStreamflowBucketV2FromSeeds = safeFetchStreamflowBucketV2FromSeeds;
