// @metaplex-foundation/genesis/dist/src/generated/accounts/streamflowBucketV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchStreamflowBucketV1FromSeeds = exports.fetchStreamflowBucketV1FromSeeds = exports.findStreamflowBucketV1Pda = exports.getStreamflowBucketV1Size = exports.getStreamflowBucketV1GpaBuilder = exports.safeFetchAllStreamflowBucketV1 = exports.fetchAllStreamflowBucketV1 = exports.safeFetchStreamflowBucketV1 = exports.fetchStreamflowBucketV1 = exports.deserializeStreamflowBucketV1 = exports.getStreamflowBucketV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getStreamflowBucketV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['state', (0, types_1.getStreamflowBucketStateSerializer)()],
        ['index', (0, serializers_1.u8)()],
        ['feeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 3 })],
        ['genesis', (0, serializers_1.publicKey)()],
        ['tokenAllocation', (0, serializers_1.u64)()],
        ['solBalance', (0, serializers_1.u64)()],
        ['tokenBalance', (0, serializers_1.u64)()],
        ['recipient', (0, serializers_1.publicKey)()],
        ['config', (0, types_1.getStreamflowConfigSerializer)()],
        ['backendSigner', (0, serializers_1.publicKey)()],
        ['fee', (0, serializers_1.u64)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    ], { description: 'StreamflowBucketV1AccountData' });
}
exports.getStreamflowBucketV1AccountDataSerializer = getStreamflowBucketV1AccountDataSerializer;
function deserializeStreamflowBucketV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getStreamflowBucketV1AccountDataSerializer());
}
exports.deserializeStreamflowBucketV1 = deserializeStreamflowBucketV1;
async function fetchStreamflowBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'StreamflowBucketV1');
    return deserializeStreamflowBucketV1(maybeAccount);
}
exports.fetchStreamflowBucketV1 = fetchStreamflowBucketV1;
async function safeFetchStreamflowBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists
        ? deserializeStreamflowBucketV1(maybeAccount)
        : null;
}
exports.safeFetchStreamflowBucketV1 = safeFetchStreamflowBucketV1;
async function fetchAllStreamflowBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'StreamflowBucketV1');
        return deserializeStreamflowBucketV1(maybeAccount);
    });
}
exports.fetchAllStreamflowBucketV1 = fetchAllStreamflowBucketV1;
async function safeFetchAllStreamflowBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeStreamflowBucketV1(maybeAccount));
}
exports.safeFetchAllStreamflowBucketV1 = safeFetchAllStreamflowBucketV1;
function getStreamflowBucketV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        state: [2, (0, types_1.getStreamflowBucketStateSerializer)()],
        index: [3, (0, serializers_1.u8)()],
        feeType: [4, (0, types_1.getFeeDiscriminantsSerializer)()],
        padding: [5, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 3 })],
        genesis: [8, (0, serializers_1.publicKey)()],
        tokenAllocation: [40, (0, serializers_1.u64)()],
        solBalance: [48, (0, serializers_1.u64)()],
        tokenBalance: [56, (0, serializers_1.u64)()],
        recipient: [64, (0, serializers_1.publicKey)()],
        config: [96, (0, types_1.getStreamflowConfigSerializer)()],
        backendSigner: [216, (0, serializers_1.publicKey)()],
        fee: [248, (0, serializers_1.u64)()],
        reserved: [256, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    })
        .deserializeUsing((account) => deserializeStreamflowBucketV1(account));
}
exports.getStreamflowBucketV1GpaBuilder = getStreamflowBucketV1GpaBuilder;
function getStreamflowBucketV1Size() {
    return 384;
}
exports.getStreamflowBucketV1Size = getStreamflowBucketV1Size;
function findStreamflowBucketV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('streamflow'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.index),
    ]);
}
exports.findStreamflowBucketV1Pda = findStreamflowBucketV1Pda;
async function fetchStreamflowBucketV1FromSeeds(context, seeds, options) {
    return fetchStreamflowBucketV1(context, findStreamflowBucketV1Pda(context, seeds), options);
}
exports.fetchStreamflowBucketV1FromSeeds = fetchStreamflowBucketV1FromSeeds;
async function safeFetchStreamflowBucketV1FromSeeds(context, seeds, options) {
    return safeFetchStreamflowBucketV1(context, findStreamflowBucketV1Pda(context, seeds), options);
}
exports.safeFetchStreamflowBucketV1FromSeeds = safeFetchStreamflowBucketV1FromSeeds;
