// @metaplex-foundation/genesis/dist/src/generated/accounts/unlockedBucketV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchUnlockedBucketV2FromSeeds = exports.fetchUnlockedBucketV2FromSeeds = exports.findUnlockedBucketV2Pda = exports.getUnlockedBucketV2Size = exports.getUnlockedBucketV2GpaBuilder = exports.safeFetchAllUnlockedBucketV2 = exports.fetchAllUnlockedBucketV2 = exports.safeFetchUnlockedBucketV2 = exports.fetchUnlockedBucketV2 = exports.deserializeUnlockedBucketV2 = exports.getUnlockedBucketV2AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getUnlockedBucketV2AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['claimed', (0, serializers_1.bool)()],
        ['claimFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
        ['claimFee', (0, serializers_1.u64)()],
        ['bucket', (0, types_1.getBucketBaseSerializer)()],
        ['recipient', (0, serializers_1.publicKey)()],
        ['claimStartCondition', (0, types_1.getConditionSerializer)()],
        ['claimEndCondition', (0, types_1.getConditionSerializer)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        ['extensions', (0, types_1.getUnlockedV2ExtensionsSerializer)()],
    ], { description: 'UnlockedBucketV2AccountData' });
}
exports.getUnlockedBucketV2AccountDataSerializer = getUnlockedBucketV2AccountDataSerializer;
function deserializeUnlockedBucketV2(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getUnlockedBucketV2AccountDataSerializer());
}
exports.deserializeUnlockedBucketV2 = deserializeUnlockedBucketV2;
async function fetchUnlockedBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'UnlockedBucketV2');
    return deserializeUnlockedBucketV2(maybeAccount);
}
exports.fetchUnlockedBucketV2 = fetchUnlockedBucketV2;
async function safeFetchUnlockedBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeUnlockedBucketV2(maybeAccount) : null;
}
exports.safeFetchUnlockedBucketV2 = safeFetchUnlockedBucketV2;
async function fetchAllUnlockedBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'UnlockedBucketV2');
        return deserializeUnlockedBucketV2(maybeAccount);
    });
}
exports.fetchAllUnlockedBucketV2 = fetchAllUnlockedBucketV2;
async function safeFetchAllUnlockedBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeUnlockedBucketV2(maybeAccount));
}
exports.safeFetchAllUnlockedBucketV2 = safeFetchAllUnlockedBucketV2;
function getUnlockedBucketV2GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        claimed: [1, (0, serializers_1.bool)()],
        claimFeeType: [2, (0, types_1.getFeeDiscriminantsSerializer)()],
        padding: [3, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
        claimFee: [8, (0, serializers_1.u64)()],
        bucket: [16, (0, types_1.getBucketBaseSerializer)()],
        recipient: [216, (0, serializers_1.publicKey)()],
        claimStartCondition: [248, (0, types_1.getConditionSerializer)()],
        claimEndCondition: [312, (0, types_1.getConditionSerializer)()],
        reserved: [376, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        extensions: [504, (0, types_1.getUnlockedV2ExtensionsSerializer)()],
    })
        .deserializeUsing((account) => deserializeUnlockedBucketV2(account));
}
exports.getUnlockedBucketV2GpaBuilder = getUnlockedBucketV2GpaBuilder;
function getUnlockedBucketV2Size() {
    return 672;
}
exports.getUnlockedBucketV2Size = getUnlockedBucketV2Size;
function findUnlockedBucketV2Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('unlocked_v2'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.bucketIndex),
    ]);
}
exports.findUnlockedBucketV2Pda = findUnlockedBucketV2Pda;
async function fetchUnlockedBucketV2FromSeeds(context, seeds, options) {
    return fetchUnlockedBucketV2(context, findUnlockedBucketV2Pda(context, seeds), options);
}
exports.fetchUnlockedBucketV2FromSeeds = fetchUnlockedBucketV2FromSeeds;
async function safeFetchUnlockedBucketV2FromSeeds(context, seeds, options) {
    return safeFetchUnlockedBucketV2(context, findUnlockedBucketV2Pda(context, seeds), options);
}
exports.safeFetchUnlockedBucketV2FromSeeds = safeFetchUnlockedBucketV2FromSeeds;
