// @metaplex-foundation/genesis/dist/src/generated/accounts/unlockedBucketV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchUnlockedBucketV1FromSeeds = exports.fetchUnlockedBucketV1FromSeeds = exports.findUnlockedBucketV1Pda = exports.getUnlockedBucketV1Size = exports.getUnlockedBucketV1GpaBuilder = exports.safeFetchAllUnlockedBucketV1 = exports.fetchAllUnlockedBucketV1 = exports.safeFetchUnlockedBucketV1 = exports.fetchUnlockedBucketV1 = exports.deserializeUnlockedBucketV1 = exports.getUnlockedBucketV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getUnlockedBucketV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['state', (0, types_1.getUnlockedBucketStateSerializer)()],
        ['index', (0, serializers_1.u8)()],
        ['feeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 1 })],
        ['solAllocationBps', (0, serializers_1.u16)()],
        ['genesis', (0, serializers_1.publicKey)()],
        ['tokenAllocation', (0, serializers_1.u64)()],
        ['solBalance', (0, serializers_1.u64)()],
        ['tokenBalance', (0, serializers_1.u64)()],
        ['start', (0, serializers_1.i64)()],
        ['recipient', (0, serializers_1.publicKey)()],
        ['backendSigner', (0, serializers_1.publicKey)()],
        ['fee', (0, serializers_1.u64)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    ], { description: 'UnlockedBucketV1AccountData' });
}
exports.getUnlockedBucketV1AccountDataSerializer = getUnlockedBucketV1AccountDataSerializer;
function deserializeUnlockedBucketV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getUnlockedBucketV1AccountDataSerializer());
}
exports.deserializeUnlockedBucketV1 = deserializeUnlockedBucketV1;
async function fetchUnlockedBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'UnlockedBucketV1');
    return deserializeUnlockedBucketV1(maybeAccount);
}
exports.fetchUnlockedBucketV1 = fetchUnlockedBucketV1;
async function safeFetchUnlockedBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeUnlockedBucketV1(maybeAccount) : null;
}
exports.safeFetchUnlockedBucketV1 = safeFetchUnlockedBucketV1;
async function fetchAllUnlockedBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'UnlockedBucketV1');
        return deserializeUnlockedBucketV1(maybeAccount);
    });
}
exports.fetchAllUnlockedBucketV1 = fetchAllUnlockedBucketV1;
async function safeFetchAllUnlockedBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeUnlockedBucketV1(maybeAccount));
}
exports.safeFetchAllUnlockedBucketV1 = safeFetchAllUnlockedBucketV1;
function getUnlockedBucketV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        state: [2, (0, types_1.getUnlockedBucketStateSerializer)()],
        index: [3, (0, serializers_1.u8)()],
        feeType: [4, (0, types_1.getFeeDiscriminantsSerializer)()],
        padding: [5, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 1 })],
        solAllocationBps: [6, (0, serializers_1.u16)()],
        genesis: [8, (0, serializers_1.publicKey)()],
        tokenAllocation: [40, (0, serializers_1.u64)()],
        solBalance: [48, (0, serializers_1.u64)()],
        tokenBalance: [56, (0, serializers_1.u64)()],
        start: [64, (0, serializers_1.i64)()],
        recipient: [72, (0, serializers_1.publicKey)()],
        backendSigner: [104, (0, serializers_1.publicKey)()],
        fee: [136, (0, serializers_1.u64)()],
        reserved: [144, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
    })
        .deserializeUsing((account) => deserializeUnlockedBucketV1(account));
}
exports.getUnlockedBucketV1GpaBuilder = getUnlockedBucketV1GpaBuilder;
function getUnlockedBucketV1Size() {
    return 272;
}
exports.getUnlockedBucketV1Size = getUnlockedBucketV1Size;
function findUnlockedBucketV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('unlocked'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.index),
    ]);
}
exports.findUnlockedBucketV1Pda = findUnlockedBucketV1Pda;
async function fetchUnlockedBucketV1FromSeeds(context, seeds, options) {
    return fetchUnlockedBucketV1(context, findUnlockedBucketV1Pda(context, seeds), options);
}
exports.fetchUnlockedBucketV1FromSeeds = fetchUnlockedBucketV1FromSeeds;
async function safeFetchUnlockedBucketV1FromSeeds(context, seeds, options) {
    return safeFetchUnlockedBucketV1(context, findUnlockedBucketV1Pda(context, seeds), options);
}
exports.safeFetchUnlockedBucketV1FromSeeds = safeFetchUnlockedBucketV1FromSeeds;
