// @metaplex-foundation/genesis/dist/src/generated/accounts/vaultDepositV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchVaultDepositV2FromSeeds = exports.fetchVaultDepositV2FromSeeds = exports.findVaultDepositV2Pda = exports.getVaultDepositV2Size = exports.getVaultDepositV2GpaBuilder = exports.safeFetchAllVaultDepositV2 = exports.fetchAllVaultDepositV2 = exports.safeFetchVaultDepositV2 = exports.fetchVaultDepositV2 = exports.deserializeVaultDepositV2 = exports.getVaultDepositV2AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getVaultDepositV2AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 2 })],
        ['index', (0, serializers_1.u32)()],
        ['depositor', (0, serializers_1.publicKey)()],
        ['bucket', (0, serializers_1.publicKey)()],
        ['timestamp', (0, serializers_1.i64)()],
        ['amountQuoteToken', (0, serializers_1.u64)()],
        ['amountQuoteTokenConsumed', (0, serializers_1.u64)()],
    ], { description: 'VaultDepositV2AccountData' });
}
exports.getVaultDepositV2AccountDataSerializer = getVaultDepositV2AccountDataSerializer;
function deserializeVaultDepositV2(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getVaultDepositV2AccountDataSerializer());
}
exports.deserializeVaultDepositV2 = deserializeVaultDepositV2;
async function fetchVaultDepositV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'VaultDepositV2');
    return deserializeVaultDepositV2(maybeAccount);
}
exports.fetchVaultDepositV2 = fetchVaultDepositV2;
async function safeFetchVaultDepositV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeVaultDepositV2(maybeAccount) : null;
}
exports.safeFetchVaultDepositV2 = safeFetchVaultDepositV2;
async function fetchAllVaultDepositV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'VaultDepositV2');
        return deserializeVaultDepositV2(maybeAccount);
    });
}
exports.fetchAllVaultDepositV2 = fetchAllVaultDepositV2;
async function safeFetchAllVaultDepositV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeVaultDepositV2(maybeAccount));
}
exports.safeFetchAllVaultDepositV2 = safeFetchAllVaultDepositV2;
function getVaultDepositV2GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        padding: [2, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 2 })],
        index: [4, (0, serializers_1.u32)()],
        depositor: [8, (0, serializers_1.publicKey)()],
        bucket: [40, (0, serializers_1.publicKey)()],
        timestamp: [72, (0, serializers_1.i64)()],
        amountQuoteToken: [80, (0, serializers_1.u64)()],
        amountQuoteTokenConsumed: [88, (0, serializers_1.u64)()],
    })
        .deserializeUsing((account) => deserializeVaultDepositV2(account));
}
exports.getVaultDepositV2GpaBuilder = getVaultDepositV2GpaBuilder;
function getVaultDepositV2Size() {
    return 96;
}
exports.getVaultDepositV2Size = getVaultDepositV2Size;
function findVaultDepositV2Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('vault_deposit_v2'),
        (0, serializers_1.publicKey)().serialize(seeds.bucket),
        (0, serializers_1.publicKey)().serialize(seeds.recipient),
    ]);
}
exports.findVaultDepositV2Pda = findVaultDepositV2Pda;
async function fetchVaultDepositV2FromSeeds(context, seeds, options) {
    return fetchVaultDepositV2(context, findVaultDepositV2Pda(context, seeds), options);
}
exports.fetchVaultDepositV2FromSeeds = fetchVaultDepositV2FromSeeds;
async function safeFetchVaultDepositV2FromSeeds(context, seeds, options) {
    return safeFetchVaultDepositV2(context, findVaultDepositV2Pda(context, seeds), options);
}
exports.safeFetchVaultDepositV2FromSeeds = safeFetchVaultDepositV2FromSeeds;
