// @metaplex-foundation/genesis/dist/src/generated/accounts/vaultBucketV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchVaultBucketV1FromSeeds = exports.fetchVaultBucketV1FromSeeds = exports.findVaultBucketV1Pda = exports.getVaultBucketV1Size = exports.getVaultBucketV1GpaBuilder = exports.safeFetchAllVaultBucketV1 = exports.fetchAllVaultBucketV1 = exports.safeFetchVaultBucketV1 = exports.fetchVaultBucketV1 = exports.deserializeVaultBucketV1 = exports.getVaultBucketV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getVaultBucketV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['state', (0, types_1.getVaultBucketStateSerializer)()],
        ['index', (0, serializers_1.u8)()],
        ['feeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['depositFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['withdrawFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['allowWithdrawals', (0, serializers_1.bool)()],
        ['genesis', (0, serializers_1.publicKey)()],
        ['quoteDepositTotal', (0, serializers_1.u64)()],
        ['depositCount', (0, serializers_1.u64)()],
        ['consumedCount', (0, serializers_1.u64)()],
        ['start', (0, serializers_1.i64)()],
        ['end', (0, serializers_1.i64)()],
        ['fee', (0, serializers_1.u64)()],
        ['depositFee', (0, serializers_1.u64)()],
        ['withdrawFee', (0, serializers_1.u64)()],
        ['backendSigner', (0, serializers_1.publicKey)()],
        ['actionsAuthority', (0, serializers_1.publicKey)()],
        ['nextSwapToBondingCurveOrdinal', (0, serializers_1.u64)()],
    ], { description: 'VaultBucketV1AccountData' });
}
exports.getVaultBucketV1AccountDataSerializer = getVaultBucketV1AccountDataSerializer;
function deserializeVaultBucketV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getVaultBucketV1AccountDataSerializer());
}
exports.deserializeVaultBucketV1 = deserializeVaultBucketV1;
async function fetchVaultBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'VaultBucketV1');
    return deserializeVaultBucketV1(maybeAccount);
}
exports.fetchVaultBucketV1 = fetchVaultBucketV1;
async function safeFetchVaultBucketV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeVaultBucketV1(maybeAccount) : null;
}
exports.safeFetchVaultBucketV1 = safeFetchVaultBucketV1;
async function fetchAllVaultBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'VaultBucketV1');
        return deserializeVaultBucketV1(maybeAccount);
    });
}
exports.fetchAllVaultBucketV1 = fetchAllVaultBucketV1;
async function safeFetchAllVaultBucketV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeVaultBucketV1(maybeAccount));
}
exports.safeFetchAllVaultBucketV1 = safeFetchAllVaultBucketV1;
function getVaultBucketV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        state: [2, (0, types_1.getVaultBucketStateSerializer)()],
        index: [3, (0, serializers_1.u8)()],
        feeType: [4, (0, types_1.getFeeDiscriminantsSerializer)()],
        depositFeeType: [5, (0, types_1.getFeeDiscriminantsSerializer)()],
        withdrawFeeType: [6, (0, types_1.getFeeDiscriminantsSerializer)()],
        allowWithdrawals: [7, (0, serializers_1.bool)()],
        genesis: [8, (0, serializers_1.publicKey)()],
        quoteDepositTotal: [40, (0, serializers_1.u64)()],
        depositCount: [48, (0, serializers_1.u64)()],
        consumedCount: [56, (0, serializers_1.u64)()],
        start: [64, (0, serializers_1.i64)()],
        end: [72, (0, serializers_1.i64)()],
        fee: [80, (0, serializers_1.u64)()],
        depositFee: [88, (0, serializers_1.u64)()],
        withdrawFee: [96, (0, serializers_1.u64)()],
        backendSigner: [104, (0, serializers_1.publicKey)()],
        actionsAuthority: [136, (0, serializers_1.publicKey)()],
        nextSwapToBondingCurveOrdinal: [168, (0, serializers_1.u64)()],
    })
        .deserializeUsing((account) => deserializeVaultBucketV1(account));
}
exports.getVaultBucketV1GpaBuilder = getVaultBucketV1GpaBuilder;
function getVaultBucketV1Size() {
    return 176;
}
exports.getVaultBucketV1Size = getVaultBucketV1Size;
function findVaultBucketV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('vault'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.index),
    ]);
}
exports.findVaultBucketV1Pda = findVaultBucketV1Pda;
async function fetchVaultBucketV1FromSeeds(context, seeds, options) {
    return fetchVaultBucketV1(context, findVaultBucketV1Pda(context, seeds), options);
}
exports.fetchVaultBucketV1FromSeeds = fetchVaultBucketV1FromSeeds;
async function safeFetchVaultBucketV1FromSeeds(context, seeds, options) {
    return safeFetchVaultBucketV1(context, findVaultBucketV1Pda(context, seeds), options);
}
exports.safeFetchVaultBucketV1FromSeeds = safeFetchVaultBucketV1FromSeeds;
