// @metaplex-foundation/genesis/dist/src/generated/accounts/vaultBucketV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchVaultBucketV2FromSeeds = exports.fetchVaultBucketV2FromSeeds = exports.findVaultBucketV2Pda = exports.getVaultBucketV2GpaBuilder = exports.safeFetchAllVaultBucketV2 = exports.fetchAllVaultBucketV2 = exports.safeFetchVaultBucketV2 = exports.fetchVaultBucketV2 = exports.deserializeVaultBucketV2 = exports.getVaultBucketV2AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getVaultBucketV2AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['depositFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['withdrawFeeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
        ['depositFee', (0, serializers_1.u64)()],
        ['depositId', (0, serializers_1.u64)()],
        ['depositCount', (0, serializers_1.u64)()],
        ['consumedCount', (0, serializers_1.u64)()],
        ['refundCount', (0, serializers_1.u64)()],
        ['nextSwapToBondingCurveOrdinal', (0, serializers_1.u64)()],
        ['quoteTokenDepositTotal', (0, serializers_1.u64)()],
        ['withdrawFee', (0, serializers_1.u64)()],
        ['bucket', (0, types_1.getBucketBaseSerializer)()],
        ['depositStartCondition', (0, types_1.getConditionSerializer)()],
        ['depositEndCondition', (0, types_1.getConditionSerializer)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        ['extensions', (0, types_1.getVaultV2ExtensionsSerializer)()],
        ['padding2', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        ['endBehaviors', (0, serializers_1.array)((0, types_1.getBehaviorSerializer)())],
    ], { description: 'VaultBucketV2AccountData' });
}
exports.getVaultBucketV2AccountDataSerializer = getVaultBucketV2AccountDataSerializer;
function deserializeVaultBucketV2(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getVaultBucketV2AccountDataSerializer());
}
exports.deserializeVaultBucketV2 = deserializeVaultBucketV2;
async function fetchVaultBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'VaultBucketV2');
    return deserializeVaultBucketV2(maybeAccount);
}
exports.fetchVaultBucketV2 = fetchVaultBucketV2;
async function safeFetchVaultBucketV2(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeVaultBucketV2(maybeAccount) : null;
}
exports.safeFetchVaultBucketV2 = safeFetchVaultBucketV2;
async function fetchAllVaultBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'VaultBucketV2');
        return deserializeVaultBucketV2(maybeAccount);
    });
}
exports.fetchAllVaultBucketV2 = fetchAllVaultBucketV2;
async function safeFetchAllVaultBucketV2(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeVaultBucketV2(maybeAccount));
}
exports.safeFetchAllVaultBucketV2 = safeFetchAllVaultBucketV2;
function getVaultBucketV2GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        depositFeeType: [1, (0, types_1.getFeeDiscriminantsSerializer)()],
        withdrawFeeType: [2, (0, types_1.getFeeDiscriminantsSerializer)()],
        padding: [3, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
        depositFee: [8, (0, serializers_1.u64)()],
        depositId: [16, (0, serializers_1.u64)()],
        depositCount: [24, (0, serializers_1.u64)()],
        consumedCount: [32, (0, serializers_1.u64)()],
        refundCount: [40, (0, serializers_1.u64)()],
        nextSwapToBondingCurveOrdinal: [48, (0, serializers_1.u64)()],
        quoteTokenDepositTotal: [56, (0, serializers_1.u64)()],
        withdrawFee: [64, (0, serializers_1.u64)()],
        bucket: [72, (0, types_1.getBucketBaseSerializer)()],
        depositStartCondition: [272, (0, types_1.getConditionSerializer)()],
        depositEndCondition: [336, (0, types_1.getConditionSerializer)()],
        reserved: [400, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        extensions: [528, (0, types_1.getVaultV2ExtensionsSerializer)()],
        padding2: [736, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        endBehaviors: [740, (0, serializers_1.array)((0, types_1.getBehaviorSerializer)())],
    })
        .deserializeUsing((account) => deserializeVaultBucketV2(account));
}
exports.getVaultBucketV2GpaBuilder = getVaultBucketV2GpaBuilder;
function findVaultBucketV2Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('vault_v2'),
        (0, serializers_1.publicKey)().serialize(seeds.genesisAccount),
        (0, serializers_1.u8)().serialize(seeds.bucketIndex),
    ]);
}
exports.findVaultBucketV2Pda = findVaultBucketV2Pda;
async function fetchVaultBucketV2FromSeeds(context, seeds, options) {
    return fetchVaultBucketV2(context, findVaultBucketV2Pda(context, seeds), options);
}
exports.fetchVaultBucketV2FromSeeds = fetchVaultBucketV2FromSeeds;
async function safeFetchVaultBucketV2FromSeeds(context, seeds, options) {
    return safeFetchVaultBucketV2(context, findVaultBucketV2Pda(context, seeds), options);
}
exports.safeFetchVaultBucketV2FromSeeds = safeFetchVaultBucketV2FromSeeds;
