// @metaplex-foundation/genesis/dist/src/generated/accounts/vaultDepositV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeFetchVaultDepositV1FromSeeds = exports.fetchVaultDepositV1FromSeeds = exports.findVaultDepositV1Pda = exports.getVaultDepositV1Size = exports.getVaultDepositV1GpaBuilder = exports.safeFetchAllVaultDepositV1 = exports.fetchAllVaultDepositV1 = exports.safeFetchVaultDepositV1 = exports.fetchVaultDepositV1 = exports.deserializeVaultDepositV1 = exports.getVaultDepositV1AccountDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const types_1 = require("../types");
function getVaultDepositV1AccountDataSerializer() {
    return (0, serializers_1.struct)([
        ['key', (0, types_1.getKeySerializer)()],
        ['bump', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 2 })],
        ['index', (0, serializers_1.u32)()],
        ['depositor', (0, serializers_1.publicKey)()],
        ['bucket', (0, serializers_1.publicKey)()],
        ['timestamp', (0, serializers_1.i64)()],
        ['amount', (0, serializers_1.u64)()],
        ['consumedAmount', (0, serializers_1.u64)()],
    ], { description: 'VaultDepositV1AccountData' });
}
exports.getVaultDepositV1AccountDataSerializer = getVaultDepositV1AccountDataSerializer;
function deserializeVaultDepositV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, getVaultDepositV1AccountDataSerializer());
}
exports.deserializeVaultDepositV1 = deserializeVaultDepositV1;
async function fetchVaultDepositV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'VaultDepositV1');
    return deserializeVaultDepositV1(maybeAccount);
}
exports.fetchVaultDepositV1 = fetchVaultDepositV1;
async function safeFetchVaultDepositV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeVaultDepositV1(maybeAccount) : null;
}
exports.safeFetchVaultDepositV1 = safeFetchVaultDepositV1;
async function fetchAllVaultDepositV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'VaultDepositV1');
        return deserializeVaultDepositV1(maybeAccount);
    });
}
exports.fetchAllVaultDepositV1 = fetchAllVaultDepositV1;
async function safeFetchAllVaultDepositV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeVaultDepositV1(maybeAccount));
}
exports.safeFetchAllVaultDepositV1 = safeFetchAllVaultDepositV1;
function getVaultDepositV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        bump: [1, (0, serializers_1.u8)()],
        padding: [2, (0, serializers_1.array)((0, serializers_1.u8)(), { size: 2 })],
        index: [4, (0, serializers_1.u32)()],
        depositor: [8, (0, serializers_1.publicKey)()],
        bucket: [40, (0, serializers_1.publicKey)()],
        timestamp: [72, (0, serializers_1.i64)()],
        amount: [80, (0, serializers_1.u64)()],
        consumedAmount: [88, (0, serializers_1.u64)()],
    })
        .deserializeUsing((account) => deserializeVaultDepositV1(account));
}
exports.getVaultDepositV1GpaBuilder = getVaultDepositV1GpaBuilder;
function getVaultDepositV1Size() {
    return 96;
}
exports.getVaultDepositV1Size = getVaultDepositV1Size;
function findVaultDepositV1Pda(context, seeds) {
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    return context.eddsa.findPda(programId, [
        (0, serializers_1.string)({ size: 'variable' }).serialize('vault_deposit'),
        (0, serializers_1.publicKey)().serialize(seeds.bucket),
        (0, serializers_1.publicKey)().serialize(seeds.recipient),
    ]);
}
exports.findVaultDepositV1Pda = findVaultDepositV1Pda;
async function fetchVaultDepositV1FromSeeds(context, seeds, options) {
    return fetchVaultDepositV1(context, findVaultDepositV1Pda(context, seeds), options);
}
exports.fetchVaultDepositV1FromSeeds = fetchVaultDepositV1FromSeeds;
async function safeFetchVaultDepositV1FromSeeds(context, seeds, options) {
    return safeFetchVaultDepositV1(context, findVaultDepositV1Pda(context, seeds), options);
}
exports.safeFetchVaultDepositV1FromSeeds = safeFetchVaultDepositV1FromSeeds;
