// @metaplex-foundation/genesis/dist/src/generated/instructions/addLaunchPoolBucketV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.addLaunchPoolBucketV2 = exports.addLaunchPoolBucketV2InstructionDiscriminator = exports.getAddLaunchPoolBucketV2InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getAddLaunchPoolBucketV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['bucketIndex', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        ['baseTokenAllocation', (0, serializers_1.u64)()],
        ['depositStartCondition', (0, types_1.getConditionSerializer)()],
        ['depositEndCondition', (0, types_1.getConditionSerializer)()],
        ['claimStartCondition', (0, types_1.getConditionSerializer)()],
        ['claimEndCondition', (0, types_1.getConditionSerializer)()],
        [
            'backendSigner',
            (() => {
                const inner = (0, types_1.getBackendSignerSerializer)();
                const sentinel = new Uint8Array([
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<BackendSigner>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        [
            'depositPenalty',
            (() => {
                const inner = (0, types_1.getLinearBpsScheduleV2Serializer)();
                const sentinel = new Uint8Array([
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255,
                    255, 255, 255, 127, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<LinearBpsScheduleV2>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        [
            'withdrawPenalty',
            (() => {
                const inner = (0, types_1.getLinearBpsScheduleV2Serializer)();
                const sentinel = new Uint8Array([
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255,
                    255, 255, 255, 127, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<LinearBpsScheduleV2>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        [
            'bonusSchedule',
            (() => {
                const inner = (0, types_1.getLinearBpsScheduleV2Serializer)();
                const sentinel = new Uint8Array([
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255,
                    255, 255, 255, 127, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<LinearBpsScheduleV2>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        [
            'depositLimit',
            (() => {
                const inner = (0, types_1.getDepositLimitSerializer)();
                const sentinel = new Uint8Array([
                    255, 255, 255, 255, 255, 255, 255, 255,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<DepositLimit>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        [
            'allowlist',
            (() => {
                const inner = (0, types_1.getAllowlistInitArgsSerializer)();
                const sentinel = new Uint8Array([
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<AllowlistInitArgs>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        [
            'claimSchedule',
            (() => {
                const inner = (0, types_1.getClaimScheduleSerializer)();
                const sentinel = new Uint8Array([
                    255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255, 255,
                    255, 255, 127, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255, 255,
                    255, 255, 127, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<ClaimSchedule>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        [
            'minimumDepositAmount',
            (() => {
                const inner = (0, types_1.getMinimumDepositAmountSerializer)();
                const sentinel = new Uint8Array([
                    255, 255, 255, 255, 255, 255, 255, 255,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<MinimumDepositAmount>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        [
            'minimumQuoteTokenThreshold',
            (() => {
                const inner = (0, types_1.getMinimumQuoteTokenThresholdSerializer)();
                const sentinel = new Uint8Array([
                    255, 255, 255, 255, 255, 255, 255, 255,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<MinimumQuoteTokenThreshold>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [
                                (0, umi_1.none)(),
                                offset + fixedSize,
                            ];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        ['endBehaviors', (0, serializers_1.array)((0, types_1.getBehaviorSerializer)())],
    ], { description: 'AddLaunchPoolBucketV2InstructionData' }), (value) => ({
        ...value,
        discriminator: 67,
        bucketIndex: value.bucketIndex ?? 0,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0],
        baseTokenAllocation: value.baseTokenAllocation ?? 1000000000,
        backendSigner: value.backendSigner ?? (0, umi_1.none)(),
        depositPenalty: value.depositPenalty ?? (0, umi_1.none)(),
        withdrawPenalty: value.withdrawPenalty ?? (0, umi_1.none)(),
        bonusSchedule: value.bonusSchedule ?? (0, umi_1.none)(),
        depositLimit: value.depositLimit ?? (0, umi_1.none)(),
        allowlist: value.allowlist ?? (0, umi_1.none)(),
        claimSchedule: value.claimSchedule ?? (0, umi_1.none)(),
        minimumDepositAmount: value.minimumDepositAmount ?? (0, umi_1.none)(),
        minimumQuoteTokenThreshold: value.minimumQuoteTokenThreshold ?? (0, umi_1.none)(),
        reserved: value.reserved ?? [0, 0, 0, 0],
        endBehaviors: value.endBehaviors ?? [],
    }));
}
exports.getAddLaunchPoolBucketV2InstructionDataSerializer = getAddLaunchPoolBucketV2InstructionDataSerializer;
// Instruction discriminator.
exports.addLaunchPoolBucketV2InstructionDiscriminator = 67;
// Instruction.
function addLaunchPoolBucketV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        baseMint: {
            index: 2,
            isWritable: false,
            value: input.baseMint ?? null,
        },
        quoteMint: {
            index: 3,
            isWritable: false,
            value: input.quoteMint ?? null,
        },
        bucketQuoteTokenAccount: {
            index: 4,
            isWritable: true,
            value: input.bucketQuoteTokenAccount ?? null,
        },
        payer: {
            index: 5,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 6,
            isWritable: false,
            value: input.authority ?? null,
        },
        quoteTokenProgram: {
            index: 7,
            isWritable: false,
            value: input.quoteTokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 8,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 9,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedArgs.bucketIndex) {
        resolvedArgs.bucketIndex = 0;
    }
    if (!resolvedAccounts.bucket.value) {
        resolvedAccounts.bucket.value = (0, accounts_1.findLaunchPoolBucketV2Pda)(context, {
            genesisAccount: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
            bucketIndex: (0, shared_1.expectSome)(resolvedArgs.bucketIndex),
        });
    }
    if (!resolvedAccounts.quoteMint.value) {
        resolvedAccounts.quoteMint.value = (0, umi_1.publicKey)('So11111111111111111111111111111111111111112');
    }
    if (!resolvedAccounts.bucketQuoteTokenAccount.value) {
        resolvedAccounts.bucketQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucket.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.quoteTokenProgram.value) {
        resolvedAccounts.quoteTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getAddLaunchPoolBucketV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.addLaunchPoolBucketV2 = addLaunchPoolBucketV2;
