// @metaplex-foundation/genesis/dist/src/generated/instructions/addPresaleBucketV2Extensions.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.addPresaleBucketV2Extensions = exports.addPresaleBucketV2ExtensionsInstructionDiscriminator = exports.getAddPresaleBucketV2ExtensionsInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getAddPresaleBucketV2ExtensionsInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 3 })],
        ['extensions', (0, serializers_1.array)((0, types_1.getPresaleV2ExtensionSerializer)())],
    ], { description: 'AddPresaleBucketV2ExtensionsInstructionData' }), (value) => ({ ...value, discriminator: 78 }));
}
exports.getAddPresaleBucketV2ExtensionsInstructionDataSerializer = getAddPresaleBucketV2ExtensionsInstructionDataSerializer;
// Instruction discriminator.
exports.addPresaleBucketV2ExtensionsInstructionDiscriminator = 78;
// Instruction.
function addPresaleBucketV2Extensions(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: false,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        payer: {
            index: 2,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 3,
            isWritable: false,
            value: input.authority ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = {
        ...input,
    };
    // Default values.
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getAddPresaleBucketV2ExtensionsInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.addPresaleBucketV2Extensions = addPresaleBucketV2Extensions;
