// @metaplex-foundation/genesis/dist/src/generated/instructions/addPresaleBucketV2Base.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.addPresaleBucketV2Base = exports.addPresaleBucketV2BaseInstructionDiscriminator = exports.getAddPresaleBucketV2BaseInstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getAddPresaleBucketV2BaseInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['bucketIndex', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        ['baseTokenAllocation', (0, serializers_1.u64)()],
        ['allocationQuoteTokenCap', (0, serializers_1.u64)()],
        ['depositStartCondition', (0, types_1.getConditionSerializer)()],
        ['depositEndCondition', (0, types_1.getConditionSerializer)()],
        ['claimStartCondition', (0, types_1.getConditionSerializer)()],
        ['claimEndCondition', (0, types_1.getConditionSerializer)()],
    ], { description: 'AddPresaleBucketV2BaseInstructionData' }), (value) => ({
        ...value,
        discriminator: 77,
        bucketIndex: value.bucketIndex ?? 0,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0],
        baseTokenAllocation: value.baseTokenAllocation ?? 1000000000,
    }));
}
exports.getAddPresaleBucketV2BaseInstructionDataSerializer = getAddPresaleBucketV2BaseInstructionDataSerializer;
// Instruction discriminator.
exports.addPresaleBucketV2BaseInstructionDiscriminator = 77;
// Instruction.
function addPresaleBucketV2Base(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        baseMint: {
            index: 2,
            isWritable: false,
            value: input.baseMint ?? null,
        },
        quoteMint: {
            index: 3,
            isWritable: false,
            value: input.quoteMint ?? null,
        },
        bucketQuoteTokenAccount: {
            index: 4,
            isWritable: true,
            value: input.bucketQuoteTokenAccount ?? null,
        },
        payer: {
            index: 5,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 6,
            isWritable: false,
            value: input.authority ?? null,
        },
        quoteTokenProgram: {
            index: 7,
            isWritable: false,
            value: input.quoteTokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 8,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 9,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedArgs.bucketIndex) {
        resolvedArgs.bucketIndex = 0;
    }
    if (!resolvedAccounts.bucket.value) {
        resolvedAccounts.bucket.value = (0, accounts_1.findPresaleBucketV2Pda)(context, {
            genesisAccount: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
            bucketIndex: (0, shared_1.expectSome)(resolvedArgs.bucketIndex),
        });
    }
    if (!resolvedAccounts.quoteMint.value) {
        resolvedAccounts.quoteMint.value = (0, umi_1.publicKey)('So11111111111111111111111111111111111111112');
    }
    if (!resolvedAccounts.bucketQuoteTokenAccount.value) {
        resolvedAccounts.bucketQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucket.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.quoteTokenProgram.value) {
        resolvedAccounts.quoteTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getAddPresaleBucketV2BaseInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.addPresaleBucketV2Base = addPresaleBucketV2Base;
