// @metaplex-foundation/genesis/dist/src/generated/instructions/addPresaleBucketV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.addPresaleBucketV1 = exports.addPresaleBucketV1InstructionDiscriminator = exports.getAddPresaleBucketV1InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getAddPresaleBucketV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u8)()],
        ['allowWithdraw', (0, serializers_1.bool)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
        ['totalAmount', (0, serializers_1.u64)()],
        ['start', (0, serializers_1.i64)()],
        ['end', (0, serializers_1.i64)()],
        ['claimStart', (0, serializers_1.i64)()],
        ['allocationSolCap', (0, serializers_1.u64)()],
        ['depositLimit', (0, serializers_1.u64)()],
        ['depositCooldown', (0, serializers_1.u64)()],
        ['perCooldownDepositLimit', (0, serializers_1.u64)()],
        ['hasAllowlist', (0, serializers_1.bool)()],
        ['allowlistMerkleTreeHeight', (0, serializers_1.u8)()],
        ['allowlistMerkleRoot', (0, serializers_1.bytes)({ size: 32 })],
        ['padding2', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        ['allowlistEnd', (0, serializers_1.i64)()],
        ['allowlistSolCap', (0, serializers_1.u64)()],
    ], { description: 'AddPresaleBucketV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 19,
        index: value.index ?? 0,
        allowWithdraw: value.allowWithdraw ?? false,
        padding: value.padding ?? [0, 0, 0, 0, 0],
        totalAmount: value.totalAmount ?? 1000000000,
        depositLimit: value.depositLimit ?? '18446744073709551615',
        depositCooldown: value.depositCooldown ?? 0,
        perCooldownDepositLimit: value.perCooldownDepositLimit ?? '18446744073709551615',
        hasAllowlist: value.hasAllowlist ?? false,
        allowlistMerkleTreeHeight: value.allowlistMerkleTreeHeight ?? 0,
        allowlistMerkleRoot: value.allowlistMerkleRoot ?? [
            0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
            0, 0, 0, 0, 0, 0, 0, 0,
        ],
        padding2: value.padding2 ?? [0, 0, 0, 0, 0, 0],
        allowlistSolCap: value.allowlistSolCap ?? '18446744073709551615',
    }));
}
exports.getAddPresaleBucketV1InstructionDataSerializer = getAddPresaleBucketV1InstructionDataSerializer;
// Instruction discriminator.
exports.addPresaleBucketV1InstructionDiscriminator = 19;
// Instruction.
function addPresaleBucketV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        mint: { index: 2, isWritable: true, value: input.mint ?? null },
        bucketTokenAccount: {
            index: 3,
            isWritable: true,
            value: input.bucketTokenAccount ?? null,
        },
        payer: {
            index: 4,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 5,
            isWritable: false,
            value: input.authority ?? null,
        },
        feeWallet: {
            index: 6,
            isWritable: true,
            value: input.feeWallet ?? null,
        },
        tokenProgram: {
            index: 7,
            isWritable: false,
            value: input.tokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 8,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 9,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        backendSigner: {
            index: 10,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.genesisAccount.value) {
        resolvedAccounts.genesisAccount.value = (0, accounts_1.findGenesisAccountV1Pda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
        });
    }
    if (!resolvedArgs.index) {
        resolvedArgs.index = 0;
    }
    if (!resolvedAccounts.bucket.value) {
        resolvedAccounts.bucket.value = (0, accounts_1.findPresaleBucketV1Pda)(context, {
            genesisAccount: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
            index: (0, shared_1.expectSome)(resolvedArgs.index),
        });
    }
    if (!resolvedAccounts.bucketTokenAccount.value) {
        resolvedAccounts.bucketTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucket.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.feeWallet.value) {
        resolvedAccounts.feeWallet.value = (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb');
    }
    if (!resolvedAccounts.tokenProgram.value) {
        resolvedAccounts.tokenProgram.value = context.programs.getPublicKey('splToken', 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
        resolvedAccounts.tokenProgram.isWritable = false;
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    if (!resolvedArgs.allowlistEnd) {
        resolvedArgs.allowlistEnd = (0, shared_1.expectSome)(resolvedArgs.end);
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getAddPresaleBucketV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.addPresaleBucketV1 = addPresaleBucketV1;
