// @metaplex-foundation/genesis/dist/src/generated/instructions/addRaydiumCpmmBucketV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.addRaydiumCpmmBucketV2 = exports.addRaydiumCpmmBucketV2InstructionDiscriminator = exports.getAddRaydiumCpmmBucketV2InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getAddRaydiumCpmmBucketV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['bucketIndex', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        ['baseTokenAllocation', (0, serializers_1.u64)()],
        [
            'backendSigner',
            (() => {
                const inner = (0, types_1.getBackendSignerSerializer)();
                const sentinel = new Uint8Array([
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<BackendSigner>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        [
            'lpLockSchedule',
            (() => {
                const inner = (0, types_1.getClaimScheduleSerializer)();
                const sentinel = new Uint8Array([
                    255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255, 255,
                    255, 255, 127, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255, 255,
                    255, 255, 127, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<ClaimSchedule>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        [
            'lpClaimAuthority',
            (() => {
                const inner = (0, serializers_1.publicKey)();
                const sentinel = new Uint8Array([
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<PublicKey>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        ['startCondition', (0, types_1.getConditionSerializer)()],
        ['reserved', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        ['endBehaviors', (0, serializers_1.array)((0, types_1.getBehaviorSerializer)())],
    ], { description: 'AddRaydiumCpmmBucketV2InstructionData' }), (value) => ({
        ...value,
        discriminator: 55,
        bucketIndex: value.bucketIndex ?? 0,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0],
        backendSigner: value.backendSigner ?? (0, umi_1.none)(),
        lpLockSchedule: value.lpLockSchedule ?? (0, umi_1.none)(),
        lpClaimAuthority: value.lpClaimAuthority ?? (0, umi_1.none)(),
        reserved: value.reserved ?? [0, 0, 0, 0],
        endBehaviors: value.endBehaviors ?? [],
    }));
}
exports.getAddRaydiumCpmmBucketV2InstructionDataSerializer = getAddRaydiumCpmmBucketV2InstructionDataSerializer;
// Instruction discriminator.
exports.addRaydiumCpmmBucketV2InstructionDiscriminator = 55;
// Instruction.
function addRaydiumCpmmBucketV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        baseMint: {
            index: 2,
            isWritable: false,
            value: input.baseMint ?? null,
        },
        quoteMint: {
            index: 3,
            isWritable: false,
            value: input.quoteMint ?? null,
        },
        bucketSigner: {
            index: 4,
            isWritable: true,
            value: input.bucketSigner ?? null,
        },
        raydiumSigner: {
            index: 5,
            isWritable: true,
            value: input.raydiumSigner ?? null,
        },
        bucketSignerQuoteTokenAccount: {
            index: 6,
            isWritable: true,
            value: input.bucketSignerQuoteTokenAccount ?? null,
        },
        raydiumSignerBaseTokenAccount: {
            index: 7,
            isWritable: true,
            value: input.raydiumSignerBaseTokenAccount ?? null,
        },
        raydiumSignerQuoteTokenAccount: {
            index: 8,
            isWritable: true,
            value: input.raydiumSignerQuoteTokenAccount ?? null,
        },
        payer: {
            index: 9,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 10,
            isWritable: false,
            value: input.authority ?? null,
        },
        baseTokenProgram: {
            index: 11,
            isWritable: false,
            value: input.baseTokenProgram ?? null,
        },
        quoteTokenProgram: {
            index: 12,
            isWritable: false,
            value: input.quoteTokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 13,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 14,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedArgs.bucketIndex) {
        resolvedArgs.bucketIndex = 0;
    }
    if (!resolvedAccounts.bucket.value) {
        resolvedAccounts.bucket.value = (0, accounts_1.findRaydiumCpmmBucketV2Pda)(context, {
            genesisAccount: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
            bucketIndex: (0, shared_1.expectSome)(resolvedArgs.bucketIndex),
        });
    }
    if (!resolvedAccounts.quoteMint.value) {
        resolvedAccounts.quoteMint.value = (0, umi_1.publicKey)('So11111111111111111111111111111111111111112');
    }
    if (!resolvedAccounts.bucketSigner.value) {
        resolvedAccounts.bucketSigner.value = (0, accounts_1.findRaydiumBucketSignerPda)(context, {
            bucket: (0, shared_1.expectPublicKey)(resolvedAccounts.bucket.value),
        });
    }
    if (!resolvedAccounts.raydiumSigner.value) {
        resolvedAccounts.raydiumSigner.value = (0, accounts_1.findRaydiumSignerPda)(context);
    }
    if (!resolvedAccounts.bucketSignerQuoteTokenAccount.value) {
        resolvedAccounts.bucketSignerQuoteTokenAccount.value =
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucketSigner.value),
            });
    }
    if (!resolvedAccounts.raydiumSignerBaseTokenAccount.value) {
        resolvedAccounts.raydiumSignerBaseTokenAccount.value =
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.raydiumSigner.value),
            });
    }
    if (!resolvedAccounts.raydiumSignerQuoteTokenAccount.value) {
        resolvedAccounts.raydiumSignerQuoteTokenAccount.value =
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.raydiumSigner.value),
            });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.baseTokenProgram.value) {
        resolvedAccounts.baseTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.quoteTokenProgram.value) {
        resolvedAccounts.quoteTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getAddRaydiumCpmmBucketV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.addRaydiumCpmmBucketV2 = addRaydiumCpmmBucketV2;
