// @metaplex-foundation/genesis/dist/src/generated/instructions/addUnlockedBucketV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.addUnlockedBucketV1 = exports.addUnlockedBucketV1InstructionDiscriminator = exports.getAddUnlockedBucketV1InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getAddUnlockedBucketV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        ['solAllocationBps', (0, serializers_1.u16)()],
        ['tokenAllocation', (0, serializers_1.u64)()],
        ['start', (0, serializers_1.i64)()],
    ], { description: 'AddUnlockedBucketV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 8,
        index: value.index ?? 0,
        padding: value.padding ?? [0, 0, 0, 0],
    }));
}
exports.getAddUnlockedBucketV1InstructionDataSerializer = getAddUnlockedBucketV1InstructionDataSerializer;
// Instruction discriminator.
exports.addUnlockedBucketV1InstructionDiscriminator = 8;
// Instruction.
function addUnlockedBucketV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        bucketTokenAccount: {
            index: 2,
            isWritable: true,
            value: input.bucketTokenAccount ?? null,
        },
        recipient: {
            index: 3,
            isWritable: false,
            value: input.recipient ?? null,
        },
        mint: { index: 4, isWritable: true, value: input.mint ?? null },
        payer: {
            index: 5,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 6,
            isWritable: false,
            value: input.authority ?? null,
        },
        feeWallet: {
            index: 7,
            isWritable: true,
            value: input.feeWallet ?? null,
        },
        tokenProgram: {
            index: 8,
            isWritable: false,
            value: input.tokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 9,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 10,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        backendSigner: {
            index: 11,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.genesisAccount.value) {
        resolvedAccounts.genesisAccount.value = (0, accounts_1.findGenesisAccountV1Pda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
        });
    }
    if (!resolvedArgs.index) {
        resolvedArgs.index = 0;
    }
    if (!resolvedAccounts.bucket.value) {
        resolvedAccounts.bucket.value = (0, accounts_1.findUnlockedBucketV1Pda)(context, {
            genesisAccount: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
            index: (0, shared_1.expectSome)(resolvedArgs.index),
        });
    }
    if (!resolvedAccounts.bucketTokenAccount.value) {
        resolvedAccounts.bucketTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucket.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.feeWallet.value) {
        resolvedAccounts.feeWallet.value = (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb');
    }
    if (!resolvedAccounts.tokenProgram.value) {
        resolvedAccounts.tokenProgram.value = context.programs.getPublicKey('splToken', 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
        resolvedAccounts.tokenProgram.isWritable = false;
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getAddUnlockedBucketV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.addUnlockedBucketV1 = addUnlockedBucketV1;
