// @metaplex-foundation/genesis/dist/src/generated/instructions/addUnlockedBucketV2Base.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.addUnlockedBucketV2Base = exports.addUnlockedBucketV2BaseInstructionDiscriminator = exports.getAddUnlockedBucketV2BaseInstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getAddUnlockedBucketV2BaseInstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['bucketIndex', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        ['baseTokenAllocation', (0, serializers_1.u64)()],
        ['claimStartCondition', (0, types_1.getConditionSerializer)()],
        ['claimEndCondition', (0, types_1.getConditionSerializer)()],
    ], { description: 'AddUnlockedBucketV2BaseInstructionData' }), (value) => ({ ...value, discriminator: 97 }));
}
exports.getAddUnlockedBucketV2BaseInstructionDataSerializer = getAddUnlockedBucketV2BaseInstructionDataSerializer;
// Instruction discriminator.
exports.addUnlockedBucketV2BaseInstructionDiscriminator = 97;
// Instruction.
function addUnlockedBucketV2Base(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        baseMint: {
            index: 2,
            isWritable: false,
            value: input.baseMint ?? null,
        },
        quoteMint: {
            index: 3,
            isWritable: false,
            value: input.quoteMint ?? null,
        },
        bucketQuoteTokenAccount: {
            index: 4,
            isWritable: true,
            value: input.bucketQuoteTokenAccount ?? null,
        },
        payer: {
            index: 5,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 6,
            isWritable: false,
            value: input.authority ?? null,
        },
        recipient: {
            index: 7,
            isWritable: false,
            value: input.recipient ?? null,
        },
        quoteTokenProgram: {
            index: 8,
            isWritable: false,
            value: input.quoteTokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 9,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 10,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getAddUnlockedBucketV2BaseInstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.addUnlockedBucketV2Base = addUnlockedBucketV2Base;
