// @metaplex-foundation/genesis/dist/src/generated/instructions/claimRaydiumCpmmLpTokensV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.claimRaydiumCpmmLpTokensV2 = exports.claimRaydiumCpmmLpTokensV2InstructionDiscriminator = exports.getClaimRaydiumCpmmLpTokensV2InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getClaimRaydiumCpmmLpTokensV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 7 })],
        ['amount', (0, serializers_1.u64)()],
    ], { description: 'ClaimRaydiumCpmmLpTokensV2InstructionData' }), (value) => ({
        ...value,
        discriminator: 57,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0, 0],
    }));
}
exports.getClaimRaydiumCpmmLpTokensV2InstructionDataSerializer = getClaimRaydiumCpmmLpTokensV2InstructionDataSerializer;
// Instruction discriminator.
exports.claimRaydiumCpmmLpTokensV2InstructionDiscriminator = 57;
// Instruction.
function claimRaydiumCpmmLpTokensV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: false,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        baseMint: {
            index: 2,
            isWritable: false,
            value: input.baseMint ?? null,
        },
        bucketSigner: {
            index: 3,
            isWritable: false,
            value: input.bucketSigner ?? null,
        },
        lpMint: {
            index: 4,
            isWritable: false,
            value: input.lpMint ?? null,
        },
        bucketSignerLpTokenAccount: {
            index: 5,
            isWritable: true,
            value: input.bucketSignerLpTokenAccount ?? null,
        },
        destinationLpTokenAccount: {
            index: 6,
            isWritable: true,
            value: input.destinationLpTokenAccount ?? null,
        },
        recipient: {
            index: 7,
            isWritable: false,
            value: input.recipient ?? null,
        },
        lpClaimAuthority: {
            index: 8,
            isWritable: false,
            value: input.lpClaimAuthority ?? null,
        },
        payer: {
            index: 9,
            isWritable: true,
            value: input.payer ?? null,
        },
        backendSigner: {
            index: 10,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
        tokenProgram: {
            index: 11,
            isWritable: false,
            value: input.tokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 12,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 13,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.bucketSigner.value) {
        resolvedAccounts.bucketSigner.value = (0, accounts_1.findRaydiumBucketSignerPda)(context, {
            bucket: (0, shared_1.expectPublicKey)(resolvedAccounts.bucket.value),
        });
    }
    if (!resolvedAccounts.bucketSignerLpTokenAccount.value) {
        resolvedAccounts.bucketSignerLpTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.lpMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucketSigner.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.tokenProgram.value) {
        resolvedAccounts.tokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getClaimRaydiumCpmmLpTokensV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.claimRaydiumCpmmLpTokensV2 = claimRaydiumCpmmLpTokensV2;
