// @metaplex-foundation/genesis/dist/src/generated/instructions/claimAuctionRefundV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.claimAuctionRefundV1 = exports.claimAuctionRefundV1InstructionDiscriminator = exports.getClaimAuctionRefundV1InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getClaimAuctionRefundV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u8)()],
        ['bidIndex', (0, serializers_1.u16)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
    ], { description: 'ClaimAuctionRefundV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 37,
        index: value.index ?? 0,
        bidIndex: value.bidIndex ?? 0,
        padding: value.padding ?? [0, 0, 0, 0],
    }));
}
exports.getClaimAuctionRefundV1InstructionDataSerializer = getClaimAuctionRefundV1InstructionDataSerializer;
// Instruction discriminator.
exports.claimAuctionRefundV1InstructionDiscriminator = 37;
// Instruction.
function claimAuctionRefundV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        mint: { index: 2, isWritable: false, value: input.mint ?? null },
        payer: {
            index: 3,
            isWritable: true,
            value: input.payer ?? null,
        },
        bid: { index: 4, isWritable: true, value: input.bid ?? null },
        backendSigner: {
            index: 5,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.genesisAccount.value) {
        resolvedAccounts.genesisAccount.value = (0, accounts_1.findGenesisAccountV1Pda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
        });
    }
    if (!resolvedArgs.index) {
        resolvedArgs.index = 0;
    }
    if (!resolvedAccounts.bucket.value) {
        resolvedAccounts.bucket.value = (0, accounts_1.findAuctionBucketV1Pda)(context, {
            genesisAccount: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
            index: (0, shared_1.expectSome)(resolvedArgs.index),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer.publicKey;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getClaimAuctionRefundV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.claimAuctionRefundV1 = claimAuctionRefundV1;
