// @metaplex-foundation/genesis/dist/src/generated/instructions/collectRaydiumCpmmFeesV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.collectRaydiumCpmmFeesV2 = exports.collectRaydiumCpmmFeesV2InstructionDiscriminator = exports.getCollectRaydiumCpmmFeesV2InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getCollectRaydiumCpmmFeesV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([['discriminator', (0, serializers_1.u8)()]], {
        description: 'CollectRaydiumCpmmFeesV2InstructionData',
    }), (value) => ({ ...value, discriminator: 58 }));
}
exports.getCollectRaydiumCpmmFeesV2InstructionDataSerializer = getCollectRaydiumCpmmFeesV2InstructionDataSerializer;
// Instruction discriminator.
exports.collectRaydiumCpmmFeesV2InstructionDiscriminator = 58;
// Instruction.
function collectRaydiumCpmmFeesV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        baseMint: {
            index: 0,
            isWritable: false,
            value: input.baseMint ?? null,
        },
        quoteMint: {
            index: 1,
            isWritable: false,
            value: input.quoteMint ?? null,
        },
        raydiumSigner: {
            index: 2,
            isWritable: true,
            value: input.raydiumSigner ?? null,
        },
        payer: {
            index: 3,
            isWritable: true,
            value: input.payer ?? null,
        },
        poolState: {
            index: 4,
            isWritable: true,
            value: input.poolState ?? null,
        },
        ammConfig: {
            index: 5,
            isWritable: false,
            value: input.ammConfig ?? null,
        },
        poolAuthority: {
            index: 6,
            isWritable: false,
            value: input.poolAuthority ?? null,
        },
        baseVault: {
            index: 7,
            isWritable: true,
            value: input.baseVault ?? null,
        },
        quoteVault: {
            index: 8,
            isWritable: true,
            value: input.quoteVault ?? null,
        },
        raydiumSignerBaseTokenAccount: {
            index: 9,
            isWritable: true,
            value: input.raydiumSignerBaseTokenAccount ?? null,
        },
        raydiumSignerQuoteTokenAccount: {
            index: 10,
            isWritable: true,
            value: input.raydiumSignerQuoteTokenAccount ?? null,
        },
        feeWallet: {
            index: 11,
            isWritable: false,
            value: input.feeWallet ?? null,
        },
        feeWalletQuoteTokenAccount: {
            index: 12,
            isWritable: true,
            value: input.feeWalletQuoteTokenAccount ?? null,
        },
        raydiumProgram: {
            index: 13,
            isWritable: false,
            value: input.raydiumProgram ?? null,
        },
        baseTokenProgram: {
            index: 14,
            isWritable: false,
            value: input.baseTokenProgram ?? null,
        },
        quoteTokenProgram: {
            index: 15,
            isWritable: false,
            value: input.quoteTokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 16,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 17,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Default values.
    if (!resolvedAccounts.quoteMint.value) {
        resolvedAccounts.quoteMint.value = (0, umi_1.publicKey)('So11111111111111111111111111111111111111112');
    }
    if (!resolvedAccounts.raydiumSigner.value) {
        resolvedAccounts.raydiumSigner.value = (0, accounts_1.findRaydiumSignerPda)(context);
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.raydiumSignerBaseTokenAccount.value) {
        resolvedAccounts.raydiumSignerBaseTokenAccount.value =
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.raydiumSigner.value),
            });
    }
    if (!resolvedAccounts.raydiumSignerQuoteTokenAccount.value) {
        resolvedAccounts.raydiumSignerQuoteTokenAccount.value =
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.raydiumSigner.value),
            });
    }
    if (!resolvedAccounts.feeWallet.value) {
        resolvedAccounts.feeWallet.value = (0, umi_1.publicKey)('A1xCjrfiQdRZHbCnovSqLeRkFTcgWJKWCYv9E8WxcfoR');
    }
    if (!resolvedAccounts.feeWalletQuoteTokenAccount.value) {
        resolvedAccounts.feeWalletQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
            owner: (0, umi_1.publicKey)('A1xCjrfiQdRZHbCnovSqLeRkFTcgWJKWCYv9E8WxcfoR'),
        });
    }
    if (!resolvedAccounts.baseTokenProgram.value) {
        resolvedAccounts.baseTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.quoteTokenProgram.value) {
        resolvedAccounts.quoteTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getCollectRaydiumCpmmFeesV2InstructionDataSerializer().serialize({});
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.collectRaydiumCpmmFeesV2 = collectRaydiumCpmmFeesV2;
