// @metaplex-foundation/genesis/dist/src/generated/instructions/depositVaultV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.depositVaultV1 = exports.depositVaultV1InstructionDiscriminator = exports.getDepositVaultV1InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getDepositVaultV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        ['amount', (0, serializers_1.u64)()],
    ], { description: 'DepositVaultV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 27,
        index: value.index ?? 0,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0],
    }));
}
exports.getDepositVaultV1InstructionDataSerializer = getDepositVaultV1InstructionDataSerializer;
// Instruction discriminator.
exports.depositVaultV1InstructionDiscriminator = 27;
// Instruction.
function depositVaultV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        depositPda: {
            index: 2,
            isWritable: true,
            value: input.depositPda ?? null,
        },
        payer: {
            index: 3,
            isWritable: true,
            value: input.payer ?? null,
        },
        recipient: {
            index: 4,
            isWritable: false,
            value: input.recipient ?? null,
        },
        feeWallet: {
            index: 5,
            isWritable: true,
            value: input.feeWallet ?? null,
        },
        systemProgram: {
            index: 6,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        backendSigner: {
            index: 7,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedArgs.index) {
        resolvedArgs.index = 0;
    }
    if (!resolvedAccounts.bucket.value) {
        resolvedAccounts.bucket.value = (0, accounts_1.findVaultBucketV1Pda)(context, {
            genesisAccount: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
            index: (0, shared_1.expectSome)(resolvedArgs.index),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.feeWallet.value) {
        resolvedAccounts.feeWallet.value = (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getDepositVaultV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = (0, accounts_1.getVaultDepositV1Size)() + umi_1.ACCOUNT_HEADER_SIZE;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.depositVaultV1 = depositVaultV1;
