// @metaplex-foundation/genesis/dist/src/generated/instructions/feeOverrideV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.feeOverrideV2 = exports.feeOverrideV2InstructionDiscriminator = exports.getFeeOverrideV2InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getFeeOverrideV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['feeOverrideType', (0, types_1.getFeeOverrideTypeSerializer)()],
        ['feeType', (0, types_1.getFeeDiscriminantsSerializer)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
        ['feeOverride', (0, serializers_1.u64)()],
    ], { description: 'FeeOverrideV2InstructionData' }), (value) => ({
        ...value,
        discriminator: 47,
        padding: value.padding ?? [0, 0, 0, 0, 0],
    }));
}
exports.getFeeOverrideV2InstructionDataSerializer = getFeeOverrideV2InstructionDataSerializer;
// Instruction discriminator.
exports.feeOverrideV2InstructionDiscriminator = 47;
// Instruction.
function feeOverrideV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        baseMint: {
            index: 2,
            isWritable: true,
            value: input.baseMint ?? null,
        },
        authority: {
            index: 3,
            isWritable: false,
            value: input.authority ?? null,
        },
        systemProgram: {
            index: 4,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.authority.value) {
        resolvedAccounts.authority.value = context.identity;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getFeeOverrideV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.feeOverrideV2 = feeOverrideV2;
