// @metaplex-foundation/genesis/dist/src/generated/instructions/graduateToRaydiumCpmmV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.graduateToRaydiumCpmmV1 = exports.graduateToRaydiumCpmmV1InstructionDiscriminator = exports.getGraduateToRaydiumCpmmV1InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getGraduateToRaydiumCpmmV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
    ], { description: 'GraduateToRaydiumCpmmV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 40,
        index: value.index ?? 0,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0],
    }));
}
exports.getGraduateToRaydiumCpmmV1InstructionDataSerializer = getGraduateToRaydiumCpmmV1InstructionDataSerializer;
// Instruction discriminator.
exports.graduateToRaydiumCpmmV1InstructionDiscriminator = 40;
// Instruction.
function graduateToRaydiumCpmmV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        bucketSigner: {
            index: 2,
            isWritable: true,
            value: input.bucketSigner ?? null,
        },
        mint: { index: 3, isWritable: false, value: input.mint ?? null },
        payer: {
            index: 4,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 5,
            isWritable: false,
            value: input.authority ?? null,
        },
        poolState: {
            index: 6,
            isWritable: true,
            value: input.poolState ?? null,
        },
        ammConfig: {
            index: 7,
            isWritable: false,
            value: input.ammConfig ?? null,
        },
        poolAuthority: {
            index: 8,
            isWritable: false,
            value: input.poolAuthority ?? null,
        },
        lpMint: {
            index: 9,
            isWritable: true,
            value: input.lpMint ?? null,
        },
        bucketSignerMintTokenAccount: {
            index: 10,
            isWritable: true,
            value: input.bucketSignerMintTokenAccount ?? null,
        },
        bucketSignerSolTokenAccount: {
            index: 11,
            isWritable: true,
            value: input.bucketSignerSolTokenAccount ?? null,
        },
        bucketSignerLpTokenAccount: {
            index: 12,
            isWritable: true,
            value: input.bucketSignerLpTokenAccount ?? null,
        },
        mintVault: {
            index: 13,
            isWritable: true,
            value: input.mintVault ?? null,
        },
        solVault: {
            index: 14,
            isWritable: true,
            value: input.solVault ?? null,
        },
        createPoolFee: {
            index: 15,
            isWritable: true,
            value: input.createPoolFee ?? null,
        },
        observationState: {
            index: 16,
            isWritable: true,
            value: input.observationState ?? null,
        },
        nativeSol: {
            index: 17,
            isWritable: false,
            value: input.nativeSol ?? null,
        },
        raydiumProgram: {
            index: 18,
            isWritable: false,
            value: input.raydiumProgram ?? null,
        },
        tokenProgram: {
            index: 19,
            isWritable: false,
            value: input.tokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 20,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 21,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        rent: {
            index: 22,
            isWritable: false,
            value: input.rent ?? null,
        },
        backendSigner: {
            index: 23,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.genesisAccount.value) {
        resolvedAccounts.genesisAccount.value = (0, accounts_1.findGenesisAccountV1Pda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
        });
    }
    if (!resolvedArgs.index) {
        resolvedArgs.index = 0;
    }
    if (!resolvedAccounts.bucket.value) {
        resolvedAccounts.bucket.value = (0, accounts_1.findRaydiumCpmmBucketV1Pda)(context, {
            genesisAccount: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
            index: (0, shared_1.expectSome)(resolvedArgs.index),
        });
    }
    if (!resolvedAccounts.bucketSigner.value) {
        resolvedAccounts.bucketSigner.value = (0, accounts_1.findRaydiumBucketSignerPda)(context, {
            bucket: (0, shared_1.expectPublicKey)(resolvedAccounts.bucket.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.bucketSignerMintTokenAccount.value) {
        resolvedAccounts.bucketSignerMintTokenAccount.value =
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucketSigner.value),
            });
    }
    if (!resolvedAccounts.nativeSol.value) {
        resolvedAccounts.nativeSol.value = (0, umi_1.publicKey)('So11111111111111111111111111111111111111112');
    }
    if (!resolvedAccounts.bucketSignerSolTokenAccount.value) {
        resolvedAccounts.bucketSignerSolTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.nativeSol.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucketSigner.value),
        });
    }
    if (!resolvedAccounts.bucketSignerLpTokenAccount.value) {
        resolvedAccounts.bucketSignerLpTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.lpMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucketSigner.value),
        });
    }
    if (!resolvedAccounts.tokenProgram.value) {
        resolvedAccounts.tokenProgram.value = context.programs.getPublicKey('splToken', 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
        resolvedAccounts.tokenProgram.isWritable = false;
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    if (!resolvedAccounts.rent.value) {
        resolvedAccounts.rent.value = (0, umi_1.publicKey)('SysvarRent111111111111111111111111111111111');
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getGraduateToRaydiumCpmmV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.graduateToRaydiumCpmmV1 = graduateToRaydiumCpmmV1;
