// @metaplex-foundation/genesis/dist/src/generated/instructions/graduateToMeteoraV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.graduateToMeteoraV1 = exports.graduateToMeteoraV1InstructionDiscriminator = exports.getGraduateToMeteoraV1InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getGraduateToMeteoraV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
    ], { description: 'GraduateToMeteoraV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 15,
        index: value.index ?? 0,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0],
    }));
}
exports.getGraduateToMeteoraV1InstructionDataSerializer = getGraduateToMeteoraV1InstructionDataSerializer;
// Instruction discriminator.
exports.graduateToMeteoraV1InstructionDiscriminator = 15;
// Instruction.
function graduateToMeteoraV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        bucketTokenAccount: {
            index: 2,
            isWritable: true,
            value: input.bucketTokenAccount ?? null,
        },
        mint: { index: 3, isWritable: false, value: input.mint ?? null },
        payer: {
            index: 4,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 5,
            isWritable: false,
            value: input.authority ?? null,
        },
        feeWallet: {
            index: 6,
            isWritable: true,
            value: input.feeWallet ?? null,
        },
        pool: { index: 7, isWritable: true, value: input.pool ?? null },
        config: {
            index: 8,
            isWritable: false,
            value: input.config ?? null,
        },
        lpMint: {
            index: 9,
            isWritable: true,
            value: input.lpMint ?? null,
        },
        tokenBMint: {
            index: 10,
            isWritable: false,
            value: input.tokenBMint ?? null,
        },
        aVault: {
            index: 11,
            isWritable: true,
            value: input.aVault ?? null,
        },
        bVault: {
            index: 12,
            isWritable: true,
            value: input.bVault ?? null,
        },
        aTokenVault: {
            index: 13,
            isWritable: true,
            value: input.aTokenVault ?? null,
        },
        bTokenVault: {
            index: 14,
            isWritable: true,
            value: input.bTokenVault ?? null,
        },
        aVaultLpMint: {
            index: 15,
            isWritable: true,
            value: input.aVaultLpMint ?? null,
        },
        bVaultLpMint: {
            index: 16,
            isWritable: true,
            value: input.bVaultLpMint ?? null,
        },
        aVaultLp: {
            index: 17,
            isWritable: true,
            value: input.aVaultLp ?? null,
        },
        bVaultLp: {
            index: 18,
            isWritable: true,
            value: input.bVaultLp ?? null,
        },
        payerTokenA: {
            index: 19,
            isWritable: true,
            value: input.payerTokenA ?? null,
        },
        payerTokenB: {
            index: 20,
            isWritable: true,
            value: input.payerTokenB ?? null,
        },
        payerPoolLp: {
            index: 21,
            isWritable: true,
            value: input.payerPoolLp ?? null,
        },
        protocolTokenAFee: {
            index: 22,
            isWritable: true,
            value: input.protocolTokenAFee ?? null,
        },
        protocolTokenBFee: {
            index: 23,
            isWritable: true,
            value: input.protocolTokenBFee ?? null,
        },
        rent: {
            index: 24,
            isWritable: false,
            value: input.rent ?? null,
        },
        mintMetadata: {
            index: 25,
            isWritable: true,
            value: input.mintMetadata ?? null,
        },
        nativeSol: {
            index: 26,
            isWritable: false,
            value: input.nativeSol ?? null,
        },
        meteoraProgram: {
            index: 27,
            isWritable: false,
            value: input.meteoraProgram ?? null,
        },
        metadataProgram: {
            index: 28,
            isWritable: false,
            value: input.metadataProgram ?? null,
        },
        vaultProgram: {
            index: 29,
            isWritable: false,
            value: input.vaultProgram ?? null,
        },
        tokenProgram: {
            index: 30,
            isWritable: false,
            value: input.tokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 31,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 32,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        backendSigner: {
            index: 33,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.genesisAccount.value) {
        resolvedAccounts.genesisAccount.value = (0, accounts_1.findGenesisAccountV1Pda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
        });
    }
    if (!resolvedArgs.index) {
        resolvedArgs.index = 0;
    }
    if (!resolvedAccounts.bucket.value) {
        resolvedAccounts.bucket.value = (0, accounts_1.findMeteoraBucketV1Pda)(context, {
            genesisAccount: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
            index: (0, shared_1.expectSome)(resolvedArgs.index),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.feeWallet.value) {
        resolvedAccounts.feeWallet.value = (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb');
    }
    if (!resolvedAccounts.rent.value) {
        resolvedAccounts.rent.value = (0, umi_1.publicKey)('SysvarRent111111111111111111111111111111111');
    }
    if (!resolvedAccounts.tokenProgram.value) {
        resolvedAccounts.tokenProgram.value = context.programs.getPublicKey('splToken', 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
        resolvedAccounts.tokenProgram.isWritable = false;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getGraduateToMeteoraV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.graduateToMeteoraV1 = graduateToMeteoraV1;
