// @metaplex-foundation/genesis/dist/src/generated/instructions/graduateToRaydiumCpmmV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.graduateToRaydiumCpmmV2 = exports.graduateToRaydiumCpmmV2InstructionDiscriminator = exports.getGraduateToRaydiumCpmmV2InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getGraduateToRaydiumCpmmV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 7 })],
    ], { description: 'GraduateToRaydiumCpmmV2InstructionData' }), (value) => ({
        ...value,
        discriminator: 56,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0, 0],
    }));
}
exports.getGraduateToRaydiumCpmmV2InstructionDataSerializer = getGraduateToRaydiumCpmmV2InstructionDataSerializer;
// Instruction discriminator.
exports.graduateToRaydiumCpmmV2InstructionDiscriminator = 56;
// Instruction.
function graduateToRaydiumCpmmV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        baseMint: {
            index: 2,
            isWritable: true,
            value: input.baseMint ?? null,
        },
        quoteMint: {
            index: 3,
            isWritable: false,
            value: input.quoteMint ?? null,
        },
        bucketSigner: {
            index: 4,
            isWritable: true,
            value: input.bucketSigner ?? null,
        },
        raydiumSigner: {
            index: 5,
            isWritable: true,
            value: input.raydiumSigner ?? null,
        },
        genesisBaseTokenAccount: {
            index: 6,
            isWritable: true,
            value: input.genesisBaseTokenAccount ?? null,
        },
        bucketSignerQuoteTokenAccount: {
            index: 7,
            isWritable: true,
            value: input.bucketSignerQuoteTokenAccount ?? null,
        },
        bucketSignerLpTokenAccount: {
            index: 8,
            isWritable: true,
            value: input.bucketSignerLpTokenAccount ?? null,
        },
        raydiumSignerBaseTokenAccount: {
            index: 9,
            isWritable: true,
            value: input.raydiumSignerBaseTokenAccount ?? null,
        },
        raydiumSignerQuoteTokenAccount: {
            index: 10,
            isWritable: true,
            value: input.raydiumSignerQuoteTokenAccount ?? null,
        },
        raydiumSignerLpTokenAccount: {
            index: 11,
            isWritable: true,
            value: input.raydiumSignerLpTokenAccount ?? null,
        },
        poolState: {
            index: 12,
            isWritable: true,
            value: input.poolState ?? null,
        },
        ammConfig: {
            index: 13,
            isWritable: false,
            value: input.ammConfig ?? null,
        },
        poolAuthority: {
            index: 14,
            isWritable: false,
            value: input.poolAuthority ?? null,
        },
        lpMint: {
            index: 15,
            isWritable: true,
            value: input.lpMint ?? null,
        },
        baseVault: {
            index: 16,
            isWritable: true,
            value: input.baseVault ?? null,
        },
        quoteVault: {
            index: 17,
            isWritable: true,
            value: input.quoteVault ?? null,
        },
        createPoolFee: {
            index: 18,
            isWritable: true,
            value: input.createPoolFee ?? null,
        },
        observationState: {
            index: 19,
            isWritable: true,
            value: input.observationState ?? null,
        },
        permission: {
            index: 20,
            isWritable: false,
            value: input.permission ?? null,
        },
        backendSigner: {
            index: 21,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
        raydiumProgram: {
            index: 22,
            isWritable: false,
            value: input.raydiumProgram ?? null,
        },
        baseTokenProgram: {
            index: 23,
            isWritable: false,
            value: input.baseTokenProgram ?? null,
        },
        quoteTokenProgram: {
            index: 24,
            isWritable: false,
            value: input.quoteTokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 25,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 26,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        rent: {
            index: 27,
            isWritable: false,
            value: input.rent ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.quoteMint.value) {
        resolvedAccounts.quoteMint.value = (0, umi_1.publicKey)('So11111111111111111111111111111111111111112');
    }
    if (!resolvedAccounts.bucketSigner.value) {
        resolvedAccounts.bucketSigner.value = (0, accounts_1.findRaydiumBucketSignerPda)(context, {
            bucket: (0, shared_1.expectPublicKey)(resolvedAccounts.bucket.value),
        });
    }
    if (!resolvedAccounts.raydiumSigner.value) {
        resolvedAccounts.raydiumSigner.value = (0, accounts_1.findRaydiumSignerPda)(context);
    }
    if (!resolvedAccounts.genesisBaseTokenAccount.value) {
        resolvedAccounts.genesisBaseTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
        });
    }
    if (!resolvedAccounts.bucketSignerQuoteTokenAccount.value) {
        resolvedAccounts.bucketSignerQuoteTokenAccount.value =
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucketSigner.value),
            });
    }
    if (!resolvedAccounts.bucketSignerLpTokenAccount.value) {
        resolvedAccounts.bucketSignerLpTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.lpMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucketSigner.value),
        });
    }
    if (!resolvedAccounts.raydiumSignerBaseTokenAccount.value) {
        resolvedAccounts.raydiumSignerBaseTokenAccount.value =
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.raydiumSigner.value),
            });
    }
    if (!resolvedAccounts.raydiumSignerQuoteTokenAccount.value) {
        resolvedAccounts.raydiumSignerQuoteTokenAccount.value =
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.raydiumSigner.value),
            });
    }
    if (!resolvedAccounts.raydiumSignerLpTokenAccount.value) {
        resolvedAccounts.raydiumSignerLpTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.lpMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.raydiumSigner.value),
        });
    }
    if (!resolvedAccounts.baseTokenProgram.value) {
        resolvedAccounts.baseTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.quoteTokenProgram.value) {
        resolvedAccounts.quoteTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    if (!resolvedAccounts.rent.value) {
        resolvedAccounts.rent.value = (0, umi_1.publicKey)('SysvarRent111111111111111111111111111111111');
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getGraduateToRaydiumCpmmV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.graduateToRaydiumCpmmV2 = graduateToRaydiumCpmmV2;
