// @metaplex-foundation/genesis/dist/src/generated/instructions/initializeV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.initializeV2 = exports.initializeV2InstructionDiscriminator = exports.getInitializeV2InstructionDataSerializer = void 0;
const mpl_token_metadata_1 = require("@metaplex-foundation/mpl-token-metadata");
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getInitializeV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['decimals', (0, serializers_1.u8)()],
        ['genesisIndex', (0, serializers_1.u16)()],
        ['fundingMode', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 3 })],
        ['totalSupplyBaseToken', (0, serializers_1.u64)()],
        ['name', (0, serializers_1.string)()],
        ['uri', (0, serializers_1.string)()],
        ['symbol', (0, serializers_1.string)()],
    ], { description: 'InitializeV2InstructionData' }), (value) => ({
        ...value,
        discriminator: 42,
        decimals: value.decimals ?? 9,
        genesisIndex: value.genesisIndex ?? 0,
        padding: value.padding ?? [0, 0, 0],
    }));
}
exports.getInitializeV2InstructionDataSerializer = getInitializeV2InstructionDataSerializer;
// Instruction discriminator.
exports.initializeV2InstructionDiscriminator = 42;
// Instruction.
function initializeV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        baseMint: {
            index: 1,
            isWritable: true,
            value: input.baseMint ?? null,
        },
        quoteMint: {
            index: 2,
            isWritable: false,
            value: input.quoteMint ?? null,
        },
        genesisBaseTokenAccount: {
            index: 3,
            isWritable: true,
            value: input.genesisBaseTokenAccount ?? null,
        },
        genesisQuoteTokenAccount: {
            index: 4,
            isWritable: true,
            value: input.genesisQuoteTokenAccount ?? null,
        },
        payer: {
            index: 5,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 6,
            isWritable: false,
            value: input.authority ?? null,
        },
        metadata: {
            index: 7,
            isWritable: true,
            value: input.metadata ?? null,
        },
        feeWallet: {
            index: 8,
            isWritable: false,
            value: input.feeWallet ?? null,
        },
        feeQuoteTokenAccount: {
            index: 9,
            isWritable: true,
            value: input.feeQuoteTokenAccount ?? null,
        },
        authoritySourceBaseTokenAccount: {
            index: 10,
            isWritable: true,
            value: input.authoritySourceBaseTokenAccount ?? null,
        },
        baseTokenProgram: {
            index: 11,
            isWritable: false,
            value: input.baseTokenProgram ?? null,
        },
        quoteTokenProgram: {
            index: 12,
            isWritable: false,
            value: input.quoteTokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 13,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        tokenMetadataProgram: {
            index: 14,
            isWritable: false,
            value: input.tokenMetadataProgram ?? null,
        },
        systemProgram: {
            index: 15,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedArgs.genesisIndex) {
        resolvedArgs.genesisIndex = 0;
    }
    if (!resolvedAccounts.genesisAccount.value) {
        resolvedAccounts.genesisAccount.value = (0, accounts_1.findGenesisAccountV2Pda)(context, {
            baseMint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
            genesisIndex: (0, shared_1.expectSome)(resolvedArgs.genesisIndex),
        });
    }
    if (!resolvedAccounts.quoteMint.value) {
        resolvedAccounts.quoteMint.value = (0, umi_1.publicKey)('So11111111111111111111111111111111111111112');
    }
    if (!resolvedAccounts.genesisBaseTokenAccount.value) {
        resolvedAccounts.genesisBaseTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
        });
    }
    if (!resolvedAccounts.genesisQuoteTokenAccount.value) {
        resolvedAccounts.genesisQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.metadata.value) {
        resolvedAccounts.metadata.value = (0, mpl_token_metadata_1.findMetadataPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
        });
    }
    if (!resolvedAccounts.feeWallet.value) {
        resolvedAccounts.feeWallet.value = (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb');
    }
    if (!resolvedAccounts.feeQuoteTokenAccount.value) {
        resolvedAccounts.feeQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.feeWallet.value),
        });
    }
    if (!resolvedAccounts.authoritySourceBaseTokenAccount.value) {
        if (resolvedAccounts.authority.value) {
            resolvedAccounts.authoritySourceBaseTokenAccount.value =
                (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                    mint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
                    owner: (0, shared_1.expectPublicKey)(resolvedAccounts.authority.value),
                });
        }
        else {
            resolvedAccounts.authoritySourceBaseTokenAccount.value =
                (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                    mint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
                    owner: (0, shared_1.expectPublicKey)(resolvedAccounts.payer.value),
                });
        }
    }
    if (!resolvedAccounts.baseTokenProgram.value) {
        resolvedAccounts.baseTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.quoteTokenProgram.value) {
        resolvedAccounts.quoteTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.tokenMetadataProgram.value) {
        resolvedAccounts.tokenMetadataProgram.value = context.programs.getPublicKey('mplTokenMetadata', 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s');
        resolvedAccounts.tokenMetadataProgram.isWritable = false;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getInitializeV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.initializeV2 = initializeV2;
