// @metaplex-foundation/genesis/dist/src/generated/instructions/initializeV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.initializeV1 = exports.initializeV1InstructionDiscriminator = exports.getInitializeV1InstructionDataSerializer = void 0;
const mpl_token_metadata_1 = require("@metaplex-foundation/mpl-token-metadata");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getInitializeV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['decimals', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        ['name', (0, serializers_1.bytes)({ size: 32 })],
        ['uri', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 128 })],
        ['symbol', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 10 })],
    ], { description: 'InitializeV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 0,
        decimals: value.decimals ?? 9,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0],
    }));
}
exports.getInitializeV1InstructionDataSerializer = getInitializeV1InstructionDataSerializer;
// Instruction discriminator.
exports.initializeV1InstructionDiscriminator = 0;
// Instruction.
function initializeV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        mint: { index: 1, isWritable: true, value: input.mint ?? null },
        payer: {
            index: 2,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 3,
            isWritable: false,
            value: input.authority ?? null,
        },
        metadata: {
            index: 4,
            isWritable: true,
            value: input.metadata ?? null,
        },
        feeWallet: {
            index: 5,
            isWritable: true,
            value: input.feeWallet ?? null,
        },
        tokenProgram: {
            index: 6,
            isWritable: false,
            value: input.tokenProgram ?? null,
        },
        tokenMetadataProgram: {
            index: 7,
            isWritable: false,
            value: input.tokenMetadataProgram ?? null,
        },
        systemProgram: {
            index: 8,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.genesisAccount.value) {
        resolvedAccounts.genesisAccount.value = (0, accounts_1.findGenesisAccountV1Pda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.metadata.value) {
        resolvedAccounts.metadata.value = (0, mpl_token_metadata_1.findMetadataPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
        });
    }
    if (!resolvedAccounts.feeWallet.value) {
        resolvedAccounts.feeWallet.value = (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb');
    }
    if (!resolvedAccounts.tokenProgram.value) {
        resolvedAccounts.tokenProgram.value = context.programs.getPublicKey('splToken', 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
        resolvedAccounts.tokenProgram.isWritable = false;
    }
    if (!resolvedAccounts.tokenMetadataProgram.value) {
        resolvedAccounts.tokenMetadataProgram.value = context.programs.getPublicKey('mplTokenMetadata', 'metaqbxxUerdq28cj1RbAWkYQm3ybzjb6a8bt518x1s');
        resolvedAccounts.tokenMetadataProgram.isWritable = false;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getInitializeV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.initializeV1 = initializeV1;
