// @metaplex-foundation/genesis/dist/src/generated/instructions/lockStreamflowV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.lockStreamflowV1 = exports.lockStreamflowV1InstructionDiscriminator = exports.getLockStreamflowV1InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getLockStreamflowV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
    ], { description: 'LockStreamflowV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 7,
        index: value.index ?? 0,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0],
    }));
}
exports.getLockStreamflowV1InstructionDataSerializer = getLockStreamflowV1InstructionDataSerializer;
// Instruction discriminator.
exports.lockStreamflowV1InstructionDiscriminator = 7;
// Instruction.
function lockStreamflowV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        bucketTokenAccount: {
            index: 2,
            isWritable: true,
            value: input.bucketTokenAccount ?? null,
        },
        recipient: {
            index: 3,
            isWritable: true,
            value: input.recipient ?? null,
        },
        authority: {
            index: 4,
            isWritable: true,
            value: input.authority ?? null,
        },
        authorityTokenAccount: {
            index: 5,
            isWritable: true,
            value: input.authorityTokenAccount ?? null,
        },
        feeWallet: {
            index: 6,
            isWritable: true,
            value: input.feeWallet ?? null,
        },
        metadata: {
            index: 7,
            isWritable: true,
            value: input.metadata ?? null,
        },
        escrowTokens: {
            index: 8,
            isWritable: true,
            value: input.escrowTokens ?? null,
        },
        recipientTokens: {
            index: 9,
            isWritable: true,
            value: input.recipientTokens ?? null,
        },
        streamflowTreasury: {
            index: 10,
            isWritable: true,
            value: input.streamflowTreasury ?? null,
        },
        streamflowTreasuryTokens: {
            index: 11,
            isWritable: true,
            value: input.streamflowTreasuryTokens ?? null,
        },
        withdrawor: {
            index: 12,
            isWritable: true,
            value: input.withdrawor ?? null,
        },
        partner: {
            index: 13,
            isWritable: true,
            value: input.partner ?? null,
        },
        partnerTokens: {
            index: 14,
            isWritable: true,
            value: input.partnerTokens ?? null,
        },
        mint: {
            index: 15,
            isWritable: false,
            value: input.mint ?? null,
        },
        feeOracle: {
            index: 16,
            isWritable: false,
            value: input.feeOracle ?? null,
        },
        rent: {
            index: 17,
            isWritable: false,
            value: input.rent ?? null,
        },
        timelockProgram: {
            index: 18,
            isWritable: false,
            value: input.timelockProgram ?? null,
        },
        tokenProgram: {
            index: 19,
            isWritable: false,
            value: input.tokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 20,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 21,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        backendSigner: {
            index: 22,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.genesisAccount.value) {
        resolvedAccounts.genesisAccount.value = (0, accounts_1.findGenesisAccountV1Pda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
        });
    }
    if (!resolvedArgs.index) {
        resolvedArgs.index = 0;
    }
    if (!resolvedAccounts.bucket.value) {
        resolvedAccounts.bucket.value = (0, accounts_1.findStreamflowBucketV1Pda)(context, {
            genesisAccount: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
            index: (0, shared_1.expectSome)(resolvedArgs.index),
        });
    }
    if (!resolvedAccounts.bucketTokenAccount.value) {
        resolvedAccounts.bucketTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucket.value),
        });
    }
    if (!resolvedAccounts.authority.value) {
        resolvedAccounts.authority.value = context.identity;
    }
    if (!resolvedAccounts.authorityTokenAccount.value) {
        resolvedAccounts.authorityTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.authority.value),
        });
    }
    if (!resolvedAccounts.feeWallet.value) {
        resolvedAccounts.feeWallet.value = (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb');
    }
    if (!resolvedAccounts.rent.value) {
        resolvedAccounts.rent.value = (0, umi_1.publicKey)('SysvarRent111111111111111111111111111111111');
    }
    if (!resolvedAccounts.tokenProgram.value) {
        resolvedAccounts.tokenProgram.value = context.programs.getPublicKey('splToken', 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
        resolvedAccounts.tokenProgram.isWritable = false;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getLockStreamflowV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.lockStreamflowV1 = lockStreamflowV1;
