// @metaplex-foundation/genesis/dist/src/generated/instructions/lockStreamflowV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.lockStreamflowV2 = exports.lockStreamflowV2InstructionDiscriminator = exports.getLockStreamflowV2InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
function getLockStreamflowV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 7 })],
    ], { description: 'LockStreamflowV2InstructionData' }), (value) => ({ ...value, discriminator: 103 }));
}
exports.getLockStreamflowV2InstructionDataSerializer = getLockStreamflowV2InstructionDataSerializer;
// Instruction discriminator.
exports.lockStreamflowV2InstructionDiscriminator = 103;
// Instruction.
function lockStreamflowV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: false,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        baseMint: {
            index: 2,
            isWritable: false,
            value: input.baseMint ?? null,
        },
        quoteMint: {
            index: 3,
            isWritable: false,
            value: input.quoteMint ?? null,
        },
        genesisBaseTokenAccount: {
            index: 4,
            isWritable: true,
            value: input.genesisBaseTokenAccount ?? null,
        },
        bucketQuoteTokenAccount: {
            index: 5,
            isWritable: true,
            value: input.bucketQuoteTokenAccount ?? null,
        },
        recipient: {
            index: 6,
            isWritable: true,
            value: input.recipient ?? null,
        },
        authority: {
            index: 7,
            isWritable: true,
            value: input.authority ?? null,
        },
        authorityTokenAccount: {
            index: 8,
            isWritable: true,
            value: input.authorityTokenAccount ?? null,
        },
        feeQuoteTokenAccount: {
            index: 9,
            isWritable: true,
            value: input.feeQuoteTokenAccount ?? null,
        },
        metadata: {
            index: 10,
            isWritable: true,
            value: input.metadata ?? null,
        },
        escrowTokens: {
            index: 11,
            isWritable: true,
            value: input.escrowTokens ?? null,
        },
        recipientTokens: {
            index: 12,
            isWritable: true,
            value: input.recipientTokens ?? null,
        },
        streamflowTreasury: {
            index: 13,
            isWritable: true,
            value: input.streamflowTreasury ?? null,
        },
        streamflowTreasuryTokens: {
            index: 14,
            isWritable: true,
            value: input.streamflowTreasuryTokens ?? null,
        },
        withdrawor: {
            index: 15,
            isWritable: true,
            value: input.withdrawor ?? null,
        },
        partner: {
            index: 16,
            isWritable: true,
            value: input.partner ?? null,
        },
        partnerTokens: {
            index: 17,
            isWritable: true,
            value: input.partnerTokens ?? null,
        },
        feeOracle: {
            index: 18,
            isWritable: false,
            value: input.feeOracle ?? null,
        },
        rent: {
            index: 19,
            isWritable: false,
            value: input.rent ?? null,
        },
        timelockProgram: {
            index: 20,
            isWritable: false,
            value: input.timelockProgram ?? null,
        },
        baseTokenProgram: {
            index: 21,
            isWritable: false,
            value: input.baseTokenProgram ?? null,
        },
        quoteTokenProgram: {
            index: 22,
            isWritable: false,
            value: input.quoteTokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 23,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 24,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        backendSigner: {
            index: 25,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.quoteMint.value) {
        resolvedAccounts.quoteMint.value = (0, umi_1.publicKey)('So11111111111111111111111111111111111111112');
    }
    if (!resolvedAccounts.authority.value) {
        resolvedAccounts.authority.value = context.identity;
    }
    if (!resolvedAccounts.feeQuoteTokenAccount.value) {
        resolvedAccounts.feeQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
            owner: (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb'),
        });
    }
    if (!resolvedAccounts.rent.value) {
        resolvedAccounts.rent.value = (0, umi_1.publicKey)('SysvarRent111111111111111111111111111111111');
    }
    if (!resolvedAccounts.baseTokenProgram.value) {
        resolvedAccounts.baseTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.quoteTokenProgram.value) {
        resolvedAccounts.quoteTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getLockStreamflowV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.lockStreamflowV2 = lockStreamflowV2;
