// @metaplex-foundation/genesis/dist/src/generated/instructions/placeAuctionBidV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.placeAuctionBidV1 = exports.placeAuctionBidV1InstructionDiscriminator = exports.getPlaceAuctionBidV1InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getPlaceAuctionBidV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u8)()],
        ['bidIndex', (0, serializers_1.u16)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 4 })],
        ['priceUlamports', (0, serializers_1.u64)()],
        ['tokenQuantity', (0, serializers_1.u64)()],
    ], { description: 'PlaceAuctionBidV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 35,
        index: value.index ?? 0,
        bidIndex: value.bidIndex ?? 0,
        padding: value.padding ?? [0, 0, 0, 0],
    }));
}
exports.getPlaceAuctionBidV1InstructionDataSerializer = getPlaceAuctionBidV1InstructionDataSerializer;
// Instruction discriminator.
exports.placeAuctionBidV1InstructionDiscriminator = 35;
// Instruction.
function placeAuctionBidV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        auctionTree: {
            index: 2,
            isWritable: true,
            value: input.auctionTree ?? null,
        },
        mint: { index: 3, isWritable: false, value: input.mint ?? null },
        bid: { index: 4, isWritable: true, value: input.bid ?? null },
        payer: {
            index: 5,
            isWritable: true,
            value: input.payer ?? null,
        },
        bidder: {
            index: 6,
            isWritable: false,
            value: input.bidder ?? null,
        },
        bidderTokenAccount: {
            index: 7,
            isWritable: true,
            value: input.bidderTokenAccount ?? null,
        },
        feeWallet: {
            index: 8,
            isWritable: true,
            value: input.feeWallet ?? null,
        },
        tokenProgram: {
            index: 9,
            isWritable: false,
            value: input.tokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 10,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 11,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        backendSigner: {
            index: 12,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.genesisAccount.value) {
        resolvedAccounts.genesisAccount.value = (0, accounts_1.findGenesisAccountV1Pda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
        });
    }
    if (!resolvedArgs.index) {
        resolvedArgs.index = 0;
    }
    if (!resolvedAccounts.bucket.value) {
        resolvedAccounts.bucket.value = (0, accounts_1.findAuctionBucketV1Pda)(context, {
            genesisAccount: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
            index: (0, shared_1.expectSome)(resolvedArgs.index),
        });
    }
    if (!resolvedAccounts.auctionTree.value) {
        resolvedAccounts.auctionTree.value = (0, accounts_1.findAuctionAVLTreeV1Pda)(context, {
            bucket: (0, shared_1.expectPublicKey)(resolvedAccounts.bucket.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.bidderTokenAccount.value) {
        resolvedAccounts.bidderTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.payer.value),
        });
    }
    if (!resolvedAccounts.feeWallet.value) {
        resolvedAccounts.feeWallet.value = (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb');
    }
    if (!resolvedAccounts.tokenProgram.value) {
        resolvedAccounts.tokenProgram.value = context.programs.getPublicKey('splToken', 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
        resolvedAccounts.tokenProgram.isWritable = false;
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getPlaceAuctionBidV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.placeAuctionBidV1 = placeAuctionBidV1;
