// @metaplex-foundation/genesis/dist/src/generated/instructions/refundVaultDepositV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.refundVaultDepositV2 = exports.refundVaultDepositV2InstructionDiscriminator = exports.getRefundVaultDepositV2InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getRefundVaultDepositV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 7 })],
    ], { description: 'RefundVaultDepositV2InstructionData' }), (value) => ({
        ...value,
        discriminator: 63,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0, 0],
    }));
}
exports.getRefundVaultDepositV2InstructionDataSerializer = getRefundVaultDepositV2InstructionDataSerializer;
// Instruction discriminator.
exports.refundVaultDepositV2InstructionDiscriminator = 63;
// Instruction.
function refundVaultDepositV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: false,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        baseMint: {
            index: 2,
            isWritable: false,
            value: input.baseMint ?? null,
        },
        quoteMint: {
            index: 3,
            isWritable: false,
            value: input.quoteMint ?? null,
        },
        bucketQuoteTokenAccount: {
            index: 4,
            isWritable: true,
            value: input.bucketQuoteTokenAccount ?? null,
        },
        recipientQuoteTokenAccount: {
            index: 5,
            isWritable: true,
            value: input.recipientQuoteTokenAccount ?? null,
        },
        depositPda: {
            index: 6,
            isWritable: true,
            value: input.depositPda ?? null,
        },
        payer: {
            index: 7,
            isWritable: true,
            value: input.payer ?? null,
        },
        recipient: {
            index: 8,
            isWritable: false,
            value: input.recipient ?? null,
        },
        actionsAuthority: {
            index: 9,
            isWritable: false,
            value: input.actionsAuthority ?? null,
        },
        quoteTokenProgram: {
            index: 10,
            isWritable: false,
            value: input.quoteTokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 11,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 12,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.quoteMint.value) {
        resolvedAccounts.quoteMint.value = (0, umi_1.publicKey)('So11111111111111111111111111111111111111112');
    }
    if (!resolvedAccounts.bucketQuoteTokenAccount.value) {
        resolvedAccounts.bucketQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucket.value),
        });
    }
    if (!resolvedAccounts.recipientQuoteTokenAccount.value) {
        resolvedAccounts.recipientQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.recipient.value),
        });
    }
    if (!resolvedAccounts.depositPda.value) {
        resolvedAccounts.depositPda.value = (0, accounts_1.findVaultDepositV2Pda)(context, {
            bucket: (0, shared_1.expectPublicKey)(resolvedAccounts.bucket.value),
            recipient: (0, shared_1.expectPublicKey)(resolvedAccounts.recipient.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.quoteTokenProgram.value) {
        resolvedAccounts.quoteTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getRefundVaultDepositV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.refundVaultDepositV2 = refundVaultDepositV2;
