// @metaplex-foundation/genesis/dist/src/generated/instructions/refundVaultDepositV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.refundVaultDepositV1 = exports.refundVaultDepositV1InstructionDiscriminator = exports.getRefundVaultDepositV1InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
function getRefundVaultDepositV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u8)()],
    ], { description: 'RefundVaultDepositV1InstructionData' }), (value) => ({ ...value, discriminator: 33, index: value.index ?? 0 }));
}
exports.getRefundVaultDepositV1InstructionDataSerializer = getRefundVaultDepositV1InstructionDataSerializer;
// Instruction discriminator.
exports.refundVaultDepositV1InstructionDiscriminator = 33;
// Instruction.
function refundVaultDepositV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: false,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        depositPda: {
            index: 2,
            isWritable: true,
            value: input.depositPda ?? null,
        },
        actionsAuthority: {
            index: 3,
            isWritable: true,
            value: input.actionsAuthority ?? null,
        },
        recipient: {
            index: 4,
            isWritable: true,
            value: input.recipient ?? null,
        },
        systemProgram: {
            index: 5,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getRefundVaultDepositV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.refundVaultDepositV1 = refundVaultDepositV1;
