// @metaplex-foundation/genesis/dist/src/generated/instructions/setBondingCurvePausedStateV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.setBondingCurvePausedStateV1 = exports.setBondingCurvePausedStateV1InstructionDiscriminator = exports.getSetBondingCurvePausedStateV1InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
function getSetBondingCurvePausedStateV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u8)()],
        ['paused', (0, serializers_1.bool)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
    ], { description: 'SetBondingCurvePausedStateV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 32,
        index: value.index ?? 0,
        padding: value.padding ?? [0, 0, 0, 0, 0],
    }));
}
exports.getSetBondingCurvePausedStateV1InstructionDataSerializer = getSetBondingCurvePausedStateV1InstructionDataSerializer;
// Instruction discriminator.
exports.setBondingCurvePausedStateV1InstructionDiscriminator = 32;
// Instruction.
function setBondingCurvePausedStateV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: false,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        actionsAuthority: {
            index: 2,
            isWritable: false,
            value: input.actionsAuthority ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = {
        ...input,
    };
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getSetBondingCurvePausedStateV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.setBondingCurvePausedStateV1 = setBondingCurvePausedStateV1;
