// @metaplex-foundation/genesis/dist/src/generated/instructions/swapBondingCurveV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.swapBondingCurveV2 = exports.swapBondingCurveV2InstructionDiscriminator = exports.getSwapBondingCurveV2InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getSwapBondingCurveV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['swapDirection', (0, types_1.getSwapDirectionSerializer)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        ['amount', (0, serializers_1.u64)()],
        ['minAmountOut', (0, serializers_1.u64)()],
    ], { description: 'SwapBondingCurveV2InstructionData' }), (value) => ({
        ...value,
        discriminator: 65,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0],
    }));
}
exports.getSwapBondingCurveV2InstructionDataSerializer = getSwapBondingCurveV2InstructionDataSerializer;
// Instruction discriminator.
exports.swapBondingCurveV2InstructionDiscriminator = 65;
// Instruction.
function swapBondingCurveV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: false,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        baseMint: {
            index: 2,
            isWritable: false,
            value: input.baseMint ?? null,
        },
        quoteMint: {
            index: 3,
            isWritable: false,
            value: input.quoteMint ?? null,
        },
        genesisBaseTokenAccount: {
            index: 4,
            isWritable: true,
            value: input.genesisBaseTokenAccount ?? null,
        },
        swapperBaseTokenAccount: {
            index: 5,
            isWritable: true,
            value: input.swapperBaseTokenAccount ?? null,
        },
        swapperQuoteTokenAccount: {
            index: 6,
            isWritable: true,
            value: input.swapperQuoteTokenAccount ?? null,
        },
        bucketQuoteTokenAccount: {
            index: 7,
            isWritable: true,
            value: input.bucketQuoteTokenAccount ?? null,
        },
        payer: {
            index: 8,
            isWritable: true,
            value: input.payer ?? null,
        },
        swapper: {
            index: 9,
            isWritable: false,
            value: input.swapper ?? null,
        },
        backendSigner: {
            index: 10,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
        feeQuoteTokenAccount: {
            index: 11,
            isWritable: true,
            value: input.feeQuoteTokenAccount ?? null,
        },
        baseTokenProgram: {
            index: 12,
            isWritable: false,
            value: input.baseTokenProgram ?? null,
        },
        quoteTokenProgram: {
            index: 13,
            isWritable: false,
            value: input.quoteTokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 14,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 15,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.quoteMint.value) {
        resolvedAccounts.quoteMint.value = (0, umi_1.publicKey)('So11111111111111111111111111111111111111112');
    }
    if (!resolvedAccounts.genesisBaseTokenAccount.value) {
        resolvedAccounts.genesisBaseTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.swapperBaseTokenAccount.value) {
        if (resolvedAccounts.swapper.value) {
            resolvedAccounts.swapperBaseTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.swapper.value),
            });
        }
        else {
            resolvedAccounts.swapperBaseTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.payer.value),
            });
        }
    }
    if (!resolvedAccounts.swapperQuoteTokenAccount.value) {
        if (resolvedAccounts.swapper.value) {
            resolvedAccounts.swapperQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.swapper.value),
            });
        }
        else {
            resolvedAccounts.swapperQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.payer.value),
            });
        }
    }
    if (!resolvedAccounts.bucketQuoteTokenAccount.value) {
        resolvedAccounts.bucketQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bucket.value),
        });
    }
    if (!resolvedAccounts.feeQuoteTokenAccount.value) {
        resolvedAccounts.feeQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
            owner: (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb'),
        });
    }
    if (!resolvedAccounts.baseTokenProgram.value) {
        resolvedAccounts.baseTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.quoteTokenProgram.value) {
        resolvedAccounts.quoteTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getSwapBondingCurveV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.swapBondingCurveV2 = swapBondingCurveV2;
