// @metaplex-foundation/genesis/dist/src/generated/instructions/swapVaultToBondingCurveV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.swapVaultToBondingCurveV1 = exports.swapVaultToBondingCurveV1InstructionDiscriminator = exports.getSwapVaultToBondingCurveV1InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getSwapVaultToBondingCurveV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['vaultBucketIndex', (0, serializers_1.u8)()],
        ['bondingCurveBucketIndex', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 5 })],
        ['swapAmount', (0, serializers_1.u64)()],
        ['ordinal', (0, serializers_1.u64)()],
    ], { description: 'SwapVaultToBondingCurveV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 31,
        vaultBucketIndex: value.vaultBucketIndex ?? 0,
        bondingCurveBucketIndex: value.bondingCurveBucketIndex ?? 0,
        padding: value.padding ?? [0, 0, 0, 0, 0],
    }));
}
exports.getSwapVaultToBondingCurveV1InstructionDataSerializer = getSwapVaultToBondingCurveV1InstructionDataSerializer;
// Instruction discriminator.
exports.swapVaultToBondingCurveV1InstructionDiscriminator = 31;
// Instruction.
function swapVaultToBondingCurveV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        vaultBucket: {
            index: 1,
            isWritable: true,
            value: input.vaultBucket ?? null,
        },
        bondingCurveBucket: {
            index: 2,
            isWritable: true,
            value: input.bondingCurveBucket ?? null,
        },
        depositPda: {
            index: 3,
            isWritable: true,
            value: input.depositPda ?? null,
        },
        bondingCurveBucketTokenAccount: {
            index: 4,
            isWritable: true,
            value: input.bondingCurveBucketTokenAccount ?? null,
        },
        recipient: {
            index: 5,
            isWritable: false,
            value: input.recipient ?? null,
        },
        recipientTokenAccount: {
            index: 6,
            isWritable: true,
            value: input.recipientTokenAccount ?? null,
        },
        payer: {
            index: 7,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 8,
            isWritable: true,
            value: input.authority ?? null,
        },
        feeWallet: {
            index: 9,
            isWritable: true,
            value: input.feeWallet ?? null,
        },
        mint: {
            index: 10,
            isWritable: false,
            value: input.mint ?? null,
        },
        tokenProgram: {
            index: 11,
            isWritable: false,
            value: input.tokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 12,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 13,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        vaultBucketBackendSigner: {
            index: 14,
            isWritable: false,
            value: input.vaultBucketBackendSigner ?? null,
        },
        bondingCurveBucketBackendSigner: {
            index: 15,
            isWritable: false,
            value: input.bondingCurveBucketBackendSigner ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.genesisAccount.value) {
        resolvedAccounts.genesisAccount.value = (0, accounts_1.findGenesisAccountV1Pda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
        });
    }
    if (!resolvedAccounts.bondingCurveBucketTokenAccount.value) {
        resolvedAccounts.bondingCurveBucketTokenAccount.value =
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bondingCurveBucket.value),
            });
    }
    if (!resolvedAccounts.recipientTokenAccount.value) {
        resolvedAccounts.recipientTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.recipient.value),
        });
    }
    if (!resolvedAccounts.authority.value) {
        resolvedAccounts.authority.value = context.identity;
    }
    if (!resolvedAccounts.feeWallet.value) {
        resolvedAccounts.feeWallet.value = (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb');
    }
    if (!resolvedAccounts.tokenProgram.value) {
        resolvedAccounts.tokenProgram.value = context.programs.getPublicKey('splToken', 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
        resolvedAccounts.tokenProgram.isWritable = false;
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getSwapVaultToBondingCurveV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.swapVaultToBondingCurveV1 = swapVaultToBondingCurveV1;
