// @metaplex-foundation/genesis/dist/src/generated/instructions/swapVaultToBondingCurveV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.swapVaultToBondingCurveV2 = exports.swapVaultToBondingCurveV2InstructionDiscriminator = exports.getSwapVaultToBondingCurveV2InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getSwapVaultToBondingCurveV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 7 })],
        ['swapAmountQuoteToken', (0, serializers_1.u64)()],
        ['ordinal', (0, serializers_1.u64)()],
    ], { description: 'SwapVaultToBondingCurveV2InstructionData' }), (value) => ({
        ...value,
        discriminator: 62,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0, 0],
    }));
}
exports.getSwapVaultToBondingCurveV2InstructionDataSerializer = getSwapVaultToBondingCurveV2InstructionDataSerializer;
// Instruction discriminator.
exports.swapVaultToBondingCurveV2InstructionDiscriminator = 62;
// Instruction.
function swapVaultToBondingCurveV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: false,
            value: input.genesisAccount ?? null,
        },
        vaultBucket: {
            index: 1,
            isWritable: true,
            value: input.vaultBucket ?? null,
        },
        bondingCurveBucket: {
            index: 2,
            isWritable: true,
            value: input.bondingCurveBucket ?? null,
        },
        baseMint: {
            index: 3,
            isWritable: false,
            value: input.baseMint ?? null,
        },
        quoteMint: {
            index: 4,
            isWritable: false,
            value: input.quoteMint ?? null,
        },
        genesisBaseTokenAccount: {
            index: 5,
            isWritable: true,
            value: input.genesisBaseTokenAccount ?? null,
        },
        recipientBaseTokenAccount: {
            index: 6,
            isWritable: true,
            value: input.recipientBaseTokenAccount ?? null,
        },
        vaultBucketQuoteTokenAccount: {
            index: 7,
            isWritable: true,
            value: input.vaultBucketQuoteTokenAccount ?? null,
        },
        bondingCurveBucketQuoteTokenAccount: {
            index: 8,
            isWritable: true,
            value: input.bondingCurveBucketQuoteTokenAccount ?? null,
        },
        depositPda: {
            index: 9,
            isWritable: true,
            value: input.depositPda ?? null,
        },
        payer: {
            index: 10,
            isWritable: false,
            value: input.payer ?? null,
        },
        actionsAuthority: {
            index: 11,
            isWritable: false,
            value: input.actionsAuthority ?? null,
        },
        recipient: {
            index: 12,
            isWritable: false,
            value: input.recipient ?? null,
        },
        bondingCurveBucketBackendSigner: {
            index: 13,
            isWritable: false,
            value: input.bondingCurveBucketBackendSigner ?? null,
        },
        feeQuoteTokenAccount: {
            index: 14,
            isWritable: true,
            value: input.feeQuoteTokenAccount ?? null,
        },
        baseTokenProgram: {
            index: 15,
            isWritable: false,
            value: input.baseTokenProgram ?? null,
        },
        quoteTokenProgram: {
            index: 16,
            isWritable: false,
            value: input.quoteTokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 17,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 18,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.quoteMint.value) {
        resolvedAccounts.quoteMint.value = (0, umi_1.publicKey)('So11111111111111111111111111111111111111112');
    }
    if (!resolvedAccounts.genesisBaseTokenAccount.value) {
        resolvedAccounts.genesisBaseTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
        });
    }
    if (!resolvedAccounts.recipientBaseTokenAccount.value) {
        resolvedAccounts.recipientBaseTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.baseMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.recipient.value),
        });
    }
    if (!resolvedAccounts.vaultBucketQuoteTokenAccount.value) {
        resolvedAccounts.vaultBucketQuoteTokenAccount.value =
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.vaultBucket.value),
            });
    }
    if (!resolvedAccounts.bondingCurveBucketQuoteTokenAccount.value) {
        resolvedAccounts.bondingCurveBucketQuoteTokenAccount.value =
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.bondingCurveBucket.value),
            });
    }
    if (!resolvedAccounts.depositPda.value) {
        resolvedAccounts.depositPda.value = (0, accounts_1.findVaultDepositV2Pda)(context, {
            bucket: (0, shared_1.expectPublicKey)(resolvedAccounts.vaultBucket.value),
            recipient: (0, shared_1.expectPublicKey)(resolvedAccounts.recipient.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.feeQuoteTokenAccount.value) {
        resolvedAccounts.feeQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
            owner: (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb'),
        });
    }
    if (!resolvedAccounts.baseTokenProgram.value) {
        resolvedAccounts.baseTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.quoteTokenProgram.value) {
        resolvedAccounts.quoteTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getSwapVaultToBondingCurveV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.swapVaultToBondingCurveV2 = swapVaultToBondingCurveV2;
