// @metaplex-foundation/genesis/dist/src/generated/instructions/triggerConditionsV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.triggerConditionsV2 = exports.triggerConditionsV2InstructionDiscriminator = exports.getTriggerConditionsV2InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
function getTriggerConditionsV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 7 })],
    ], { description: 'TriggerConditionsV2InstructionData' }), (value) => ({
        ...value,
        discriminator: 43,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0, 0],
    }));
}
exports.getTriggerConditionsV2InstructionDataSerializer = getTriggerConditionsV2InstructionDataSerializer;
// Instruction discriminator.
exports.triggerConditionsV2InstructionDiscriminator = 43;
// Instruction.
function triggerConditionsV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: false,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        baseMint: {
            index: 2,
            isWritable: false,
            value: input.baseMint ?? null,
        },
        payer: {
            index: 3,
            isWritable: true,
            value: input.payer ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getTriggerConditionsV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.triggerConditionsV2 = triggerConditionsV2;
