// @metaplex-foundation/genesis/dist/src/generated/instructions/triggerBehaviorsV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.triggerBehaviorsV2 = exports.triggerBehaviorsV2InstructionDiscriminator = exports.getTriggerBehaviorsV2InstructionDataSerializer = void 0;
const mpl_toolbox_1 = require("@metaplex-foundation/mpl-toolbox");
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
function getTriggerBehaviorsV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 7 })],
    ], { description: 'TriggerBehaviorsV2InstructionData' }), (value) => ({
        ...value,
        discriminator: 44,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0, 0],
    }));
}
exports.getTriggerBehaviorsV2InstructionDataSerializer = getTriggerBehaviorsV2InstructionDataSerializer;
// Instruction discriminator.
exports.triggerBehaviorsV2InstructionDiscriminator = 44;
// Instruction.
function triggerBehaviorsV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        genesisQuoteTokenAccount: {
            index: 1,
            isWritable: true,
            value: input.genesisQuoteTokenAccount ?? null,
        },
        primaryBucket: {
            index: 2,
            isWritable: true,
            value: input.primaryBucket ?? null,
        },
        primaryBucketQuoteTokenAccount: {
            index: 3,
            isWritable: true,
            value: input.primaryBucketQuoteTokenAccount ?? null,
        },
        baseMint: {
            index: 4,
            isWritable: false,
            value: input.baseMint ?? null,
        },
        quoteMint: {
            index: 5,
            isWritable: false,
            value: input.quoteMint ?? null,
        },
        payer: {
            index: 6,
            isWritable: true,
            value: input.payer ?? null,
        },
        feeQuoteTokenAccount: {
            index: 7,
            isWritable: true,
            value: input.feeQuoteTokenAccount ?? null,
        },
        baseTokenProgram: {
            index: 8,
            isWritable: false,
            value: input.baseTokenProgram ?? null,
        },
        quoteTokenProgram: {
            index: 9,
            isWritable: false,
            value: input.quoteTokenProgram ?? null,
        },
        associatedTokenProgram: {
            index: 10,
            isWritable: false,
            value: input.associatedTokenProgram ?? null,
        },
        systemProgram: {
            index: 11,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.quoteMint.value) {
        resolvedAccounts.quoteMint.value = (0, umi_1.publicKey)('So11111111111111111111111111111111111111112');
    }
    if (!resolvedAccounts.genesisQuoteTokenAccount.value) {
        resolvedAccounts.genesisQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
            owner: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
        });
    }
    if (!resolvedAccounts.primaryBucketQuoteTokenAccount.value) {
        resolvedAccounts.primaryBucketQuoteTokenAccount.value =
            (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
                mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
                owner: (0, shared_1.expectPublicKey)(resolvedAccounts.primaryBucket.value),
            });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.feeQuoteTokenAccount.value) {
        resolvedAccounts.feeQuoteTokenAccount.value = (0, mpl_toolbox_1.findAssociatedTokenPda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.quoteMint.value),
            owner: (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb'),
        });
    }
    if (!resolvedAccounts.baseTokenProgram.value) {
        resolvedAccounts.baseTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.quoteTokenProgram.value) {
        resolvedAccounts.quoteTokenProgram.value = (0, umi_1.publicKey)('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
    }
    if (!resolvedAccounts.associatedTokenProgram.value) {
        resolvedAccounts.associatedTokenProgram.value = (0, umi_1.publicKey)('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getTriggerBehaviorsV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.triggerBehaviorsV2 = triggerBehaviorsV2;
