// @metaplex-foundation/genesis/dist/src/generated/instructions/updateRaydiumCpmmBucketV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateRaydiumCpmmBucketV2 = exports.updateRaydiumCpmmBucketV2InstructionDiscriminator = exports.getUpdateRaydiumCpmmBucketV2InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getUpdateRaydiumCpmmBucketV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 7 })],
        [
            'baseTokenAllocation',
            (() => {
                const inner = (0, serializers_1.u64)();
                const sentinel = new Uint8Array([
                    255, 255, 255, 255, 255, 255, 255, 255,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<bigint>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
        [
            'startCondition',
            (() => {
                const inner = (0, types_1.getConditionSerializer)();
                const sentinel = new Uint8Array([
                    255, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
                    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 255, 255, 255,
                    255, 255, 127,
                ]);
                if (inner.fixedSize == null) {
                    throw new Error('Fixed-size options require an inner serializer with a fixed size.');
                }
                if (inner.fixedSize !== sentinel.length) {
                    throw new Error('Fixed-size option sentinel length must match the inner serializer fixed size.');
                }
                const { fixedSize } = inner;
                const normalize = (input) => {
                    if (input == null) {
                        return null;
                    }
                    if ((0, umi_1.isOption)(input)) {
                        return input.__option === 'None' ? null : input.value;
                    }
                    return input;
                };
                return {
                    description: 'Option<Condition>',
                    fixedSize,
                    maxSize: fixedSize,
                    serialize: (value) => {
                        const normalized = normalize(value);
                        return normalized == null
                            ? sentinel.slice()
                            : inner.serialize(normalized);
                    },
                    deserialize: (bytes, offset = 0) => {
                        const slice = bytes.slice(offset, offset + fixedSize);
                        const isSentinel = slice.every((byte, i) => byte === sentinel[i]);
                        if (isSentinel) {
                            return [(0, umi_1.none)(), offset + fixedSize];
                        }
                        const [value, newOffset] = inner.deserialize(bytes, offset);
                        return [(0, umi_1.some)(value), newOffset];
                    },
                };
            })(),
        ],
    ], { description: 'UpdateRaydiumCpmmBucketV2InstructionData' }), (value) => ({ ...value, discriminator: 96 }));
}
exports.getUpdateRaydiumCpmmBucketV2InstructionDataSerializer = getUpdateRaydiumCpmmBucketV2InstructionDataSerializer;
// Instruction discriminator.
exports.updateRaydiumCpmmBucketV2InstructionDiscriminator = 96;
// Instruction.
function updateRaydiumCpmmBucketV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: true,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        authority: {
            index: 2,
            isWritable: false,
            value: input.authority ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.authority.value) {
        resolvedAccounts.authority.value = context.identity;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getUpdateRaydiumCpmmBucketV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.updateRaydiumCpmmBucketV2 = updateRaydiumCpmmBucketV2;
