// @metaplex-foundation/genesis/dist/src/generated/instructions/updatePresaleBucketCapsV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.updatePresaleBucketCapsV1 = exports.updatePresaleBucketCapsV1InstructionDiscriminator = exports.getUpdatePresaleBucketCapsV1InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getUpdatePresaleBucketCapsV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
        ['newAllocationSolCap', (0, serializers_1.u64)()],
        ['newAllowlistSolCap', (0, serializers_1.u64)()],
    ], { description: 'UpdatePresaleBucketCapsV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 41,
        index: value.index ?? 0,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0],
    }));
}
exports.getUpdatePresaleBucketCapsV1InstructionDataSerializer = getUpdatePresaleBucketCapsV1InstructionDataSerializer;
// Instruction discriminator.
exports.updatePresaleBucketCapsV1InstructionDiscriminator = 41;
// Instruction.
function updatePresaleBucketCapsV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: false,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        mint: { index: 2, isWritable: false, value: input.mint ?? null },
        authority: {
            index: 3,
            isWritable: false,
            value: input.authority ?? null,
        },
        backendSigner: {
            index: 4,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.genesisAccount.value) {
        resolvedAccounts.genesisAccount.value = (0, accounts_1.findGenesisAccountV1Pda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
        });
    }
    if (!resolvedArgs.index) {
        resolvedArgs.index = 0;
    }
    if (!resolvedAccounts.bucket.value) {
        resolvedAccounts.bucket.value = (0, accounts_1.findPresaleBucketV1Pda)(context, {
            genesisAccount: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
            index: (0, shared_1.expectSome)(resolvedArgs.index),
        });
    }
    if (!resolvedAccounts.authority.value) {
        resolvedAccounts.authority.value = context.identity;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getUpdatePresaleBucketCapsV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.updatePresaleBucketCapsV1 = updatePresaleBucketCapsV1;
