// @metaplex-foundation/genesis/dist/src/generated/instructions/withdrawLaunchPoolV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.withdrawLaunchPoolV1 = exports.withdrawLaunchPoolV1InstructionDiscriminator = exports.getWithdrawLaunchPoolV1InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getWithdrawLaunchPoolV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['index', (0, serializers_1.u8)()],
        ['padding', (0, serializers_1.array)((0, serializers_1.u8)(), { size: 6 })],
    ], { description: 'WithdrawLaunchPoolV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 3,
        index: value.index ?? 0,
        padding: value.padding ?? [0, 0, 0, 0, 0, 0],
    }));
}
exports.getWithdrawLaunchPoolV1InstructionDataSerializer = getWithdrawLaunchPoolV1InstructionDataSerializer;
// Instruction discriminator.
exports.withdrawLaunchPoolV1InstructionDiscriminator = 3;
// Instruction.
function withdrawLaunchPoolV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('genesis', 'GNS1S5J5AspKXgpjz6SvKL66kPaKWAhaGRhCqPRxii2B');
    // Accounts.
    const resolvedAccounts = {
        genesisAccount: {
            index: 0,
            isWritable: false,
            value: input.genesisAccount ?? null,
        },
        bucket: {
            index: 1,
            isWritable: true,
            value: input.bucket ?? null,
        },
        mint: { index: 2, isWritable: false, value: input.mint ?? null },
        depositPda: {
            index: 3,
            isWritable: true,
            value: input.depositPda ?? null,
        },
        payer: {
            index: 4,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 5,
            isWritable: false,
            value: input.authority ?? null,
        },
        recipientTokenAccount: {
            index: 6,
            isWritable: true,
            value: input.recipientTokenAccount ?? null,
        },
        feeWallet: {
            index: 7,
            isWritable: true,
            value: input.feeWallet ?? null,
        },
        tokenProgram: {
            index: 8,
            isWritable: false,
            value: input.tokenProgram ?? null,
        },
        backendSigner: {
            index: 9,
            isWritable: false,
            value: input.backendSigner ?? null,
        },
        penaltyWallet: {
            index: 10,
            isWritable: true,
            value: input.penaltyWallet ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.genesisAccount.value) {
        resolvedAccounts.genesisAccount.value = (0, accounts_1.findGenesisAccountV1Pda)(context, {
            mint: (0, shared_1.expectPublicKey)(resolvedAccounts.mint.value),
        });
    }
    if (!resolvedArgs.index) {
        resolvedArgs.index = 0;
    }
    if (!resolvedAccounts.bucket.value) {
        resolvedAccounts.bucket.value = (0, accounts_1.findLaunchPoolBucketV1Pda)(context, {
            genesisAccount: (0, shared_1.expectPublicKey)(resolvedAccounts.genesisAccount.value),
            index: (0, shared_1.expectSome)(resolvedArgs.index),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.feeWallet.value) {
        resolvedAccounts.feeWallet.value = (0, umi_1.publicKey)('9kFjQsxtpBsaw8s7aUyiY3wazYDNgFP4Lj5rsBVVF8tb');
    }
    if (!resolvedAccounts.tokenProgram.value) {
        resolvedAccounts.tokenProgram.value = context.programs.getPublicKey('splToken', 'TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
        resolvedAccounts.tokenProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getWithdrawLaunchPoolV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.withdrawLaunchPoolV1 = withdrawLaunchPoolV1;
